// sanity/lib/cjs/form/inputs/files/FileInput/FileDetails.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FileDetails = FileDetails;

var _icons = require("@sanity/icons");

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _helper = require("../../common/helper");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function FileDetails(props) {
  const {
    originalFilename,
    size,
    children,
    muted,
    disabled,
    onClick,
    isMenuOpen,
    onMenuOpen
  } = props;
  const [menuElement, setMenuRef] = (0, _react.useState)(null);

  const handleClick = _react.default.useCallback(() => onMenuOpen(true), [onMenuOpen]);

  (0, _ui.useClickOutside)(_react.default.useCallback(() => onMenuOpen(false), [onMenuOpen]), [menuElement]);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    wrap: "nowrap",
    justify: "space-between",
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: muted || disabled ? undefined : 'button',
    tabIndex: disabled ? undefined : 0,
    __unstable_focusRing: true,
    radius: 2,
    padding: 2,
    tone: "inherit",
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    wrap: "nowrap",
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    tone: "transparent",
    shadow: 1,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: muted
  }, /*#__PURE__*/_react.default.createElement(_icons.BinaryDocumentIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2,
    marginLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 2,
    textOverflow: "ellipsis",
    muted: muted,
    "data-testid": "file-name"
  }, originalFilename), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true,
    "data-testid": "file-size"
  }, (0, _helper.formatBytes)(size))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Popover, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Menu, {
      ref: setMenuRef
    }, children),
    portal: true,
    open: isMenuOpen
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    icon: _icons.EllipsisVerticalIcon,
    mode: "bleed",
    "data-testid": "options-menu-button",
    onClick: handleClick
  })))));
}
