// sanity/lib/cjs/form/inputs/PortableText/toolbar/ActionMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionMenu = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _portableTextEditor = require("@sanity/portable-text-editor");

var _collapseMenu = require("../../../../components/collapseMenu");

var _hooks = require("./hooks");

var _helpers = require("./helpers");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const CollapseMenuMemo = /*#__PURE__*/(0, _react.memo)(_collapseMenu.CollapseMenu);
const MENU_POPOVER_PROPS = {
  constrainSize: true,
  portal: true
};
const COLLAPSE_BUTTON_PROPS = {
  padding: 2,
  mode: 'bleed'
};
const ActionMenu = /*#__PURE__*/(0, _react.memo)(function ActionMenu(props) {
  const {
    disabled: disabledProp,
    groups,
    isFullscreen,
    collapsed
  } = props;
  const focusBlock = (0, _hooks.useFocusBlock)();
  const features = (0, _hooks.useFeatures)();
  const editor = (0, _portableTextEditor.usePortableTextEditor)();
  const isVoidBlock = (focusBlock === null || focusBlock === void 0 ? void 0 : focusBlock._type) !== features.types.block.name;
  const isEmptyTextBlock = !isVoidBlock && (focusBlock === null || focusBlock === void 0 ? void 0 : focusBlock.children.length) === 1 && (focusBlock === null || focusBlock === void 0 ? void 0 : focusBlock.children[0].text) === '';
  const disabled = disabledProp || isVoidBlock;
  const actions = (0, _react.useMemo)(() => groups.reduce((acc, group) => {
    return acc.concat(group.actions.map( // eslint-disable-next-line max-nested-callbacks
    (action, actionIndex) => {
      if (actionIndex === 0) return { ...action,
        firstInGroup: true
      };
      return action;
    }));
  }, []), [groups]);
  const activeKeys = (0, _hooks.useActiveActionKeys)({
    actions
  });
  const handleMenuClose = (0, _react.useCallback)(() => {
    _portableTextEditor.PortableTextEditor.focus(editor);
  }, [editor]);
  const children = (0, _react.useMemo)(() => actions.map(action => {
    const annotationDisabled = action.type === 'annotation' && isEmptyTextBlock;
    const active = activeKeys.includes(action.key);
    return /*#__PURE__*/_react.default.createElement(_collapseMenu.CollapseMenuButton, _extends({
      disabled: disabled || annotationDisabled
    }, COLLAPSE_BUTTON_PROPS, {
      dividerBefore: action.firstInGroup,
      icon: (0, _helpers.getActionIcon)(action, active),
      key: action.key // eslint-disable-next-line react/jsx-no-bind
      ,
      onClick: () => action.handle(active),
      selected: active,
      text: action.title || action.key,
      tooltipText: action.title || action.key,
      tooltipProps: {
        disabled: disabled || annotationDisabled,
        placement: isFullscreen ? 'bottom' : 'top',
        portal: 'default'
      }
    }));
  }), [actions, activeKeys, disabled, isEmptyTextBlock, isFullscreen]);
  const menuButtonProps = (0, _react.useMemo)(() => ({
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.EllipsisVerticalIcon,
      mode: "bleed",
      padding: 2,
      disabled: disabled
    }),
    popover: MENU_POPOVER_PROPS
  }), [disabled]);
  return /*#__PURE__*/_react.default.createElement(CollapseMenuMemo, {
    collapsed: collapsed,
    disableRestoreFocusOnClose: true,
    gap: 1,
    menuButtonProps: menuButtonProps,
    onMenuClose: handleMenuClose
  }, children);
});
exports.ActionMenu = ActionMenu;
