// sanity/lib/esm/form/inputs/files/FileInput/FileDetails.js
import { BinaryDocumentIcon, EllipsisVerticalIcon } from '@sanity/icons';
import React, { useState } from 'react';
import { Box, Button, Card, Flex, Menu, Popover, Stack, Text, useClickOutside } from '@sanity/ui';
import { formatBytes } from '../../common/helper';
export function FileDetails(props) {
  const {
    originalFilename,
    size,
    children,
    muted,
    disabled,
    onClick,
    isMenuOpen,
    onMenuOpen
  } = props;
  const [menuElement, setMenuRef] = useState(null);
  const handleClick = React.useCallback(() => onMenuOpen(true), [onMenuOpen]);
  useClickOutside(React.useCallback(() => onMenuOpen(false), [onMenuOpen]), [menuElement]);
  return /*#__PURE__*/React.createElement(Flex, {
    wrap: "nowrap",
    justify: "space-between",
    align: "center"
  }, /*#__PURE__*/React.createElement(Card, {
    as: muted || disabled ? undefined : 'button',
    tabIndex: disabled ? undefined : 0,
    __unstable_focusRing: true,
    radius: 2,
    padding: 2,
    tone: "inherit",
    onClick: onClick
  }, /*#__PURE__*/React.createElement(Flex, {
    wrap: "nowrap",
    align: "center"
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 3,
    tone: "transparent",
    shadow: 1,
    radius: 1
  }, /*#__PURE__*/React.createElement(Text, {
    muted: muted
  }, /*#__PURE__*/React.createElement(BinaryDocumentIcon, null))), /*#__PURE__*/React.createElement(Stack, {
    flex: 1,
    space: 2,
    marginLeft: 3
  }, /*#__PURE__*/React.createElement(Text, {
    size: 2,
    textOverflow: "ellipsis",
    muted: muted,
    "data-testid": "file-name"
  }, originalFilename), /*#__PURE__*/React.createElement(Text, {
    size: 1,
    muted: true,
    "data-testid": "file-size"
  }, formatBytes(size))))), /*#__PURE__*/React.createElement(Box, {
    padding: 2
  }, /*#__PURE__*/React.createElement(Flex, {
    justify: "center"
  }, /*#__PURE__*/React.createElement(Popover, {
    content: /*#__PURE__*/React.createElement(Menu, {
      ref: setMenuRef
    }, children),
    portal: true,
    open: isMenuOpen
  }, /*#__PURE__*/React.createElement(Button, {
    icon: EllipsisVerticalIcon,
    mode: "bleed",
    "data-testid": "options-menu-button",
    onClick: handleClick
  })))));
}
