// sanity/lib/cjs/form/inputs/DateInputs/DateInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateInput = DateInput;

var _react = _interopRequireWildcard(require("react"));

var _legacyDateFormat = require("@sanity/util/legacyDateFormat");

var _patch = require("../../patch");

var _changeIndicators = require("../../../components/changeIndicators");

var _CommonDateTimeInput = require("./CommonDateTimeInput");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// This is the format dates are stored on
const VALUE_FORMAT = 'YYYY-MM-DD'; // default to how they are stored

const DEFAULT_DATE_FORMAT = VALUE_FORMAT;

function parseOptions() {
  let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    calendarTodayLabel: options.calendarTodayLabel || 'Today'
  };
}

const deserialize = value => (0, _legacyDateFormat.parse)(value, VALUE_FORMAT);

const serialize = date => (0, _legacyDateFormat.format)(date, VALUE_FORMAT);

function DateInput(props) {
  const {
    id,
    readOnly,
    onChange,
    schemaType,
    focusRef,
    path,
    focused,
    changed,
    value
  } = props;
  const {
    dateFormat
  } = parseOptions(schemaType.options);
  const handleChange = (0, _react.useCallback)(nextDate => {
    onChange(nextDate === null ? (0, _patch.unset)() : (0, _patch.set)(nextDate));
  }, [onChange]);
  const formatInputValue = (0, _react.useCallback)(date => (0, _legacyDateFormat.format)(date, dateFormat), [dateFormat]);
  const parseInputValue = (0, _react.useCallback)(inputValue => (0, _legacyDateFormat.parse)(inputValue, dateFormat), [dateFormat]);
  return /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/_react.default.createElement(_CommonDateTimeInput.CommonDateTimeInput, {
    deserialize: deserialize,
    formatInputValue: formatInputValue,
    id: id,
    onChange: handleChange,
    parseInputValue: parseInputValue,
    placeholder: schemaType.placeholder,
    readOnly: readOnly,
    ref: focusRef,
    selectTime: false,
    serialize: serialize,
    value: value
  }));
}
