// sanity/lib/esm/cli/actions/graphql/deleteApiAction.js
export default async function deleteGraphQLApi(args, context) {
  const {
    apiClient,
    output,
    prompt
  } = context;
  const flags = args.extOptions;
  let client = apiClient({
    requireUser: true,
    requireProject: true
  }).config({
    apiVersion: '1'
  });
  const projectId = flags.project ? "".concat(flags.project) : client.config().projectId;
  const dataset = flags.dataset ? "".concat(flags.dataset) : client.config().dataset;
  const tag = flags.tag ? "".concat(flags.tag) : 'default';
  const confirmMessage = tag === 'default' ? "Are you absolutely sure you want to delete the current GraphQL API connected to the \"".concat(dataset, "\" dataset in project ").concat(projectId, "?") : "Are you absolutely sure you want to delete the GraphQL API connected to the \"".concat(dataset, "\" dataset in project ").concat(projectId, ", tagged \"").concat(tag, "\"?");
  const confirmedDelete = await prompt.single({
    type: 'confirm',
    message: confirmMessage,
    default: false
  });

  if (!confirmedDelete) {
    return;
  }

  if (projectId !== client.config().projectId) {
    client = client.clone().config({
      projectId
    });
  }

  try {
    await client.request({
      url: "/apis/graphql/".concat(dataset, "/").concat(tag),
      method: 'DELETE'
    });
  } catch (err) {
    throw err;
  }

  output.print('GraphQL API deleted');
}
