// @sanity/cli/lib/actions/init-project/prompts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.promptForAclMode = promptForAclMode;
exports.promptForDefaultConfig = promptForDefaultConfig;
exports.promptForTypeScript = promptForTypeScript;
exports.promptImplicitReconfigure = promptImplicitReconfigure;

function promptForTypeScript(prompt) {
  return prompt.single({
    type: 'confirm',
    message: 'Do you want to use TypeScript?',
    default: false
  });
}

function promptForDefaultConfig(prompt) {
  return prompt.single({
    type: 'confirm',
    message: 'Use the default dataset configuration?',
    default: true
  });
}

function promptImplicitReconfigure(prompt) {
  return prompt.single({
    type: 'confirm',
    message: 'The current folder contains a configured Sanity studio. Would you like to reconfigure it?',
    default: true
  });
}

async function promptForAclMode(prompt, output) {
  const mode = await prompt.single({
    type: 'list',
    message: 'Choose dataset visibility – this can be changed later',
    choices: [{
      value: 'public',
      name: 'Public (world readable)'
    }, {
      value: 'private',
      name: 'Private (authenticated requests only)'
    }]
  });

  if (mode === 'private') {
    output.print('Please note that while documents are private, assets (files and images) are still public\n');
  }

  return mode;
}
