// sanity/lib/cjs/desk/panes/document/inspectDialog/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isDocumentLike = isDocumentLike;
exports.isExpanded = isExpanded;
exports.maybeSelectAll = maybeSelectAll;
exports.select = select;
exports.selectElement = selectElement;
exports.toggleExpanded = toggleExpanded;

var _hashlru = _interopRequireDefault(require("hashlru"));

var _util = require("../../../../util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const lru = (0, _hashlru.default)(1000);

function isExpanded(keyPath, value) {
  const cached = lru.get(keyPath);

  if (cached === undefined) {
    lru.set(keyPath, Array.isArray(value) || (0, _util.isRecord)(value));
    return isExpanded(keyPath, value);
  }

  return cached;
}

function toggleExpanded(event) {
  const {
    path
  } = event;
  const current = lru.get(path);

  if (current === undefined) {
    // something is wrong
    return;
  }

  lru.set(path, !current);
}

function selectElement(element) {
  const sel = window.getSelection();

  if (sel) {
    const range = document.createRange();
    sel.removeAllRanges();
    range.selectNodeContents(element);
    sel.addRange(range);
  }
}

function select(event) {
  selectElement(event.currentTarget);
}

function maybeSelectAll(event) {
  const selectAll = event.keyCode === 65 && (event.metaKey || event.ctrlKey);

  if (!selectAll) {
    return;
  }

  event.preventDefault();
  selectElement(event.currentTarget);
}

function isDocumentLike(value) {
  return (0, _util.isRecord)(value) && (0, _util.isString)(value._id) && (0, _util.isString)(value._type);
}
