// sanity/lib/cjs/form/inputs/files/ImageToolInput/imagetool/ImageLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageLoader = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ImageLoader extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      isLoading: true,
      image: null,
      error: null
    });
  }

  UNSAFE_componentWillMount() {
    this.loadImage(this.props.src);
  }

  loadImage(src) {
    const image = new Image();
    this.setState({
      image: null,
      error: null
    });

    image.onload = () => {
      this.setState({
        image: image,
        error: null,
        isLoading: false
      });
    };

    image.onerror = () => {
      this.setState({
        error: new Error("Could not load image from ".concat(JSON.stringify(this.props.src))),
        isLoading: false
      });
    };

    image.referrerPolicy = 'strict-origin-when-cross-origin';
    image.src = src;
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.src !== this.props.src) {
      this.loadImage(nextProps.src);
    }
  }

  render() {
    const {
      error,
      image,
      isLoading
    } = this.state;
    return this.props.children({
      image,
      error,
      isLoading
    });
  }

}

exports.ImageLoader = ImageLoader;
