// sanity/lib/cjs/cli/actions/graphql/gen2/generateTypeFilters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateTypeFilters = generateTypeFilters;

var _helpers = require("../helpers");

var _booleanFilters = require("./filters/booleanFilters");

var _dateFilters = require("./filters/dateFilters");

var _dateTimeFilters = require("./filters/dateTimeFilters");

var _documentFilters = require("./filters/documentFilters");

var _floatFilters = require("./filters/floatFilters");

var _idFilters = require("./filters/idFilters");

var _integerFilters = require("./filters/integerFilters");

var _stringFilters = require("./filters/stringFilters");

const typeAliases = {
  Url: 'String',
  Text: 'String',
  Email: 'String'
};
const filterCreators = {
  ID: _idFilters.createIdFilters,
  String: _stringFilters.createStringFilters,
  Float: _floatFilters.createFloatFilters,
  Integer: _integerFilters.createIntegerFilters,
  Boolean: _booleanFilters.createBooleanFilters,
  Datetime: _dateTimeFilters.createDateTimeFilters,
  Date: _dateFilters.createDateFilters,
  Document: _documentFilters.createDocumentFilters
};

function generateTypeFilters(types) {
  const builtInTypeKeys = Object.keys(filterCreators);
  const builtinTypeValues = Object.values(filterCreators);
  const objectTypes = types.filter(_helpers.isNonUnion).filter(type => type.type === 'Object' && !['Block', 'Span'].includes(type.name) && !type.interfaces && !builtInTypeKeys.includes(type.type));
  const unionTypes = types.filter(_helpers.isUnion).map(type => type.name);
  const documentTypes = types.filter(_helpers.isNonUnion).filter(type => type.type === 'Object' && type.interfaces && type.interfaces.includes('Document'));
  const builtinTypeFilters = createBuiltinTypeFilters(builtinTypeValues);
  const objectTypeFilters = createObjectTypeFilters(objectTypes, {
    unionTypes
  });
  const documentTypeFilters = createDocumentTypeFilters(documentTypes, {
    unionTypes
  });
  return [...builtinTypeFilters, ...objectTypeFilters, ...documentTypeFilters];
}

function createBuiltinTypeFilters(builtinTypeValues) {
  return builtinTypeValues.map(filterCreator => filterCreator());
}

function createObjectTypeFilters(objectTypes, options) {
  return objectTypes.map(objectType => ({
    name: "".concat(objectType.name, "Filter"),
    kind: 'InputObject',
    fields: createFieldFilters(objectType, options)
  }));
}

function createDocumentTypeFilters(documentTypes, options) {
  return documentTypes.map(documentType => ({
    name: "".concat(documentType.name, "Filter"),
    kind: 'InputObject',
    fields: [...getDocumentFilters(), ...createFieldFilters(documentType, options)]
  }));
}

function createFieldFilters(objectType, options) {
  const {
    unionTypes
  } = options;
  return objectType.fields.filter(field => field.type !== 'JSON' && field.kind !== 'List' && !unionTypes.includes(field.type)).map(field => ({
    fieldName: field.fieldName,
    type: "".concat(typeAliases[field.type] || field.type, "Filter"),
    isReference: field.isReference
  }));
}

function getDocumentFilters() {
  return [{
    fieldName: '_',
    type: 'DocumentFilter',
    description: 'Apply filters on document level'
  }];
}
