// sanity/lib/dts/src/datastores/history/history/Reconstruction.js
/**
 * A reconstruction represents a single reconstruction of a
 */
export class Reconstruction {
    constructor(timeline, doc, start, end) {
        this.timeline = timeline;
        this.start = start;
        this.end = end;
        this.doc = doc;
    }
    same(start, end) {
        return this.start === start && this.end === end;
    }
    /** Returns the attributes as seen at the end of the range. */
    endAttributes() {
        return getAttrs(this.endDocument());
    }
    endDocument() {
        if (!this._endDocument) {
            this._endDocument = this.timeline.replayBackwardsUntil(this.end.end, this.doc);
        }
        return this._endDocument;
    }
    /** Returns the attributes as seen at the end of the range. */
    startAttributes() {
        return getAttrs(this.startDocument());
    }
    startDocument() {
        if (!this.start)
            throw new Error('start required');
        if (!this._startDocument) {
            this._startDocument = this.timeline.replayBackwardsBetween(this.start.end, this.end.end - 1, this.endDocument());
        }
        return this._startDocument;
    }
    diff() {
        if (!this._diff) {
            if (!this.start)
                throw new Error('start required');
            this._diff = this.timeline.calculateDiff(this.startDocument(), this.endDocument(), this.start.index + 1, this.end.index);
        }
        return this._diff;
    }
}
function getAttrs(doc) {
    return doc.draft || doc.published;
}
