// sanity/lib/cjs/templates/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  prepareTemplates: true,
  defaultTemplateForType: true,
  defaultTemplatesForSchema: true,
  isBuilder: true,
  resolveInitialValue: true,
  resolveInitialValueForType: true,
  resolveInitialObjectValue: true
};
Object.defineProperty(exports, "defaultTemplateForType", {
  enumerable: true,
  get: function () {
    return _prepare.defaultTemplateForType;
  }
});
Object.defineProperty(exports, "defaultTemplatesForSchema", {
  enumerable: true,
  get: function () {
    return _prepare.defaultTemplatesForSchema;
  }
});
Object.defineProperty(exports, "isBuilder", {
  enumerable: true,
  get: function () {
    return _resolve.isBuilder;
  }
});
Object.defineProperty(exports, "prepareTemplates", {
  enumerable: true,
  get: function () {
    return _prepare.prepareTemplates;
  }
});
Object.defineProperty(exports, "resolveInitialObjectValue", {
  enumerable: true,
  get: function () {
    return _resolve.resolveInitialObjectValue;
  }
});
Object.defineProperty(exports, "resolveInitialValue", {
  enumerable: true,
  get: function () {
    return _resolve.resolveInitialValue;
  }
});
Object.defineProperty(exports, "resolveInitialValueForType", {
  enumerable: true,
  get: function () {
    return _resolve.resolveInitialValueForType;
  }
});

var _prepare = require("./prepare");

var _resolve = require("./resolve");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
