// sanity/lib/cjs/templates/prepare.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultTemplateForType = defaultTemplateForType;
exports.defaultTemplatesForSchema = defaultTemplatesForSchema;
exports.prepareTemplates = prepareTemplates;

var _resolve = require("./resolve");

var _validate = require("./validate");

function maybeSerialize(template) {
  return (0, _resolve.isBuilder)(template) ? template.serialize() : template;
}

function isNonNullable(t) {
  return !!t;
}

function prepareTemplates(schema, initialValueTemplates) {
  const serialized = initialValueTemplates.map(maybeSerialize);
  return (0, _validate.validateTemplates)(schema, serialized);
}

function defaultTemplateForType(schemaType) {
  return {
    id: schemaType.name,
    schemaType: schemaType.name,
    title: schemaType.title || schemaType.name,
    icon: schemaType.icon,
    value: schemaType.initialValue || {
      _type: schemaType.name
    }
  };
}

function defaultTemplatesForSchema(schema) {
  const schemaTypes = schema.getTypeNames().filter(typeName => !/^sanity\./.test(typeName)).map(typeName => schema.get(typeName)).filter(isNonNullable).filter(schemaType => {
    var _schemaType$type;

    return ((_schemaType$type = schemaType.type) === null || _schemaType$type === void 0 ? void 0 : _schemaType$type.name) === 'document';
  });
  return prepareTemplates(schema, schemaTypes.map(schemaType => defaultTemplateForType(schemaType)));
}
