// sanity/lib/esm/desk/components/DraftStatus.js
import React from 'react';
import { EditIcon } from '@sanity/icons';
import { Box, Text, Tooltip } from '@sanity/ui';
import { TextWithTone } from '../../components/TextWithTone';
import { TimeAgo } from './TimeAgo';
export function DraftStatus(props) {
  const {
    document
  } = props;
  const updatedAt = document && '_updatedAt' in document && document._updatedAt;
  return /*#__PURE__*/React.createElement(Tooltip, {
    portal: true,
    content: /*#__PURE__*/React.createElement(Box, {
      padding: 2
    }, /*#__PURE__*/React.createElement(Text, {
      size: 1
    }, document ? /*#__PURE__*/React.createElement(React.Fragment, null, "Edited ", updatedAt && /*#__PURE__*/React.createElement(TimeAgo, {
      time: updatedAt
    })) : /*#__PURE__*/React.createElement(React.Fragment, null, "No unpublished edits")))
  }, /*#__PURE__*/React.createElement(TextWithTone, {
    tone: "caution",
    dimmed: !document,
    muted: !document,
    size: 1
  }, /*#__PURE__*/React.createElement(EditIcon, null)));
}
