// sanity/lib/esm/desk/components/MissingDocumentTypesMessage.js
import React from 'react';
import { Box, Card, Container, Flex, Stack, Text } from '@sanity/ui';
import { WarningOutlineIcon } from '@sanity/icons';
import { useSource } from '../../studio';
export function MissingDocumentTypesMessage() {
  const {
    name: sourceName
  } = useSource();
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 0
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 4,
    radius: 2,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Box, null, /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, /*#__PURE__*/React.createElement(WarningOutlineIcon, null))), /*#__PURE__*/React.createElement(Stack, {
    flex: 1,
    marginLeft: 3,
    space: 3
  }, /*#__PURE__*/React.createElement(Text, {
    as: "h1",
    size: 1,
    weight: "bold"
  }, "No schema types in the ", /*#__PURE__*/React.createElement("em", null, sourceName), " source!"), /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Please add schema types in your source configuration."), /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, /*#__PURE__*/React.createElement("a", {
    href: "https://beta.sanity.io/docs/platform/studio/config",
    target: "_blank",
    rel: "noreferrer"
  }, "Learn how to add a schema types \u2192"))))))));
}
