// sanity/lib/dts/src/_exports/_unstable.js
export { resolveConfig } from '../config';
export * from '../components/DefaultDocument';
export * from '../components/InsufficientPermissionsMessage';
export * from '../components/IntentButton';
export * from '../components/PreviewCard';
export * from '../components/TextWithTone';
export * from '../components/UserAvatar';
export * from '../components/WithReferringDocuments';
export * from '../components/changeIndicators';
export * from '../components/collapseMenu';
export * from '../components/globalErrorHandler';
export * from '../components/popoverDialog';
export * from '../components/previews';
export * from '../components/progress';
export * from '../components/react-track-elements';
export * from '../components/rovingFocus';
export * from '../components/scroll';
export * from '../components/transitional';
export * from '../components/validation';
export * from '../components/zOffsets';
export * from '../datastores';
export * from '../desk/structureBuilder';
export { ConfirmDeleteDialog } from '../desk/components/confirmDeleteDialog';
export * from '../field';
export * from '../field/diff/components/ChangeBreadcrumb';
export * from '../field/diff/components/ChangeTitleSegment';
export * from '../field/diff/components/DiffErrorBoundary';
export * from '../field/diff/components/DiffInspectWrapper';
export * from '../field/diff/components/FallbackDiff';
export * from '../field/diff/components/FieldChange';
export * from '../field/diff/components/GroupChange';
export * from '../field/diff/components/MetaInfo';
export * from '../field/diff/components/NoChanges';
export * from '../field/diff/components/RevertChangesButton';
export * from '../field/diff/components/ValueError';
export { PatchEvent } from '../form';
export * from '../hooks/useTimeAgo';
export { DocumentPreviewPresence, FieldPresence, FieldPresenceWithOverlay, FieldPresenceWithoutOverlay, OverlayDisabled, PresenceOverlay, PresenceScope, } from '../presence';
export { SanityDefaultPreview, SanityPreview } from '../preview';
export * from '../router';
export * from '../schema';
export * from '../user-color';
