// sanity/lib/cjs/preview/components/SanityPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SanityPreview = SanityPreview;

var _types = require("@sanity/types");

var _react = _interopRequireWildcard(require("react"));

var _useDocumentPreview = require("../useDocumentPreview");

var _useVisibility = require("../useVisibility");

var _constants = require("./_constants");

var _resolvePreviewComponent2 = require("./_resolvePreviewComponent");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function SanityPreview(props) {
  const {
    layout = 'default',
    ordering,
    schemaType,
    style: styleProp,
    value: valueProp,
    ...restProps
  } = props;
  const component = (0, _resolvePreviewComponent2._resolvePreviewComponent)(schemaType);
  const [element, setElement] = (0, _react.useState)(null);
  const isPTE = ['inline', 'block', 'blockImage'].includes(layout); // Subscribe to visiblity

  const visibility = (0, _useVisibility.useVisibility)({
    // NOTE: disable when PTE preview
    element: isPTE ? null : element,
    hideDelay: _constants._HIDE_DELAY
  }); // Subscribe document preview value

  const {
    error,
    value
  } = (0, _useDocumentPreview.unstable_useDocumentPreview)({
    enabled: isPTE || visibility,
    ordering,
    schemaType,
    value: valueProp
  });
  const setRef = (0, _react.useCallback)(refValue => {
    setElement(refValue);
  }, []);
  const style = (0, _react.useMemo)(() => ({ ...styleProp,
    minWidth: (styleProp === null || styleProp === void 0 ? void 0 : styleProp.minWidth) || 1,
    minHeight: (styleProp === null || styleProp === void 0 ? void 0 : styleProp.minHeight) || 1
  }), [styleProp]);
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: setRef,
    style: style
  }, /*#__PURE__*/(0, _react.createElement)(component, { ...restProps,
    description: value === null || value === void 0 ? void 0 : value.description,
    error,
    isPlaceholder: !value,
    layout,
    media: (0, _types.isImage)(valueProp) ? valueProp : value === null || value === void 0 ? void 0 : value.media,
    schemaType,
    subtitle: value === null || value === void 0 ? void 0 : value.subtitle,
    title: value === null || value === void 0 ? void 0 : value.title,
    value
  }));
}
