// sanity/lib/esm/form/inputs/files/common/UploadPlaceholder.js
import React from 'react';
import { UploadIcon } from '@sanity/icons';
import { Flex, Inline } from '@sanity/ui';
import { FileInputButton } from './FileInputButton/FileInputButton';
import { PlaceholderText } from './PlaceholderText';
export const UploadPlaceholder = /*#__PURE__*/React.memo(function UploadPlaceholder(_ref) {
  let {
    browse,
    onUpload,
    readOnly,
    type,
    hoveringFiles,
    acceptedFiles,
    rejectedFilesCount,
    accept,
    directUploads
  } = _ref;
  return /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    justify: "space-between",
    gap: 4,
    direction: ['column', 'column', 'row'],
    paddingY: [2, 2, 0]
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    justify: "center",
    gap: 2,
    flex: 1
  }, /*#__PURE__*/React.createElement(PlaceholderText, {
    readOnly: readOnly,
    hoveringFiles: hoveringFiles,
    acceptedFiles: acceptedFiles,
    rejectedFilesCount: rejectedFilesCount,
    type: type,
    directUploads: directUploads
  })), /*#__PURE__*/React.createElement(Inline, {
    space: 2
  }, /*#__PURE__*/React.createElement(FileInputButton, {
    icon: UploadIcon,
    mode: "ghost",
    onSelect: onUpload,
    accept: accept,
    text: "Upload",
    "data-testid": "file-input-upload-button",
    disabled: readOnly || !directUploads
  }), browse));
});
