// sanity/lib/cjs/util/draftUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DRAFTS_FOLDER = void 0;
exports.collate = collate;
exports.createDraftFrom = createDraftFrom;
exports.createPublishedFrom = createPublishedFrom;
exports.getDraftId = getDraftId;
exports.getIdPair = getIdPair;
exports.getPublishedId = getPublishedId;
exports.isDraft = isDraft;
exports.isDraftId = isDraftId;
exports.isPublishedId = isPublishedId;
exports.newDraftFrom = newDraftFrom;
exports.removeDupes = removeDupes;

var _isNonNullable = require("./isNonNullable");

const DRAFTS_FOLDER = 'drafts';
exports.DRAFTS_FOLDER = DRAFTS_FOLDER;
const DRAFTS_PREFIX = "".concat(DRAFTS_FOLDER, ".");

function isDraft(document) {
  return isDraftId(document._id);
}

function isDraftId(id) {
  return id.startsWith(DRAFTS_PREFIX);
}

function getIdPair(id) {
  return {
    draftId: getDraftId(id),
    publishedId: getPublishedId(id)
  };
}

function isPublishedId(id) {
  return !isDraftId(id);
}

function getDraftId(id) {
  return isDraftId(id) ? id : DRAFTS_PREFIX + id;
}

function getPublishedId(id) {
  return isDraftId(id) ? id.slice(DRAFTS_PREFIX.length) : id;
}

function createDraftFrom(document) {
  return { ...document,
    _id: getDraftId(document._id)
  };
}

function newDraftFrom(document) {
  return { ...document,
    _id: DRAFTS_PREFIX
  };
}

function createPublishedFrom(document) {
  return { ...document,
    _id: getPublishedId(document._id)
  };
}
/**
 * Takes a list of documents and collates draft/published pairs into single entries
 * `{id: <published id>, draft?: <draft document>, published?: <published document>}`
 *
 * Note: because Map is ordered by insertion key the resulting array will be ordered by whichever
 * version appeared first
 */


function collate(documents) {
  const byId = documents.reduce((res, doc) => {
    const publishedId = getPublishedId(doc._id);
    let entry = res.get(publishedId);

    if (!entry) {
      entry = {
        id: publishedId,
        type: doc._type,
        published: undefined,
        draft: undefined
      };
      res.set(publishedId, entry);
    }

    entry[publishedId === doc._id ? 'published' : 'draft'] = doc;
    return res;
  }, new Map());
  return Array.from(byId.values());
} // Removes published documents that also has a draft


function removeDupes(documents) {
  return collate(documents).map(entry => entry.draft || entry.published).filter(_isNonNullable.isNonNullable);
}
