// sanity/lib/esm/datastores/document/getPairListener.js
/* eslint-disable @typescript-eslint/no-use-before-define */
import { defer, of as observableOf } from 'rxjs';
import { concatMap, map } from 'rxjs/operators';
export function getPairListener(client, idPair) {
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  const {
    publishedId,
    draftId
  } = idPair;
  return defer(() => client.observable.listen("*[_id == $publishedId || _id == $draftId]", {
    publishedId,
    draftId
  }, {
    includeResult: false,
    events: ['welcome', 'mutation', 'reconnect'],
    effectFormat: 'mendoza',
    tag: options.tag || 'document.pair-listener'
  })).pipe(concatMap(event => event.type === 'welcome' ? fetchInitialDocumentSnapshots().pipe(concatMap(snapshots => [createSnapshotEvent(draftId, snapshots.draft), createSnapshotEvent(publishedId, snapshots.published)])) : observableOf(event)));

  function fetchInitialDocumentSnapshots() {
    return client.observable.getDocuments([draftId, publishedId], {
      tag: 'document.snapshots'
    }).pipe(map(_ref => {
      let [draft, published] = _ref;
      return {
        draft,
        published
      };
    }));
  }
}

function createSnapshotEvent(documentId, document) {
  return {
    type: 'snapshot',
    documentId,
    document
  };
}
