// sanity/lib/cjs/form/inputs/PortableText/hooks/useScrollToOpenedMember.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useScrollToOpenedMember = useScrollToOpenedMember;

var _portableTextEditor = require("@sanity/portable-text-editor");

var _react = require("react");

var _scrollIntoViewIfNeeded = _interopRequireDefault(require("scroll-into-view-if-needed"));

var _usePortableTextMembers = require("./usePortableTextMembers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This hook will scroll to the "opened" portable text object's editor dom node.
// If the opened item is a regular text block, place the cursor there as well.
function useScrollToOpenedMember(props) {
  const {
    editorRootPath,
    scrollElement,
    hasFocus,
    onCloseItem
  } = props;
  const portableTextMemberItems = (0, _usePortableTextMembers.usePortableTextMemberItems)();
  const editor = (0, _portableTextEditor.usePortableTextEditor)(); // Find the opened item with the highest path

  const memberItem = (0, _react.useMemo)(() => {
    return portableTextMemberItems.filter(item => item.member.open).sort((a, b) => b.member.item.path.length - a.member.item.path.length)[0];
  }, [portableTextMemberItems]);
  (0, _react.useEffect)(() => {
    var _memberItem$elementRe;

    // If the editor is already is focused, don't interfere.
    if (hasFocus) {
      return;
    }

    if (memberItem !== null && memberItem !== void 0 && (_memberItem$elementRe = memberItem.elementRef) !== null && _memberItem$elementRe !== void 0 && _memberItem$elementRe.current) {
      var _memberItem$elementRe2;

      (0, _scrollIntoViewIfNeeded.default)((_memberItem$elementRe2 = memberItem.elementRef) === null || _memberItem$elementRe2 === void 0 ? void 0 : _memberItem$elementRe2.current, {
        boundary: scrollElement,
        scrollMode: 'if-needed'
      }); // If this is a text block, place the cursor in the beginning of it.

      if (memberItem.kind === 'textBlock') {
        const relativePath = memberItem.member.item.path.slice(editorRootPath.length);

        _portableTextEditor.PortableTextEditor.select(editor, {
          anchor: {
            path: relativePath,
            offset: 0
          },
          focus: {
            path: relativePath,
            offset: 0
          }
        });

        _portableTextEditor.PortableTextEditor.focus(editor); // "auto-close" regular text blocks or they get sticky here when trying to focus on an other field
        // There is no natural way of closing them (however opening something else would close them)


        onCloseItem();
      }
    }
  }, [editor, hasFocus, editorRootPath.length, memberItem, scrollElement, onCloseItem]);
}
