// sanity/lib/cjs/desk/panes/document/DocumentPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentPane = void 0;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _paths = require("@sanity/util/paths");

var _pane = require("../../components/pane");

var _usePaneLayout = require("../../components/pane/usePaneLayout");

var _error = require("../error");

var _loading = require("../loading");

var _changeIndicators = require("../../../components/changeIndicators");

var _zOffsets = require("../../../components/zOffsets");

var _environment = require("../../../environment");

var _hooks = require("../../../hooks");

var _studio = require("../../../studio");

var _components = require("../../components");

var _datastores = require("../../../datastores");

var _form = require("../../../form");

var _useDeskTool = require("../../useDeskTool");

var _changesPanel = require("./changesPanel");

var _documentPanel = require("./documentPanel");

var _DocumentOperationResults = require("./DocumentOperationResults");

var _keyboardShortcuts = require("./keyboardShortcuts");

var _statusBar = require("./statusBar");

var _useDocumentPane = require("./useDocumentPane");

var _DocumentPaneProvider = require("./DocumentPaneProvider");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const DOCUMENT_PANEL_MIN_WIDTH = 320;
const DOCUMENT_PANEL_INITIAL_MIN_WIDTH = 600;
const CHANGES_PANEL_MIN_WIDTH = 320;
const DIALOG_PROVIDER_POSITION = [// We use the `position: fixed` for dialogs on narrow screens (< 512px).
'fixed', // And we use the `position: absolute` strategy (within panes) on wide screens.
'absolute'];
const StyledChangeConnectorRoot = (0, _styledComponents.default)(_changeIndicators.ChangeConnectorRoot)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  flex: 1;\n  display: flex;\n  min-height: 0;\n  min-width: 0;\n"])));
const DocumentPane = /*#__PURE__*/(0, _react.memo)(function DocumentPane(props) {
  const {
    name: parentSourceName
  } = (0, _studio.useSource)();
  return /*#__PURE__*/_react.default.createElement(_studio.SourceProvider, {
    name: props.pane.source || parentSourceName
  }, /*#__PURE__*/_react.default.createElement(DocumentPaneInner, props));
});
exports.DocumentPane = DocumentPane;

function DocumentPaneInner(props) {
  var _routerPanesState;

  const {
    pane,
    paneKey
  } = props;
  const {
    resolveNewDocumentOptions
  } = (0, _studio.useSource)().document;
  const paneRouter = (0, _components.usePaneRouter)();
  const options = usePaneOptions(pane.options, paneRouter.params);
  const {
    documentType,
    isLoaded: isDocumentLoaded
  } = (0, _datastores.useDocumentType)(options.id, options.type);
  const templateItems = (0, _react.useMemo)(() => {
    return resolveNewDocumentOptions({
      type: 'global'
    });
  }, [resolveNewDocumentOptions]);
  const [templatePermissions, isTemplatePermissionsLoading] = (0, _datastores.useTemplatePermissions)({
    templateItems
  });
  const isLoaded = isDocumentLoaded && !isTemplatePermissionsLoading;
  const providerProps = (0, _react.useMemo)(() => {
    return isLoaded && documentType && options.type !== documentType ? mergeDocumentType(props, options, documentType) : props;
  }, [props, documentType, isLoaded, options]);
  const {
    ReferenceChildLink,
    handleEditReference,
    groupIndex,
    routerPanesState
  } = paneRouter;
  const childParams = ((_routerPanesState = routerPanesState[groupIndex + 1]) === null || _routerPanesState === void 0 ? void 0 : _routerPanesState[0].params) || {};
  const routerPanesStateLength = routerPanesState.length;
  const {
    parentRefPath
  } = childParams;
  const activePath = (0, _react.useMemo)(() => {
    return parentRefPath ? {
      path: (0, _paths.fromString)(parentRefPath),
      state: // eslint-disable-next-line no-nested-ternary
      groupIndex >= routerPanesStateLength - 1 ? 'none' : groupIndex >= routerPanesStateLength - 2 ? 'selected' : 'pressed'
    } : {
      path: [],
      state: 'none'
    };
  }, [parentRefPath, groupIndex, routerPanesStateLength]);

  if (options.type === '*' && !isLoaded) {
    return /*#__PURE__*/_react.default.createElement(_loading.LoadingPane, {
      flex: 2.5,
      minWidth: 320,
      paneKey: paneKey,
      title: "Loading document\u2026"
    });
  }

  if (!documentType) {
    return /*#__PURE__*/_react.default.createElement(_error.ErrorPane, {
      flex: 2.5,
      minWidth: 320,
      paneKey: paneKey,
      title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "The document was not found")
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p"
    }, "The document type is not defined, and a document with the ", /*#__PURE__*/_react.default.createElement("code", null, options.id), ' ', "identifier could not be found.")));
  }

  return /*#__PURE__*/_react.default.createElement(_DocumentPaneProvider.DocumentPaneProvider, providerProps, /*#__PURE__*/_react.default.createElement(_form.ReferenceInputOptionsProvider, {
    EditReferenceLinkComponent: ReferenceChildLink,
    onEditReference: handleEditReference,
    initialValueTemplateItems: templatePermissions,
    activePath: activePath
  }, /*#__PURE__*/_react.default.createElement(InnerDocumentPane, null)));
}

function usePaneOptions(options) {
  let params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const templates = (0, _hooks.useTemplates)();
  return (0, _react.useMemo)(() => {
    // The document type is provided, so return
    if (options.type && options.type !== '*') {
      return options;
    } // Attempt to derive document type from the template configuration


    const templateName = options.template || params.template;
    const template = templateName ? templates.find(t => t.id === templateName) : undefined;
    const documentType = template === null || template === void 0 ? void 0 : template.schemaType; // No document type was found in a template

    if (!documentType) {
      return options;
    } // The template provided the document type, so modify the pane’s `options` property


    return { ...options,
      type: documentType
    };
  }, [options, params.template, templates]);
}

function mergeDocumentType(props, options, documentType) {
  return { ...props,
    pane: { ...props.pane,
      options: { ...options,
        type: documentType
      }
    }
  };
}

function InnerDocumentPane() {
  const {
    changesOpen,
    documentType,
    onFocus,
    onPathOpen,
    onHistoryOpen,
    onKeyUp,
    inspectOpen,
    paneKey,
    schemaType,
    value
  } = (0, _useDocumentPane.useDocumentPane)();
  const {
    features
  } = (0, _useDeskTool.useDeskTool)();
  const {
    collapsed: layoutCollapsed
  } = (0, _usePaneLayout.usePaneLayout)();
  const zOffsets = (0, _zOffsets.useZIndex)();
  const [rootElement, setRootElement] = (0, _react.useState)(null);
  const [footerElement, setFooterElement] = (0, _react.useState)(null);
  const [actionsBoxElement, setActionsBoxElement] = (0, _react.useState)(null);
  const footerRect = (0, _ui.useElementRect)(footerElement);
  const footerH = footerRect === null || footerRect === void 0 ? void 0 : footerRect.height;
  const documentPanel = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_documentPanel.DocumentPanel, {
    footerHeight: footerH || null,
    rootElement: rootElement,
    isInspectOpen: inspectOpen
  }), [footerH, rootElement, inspectOpen]);
  const footer = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_pane.PaneFooter, {
    ref: setFooterElement
  }, /*#__PURE__*/_react.default.createElement(_statusBar.DocumentStatusBar, {
    actionsBoxRef: setActionsBoxElement
  })), []);
  const changesPanel = (0, _react.useMemo)(() => {
    if (!features.reviewChanges) return null;
    if (!changesOpen) return null;
    return /*#__PURE__*/_react.default.createElement(_ui.BoundaryElementProvider, {
      element: rootElement
    }, /*#__PURE__*/_react.default.createElement(_changesPanel.ChangesPanel, null));
  }, [changesOpen, features.reviewChanges, rootElement]);
  const onConnectorSetFocus = (0, _react.useCallback)(path => {
    onPathOpen(path);
    onFocus(path);
  }, [onPathOpen, onFocus]);
  const children = (0, _react.useMemo)(() => {
    if (!schemaType) {
      return /*#__PURE__*/_react.default.createElement(_error.ErrorPane, {
        flex: 2.5,
        minWidth: 320,
        paneKey: paneKey,
        title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Unknown document type: ", /*#__PURE__*/_react.default.createElement("code", null, documentType)),
        tone: "caution"
      }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
        space: 4
      }, documentType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "This document has the schema type ", /*#__PURE__*/_react.default.createElement("code", null, documentType), ", which is not defined as a type in the local content studio schema."), !documentType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "This document does not exist, and no schema type was specified for it."), _environment.isDev && value && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "Here is the JSON representation of the document:"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
        padding: 3,
        overflow: "auto",
        radius: 2,
        shadow: 1,
        tone: "inherit"
      }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
        language: "json",
        size: [1, 1, 2]
      }, JSON.stringify(value, null, 2))))));
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.DialogProvider, {
      position: DIALOG_PROVIDER_POSITION,
      zOffset: zOffsets.portal
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      direction: "column",
      flex: 1,
      height: layoutCollapsed ? undefined : 'fill'
    }, /*#__PURE__*/_react.default.createElement(StyledChangeConnectorRoot, {
      "data-testid": "change-connector-root",
      isReviewChangesOpen: changesOpen,
      onOpenReviewChanges: onHistoryOpen,
      onSetFocus: onConnectorSetFocus
    }, documentPanel, changesPanel))), footer, /*#__PURE__*/_react.default.createElement(_DocumentOperationResults.DocumentOperationResults, null));
  }, [changesOpen, changesPanel, documentPanel, documentType, footer, onConnectorSetFocus, onHistoryOpen, layoutCollapsed, paneKey, schemaType, value, zOffsets.portal]);
  const currentMinWidth = changesOpen ? DOCUMENT_PANEL_INITIAL_MIN_WIDTH + CHANGES_PANEL_MIN_WIDTH : DOCUMENT_PANEL_INITIAL_MIN_WIDTH;
  const minWidth = changesOpen ? DOCUMENT_PANEL_MIN_WIDTH + CHANGES_PANEL_MIN_WIDTH : DOCUMENT_PANEL_MIN_WIDTH;
  return /*#__PURE__*/_react.default.createElement(_keyboardShortcuts.DocumentActionShortcuts, {
    actionsBoxElement: actionsBoxElement,
    currentMinWidth: currentMinWidth,
    "data-testid": "document-pane",
    flex: 2.5,
    id: paneKey,
    minWidth: minWidth,
    onKeyUp: onKeyUp,
    rootRef: setRootElement
  }, children);
}
