// sanity/lib/cjs/desk/panes/document/menuItems.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProductionPreviewItem = exports.getMenuItems = void 0;

var _icons = require("@sanity/icons");

const getHistoryMenuItem = params => {
  const {
    features,
    hasValue,
    changesOpen
  } = params;
  if (!features.reviewChanges) return null;
  return {
    action: 'reviewChanges',
    title: 'Review changes',
    icon: _icons.RestoreIcon,
    isDisabled: changesOpen || !hasValue
  };
};

const getInspectItem = _ref => {
  let {
    hasValue
  } = _ref;
  return {
    action: 'inspect',
    title: 'Inspect',
    icon: _icons.BinaryDocumentIcon,
    isDisabled: !hasValue,
    shortcut: 'Ctrl+Alt+I'
  };
};

const getProductionPreviewItem = _ref2 => {
  let {
    previewUrl
  } = _ref2;
  if (!previewUrl) return null;
  return {
    action: 'production-preview',
    title: 'Open preview',
    icon: _icons.EarthAmericasIcon,
    shortcut: 'Ctrl+Alt+O'
  };
};

exports.getProductionPreviewItem = getProductionPreviewItem;

const getMenuItems = params => {
  const items = [getProductionPreviewItem, getHistoryMenuItem, getInspectItem].filter(Boolean).map(fn => fn(params));
  return items.filter(i => i !== null);
};

exports.getMenuItems = getMenuItems;
