// sanity/lib/cjs/presence/overlay/AbsoluteOverlay.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AbsoluteOverlay = AbsoluteOverlay;

var _sortBy2 = _interopRequireDefault(require("lodash/sortBy"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const OVERLAY_STYLE = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  pointerEvents: 'none'
};
const ITEM_STYLE = {
  display: 'inline-block',
  transitionProperty: 'all',
  transitionDuration: '2000ms',
  transitionTimingFunction: 'cubic-bezier(0.85, 0, 0.15, 1)',
  background: 'rgba(255, 0, 0, 0.25)',
  overflow: 'hidden',
  textAlign: 'right',
  pointerEvents: 'all',
  position: 'absolute',
  outline: '1px solid #f00'
}; // This renders regions as they are reported from the RegionReporter. Useful for debugging

function AbsoluteOverlay(props) {
  const {
    regions,
    trackerRef,
    children,
    ...rest
  } = props;
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: trackerRef,
    style: {
      position: 'relative'
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, children), /*#__PURE__*/_react.default.createElement("div", {
    style: OVERLAY_STYLE
  }, (0, _sortBy2.default)(regions.filter(r => {
    var _r$data$presence;

    return ((_r$data$presence = r.data.presence) === null || _r$data$presence === void 0 ? void 0 : _r$data$presence.length) > 0;
  }), region => -region.rect.top).map(region => {
    return /*#__PURE__*/_react.default.createElement("div", {
      key: region.id,
      style: { ...ITEM_STYLE,
        ...region.rect
      }
    }, region.id);
  })));
}
