// sanity/lib/cjs/form/studio/inputResolver/defaultInputs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultInputs = void 0;

var _BooleanInput = require("../../inputs/BooleanInput");

var _EmailInput = require("../../inputs/EmailInput");

var _NumberInput = require("../../inputs/NumberInput");

var _ObjectInput = require("../../inputs/ObjectInput");

var _StringInput = require("../../inputs/StringInput");

var _DateInputs = require("../../inputs/DateInputs");

var _TextInput = require("../../inputs/TextInput");

var _UrlInput = require("../../inputs/UrlInput");

var _SlugInput = require("../../inputs/Slug/SlugInput");

var _StudioArrayInput = require("../inputs/StudioArrayInput");

var _StudioImageInput = require("../inputs/StudioImageInput");

var _StudioFileInput = require("../inputs/StudioFileInput");

var _StudioCrossDatasetReferenceInput = require("../inputs/crossDatasetReference/StudioCrossDatasetReferenceInput");

const defaultInputs = {
  document: _ObjectInput.ObjectInput,
  object: _ObjectInput.ObjectInput,
  array: _StudioArrayInput.StudioArrayInput,
  boolean: _BooleanInput.BooleanInput,
  number: _NumberInput.NumberInput,
  text: _TextInput.TextInput,
  email: _EmailInput.EmailInput,
  datetime: _DateInputs.DateTimeInput,
  date: _DateInputs.DateInput,
  url: _UrlInput.UrlInput,
  image: _StudioImageInput.StudioImageInput,
  file: _StudioFileInput.StudioFileInput,
  string: _StringInput.StringInput,
  slug: _SlugInput.SlugInput,
  crossDatasetReference: _StudioCrossDatasetReferenceInput.StudioCrossDatasetReferenceInput
};
exports.defaultInputs = defaultInputs;
