// sanity/lib/cjs/components/previews/general/DetailPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DetailPreview = DetailPreview;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _Media = require("../_common/Media");

var _constants = require("../constants");

var _helpers = require("../helpers");

var _DetailPreview = require("./DetailPreview.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_MEDIA_DIMENSIONS = { ..._constants.PREVIEW_MEDIA_SIZE.detail,
  fit: 'crop',
  aspect: 1,
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
};

function DetailPreview(props) {
  const {
    title,
    subtitle,
    description,
    mediaDimensions = DEFAULT_MEDIA_DIMENSIONS,
    media,
    status,
    children,
    isPlaceholder
  } = props;

  const statusNode = status && /*#__PURE__*/_react.default.createElement(_DetailPreview.StatusBox, {
    marginLeft: 3,
    paddingRight: 1
  }, (0, _helpers.renderPreviewNode)(status, 'detail'));

  if (isPlaceholder) {
    return /*#__PURE__*/_react.default.createElement(_DetailPreview.RootFlex, {
      "data-testid": "detail-preview"
    }, media !== false && /*#__PURE__*/_react.default.createElement(_DetailPreview.MediaSkeleton, {
      "data-testid": "detail-preview__media"
    }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1,
      paddingLeft: media === false ? 1 : 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center",
      "data-testid": "detail-preview__header"
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      flex: 1,
      space: 2
    }, /*#__PURE__*/_react.default.createElement(_DetailPreview.TitleSkeleton, null), /*#__PURE__*/_react.default.createElement(_DetailPreview.SubtitleSkeleton, null)), statusNode), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 3
    }, /*#__PURE__*/_react.default.createElement(_DetailPreview.DescriptionSkeleton, null))));
  }

  return /*#__PURE__*/_react.default.createElement(_DetailPreview.RootFlex, {
    "data-testid": "detail-preview"
  }, media !== false && /*#__PURE__*/_react.default.createElement(_Media.Media, {
    dimensions: mediaDimensions,
    layout: "detail",
    media: media
  }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    paddingLeft: media === false ? 1 : 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    "data-testid": "detail-preview__header"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis",
    style: {
      color: 'inherit'
    }
  }, title && (0, _helpers.renderPreviewNode)(title, 'detail'), !title && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Untitled")), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, (0, _helpers.renderPreviewNode)(subtitle, 'detail'))), statusNode), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_DetailPreview.DescriptionText, {
    muted: true,
    size: 1
  }, (0, _helpers.renderPreviewNode)(description, 'detail')))), children);
}
