// sanity/lib/cjs/desk/actions/documentActions/UnpublishAction.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnpublishAction = void 0;

var _icons = require("@sanity/icons");

var _react = _interopRequireWildcard(require("react"));

var _InsufficientPermissionsMessage = require("../../../components/InsufficientPermissionsMessage");

var _datastores = require("../../../datastores");

var _hooks = require("../../../hooks");

var _components = require("../../components");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const DISABLED_REASON_TITLE = {
  NOT_PUBLISHED: 'This document is not published'
};

const UnpublishAction = _ref => {
  let {
    id,
    type,
    draft,
    onComplete,
    liveEdit
  } = _ref;
  const {
    unpublish
  } = (0, _hooks.useDocumentOperation)(id, type);
  const [isConfirmDialogOpen, setConfirmDialogOpen] = (0, _react.useState)(false);
  const [permissions, isPermissionsLoading] = (0, _datastores.useDocumentPairPermissions)({
    id,
    type,
    permission: 'unpublish'
  });
  const currentUser = (0, _datastores.useCurrentUser)();
  const handleCancel = (0, _react.useCallback)(() => {
    setConfirmDialogOpen(false);
    onComplete();
  }, [onComplete]);
  const handleConfirm = (0, _react.useCallback)(() => {
    setConfirmDialogOpen(false);
    unpublish.execute();
    onComplete();
  }, [onComplete, unpublish]);
  const modal = (0, _react.useMemo)(() => {
    if (isConfirmDialogOpen) {
      return {
        type: 'dialog',
        onClose: onComplete,
        content: /*#__PURE__*/_react.default.createElement(_components.ConfirmDeleteDialog, {
          id: (draft === null || draft === void 0 ? void 0 : draft._id) || id,
          type: type,
          action: "unpublish",
          onCancel: handleCancel,
          onConfirm: handleConfirm
        })
      };
    }

    return null;
  }, [draft, id, handleCancel, handleConfirm, isConfirmDialogOpen, onComplete, type]);

  if (liveEdit) {
    return null;
  }

  if (!isPermissionsLoading && !(permissions !== null && permissions !== void 0 && permissions.granted)) {
    return {
      tone: 'critical',
      icon: _icons.UnpublishIcon,
      label: 'Unpublish',
      title: /*#__PURE__*/_react.default.createElement(_InsufficientPermissionsMessage.InsufficientPermissionsMessage, {
        operationLabel: "unpublish this document",
        currentUser: currentUser
      }),
      disabled: true
    };
  }

  return {
    tone: 'critical',
    icon: _icons.UnpublishIcon,
    disabled: Boolean(unpublish.disabled) || isPermissionsLoading,
    label: 'Unpublish',
    title: unpublish.disabled ? DISABLED_REASON_TITLE[unpublish.disabled] : '',
    onHandle: () => setConfirmDialogOpen(true),
    modal
  };
};

exports.UnpublishAction = UnpublishAction;
UnpublishAction.action = 'unpublish';
