// sanity/lib/cjs/form/utils/asserters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertType = assertType;
exports.isArrayInputProps = isArrayInputProps;
exports.isBooleanField = isBooleanField;
exports.isObjectField = isObjectField;
exports.isObjectInputProps = isObjectInputProps;
exports.isPrimitiveField = isPrimitiveField;

var _types = require("@sanity/types");

// eslint-disable-next-line no-empty-function,@typescript-eslint/no-empty-function
function assertType(v) {}

function isObjectInputProps(inputProps) {
  return (0, _types.isObjectSchemaType)(inputProps.schemaType);
}

function isArrayInputProps(inputProps) {
  return (0, _types.isArraySchemaType)(inputProps.schemaType);
}

function isPrimitiveField(fieldProps) {
  return !(0, _types.isObjectSchemaType)(fieldProps.schemaType) && !(0, _types.isArraySchemaType)(fieldProps.schemaType);
}

function isBooleanField(fieldProps) {
  return (0, _types.isBooleanSchemaType)(fieldProps.schemaType);
}

function isObjectField(fieldProps) {
  return (0, _types.isObjectSchemaType)(fieldProps.schemaType);
}
