// sanity/lib/cjs/desk/panes/document/changesPanel/ChangesPanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChangesPanel = ChangesPanel;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _field = require("../../../../field");

var _components = require("../../../components");

var _timeline = require("../timeline");

var _useDocumentPane = require("../useDocumentPane");

var _scroll = require("../../../../components/scroll");

var _changeIndicators = require("../../../../components/changeIndicators");

var _UserAvatar = require("../../../../components/UserAvatar");

var _helpers = require("./helpers");

var _LoadingContent = require("./content/LoadingContent");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const Scroller = (0, _styledComponents.default)(_scroll.ScrollContainer)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 100%;\n  overflow: auto;\n  position: relative;\n  scroll-behavior: smooth;\n"])));

function ChangesPanel() {
  const {
    documentId,
    onHistoryClose,
    historyController,
    schemaType,
    value
  } = (0, _useDocumentPane.useDocumentPane)();
  const {
    collapsed
  } = (0, _components.usePane)();
  const scrollRef = (0, _react.useRef)(null);
  const historyState = historyController.selectionState;
  const loading = historyState === 'loading';
  const since = historyController.sinceTime;
  const diff = historyController.currentObjectDiff();
  const isComparingCurrent = !historyController.onOlderRevision();

  const documentContext = _react.default.useMemo(() => ({
    documentId,
    schemaType,
    FieldWrapper: _changeIndicators.ChangeFieldWrapper,
    rootDiff: diff,
    isComparingCurrent,
    value
  }), [documentId, diff, isComparingCurrent, schemaType, value]);

  const changeAnnotations = _react.default.useMemo(() => diff ? (0, _helpers.collectLatestAuthorAnnotations)(diff) : [], [diff]);

  if (collapsed) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    flex: 1,
    style: {
      borderLeft: '1px dashed var(--card-border-color)',
      overflow: 'hidden',
      minWidth: 320
    },
    "data-testid": "review-changes-pane"
  }, /*#__PURE__*/_react.default.createElement(_components.PaneHeader, {
    actions: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.CloseIcon,
      mode: "bleed",
      onClick: onHistoryClose,
      padding: 3,
      title: "Hide changes panel"
    }),
    subActions: changeAnnotations.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingRight: 1
    }, /*#__PURE__*/_react.default.createElement(_field.DiffTooltip, {
      annotations: changeAnnotations,
      description: "Changes by",
      placement: "bottom-end"
    }, /*#__PURE__*/_react.default.createElement(_ui.AvatarStack, {
      maxLength: 4
    }, changeAnnotations.map(_ref => {
      let {
        author
      } = _ref;
      return /*#__PURE__*/_react.default.createElement(_UserAvatar.UserAvatar, {
        key: author,
        user: author
      });
    })))),
    tabs: /*#__PURE__*/_react.default.createElement(_timeline.TimelineMenu, {
      mode: "since",
      chunk: since
    }),
    title: "Changes"
  }), /*#__PURE__*/_react.default.createElement(_components.PaneContent, null, /*#__PURE__*/_react.default.createElement(_ui.BoundaryElementProvider, {
    element: scrollRef.current
  }, /*#__PURE__*/_react.default.createElement(Scroller, {
    "data-ui": "Scroller",
    ref: scrollRef
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(Content, {
    diff: diff,
    documentContext: documentContext,
    loading: loading
  }))))));
}

function Content(_ref2) {
  let {
    diff,
    documentContext,
    loading
  } = _ref2;
  const {
    schemaType
  } = (0, _useDocumentPane.useDocumentPane)();

  if (loading) {
    return /*#__PURE__*/_react.default.createElement(_LoadingContent.LoadingContent, null);
  }

  if (!diff) {
    return /*#__PURE__*/_react.default.createElement(_field.NoChanges, null);
  }

  return /*#__PURE__*/_react.default.createElement(_field.DocumentChangeContext.Provider, {
    value: documentContext
  }, /*#__PURE__*/_react.default.createElement(_field.ChangeList, {
    diff: diff,
    schemaType: schemaType
  }));
}
