// sanity/lib/esm/form/studio/StudioFormBuilderProvider.js
/* eslint-disable camelcase */
import React from 'react';
import { useSource } from '../../studio';
import { FormBuilderProvider } from '../FormBuilderProvider';

/**
 * Default wiring for `FormBuilderProvider` when used with Sanity
 *
 * @alpha This API might change.
 */
export function StudioFormBuilderProvider(props) {
  const {
    __internal_patchChannel: patchChannel,
    autoFocus,
    changesOpen,
    children,
    collapsedFieldSets,
    collapsedPaths,
    focusPath,
    focused,
    groups,
    id,
    members,
    onChange,
    onPathBlur,
    onPathFocus,
    onPathOpen,
    onFieldGroupSelect,
    onSetPathCollapsed,
    onSetFieldSetCollapsed,
    presence,
    readOnly,
    schemaType,
    validation,
    value
  } = props;
  const {
    file,
    image,
    renderField,
    renderInput,
    renderItem,
    renderPreview
  } = useSource().form;
  return /*#__PURE__*/React.createElement(FormBuilderProvider, {
    __internal_patchChannel: patchChannel,
    autoFocus: autoFocus,
    changesOpen: changesOpen,
    collapsedFieldSets: collapsedFieldSets,
    collapsedPaths: collapsedPaths,
    file: file,
    focusPath: focusPath,
    focused: focused,
    groups: groups,
    id: id,
    image: image,
    members: members,
    onChange: onChange,
    onPathBlur: onPathBlur,
    onPathFocus: onPathFocus,
    onPathOpen: onPathOpen,
    onFieldGroupSelect: onFieldGroupSelect,
    onSetPathCollapsed: onSetPathCollapsed,
    onSetFieldSetCollapsed: onSetFieldSetCollapsed,
    presence: presence,
    readOnly: readOnly,
    renderField: renderField,
    renderInput: renderInput,
    renderItem: renderItem,
    renderPreview: renderPreview,
    schemaType: schemaType,
    validation: validation,
    value: value
  }, children);
}
