// sanity/lib/cjs/_exports/_unstable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  resolveConfig: true,
  ConfirmDeleteDialog: true,
  PatchEvent: true,
  DocumentPreviewPresence: true,
  FieldPresence: true,
  FieldPresenceWithOverlay: true,
  FieldPresenceWithoutOverlay: true,
  OverlayDisabled: true,
  PresenceOverlay: true,
  PresenceScope: true,
  SanityDefaultPreview: true,
  SanityPreview: true
};
Object.defineProperty(exports, "ConfirmDeleteDialog", {
  enumerable: true,
  get: function () {
    return _confirmDeleteDialog.ConfirmDeleteDialog;
  }
});
Object.defineProperty(exports, "DocumentPreviewPresence", {
  enumerable: true,
  get: function () {
    return _presence.DocumentPreviewPresence;
  }
});
Object.defineProperty(exports, "FieldPresence", {
  enumerable: true,
  get: function () {
    return _presence.FieldPresence;
  }
});
Object.defineProperty(exports, "FieldPresenceWithOverlay", {
  enumerable: true,
  get: function () {
    return _presence.FieldPresenceWithOverlay;
  }
});
Object.defineProperty(exports, "FieldPresenceWithoutOverlay", {
  enumerable: true,
  get: function () {
    return _presence.FieldPresenceWithoutOverlay;
  }
});
Object.defineProperty(exports, "OverlayDisabled", {
  enumerable: true,
  get: function () {
    return _presence.OverlayDisabled;
  }
});
Object.defineProperty(exports, "PatchEvent", {
  enumerable: true,
  get: function () {
    return _form.PatchEvent;
  }
});
Object.defineProperty(exports, "PresenceOverlay", {
  enumerable: true,
  get: function () {
    return _presence.PresenceOverlay;
  }
});
Object.defineProperty(exports, "PresenceScope", {
  enumerable: true,
  get: function () {
    return _presence.PresenceScope;
  }
});
Object.defineProperty(exports, "SanityDefaultPreview", {
  enumerable: true,
  get: function () {
    return _preview.SanityDefaultPreview;
  }
});
Object.defineProperty(exports, "SanityPreview", {
  enumerable: true,
  get: function () {
    return _preview.SanityPreview;
  }
});
Object.defineProperty(exports, "resolveConfig", {
  enumerable: true,
  get: function () {
    return _config.resolveConfig;
  }
});

var _config = require("../config");

var _DefaultDocument = require("../components/DefaultDocument");

Object.keys(_DefaultDocument).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _DefaultDocument[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _DefaultDocument[key];
    }
  });
});

var _InsufficientPermissionsMessage = require("../components/InsufficientPermissionsMessage");

Object.keys(_InsufficientPermissionsMessage).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _InsufficientPermissionsMessage[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _InsufficientPermissionsMessage[key];
    }
  });
});

var _IntentButton = require("../components/IntentButton");

Object.keys(_IntentButton).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IntentButton[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _IntentButton[key];
    }
  });
});

var _PreviewCard = require("../components/PreviewCard");

Object.keys(_PreviewCard).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _PreviewCard[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _PreviewCard[key];
    }
  });
});

var _TextWithTone = require("../components/TextWithTone");

Object.keys(_TextWithTone).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _TextWithTone[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _TextWithTone[key];
    }
  });
});

var _UserAvatar = require("../components/UserAvatar");

Object.keys(_UserAvatar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _UserAvatar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _UserAvatar[key];
    }
  });
});

var _WithReferringDocuments = require("../components/WithReferringDocuments");

Object.keys(_WithReferringDocuments).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _WithReferringDocuments[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _WithReferringDocuments[key];
    }
  });
});

var _changeIndicators = require("../components/changeIndicators");

Object.keys(_changeIndicators).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _changeIndicators[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _changeIndicators[key];
    }
  });
});

var _collapseMenu = require("../components/collapseMenu");

Object.keys(_collapseMenu).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _collapseMenu[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _collapseMenu[key];
    }
  });
});

var _globalErrorHandler = require("../components/globalErrorHandler");

Object.keys(_globalErrorHandler).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _globalErrorHandler[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _globalErrorHandler[key];
    }
  });
});

var _popoverDialog = require("../components/popoverDialog");

Object.keys(_popoverDialog).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _popoverDialog[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _popoverDialog[key];
    }
  });
});

var _previews = require("../components/previews");

Object.keys(_previews).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _previews[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _previews[key];
    }
  });
});

var _progress = require("../components/progress");

Object.keys(_progress).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _progress[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _progress[key];
    }
  });
});

var _reactTrackElements = require("../components/react-track-elements");

Object.keys(_reactTrackElements).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reactTrackElements[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reactTrackElements[key];
    }
  });
});

var _rovingFocus = require("../components/rovingFocus");

Object.keys(_rovingFocus).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rovingFocus[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rovingFocus[key];
    }
  });
});

var _scroll = require("../components/scroll");

Object.keys(_scroll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _scroll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _scroll[key];
    }
  });
});

var _transitional = require("../components/transitional");

Object.keys(_transitional).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _transitional[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _transitional[key];
    }
  });
});

var _validation = require("../components/validation");

Object.keys(_validation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validation[key];
    }
  });
});

var _zOffsets = require("../components/zOffsets");

Object.keys(_zOffsets).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zOffsets[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zOffsets[key];
    }
  });
});

var _datastores = require("../datastores");

Object.keys(_datastores).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datastores[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _datastores[key];
    }
  });
});

var _structureBuilder = require("../desk/structureBuilder");

Object.keys(_structureBuilder).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _structureBuilder[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _structureBuilder[key];
    }
  });
});

var _confirmDeleteDialog = require("../desk/components/confirmDeleteDialog");

var _field = require("../field");

Object.keys(_field).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _field[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _field[key];
    }
  });
});

var _ChangeBreadcrumb = require("../field/diff/components/ChangeBreadcrumb");

Object.keys(_ChangeBreadcrumb).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChangeBreadcrumb[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChangeBreadcrumb[key];
    }
  });
});

var _ChangeTitleSegment = require("../field/diff/components/ChangeTitleSegment");

Object.keys(_ChangeTitleSegment).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChangeTitleSegment[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChangeTitleSegment[key];
    }
  });
});

var _DiffErrorBoundary = require("../field/diff/components/DiffErrorBoundary");

Object.keys(_DiffErrorBoundary).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _DiffErrorBoundary[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _DiffErrorBoundary[key];
    }
  });
});

var _DiffInspectWrapper = require("../field/diff/components/DiffInspectWrapper");

Object.keys(_DiffInspectWrapper).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _DiffInspectWrapper[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _DiffInspectWrapper[key];
    }
  });
});

var _FallbackDiff = require("../field/diff/components/FallbackDiff");

Object.keys(_FallbackDiff).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _FallbackDiff[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _FallbackDiff[key];
    }
  });
});

var _FieldChange = require("../field/diff/components/FieldChange");

Object.keys(_FieldChange).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _FieldChange[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _FieldChange[key];
    }
  });
});

var _GroupChange = require("../field/diff/components/GroupChange");

Object.keys(_GroupChange).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _GroupChange[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _GroupChange[key];
    }
  });
});

var _MetaInfo = require("../field/diff/components/MetaInfo");

Object.keys(_MetaInfo).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _MetaInfo[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _MetaInfo[key];
    }
  });
});

var _NoChanges = require("../field/diff/components/NoChanges");

Object.keys(_NoChanges).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _NoChanges[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _NoChanges[key];
    }
  });
});

var _RevertChangesButton = require("../field/diff/components/RevertChangesButton");

Object.keys(_RevertChangesButton).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _RevertChangesButton[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _RevertChangesButton[key];
    }
  });
});

var _ValueError = require("../field/diff/components/ValueError");

Object.keys(_ValueError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ValueError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ValueError[key];
    }
  });
});

var _form = require("../form");

var _useTimeAgo = require("../hooks/useTimeAgo");

Object.keys(_useTimeAgo).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useTimeAgo[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useTimeAgo[key];
    }
  });
});

var _presence = require("../presence");

var _preview = require("../preview");

var _router = require("../router");

Object.keys(_router).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _router[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _router[key];
    }
  });
});

var _schema = require("../schema");

Object.keys(_schema).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _schema[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _schema[key];
    }
  });
});

var _userColor = require("../user-color");

Object.keys(_userColor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _userColor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _userColor[key];
    }
  });
});
