// sanity/lib/cjs/_exports/desk.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.ComponentBuilder;
  }
});
Object.defineProperty(exports, "ComponentViewBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.ComponentViewBuilder;
  }
});
Object.defineProperty(exports, "ConfirmDeleteDialog", {
  enumerable: true,
  get: function () {
    return _components.ConfirmDeleteDialog;
  }
});
Object.defineProperty(exports, "DeleteAction", {
  enumerable: true,
  get: function () {
    return _actions.DeleteAction;
  }
});
Object.defineProperty(exports, "DiscardChangesAction", {
  enumerable: true,
  get: function () {
    return _actions.DiscardChangesAction;
  }
});
Object.defineProperty(exports, "DocumentBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.DocumentBuilder;
  }
});
Object.defineProperty(exports, "DocumentListBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.DocumentListBuilder;
  }
});
Object.defineProperty(exports, "DocumentListItemBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.DocumentListItemBuilder;
  }
});
Object.defineProperty(exports, "DocumentTypeListBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.DocumentTypeListBuilder;
  }
});
Object.defineProperty(exports, "DuplicateAction", {
  enumerable: true,
  get: function () {
    return _actions.DuplicateAction;
  }
});
Object.defineProperty(exports, "FormViewBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.FormViewBuilder;
  }
});
Object.defineProperty(exports, "GenericListBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.GenericListBuilder;
  }
});
Object.defineProperty(exports, "GenericViewBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.GenericViewBuilder;
  }
});
Object.defineProperty(exports, "HistoryRestoreAction", {
  enumerable: true,
  get: function () {
    return _actions.HistoryRestoreAction;
  }
});
Object.defineProperty(exports, "InitialValueTemplateItemBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.InitialValueTemplateItemBuilder;
  }
});
Object.defineProperty(exports, "ListBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.ListBuilder;
  }
});
Object.defineProperty(exports, "ListItemBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.ListItemBuilder;
  }
});
Object.defineProperty(exports, "MenuItemBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.MenuItemBuilder;
  }
});
Object.defineProperty(exports, "MenuItemGroupBuilder", {
  enumerable: true,
  get: function () {
    return _structureBuilder.MenuItemGroupBuilder;
  }
});
Object.defineProperty(exports, "PublishAction", {
  enumerable: true,
  get: function () {
    return _actions.PublishAction;
  }
});
Object.defineProperty(exports, "Timeline", {
  enumerable: true,
  get: function () {
    return _datastores.Timeline;
  }
});
Object.defineProperty(exports, "TimelineController", {
  enumerable: true,
  get: function () {
    return _datastores.TimelineController;
  }
});
Object.defineProperty(exports, "UnpublishAction", {
  enumerable: true,
  get: function () {
    return _actions.UnpublishAction;
  }
});
Object.defineProperty(exports, "deskTool", {
  enumerable: true,
  get: function () {
    return _desk.deskTool;
  }
});
Object.defineProperty(exports, "useDocumentPane", {
  enumerable: true,
  get: function () {
    return _desk.useDocumentPane;
  }
});
Object.defineProperty(exports, "usePaneRouter", {
  enumerable: true,
  get: function () {
    return _components.usePaneRouter;
  }
});

var _datastores = require("../datastores");

var _desk = require("../desk");

var _actions = require("../desk/actions");

var _components = require("../desk/components");

var _structureBuilder = require("../desk/structureBuilder");
