// sanity/lib/cjs/_exports/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  SANITY_VERSION: true,
  createConfig: true,
  createPlugin: true,
  useCurrentUser: true,
  useUser: true,
  isDev: true,
  isProd: true,
  PatchEvent: true,
  useClient: true,
  useConnectionState: true,
  useDataset: true,
  useDocumentOperation: true,
  useEditState: true,
  useProjectId: true,
  useSchema: true,
  useSyncState: true,
  useTemplates: true,
  useTools: true,
  useValidationStatus: true,
  renderStudio: true,
  SourceProvider: true,
  Studio: true,
  StudioLayout: true,
  StudioProvider: true,
  ToolLink: true,
  ToolMenu: true,
  useColorScheme: true,
  useSource: true,
  useWorkspace: true,
  useWorkspaces: true,
  WorkspaceProvider: true,
  defaultTheme: true,
  getDraftId: true,
  getPublishedId: true
};
Object.defineProperty(exports, "PatchEvent", {
  enumerable: true,
  get: function () {
    return _form.PatchEvent;
  }
});
Object.defineProperty(exports, "SANITY_VERSION", {
  enumerable: true,
  get: function () {
    return _version.SANITY_VERSION;
  }
});
Object.defineProperty(exports, "SourceProvider", {
  enumerable: true,
  get: function () {
    return _studio.SourceProvider;
  }
});
Object.defineProperty(exports, "Studio", {
  enumerable: true,
  get: function () {
    return _studio.Studio;
  }
});
Object.defineProperty(exports, "StudioLayout", {
  enumerable: true,
  get: function () {
    return _studio.StudioLayout;
  }
});
Object.defineProperty(exports, "StudioProvider", {
  enumerable: true,
  get: function () {
    return _studio.StudioProvider;
  }
});
Object.defineProperty(exports, "ToolLink", {
  enumerable: true,
  get: function () {
    return _studio.ToolLink;
  }
});
Object.defineProperty(exports, "ToolMenu", {
  enumerable: true,
  get: function () {
    return _studio.ToolMenu;
  }
});
Object.defineProperty(exports, "WorkspaceProvider", {
  enumerable: true,
  get: function () {
    return _studio.WorkspaceProvider;
  }
});
Object.defineProperty(exports, "createConfig", {
  enumerable: true,
  get: function () {
    return _config.createConfig;
  }
});
Object.defineProperty(exports, "createPlugin", {
  enumerable: true,
  get: function () {
    return _config.createPlugin;
  }
});
Object.defineProperty(exports, "defaultTheme", {
  enumerable: true,
  get: function () {
    return _theme.defaultTheme;
  }
});
Object.defineProperty(exports, "getDraftId", {
  enumerable: true,
  get: function () {
    return _util.getDraftId;
  }
});
Object.defineProperty(exports, "getPublishedId", {
  enumerable: true,
  get: function () {
    return _util.getPublishedId;
  }
});
Object.defineProperty(exports, "isDev", {
  enumerable: true,
  get: function () {
    return _environment.isDev;
  }
});
Object.defineProperty(exports, "isProd", {
  enumerable: true,
  get: function () {
    return _environment.isProd;
  }
});
Object.defineProperty(exports, "renderStudio", {
  enumerable: true,
  get: function () {
    return _studio.renderStudio;
  }
});
Object.defineProperty(exports, "useClient", {
  enumerable: true,
  get: function () {
    return _hooks.useClient;
  }
});
Object.defineProperty(exports, "useColorScheme", {
  enumerable: true,
  get: function () {
    return _studio.useColorScheme;
  }
});
Object.defineProperty(exports, "useConnectionState", {
  enumerable: true,
  get: function () {
    return _hooks.useConnectionState;
  }
});
Object.defineProperty(exports, "useCurrentUser", {
  enumerable: true,
  get: function () {
    return _datastores.useCurrentUser;
  }
});
Object.defineProperty(exports, "useDataset", {
  enumerable: true,
  get: function () {
    return _hooks.useDataset;
  }
});
Object.defineProperty(exports, "useDocumentOperation", {
  enumerable: true,
  get: function () {
    return _hooks.useDocumentOperation;
  }
});
Object.defineProperty(exports, "useEditState", {
  enumerable: true,
  get: function () {
    return _hooks.useEditState;
  }
});
Object.defineProperty(exports, "useProjectId", {
  enumerable: true,
  get: function () {
    return _hooks.useProjectId;
  }
});
Object.defineProperty(exports, "useSchema", {
  enumerable: true,
  get: function () {
    return _hooks.useSchema;
  }
});
Object.defineProperty(exports, "useSource", {
  enumerable: true,
  get: function () {
    return _studio.useSource;
  }
});
Object.defineProperty(exports, "useSyncState", {
  enumerable: true,
  get: function () {
    return _hooks.useSyncState;
  }
});
Object.defineProperty(exports, "useTemplates", {
  enumerable: true,
  get: function () {
    return _hooks.useTemplates;
  }
});
Object.defineProperty(exports, "useTools", {
  enumerable: true,
  get: function () {
    return _hooks.useTools;
  }
});
Object.defineProperty(exports, "useUser", {
  enumerable: true,
  get: function () {
    return _datastores.useUser;
  }
});
Object.defineProperty(exports, "useValidationStatus", {
  enumerable: true,
  get: function () {
    return _hooks.useValidationStatus;
  }
});
Object.defineProperty(exports, "useWorkspace", {
  enumerable: true,
  get: function () {
    return _studio.useWorkspace;
  }
});
Object.defineProperty(exports, "useWorkspaces", {
  enumerable: true,
  get: function () {
    return _studio.useWorkspaces;
  }
});

var _types = require("@sanity/types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

var _version = require("../version");

var _DefaultDocument = require("../components/DefaultDocument");

Object.keys(_DefaultDocument).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _DefaultDocument[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _DefaultDocument[key];
    }
  });
});

var _config = require("../config");

var _datastores = require("../datastores");

var _environment = require("../environment");

var _form = require("../form");

var _hooks = require("../hooks");

var _studio = require("../studio");

var _theme = require("../theme");

var _util = require("../util");
