// sanity/lib/cjs/components/previews/__workshop__/PortableTextPreviewStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PortableTextPreviewStory;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _uiWorkshop = require("@sanity/ui-workshop");

var _react = _interopRequireWildcard(require("react"));

var _InlinePreview = require("../portableText/InlinePreview");

var _BlockPreview = require("../portableText/BlockPreview");

var _BlockImagePreview = require("../portableText/BlockImagePreview");

var _constants = require("../constants");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const MEDIA_OPTIONS = {
  None: 'none',
  Image: 'image',
  Icon: 'icon'
};
const LAYOUT_OPTIONS = {
  Block: 'block',
  'Block image': 'blockImage',
  Inline: 'inline'
}; // const mediaValues: Record<string, React.ReactNode> = {
//   image: <img src="https://source.unsplash.com/70x70/?abstract" />,
//   icon: <DocumentIcon />,
// }

const previewComponents = {
  block: _BlockPreview.BlockPreview,
  blockImage: _BlockImagePreview.BlockImagePreview,
  inline: _InlinePreview.InlinePreview
};
const padding = {
  block: 1,
  blockImage: 0,
  inline: 1
};

function PortableTextPreviewStory() {
  const layout = (0, _uiWorkshop.useSelect)('Layout', LAYOUT_OPTIONS, 'block');
  const withActions = (0, _uiWorkshop.useBoolean)('With actions', false);
  const isPlaceholder = (0, _uiWorkshop.useBoolean)('Is placeholder', false);
  const mediaKey = (0, _uiWorkshop.useSelect)('Media', MEDIA_OPTIONS) || 'none';
  const title = (0, _uiWorkshop.useString)('Title', 'Title', 'Props');
  const subtitle = (0, _uiWorkshop.useString)('Subtitle', 'Subtitle', 'Props');
  const status = (0, _uiWorkshop.useBoolean)('Status', false);
  const media = (0, _react.useMemo)(() => {
    const {
      width,
      height
    } = layout ? _constants.PREVIEW_MEDIA_SIZE[layout] : _constants.PREVIEW_MEDIA_SIZE.default;

    if (mediaKey === 'image') {
      return /*#__PURE__*/_react.default.createElement("img", {
        src: "https://source.unsplash.com/".concat(width * 2, "x").concat(height * 2, "/?abstract")
      });
    }

    if (mediaKey === 'icon') {
      return /*#__PURE__*/_react.default.createElement(_icons.DocumentIcon, null);
    }

    if (mediaKey === 'text') {
      return 'T';
    }

    return false;
  }, [layout, mediaKey]);
  const component = layout && previewComponents[layout];

  if (!component) {
    return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center",
      height: "fill",
      justify: "center",
      padding: 4,
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "Unknown layout: ", layout));
  }

  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    border: true,
    padding: padding[layout],
    radius: 1,
    style: {
      lineHeight: 0
    }
  }, /*#__PURE__*/(0, _react.createElement)(component, {
    actions: withActions && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      fontSize: 1,
      icon: _icons.EllipsisVerticalIcon,
      mode: "bleed",
      paddingX: 2
    }),
    isPlaceholder,
    media,
    status: status && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)),
    title,
    subtitle
  })))));
}
