// sanity/lib/cjs/desk/panes/unknown/UnknownPaneType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownPane = UnknownPane;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _util = require("../../../util");

var _pane = require("../../components/pane");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @internal
 */
function UnknownPane(props) {
  const {
    isSelected,
    pane,
    paneKey
  } = props;
  const type = (0, _util.isRecord)(pane) && pane.type || null;
  return /*#__PURE__*/_react.default.createElement(_pane.Pane, {
    id: paneKey,
    selected: isSelected
  }, /*#__PURE__*/_react.default.createElement(_pane.PaneHeader, {
    title: "Unknown pane type"
  }), /*#__PURE__*/_react.default.createElement(_pane.PaneContent, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, typeof type === 'string' ? /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Structure item of type ", /*#__PURE__*/_react.default.createElement("code", null, type), " is not a known entity.") : /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Structure item is missing required ", /*#__PURE__*/_react.default.createElement("code", null, "type"), " property."))));
}
