// sanity/lib/cjs/datastores/grants/debug/exampleGrants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.viewer = exports.restricted = exports.requiresApproval = exports.editor = exports.developer = exports.contributor = exports.administrator = void 0;
// NOTE: these example grants are used in other test files.
// if you change this file, check to see if other files need to be updated too
const administrator = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.administrator = administrator;
const editor = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.editor = editor;
const developer = [{
  filter: '_id in path("**")',
  permissions: ['read', 'create', 'history', 'update']
}];
exports.developer = developer;
const contributor = [{
  filter: '_id in path("**")',
  permissions: ['read']
}, {
  filter: '_id in path("drafts.**")',
  permissions: ['create', 'history', 'update']
}];
exports.contributor = contributor;
const viewer = [{
  filter: '_id in path("**")',
  permissions: ['read', 'history']
}];
exports.viewer = viewer;
const requiresApproval = [{
  filter: '!locked',
  permissions: ['read', 'create', 'update']
}];
exports.requiresApproval = requiresApproval;
const restricted = [{
  filter: '_id in path("drafts.**") && _type in ["stringsTest", "book", "author", "referenceTest"]',
  permissions: ['read', 'create', 'history', 'update']
}, {
  filter: '_id in path("**")',
  permissions: ['read']
}];
exports.restricted = restricted;
