// @sanity/diff/lib/dts/src/calculate/diffSimple.js
export function diffNumber(fromInput, toInput, options) {
    const fromValue = fromInput.value;
    const toValue = toInput.value;
    const type = fromInput.type;
    if (fromValue === toValue)
        return {
            type,
            action: 'unchanged',
            fromValue,
            toValue,
            isChanged: false,
        };
    return {
        type: fromInput.type,
        action: 'changed',
        isChanged: true,
        fromValue: fromValue,
        toValue: toValue,
        annotation: toInput.annotation,
    };
}
export function diffBoolean(fromInput, toInput, options) {
    const fromValue = fromInput.value;
    const toValue = toInput.value;
    const type = fromInput.type;
    if (fromValue === toValue)
        return {
            type,
            action: 'unchanged',
            fromValue,
            toValue,
            isChanged: false,
        };
    return {
        type: fromInput.type,
        action: 'changed',
        isChanged: true,
        fromValue: fromValue,
        toValue: toValue,
        annotation: toInput.annotation,
    };
}
