// sanity/lib/cjs/form/FormInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormInput = void 0;

var _react = _interopRequireWildcard(require("react"));

var _paths = require("@sanity/util/paths");

var _types = require("@sanity/types");

var _asserters = require("./utils/asserters");

var _members = require("./members");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const pass = _ref => {
  let {
    children
  } = _ref;
  return children;
};

function hasAbsolutePath(a) {
  return 'absolutePath' in a;
}

const FormInput = /*#__PURE__*/(0, _react.memo)(function FormInput(props) {
  const absolutePath = (0, _react.useMemo)(() => {
    return hasAbsolutePath(props) ? props.absolutePath : props.path.concat(props.relativePath);
  }, [props]);
  return /*#__PURE__*/_react.default.createElement(FormInputInner, _extends({}, props, {
    absolutePath: absolutePath,
    destinationRenderField: props.renderField,
    destinationRenderInput: props.renderInput,
    destinationRenderItem: props.renderItem,
    destinationRenderPreview: props.renderPreview
  }));
});
/**
 * An input that takes input props for object or array and renders an input for a given sub-path
 */

exports.FormInput = FormInput;
const FormInputInner = /*#__PURE__*/(0, _react.memo)(function FormInputInner(props) {
  const {
    absolutePath,
    destinationRenderInput,
    destinationRenderItem,
    destinationRenderField,
    destinationRenderPreview
  } = props;
  const renderInput = (0, _react.useCallback)(inputProps => {
    const isDestinationReached = (0, _paths.isEqual)(inputProps.path, absolutePath) || (0, _paths.startsWith)(absolutePath, inputProps.path);

    if (isDestinationReached) {
      // we have reached the destination node and can now render with the passed renderInput
      return destinationRenderInput(inputProps);
    }

    if (!(0, _asserters.isObjectInputProps)(inputProps) && !(0, _asserters.isArrayInputProps)(inputProps)) {
      throw new Error("Expected either object input props or array input props for: ".concat(JSON.stringify(inputProps.path)));
    } // we have not yet reached the destination path, so we'll continue recurse until we get there


    return /*#__PURE__*/_react.default.createElement(FormInputInner, _extends({}, inputProps, {
      absolutePath: absolutePath,
      destinationRenderInput: destinationRenderInput,
      destinationRenderItem: destinationRenderItem,
      destinationRenderField: destinationRenderField,
      destinationRenderPreview: destinationRenderPreview
    }));
  }, [absolutePath, destinationRenderField, destinationRenderInput, destinationRenderItem, destinationRenderPreview]);
  const renderField = (0, _react.useCallback)(fieldProps => {
    // we want to render the field around the input if either of these are true:
    // 1. we have reached the destination path and the `includeField`-prop is passed as true
    // 2. we are currently at a node somewhere below/inside the destination path
    const shouldRenderField = (0, _paths.startsWith)(absolutePath, fieldProps.path) && (props.includeField || !(0, _paths.isEqual)(absolutePath, fieldProps.path));
    return shouldRenderField ? destinationRenderField(fieldProps) : pass(fieldProps);
  }, [absolutePath, destinationRenderField, props.includeField]);

  if ((0, _asserters.isArrayInputProps)(props)) {
    const childPath = (0, _paths.trimLeft)(props.path, absolutePath);
    const itemMember = props.members.find(member => member.kind == 'item' && (0, _types.isKeySegment)(childPath[0]) && member.key === childPath[0]._key);

    if (!itemMember) {
      const relativePath = (0, _paths.trimLeft)(props.path, absolutePath);
      const key = relativePath[0]._key;
      return /*#__PURE__*/_react.default.createElement("div", null, "No array item with _key ", /*#__PURE__*/_react.default.createElement("code", null, "\"", key, "\""), " found at path ", JSON.stringify(props.path));
    }

    if (itemMember.kind === 'error') {
      return /*#__PURE__*/_react.default.createElement(_members.MemberItemError, {
        member: itemMember
      });
    }

    return /*#__PURE__*/_react.default.createElement(_members.MemberItem, {
      member: itemMember,
      renderInput: renderInput,
      renderField: renderField,
      renderItem: pass,
      renderPreview: destinationRenderPreview
    });
  }

  if ((0, _asserters.isObjectInputProps)(props)) {
    const childPath = (0, _paths.trimLeft)(props.path, absolutePath);
    const fieldMember = props.members.find(member => member.kind == 'field' && childPath[0] === member.name);

    if (!fieldMember) {
      const fieldName = childPath[0];
      return /*#__PURE__*/_react.default.createElement("div", null, "Field ", JSON.stringify(fieldName), " not found among members \u2013 please verify that it's both defined in the schema and that it has not been conditionally hidden.");
    }

    return /*#__PURE__*/_react.default.createElement(_members.MemberField, {
      member: fieldMember,
      renderInput: renderInput,
      renderField: renderField,
      renderItem: pass,
      renderPreview: destinationRenderPreview
    });
  }

  throw new Error('FormInput can only be used with arrays or objects');
});
