// sanity/lib/cjs/form/FormBuilderProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormBuilderProvider = FormBuilderProvider;

var _react = _interopRequireWildcard(require("react"));

var _FormBuilderContext = require("./FormBuilderContext");

var _ArrayFunctions = require("./inputs/arrays/common/ArrayFunctions");

var _Markers = require("./inputs/PortableText/_legacyDefaultParts/Markers");

var _CustomMarkers = require("./inputs/PortableText/_legacyDefaultParts/CustomMarkers");

var _FormCallbacks = require("./studio/contexts/FormCallbacks");

var _Presence = require("./studio/contexts/Presence");

var _Validation = require("./studio/contexts/Validation");

var _defaults = require("./defaults");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint-disable camelcase */
const missingPatchChannel = {
  publish: () => undefined,
  subscribe: () => {
    console.warn('No patch channel provided to form-builder. If you need input based patch updates, please provide one');
    return () => undefined;
  }
};

function FormBuilderProvider(props) {
  const {
    __internal_patchChannel: patchChannel = missingPatchChannel,
    autoFocus,
    changesOpen,
    children,
    collapsedFieldSets,
    collapsedPaths,
    file,
    filterField,
    focusPath,
    focused,
    groups,
    id,
    image,
    members,
    onChange,
    onFieldGroupSelect,
    onPathBlur,
    onPathFocus,
    onPathOpen,
    onSetFieldSetCollapsed,
    onSetPathCollapsed,
    presence,
    readOnly,
    renderField,
    renderInput,
    renderItem,
    renderPreview,
    schemaType,
    unstable,
    validation,
    value: documentValue
  } = props;
  const documentValueRef = (0, _react.useRef)(documentValue);
  (0, _react.useEffect)(() => {
    documentValueRef.current = documentValue;
  }, [documentValue]);

  const __internal = (0, _react.useMemo)(() => ({
    patchChannel,
    // eslint-disable-line camelcase
    components: {
      ArrayFunctions: (unstable === null || unstable === void 0 ? void 0 : unstable.ArrayFunctions) || _ArrayFunctions.DefaultArrayInputFunctions,
      CustomMarkers: (unstable === null || unstable === void 0 ? void 0 : unstable.CustomMarkers) || _CustomMarkers.DefaultCustomMarkers,
      Markers: (unstable === null || unstable === void 0 ? void 0 : unstable.Markers) || _Markers.DefaultMarkers
    },
    file: {
      assetSources: file !== null && file !== void 0 && file.assetSources ? _ensureArrayOfSources(file.assetSources) || _defaults.defaultFileAssetSources : _defaults.defaultFileAssetSources,
      directUploads: (file === null || file === void 0 ? void 0 : file.directUploads) !== false
    },
    filterField: filterField || (() => true),
    image: {
      assetSources: image !== null && image !== void 0 && image.assetSources ? _ensureArrayOfSources(image.assetSources) || _defaults.defaultImageAssetSources : _defaults.defaultImageAssetSources,
      directUploads: (image === null || image === void 0 ? void 0 : image.directUploads) !== false
    },
    getDocument: () => documentValueRef.current,
    onChange
  }), [file, filterField, image, onChange, patchChannel, unstable]);

  const formBuilder = (0, _react.useMemo)(() => ({
    __internal,
    autoFocus,
    changesOpen,
    collapsedFieldSets,
    collapsedPaths,
    focusPath,
    focused,
    groups,
    id,
    members,
    readOnly,
    renderField,
    renderInput,
    renderItem,
    renderPreview,
    schemaType,
    value: documentValue
  }), [__internal, autoFocus, changesOpen, collapsedFieldSets, collapsedPaths, documentValue, focusPath, focused, groups, id, members, readOnly, renderField, renderInput, renderItem, renderPreview, schemaType]);
  return /*#__PURE__*/_react.default.createElement(_FormBuilderContext.FormBuilderContext.Provider, {
    value: formBuilder
  }, /*#__PURE__*/_react.default.createElement(_FormCallbacks.FormCallbacksProvider, {
    onChange: onChange,
    onFieldGroupSelect: onFieldGroupSelect,
    onPathBlur: onPathBlur,
    onPathFocus: onPathFocus,
    onPathOpen: onPathOpen,
    onSetPathCollapsed: onSetPathCollapsed,
    onSetFieldSetCollapsed: onSetFieldSetCollapsed
  }, /*#__PURE__*/_react.default.createElement(_Presence.PresenceProvider, {
    presence: presence
  }, /*#__PURE__*/_react.default.createElement(_Validation.ValidationProvider, {
    validation: validation
  }, children))));
}

function _ensureArrayOfSources(sources) {
  if (Array.isArray(sources)) {
    return sources;
  }

  console.warn('Configured asset sources is not an array:', sources);
  return null;
}
