// sanity/lib/cjs/form/inputs/files/common/FileInputButton/FileInputButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FileInputButton = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactIs = require("react-is");

var _autoId = require("@reach/auto-id");

var _ui = require("@sanity/ui");

var _styles = require("./styles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const FileInputButton = /*#__PURE__*/_react.default.forwardRef(function FileInputButton(props, forwardedRef) {
  const {
    icon,
    id: idProp,
    accept,
    capture,
    fontSize,
    multiple,
    onSelect,
    padding = 3,
    space = 3,
    textAlign,
    text,
    disabled,
    ...rest
  } = props;
  const id = (0, _autoId.useId)(idProp);
  const theme = (0, _ui.useTheme)();

  const handleChange = _react.default.useCallback(event => {
    if (onSelect && event.target.files) {
      onSelect(Array.from(event.target.files));
    }
  }, [onSelect]);

  const content = /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    justify: "center",
    padding: padding
  }, icon && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginRight: text ? space : undefined
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: fontSize
  }, /*#__PURE__*/(0, _react.isValidElement)(icon) && icon, (0, _reactIs.isValidElementType)(icon) && /*#__PURE__*/(0, _react.createElement)(icon))), text && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    align: textAlign,
    size: fontSize,
    textOverflow: "ellipsis",
    weight: theme.sanity.button.textWeight
  }, text));

  return /*#__PURE__*/_react.default.createElement(_styles.FileButton, _extends({}, rest, {
    htmlFor: id,
    padding: 0,
    fontSize: 2,
    disabled: disabled
  }), content, /*#__PURE__*/_react.default.createElement("input", {
    "data-testid": "file-button-input",
    accept: accept,
    capture: capture,
    id: id,
    multiple: multiple,
    onChange: handleChange,
    ref: forwardedRef,
    type: "file",
    value: "",
    disabled: disabled
  }));
});

exports.FileInputButton = FileInputButton;
