// sanity/lib/cjs/cli/util/colorizeJson.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.colorizeJson = colorizeJson;

var _jsonLexer = _interopRequireDefault(require("json-lexer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const identity = inp => inp;

function colorizeJson(input, chalk) {
  const formatters = {
    punctuator: chalk.white,
    key: chalk.white,
    string: chalk.green,
    number: chalk.yellow,
    literal: chalk.bold,
    whitespace: identity
  };
  const json = JSON.stringify(input, null, 2);
  return (0, _jsonLexer.default)(json).map((token, i, arr) => {
    // Note how the following only works because we pretty-print the JSON
    const prevToken = i === 0 ? token : arr[i - 1];

    if (token.type === 'string' && prevToken.type === 'whitespace' && /^\n\s+$/.test(prevToken.value)) {
      return { ...token,
        type: 'key'
      };
    }

    return token;
  }).map(token => {
    const formatter = formatters[token.type] || identity;
    return formatter(token.raw);
  }).join('');
}
