// @sanity/dashboard/lib/legacyParts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DefaultPreview", {
  enumerable: true,
  get: function get() {
    return _default.default;
  }
});
Object.defineProperty(exports, "WidgetContainer", {
  enumerable: true,
  get: function get() {
    return _widgetContainer.default;
  }
});
Object.defineProperty(exports, "dashboardConfig", {
  enumerable: true,
  get: function get() {
    return _config.default;
  }
});
Object.defineProperty(exports, "definitions", {
  enumerable: true,
  get: function get() {
    return _widget.default;
  }
});
Object.defineProperty(exports, "sanityClient", {
  enumerable: true,
  get: function get() {
    return _client.default;
  }
});
Object.defineProperty(exports, "userStore", {
  enumerable: true,
  get: function get() {
    return _user.default;
  }
});

var _widgetContainer = _interopRequireDefault(require("part:@sanity/dashboard/widget-container"));

var _config = _interopRequireDefault(require("part:@sanity/dashboard/config?"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _widget = _interopRequireDefault(require("all:part:@sanity/dashboard/widget?"));

var _default = _interopRequireDefault(require("part:@sanity/components/previews/default"));

var _user = _interopRequireDefault(require("part:@sanity/base/user"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
