// sanity/lib/cjs/field/types/portableText/diff/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UNKNOWN_TYPE_NAME = void 0;
exports.blockToSymbolizedText = blockToSymbolizedText;
exports.createPortableTextDiff = createPortableTextDiff;
exports.escapeRegExp = escapeRegExp;
exports.findAnnotationDiff = findAnnotationDiff;
exports.findChildDiff = findChildDiff;
exports.findSpanDiffFromChild = findSpanDiffFromChild;
exports.getAllMarkDefs = getAllMarkDefs;
exports.getAnnotations = getAnnotations;
exports.getChildSchemaType = getChildSchemaType;
exports.getDecorators = getDecorators;
exports.getInlineObjects = getInlineObjects;
exports.hasPTMemberType = hasPTMemberType;
exports.isDecorator = isDecorator;
exports.isEmptyObject = isEmptyObject;
exports.isHeader = isHeader;
exports.isPTSchemaType = isPTSchemaType;

var _orderBy2 = _interopRequireDefault(require("lodash/orderBy"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _flatten2 = _interopRequireDefault(require("lodash/flatten"));

var _diffMatchPatch = require("diff-match-patch");

var TextSymbols = _interopRequireWildcard(require("./symbols"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const dmp = new _diffMatchPatch.diff_match_patch();
const UNKNOWN_TYPE_NAME = '_UNKOWN_TYPE_';
exports.UNKNOWN_TYPE_NAME = UNKNOWN_TYPE_NAME;

function hasPTMemberType(schemaType) {
  return schemaType.of.some(isPTSchemaType);
}

const startMarkSymbols = TextSymbols.DECORATOR_SYMBOLS.map(set => set[0]).concat(TextSymbols.ANNOTATION_SYMBOLS.map(set => set[0]));
const endMarkSymbols = TextSymbols.DECORATOR_SYMBOLS.map(set => set[1]).concat(TextSymbols.ANNOTATION_SYMBOLS.map(set => set[1]));
const allSymbols = startMarkSymbols.concat(endMarkSymbols).concat(TextSymbols.INLINE_SYMBOLS).concat(TextSymbols.CHILD_SYMBOL).concat(TextSymbols.SEGMENT_START_SYMBOL);
const symbolRegex = new RegExp("".concat(allSymbols.join('|')), 'g');
const segmentRegex = new RegExp("".concat(allSymbols.join('|'), "|\n"), 'g');

function isPTSchemaType(schemaType) {
  return schemaType.jsonType === 'object' && schemaType.name === 'block';
}

function isHeader(node) {
  return !!node.style && ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(node.style);
}

function findChildDiff(diff, child) {
  const childrenDiff = diff.fields.children;
  return childrenDiff.items.filter(item => item.diff.isChanged && (item.diff.toValue === child || item.diff.fromValue === child)).map(item => item.diff).map(childDiff => childDiff)[0];
}

function getChildSchemaType(fields, child) {
  const childrenField = fields.find(f => f.name === 'children');
  const cSchemaType = childrenField && childrenField.type && childrenField.type.jsonType === 'array' && childrenField.type.of.find(type => type.name === child._type) || undefined;
  return cSchemaType;
}

function getDecorators(spanSchemaType) {
  if (spanSchemaType.decorators) {
    return (0, _orderBy2.default)(spanSchemaType.decorators, ['value'], ['asc']);
  }

  return [];
}

function getAnnotations(spanSchemaType) {
  if (spanSchemaType.annotations) {
    return (0, _orderBy2.default)(spanSchemaType.annotations, ['name'], ['asc']);
  }

  return [];
}

function isDecorator(name, schemaType) {
  return getDecorators(schemaType).some(dec => dec.value === name);
}

function blockToSymbolizedText(diff, block, decoratorMap, annotationMap, inlineMap) {
  if (!block) {
    return '';
  }

  return block.children.map(child => {
    var _child$text;

    let returned = ((_child$text = child.text) === null || _child$text === void 0 ? void 0 : _child$text.replace(symbolRegex, '')) || ''; // Make sure symbols aren't in the text already

    if (child._type === 'span') {
      // Attatch stringdiff segments
      const spanDiff = findSpanDiffFromChild(diff, child);
      const textDiff = spanDiff === null || spanDiff === void 0 ? void 0 : spanDiff.fields.text;

      if (textDiff && textDiff.toValue === child.text && textDiff.type === 'string' && textDiff.action !== 'unchanged') {
        returned = textDiff.segments.filter(seg => seg.action !== 'removed').map(seg => seg.text.replace(symbolRegex, '')).join(TextSymbols.SEGMENT_START_SYMBOL);
      }

      if (child.marks) {
        child.marks.forEach(mark => {
          const _isDecorator = !!decoratorMap[mark];

          if (_isDecorator) {
            returned = "".concat(decoratorMap[mark][0]).concat(returned).concat(decoratorMap[mark][1]);
          } else if (annotationMap[mark]) {
            returned = "".concat(annotationMap[mark][0]).concat(returned).concat(annotationMap[mark][1]);
          }
        });
      }
    } else {
      returned = inlineMap[child._key];
    }

    return "".concat(TextSymbols.CHILD_SYMBOL).concat(returned);
  }).join('');
}

function createPortableTextDiff(diff, schemaType) {
  const displayValue = diff.action === 'removed' ? diff.fromValue : diff.toValue;
  const _diff = { ...diff,
    origin: diff,
    displayValue
  };

  if (displayValue) {
    const annotationMap = {};
    const decoratorMap = {};
    const inlineMap = {};
    const spanSchemaType = getChildSchemaType(schemaType.fields, {
      _key: 'bogus',
      _type: 'span'
    });

    if (spanSchemaType) {
      getDecorators(spanSchemaType).forEach((dec, index) => {
        decoratorMap[dec.value] = TextSymbols.DECORATOR_SYMBOLS[index];
      });
    }

    const allMarkDefs = getAllMarkDefs(_diff.origin);
    allMarkDefs.forEach((markDef, index) => {
      annotationMap[markDef._key] = TextSymbols.ANNOTATION_SYMBOLS[index];
    });
    const inlines = getInlineObjects(_diff.origin);
    inlines.forEach((inline, index) => {
      inlineMap[inline._key] = TextSymbols.INLINE_SYMBOLS[index];
    });
    const fromText = blockToSymbolizedText(_diff.origin, _diff.fromValue, decoratorMap, annotationMap, inlineMap);
    const toText = blockToSymbolizedText(_diff.origin, _diff.toValue, decoratorMap, annotationMap, inlineMap);
    const toPseudoValue = { ...displayValue,
      children: [{
        _type: 'span',
        _key: 'pseudoSpanKey',
        text: toText,
        marks: []
      }]
    };
    const fromPseudoValue = {
      displayValue,
      children: [{
        _type: 'span',
        _key: 'pseudoSpanKey',
        text: fromText,
        marks: []
      }]
    };
    const pseudoDiff = {
      origin: diff,
      action: 'changed',
      type: 'object',
      displayValue: toPseudoValue,
      fromValue: fromPseudoValue,
      toValue: toPseudoValue,
      isChanged: true,
      fields: {
        children: {
          action: 'changed',
          type: 'array',
          isChanged: true,
          items: [{
            diff: {
              action: 'changed',
              type: 'object',
              isChanged: true,
              fields: {
                text: {
                  type: 'string',
                  action: 'changed',
                  isChanged: true,
                  fromValue: fromText,
                  toValue: toText,
                  segments: buildSegments(fromText, toText).map(seg => ({ ...seg,
                    ...(_diff.action !== 'unchanged' && _diff.annotation ? {
                      annotation: _diff.annotation
                    } // Fallback if we can't find a spesific original diff
                    : {})
                  }))
                }
              },
              fromValue: fromPseudoValue.children[0],
              toValue: toPseudoValue.children[0]
            },
            annotation: null,
            fromIndex: 0,
            toIndex: 0,
            hasMoved: false
          }],
          fromValue: fromPseudoValue.children,
          toValue: toPseudoValue.children
        }
      }
    };
    return pseudoDiff;
  }

  throw new Error('Can not display this diff');
}

function buildSegments(fromInput, toInput) {
  const segments = [];
  const dmpDiffs = dmp.diff_main(fromInput, toInput);
  dmp.diff_cleanupEfficiency(dmpDiffs);
  let fromIdx = 0;
  let toIdx = 0;

  for (const [op, text] of dmpDiffs) {
    switch (op) {
      case _diffMatchPatch.DIFF_EQUAL:
        segments.push({
          type: 'stringSegment',
          action: 'unchanged',
          text
        });
        fromIdx += text.length;
        toIdx += text.length;
        break;

      case _diffMatchPatch.DIFF_DELETE:
        segments.push({
          type: 'stringSegment',
          action: 'removed',
          text: fromInput.substring(fromIdx, fromIdx + text.length),
          annotation: null
        });
        fromIdx += text.length;
        break;

      case _diffMatchPatch.DIFF_INSERT:
        segments.push({
          type: 'stringSegment',
          action: 'added',
          text: toInput.substring(toIdx, toIdx + text.length),
          annotation: null
        });
        toIdx += text.length;
        break;

      default: // Do nothing

    }
  } // Clean up so that marks / symbols are treated as an own segment


  return (0, _flatten2.default)(segments.map(seg => {
    const newSegments = [];

    if (seg.text.length > 1) {
      const markMatches = [...seg.text.matchAll(segmentRegex)];
      let lastIndex = -1;
      markMatches.forEach(match => {
        const index = match.index || 0;

        if (index > lastIndex) {
          newSegments.push({ ...seg,
            text: seg.text.substring(lastIndex + 1, index)
          });
          newSegments.push({ ...seg,
            text: match[0]
          });
        }

        if (match === markMatches[markMatches.length - 1]) {
          newSegments.push({ ...seg,
            text: seg.text.substring(index + 1)
          });
        }

        lastIndex = index;
      });

      if (markMatches.length === 0) {
        newSegments.push(seg);
      }
    } else {
      newSegments.push(seg);
    }

    return newSegments;
  }));
}

function getInlineObjects(diff) {
  const allChildren = [...(diff.toValue ? diff.toValue.children.filter(cld => cld._type !== 'span') : [])];
  const previousChildren = diff.fromValue ? diff.fromValue.children.filter(cld => cld._type !== 'span') : [];
  previousChildren.forEach(oCld => {
    if (!allChildren.some(cld => oCld._key === cld._key)) {
      allChildren.push(oCld);
    }
  });
  return (0, _orderBy2.default)(allChildren, ['_key'], ['asc']);
}

function findSpanDiffFromChild(diff, child) {
  // Find span in original diff which has a string segment similar to the one from the input
  const candidate = diff.fields.children && diff.fields.children.action !== 'unchanged' && diff.fields.children.type === 'array' && diff.fields.children.items.find(item => item.diff && item.diff.type === 'object' && (item.diff.action === 'removed' ? item.diff.fromValue && item.diff.fromValue._key === child._key : (item.diff.toValue && item.diff.toValue._key) === child._key));

  if (candidate) {
    return candidate.diff;
  }

  return undefined;
}

function findAnnotationDiff(diff, markDefKey) {
  var _diff$fields$markDefs;

  return diff.fields.markDefs && diff.fields.markDefs.isChanged && diff.fields.markDefs.type === 'array' && ((_diff$fields$markDefs = diff.fields.markDefs.items.find(item => item.diff && item.diff.type === 'object' && (item.diff.toValue && item.diff.toValue._key && item.diff.toValue._key === markDefKey || item.diff.fromValue && item.diff.fromValue._key && item.diff.fromValue._key === markDefKey))) === null || _diff$fields$markDefs === void 0 ? void 0 : _diff$fields$markDefs.diff) || undefined;
}

function isEmptyObject(object) {
  return object && (0, _isEqual2.default)(Object.keys(object), ['_key', '_type']) || false;
}

function escapeRegExp(text) {
  return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
}

function getAllMarkDefs(diff) {
  const allDefs = [...(diff.toValue && diff.toValue.markDefs ? diff.toValue.markDefs : [])];
  const oldDefs = diff.fromValue && diff.fromValue.markDefs ? diff.fromValue.markDefs : [];
  oldDefs.forEach(oDef => {
    if (!allDefs.some(def => oDef._key === def._key)) {
      allDefs.push(oDef);
    }
  });
  return (0, _orderBy2.default)(allDefs, ['_key'], ['asc']);
}
