// @sanity/cli/lib/packageManager/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cliPackageManager = void 0;
Object.defineProperty(exports, "getCliUpgradeCommand", {
  enumerable: true,
  get: function () {
    return _getUpgradeCommand.getCliUpgradeCommand;
  }
});
Object.defineProperty(exports, "getInstallCommand", {
  enumerable: true,
  get: function () {
    return _getInstallCommand.getInstallCommand;
  }
});
Object.defineProperty(exports, "getPackageManagerChoice", {
  enumerable: true,
  get: function () {
    return _packageManagerChoice.getPackageManagerChoice;
  }
});
Object.defineProperty(exports, "getYarnStub", {
  enumerable: true,
  get: function () {
    return _yarnStub.getYarnStub;
  }
});
Object.defineProperty(exports, "installDeclaredPackages", {
  enumerable: true,
  get: function () {
    return _installPackages.installDeclaredPackages;
  }
});
Object.defineProperty(exports, "installNewPackages", {
  enumerable: true,
  get: function () {
    return _installPackages.installNewPackages;
  }
});
Object.defineProperty(exports, "uninstallPackages", {
  enumerable: true,
  get: function () {
    return _uninstallPackages.uninstallPackages;
  }
});

var _getUpgradeCommand = require("./getUpgradeCommand");

var _uninstallPackages = require("./uninstallPackages");

var _installPackages = require("./installPackages");

var _packageManagerChoice = require("./packageManagerChoice");

var _getInstallCommand = require("./getInstallCommand");

var _yarnStub = require("./yarnStub");

// Exported for internal CLI usage
// Exported for use in `sanity` (formerly `@sanity/core`)

/**
 * @internal
 */
const cliPackageManager = {
  getInstallCommand: _getInstallCommand.getInstallCommand,
  getPackageManagerChoice: _packageManagerChoice.getPackageManagerChoice,
  installNewPackages: _installPackages.installNewPackages
};
/**
 * @internal
 */

exports.cliPackageManager = cliPackageManager;
