// @sanity/cli/lib/packageManager/installPackages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.installDeclaredPackages = installDeclaredPackages;
exports.installNewPackages = installNewPackages;

var _execa = _interopRequireDefault(require("execa"));

var _packageManagerChoice = require("./packageManagerChoice");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function installDeclaredPackages(cwd, packageManager, context) {
  var _result, _result2;

  const {
    output
  } = context;
  const execOptions = {
    encoding: 'utf8',
    env: (0, _packageManagerChoice.getPartialEnvWithNpmPath)(cwd),
    cwd,
    stdio: 'inherit'
  };
  const npmArgs = ['install', '--legacy-peer-deps'];
  let result;

  if (packageManager === 'npm') {
    output.print("Running 'npm ".concat(npmArgs.join(' '), "'"));
    result = await (0, _execa.default)('npm', npmArgs, execOptions);
  } else if (packageManager === 'yarn') {
    const yarnArgs = ['install'];
    output.print("Running 'yarn ".concat(yarnArgs.join(' '), "'"));
    result = await (0, _execa.default)('yarn', yarnArgs, execOptions);
  } else if (packageManager === 'pnpm') {
    const pnpmArgs = ['install'];
    output.print("Running 'pnpm ".concat(pnpmArgs.join(' '), "'"));
    result = await (0, _execa.default)('pnpm', pnpmArgs, execOptions);
  } else if (packageManager === 'manual') {
    output.print("Manual installation selected - run 'npm ".concat(npmArgs.join(' '), "' or similar"));
  }

  if ((_result = result) !== null && _result !== void 0 && _result.exitCode || (_result2 = result) !== null && _result2 !== void 0 && _result2.failed) {
    throw new Error('Dependency installation failed');
  }
}

async function installNewPackages(options, context) {
  var _result3, _result4;

  const {
    packageManager,
    packages
  } = options;
  const {
    output,
    workDir
  } = context;
  const execOptions = {
    encoding: 'utf8',
    env: (0, _packageManagerChoice.getPartialEnvWithNpmPath)(workDir),
    cwd: workDir,
    stdio: 'inherit'
  };
  const npmArgs = ['install', '--legacy-peer-deps', '--save', ...packages];
  let result;

  if (packageManager === 'npm') {
    output.print("Running 'npm ".concat(npmArgs.join(' '), "'"));
    result = await (0, _execa.default)('npm', npmArgs, execOptions);
  } else if (packageManager === 'yarn') {
    const yarnArgs = ['add', ...packages];
    output.print("Running 'yarn ".concat(yarnArgs.join(' '), "'"));
    result = await (0, _execa.default)('yarn', yarnArgs, execOptions);
  } else if (packageManager === 'pnpm') {
    const pnpmArgs = ['add', '--save-prod', ...packages];
    output.print("Running 'pnpm ".concat(pnpmArgs.join(' '), "'"));
    result = await (0, _execa.default)('pnpm', pnpmArgs, execOptions);
  } else if (packageManager === 'manual') {
    output.print("Manual installation selected - run 'npm ".concat(npmArgs.join(' '), "' or equivalent"));
  }

  if ((_result3 = result) !== null && _result3 !== void 0 && _result3.exitCode || (_result4 = result) !== null && _result4 !== void 0 && _result4.failed) {
    throw new Error('Package installation failed');
  }
}
