// sanity/lib/dts/src/components/previews/__workshop__/InlinePreviewStory.js
import { DocumentIcon } from '@sanity/icons';
import { Card, Container, Flex, Heading, Stack, Text } from '@sanity/ui';
import { useBoolean, useString } from '@sanity/ui-workshop';
import React from 'react';
import styled from 'styled-components';
import { InlinePreview } from '../portableText/InlinePreview';
const PreviewWrapper = styled.span `
  display: inline-block;
  vertical-align: top;
  background-color: var(--card-bg-color);
  box-shadow: inset 0 0 0 1px var(--card-border-color);
  border-radius: 2.66px;
  padding: 2px;
  margin-top: 0.0625em;
  height: calc(1em - 1px);

  & > span {
    vertical-align: top;
    margin: 0;
  }
`;
export default function InlinePreviewStory() {
    const title = useString('Title', 'Inline object', 'Props');
    const withImage = useBoolean('With image', true, 'Props');
    return (React.createElement(Card, { height: "fill" },
        React.createElement(Flex, { align: "center", height: "fill", justify: "center", padding: 4, sizing: "border" },
            React.createElement(Container, { width: 1 },
                React.createElement(Stack, { space: 4 },
                    React.createElement(Card, { tone: "primary" },
                        React.createElement(Heading, { size: 5 },
                            "Hello",
                            ' ',
                            React.createElement(PreviewWrapper, null,
                                React.createElement(InlinePreview
                                // eslint-disable-next-line react/jsx-no-bind
                                , { 
                                    // eslint-disable-next-line react/jsx-no-bind
                                    media: () => withImage ? (React.createElement("img", { src: "https://source.unsplash.com/60x60/?face" })) : (React.createElement(DocumentIcon, null)), title: title })),
                            ' ',
                            "world")),
                    React.createElement(Card, { tone: "primary" },
                        React.createElement(Heading, null,
                            "Hello",
                            ' ',
                            React.createElement(PreviewWrapper, null,
                                React.createElement(InlinePreview
                                // eslint-disable-next-line react/jsx-no-bind
                                , { 
                                    // eslint-disable-next-line react/jsx-no-bind
                                    media: () => withImage ? (React.createElement("img", { src: "https://source.unsplash.com/60x60/?face" })) : (React.createElement(DocumentIcon, null)), title: title })),
                            ' ',
                            "world")),
                    React.createElement(Card, { tone: "primary" },
                        React.createElement(Text, null,
                            "Hello",
                            ' ',
                            React.createElement(PreviewWrapper, null,
                                React.createElement(InlinePreview
                                // eslint-disable-next-line react/jsx-no-bind
                                , { 
                                    // eslint-disable-next-line react/jsx-no-bind
                                    media: () => withImage ? (React.createElement("img", { src: "https://source.unsplash.com/60x60/?face" })) : (React.createElement(DocumentIcon, null)), title: title })),
                            ' ',
                            "world")))))));
}
