// sanity/lib/dts/src/components/previews/__workshop__/TemplatePreviewStory.js
import { DocumentIcon } from '@sanity/icons';
import { Card, Container, Flex, Text } from '@sanity/ui';
import { useBoolean, useSelect, useString, useText } from '@sanity/ui-workshop';
import React, { createElement } from 'react';
import { TemplatePreview } from '../template/TemplatePreview';
const MEDIA_OPTIONS = {
    None: 'none',
    Image: 'image',
    Icon: 'icon',
};
const LAYOUT_OPTIONS = {
    'Create document': 'createDocument',
};
const mediaValues = {
    image: React.createElement("img", { src: "https://source.unsplash.com/70x70/?abstract" }),
    icon: React.createElement(DocumentIcon, null),
};
const previewComponents = {
    createDocument: TemplatePreview,
};
export default function TemplatePreviewStory() {
    const layout = useSelect('Layout', LAYOUT_OPTIONS, 'createDocument');
    const isPlaceholder = useBoolean('Is placeholder', false);
    const mediaKey = useSelect('Media', MEDIA_OPTIONS) || 'none';
    const title = useString('Title', 'Title', 'Props');
    const subtitle = useString('Subtitle', 'Subtitle', 'Props');
    const description = useText('Description', undefined, 'Props');
    const media = mediaValues[mediaKey] || false;
    const component = layout && previewComponents[layout];
    if (!component) {
        return (React.createElement(Flex, { align: "center", height: "fill", justify: "center", padding: 4, sizing: "border" },
            React.createElement(Text, null,
                "Unknown layout: ",
                layout)));
    }
    return (React.createElement(Card, { height: "fill", tone: "transparent" },
        React.createElement(Flex, { align: "center", height: "fill", justify: "center", padding: 4, sizing: "border" },
            React.createElement(Container, { width: 0 },
                React.createElement(Card, { padding: 3, radius: 2 }, createElement(component, {
                    description,
                    isPlaceholder,
                    media,
                    title,
                    subtitle,
                }))))));
}
