// @sanity/cli/lib/dts/src/actions/init-project/prompts.js
export function promptForTypeScript(prompt) {
    return prompt.single({
        type: 'confirm',
        message: 'Do you want to use TypeScript?',
        default: false,
    });
}
export function promptForDefaultConfig(prompt) {
    return prompt.single({
        type: 'confirm',
        message: 'Use the default dataset configuration?',
        default: true,
    });
}
export function promptImplicitReconfigure(prompt) {
    return prompt.single({
        type: 'confirm',
        message: 'The current folder contains a configured Sanity studio. Would you like to reconfigure it?',
        default: true,
    });
}
export async function promptForAclMode(prompt, output) {
    const mode = await prompt.single({
        type: 'list',
        message: 'Choose dataset visibility – this can be changed later',
        choices: [
            {
                value: 'public',
                name: 'Public (world readable)',
            },
            {
                value: 'private',
                name: 'Private (authenticated requests only)',
            },
        ],
    });
    if (mode === 'private') {
        output.print('Please note that while documents are private, assets (files and images) are still public\n');
    }
    return mode;
}
