// sanity/lib/cjs/form/inputs/ReferenceInput/ArrayItemReferenceInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayItemReferenceInput = ArrayItemReferenceInput;

var _react = _interopRequireWildcard(require("react"));

var _icons = require("@sanity/icons");

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _ui = require("@sanity/ui");

var _reactRx = require("react-rx");

var _uuid = require("@sanity/uuid");

var _autoId = require("@reach/auto-id");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _formField = require("../../components/formField");

var _PreviewCard = require("../../../components/PreviewCard");

var _presence = require("../../../presence");

var _router = require("../../../router");

var _patch = require("../../patch");

var _empty = require("../../utils/empty");

var _isNonNullable = require("../../utils/isNonNullable");

var _AlertStrip = require("../../components/AlertStrip");

var _RowWrapper = require("../arrays/ArrayOfObjectsInput/item/components/RowWrapper");

var _DragHandle = require("../arrays/common/DragHandle");

var _InsertMenu = require("../arrays/ArrayOfObjectsInput/InsertMenu");

var _useOnClickOutside = require("../../hooks/useOnClickOutside");

var _util = require("../../../util");

var _OptionPreview = require("./OptionPreview");

var _useReferenceInfo = require("./useReferenceInfo");

var _PreviewReferenceValue = require("./PreviewReferenceValue");

var _CreateButton = require("./CreateButton");

var _ReferenceAutocomplete = require("./ReferenceAutocomplete");

var _AutocompleteContainer = require("./AutocompleteContainer");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const StyledPreviewCard = (0, _styledComponents.default)(_PreviewCard.PreviewCard)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  /* this is a hack to avoid layout jumps while previews are loading\n  there's probably better ways of solving this */\n  min-height: 36px;\n"])));
const INITIAL_SEARCH_STATE = {
  hits: [],
  isLoading: false
};

const NO_FILTER = () => true;

const dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  paddingX: 1,
  paddingY: 3
});

function valueHasRef(value) {
  return typeof value._ref === 'string';
}

function ArrayItemReferenceInput(props) {
  var _loadableReferenceInf, _value$_strengthenOnP, _loadableReferenceInf2, _loadableReferenceInf3, _loadableReferenceInf4, _loadableReferenceInf5, _loadableReferenceInf6, _loadableReferenceInf7, _loadableReferenceInf8;

  const {
    schemaType: type,
    value,
    validation,
    liveEdit,
    onSearch,
    onChange,
    insertableTypes,
    focusPath = _empty.EMPTY_ARRAY,
    presence,
    createOptions,
    isSortable,
    level,
    onInsert,
    selectedState,
    editReferenceLinkComponent: EditReferenceLink,
    onEditReference,
    getReferenceInfo,
    onBlur,
    onFocus,
    onFocusPath,
    readOnly,
    focusRef,
    renderPreview
  } = props;
  const forwardedRef = (0, _ui.useForwardedRef)(focusRef);
  const [searchState, setSearchState] = (0, _react.useState)(INITIAL_SEARCH_STATE);

  const handleCreateNew = option => {
    const id = (0, _uuid.uuid)();
    const patches = [(0, _patch.setIfMissing)({}), (0, _patch.set)(type.name, ['_type']), (0, _patch.set)(id, ['_ref']), (0, _patch.set)(true, ['_weak']), (0, _patch.set)({
      type: option.type,
      weak: type.weak,
      template: option.template
    }, ['_strengthenOnPublish'])];
    onChange(patches);
    onEditReference({
      id,
      type: option.type,
      template: option.template
    });
    onFocusPath([]);
  };

  const handleDuplicate = (0, _react.useCallback)(() => {
    if (value !== null && value !== void 0 && value._key) {// todo
      // onInsert?.({
      //   items: [{...value, _key: randomKey()}],
      //   position: 'after',
      // })
    }
  }, [onInsert, value]);
  const handleInsert = (0, _react.useCallback)(pos => {
    if (value !== null && value !== void 0 && value._key) {// todo
      // onInsert?.({
      //   item: {_type: type.name, _key: randomKey()},
      //   position: pos,
      // })
    }
  }, [onInsert, type.name, value === null || value === void 0 ? void 0 : value._key]);
  const handleChange = (0, _react.useCallback)(id => {
    if (!id) {
      onChange((0, _patch.unset)());
      onFocusPath([]);
      return;
    }

    const hit = searchState.hits.find(h => h.id === id);

    if (!hit) {
      throw new Error('Selected an item that wasnt part of the result set');
    }

    const patches = [(0, _patch.setIfMissing)({}), (0, _patch.set)(type.name, ['_type']), (0, _patch.set)((0, _util.getPublishedId)(id), ['_ref']), // if there's no published version of this document, set the reference to weak
    hit.published && !type.weak ? (0, _patch.unset)(['_weak']) : (0, _patch.set)(true, ['_weak']), hit.published ? (0, _patch.unset)(['_strengthenOnPublish']) : (0, _patch.set)({
      type: hit === null || hit === void 0 ? void 0 : hit.type,
      weak: type.weak
    }, ['_strengthenOnPublish'])].filter(_isNonNullable.isNonNullable);
    onChange(patches);
    onFocusPath([]);
  }, [searchState.hits, type.name, type.weak, onChange, onFocusPath]);
  const handleClear = (0, _react.useCallback)(() => {
    onChange((0, _patch.unset)());
  }, [onChange]);
  const handlePreviewKeyPress = (0, _react.useCallback)(event => {
    if (event.key !== 'Enter' && event.key !== 'Space') {
      // enable "search for reference"-mode
      onFocusPath(['_ref']);
    }
  }, [onFocusPath]);
  const handleCancelEdit = (0, _react.useCallback)(() => {
    if (!(value !== null && value !== void 0 && value._ref)) {
      onChange((0, _patch.unset)());
    }

    onFocusPath([]);
  }, [onChange, onFocus, value === null || value === void 0 ? void 0 : value._ref]);
  const handleAutocompleteKeyDown = (0, _react.useCallback)(event => {
    // escape
    if (event.keyCode === 27) {
      handleCancelEdit();
    }
  }, [handleCancelEdit]);
  const getReferenceInfoMemo = (0, _react.useCallback)(id => getReferenceInfo(id, type), [getReferenceInfo, type]);
  const loadableReferenceInfo = (0, _useReferenceInfo.useReferenceInfo)(value === null || value === void 0 ? void 0 : value._ref, getReferenceInfoMemo);
  const refTypeName = ((_loadableReferenceInf = loadableReferenceInfo.result) === null || _loadableReferenceInf === void 0 ? void 0 : _loadableReferenceInf.type) || (value === null || value === void 0 ? void 0 : (_value$_strengthenOnP = value._strengthenOnPublish) === null || _value$_strengthenOnP === void 0 ? void 0 : _value$_strengthenOnP.type);
  const refType = refTypeName ? type.to.find(toType => toType.name === refTypeName) : undefined; // --- focus handling

  const hasFocusAtRef = focusPath.length === 1 && (focusPath[0] === '_ref' || focusPath[0] === '$'); // const focusElementRef = useForwardedRef(forwardedRef)
  // useDidUpdate({hasFocusAt: hasFocusAtRef, ref: value?._ref}, (prev, current) => {
  //   const refUpdated = prev?.ref !== current.ref
  //   const focusAtUpdated = prev?.hasFocusAt !== current.hasFocusAt
  //   if ((focusAtUpdated || refUpdated) && current.hasFocusAt) {
  //     // if search mode changed and we're having focus always ensure the
  //     // ref element gets focus
  //     focusElementRef.current?.focus()
  //   }
  // })

  const weakIs = value !== null && value !== void 0 && value._weak ? 'weak' : 'strong';
  const weakShouldBe = type.weak === true ? 'weak' : 'strong';
  const hasRef = value && valueHasRef(value); // If the reference value is marked with _strengthenOnPublish,
  // we allow weak references if the reference points to a document that has a draft but not a published
  // In all other cases we should display a "weak mismatch" warning

  const weakWarningOverride = hasRef && !loadableReferenceInfo.isLoading && (value === null || value === void 0 ? void 0 : value._strengthenOnPublish);
  const handleFixStrengthMismatch = (0, _react.useCallback)(() => {
    onChange(type.weak === true ? (0, _patch.set)(true, ['_weak']) : (0, _patch.unset)(['_weak']));
  }, [onChange, type]);
  const referenceExists = hasRef && ((_loadableReferenceInf2 = loadableReferenceInfo.result) === null || _loadableReferenceInf2 === void 0 ? void 0 : (_loadableReferenceInf3 = _loadableReferenceInf2.preview) === null || _loadableReferenceInf3 === void 0 ? void 0 : (_loadableReferenceInf4 = _loadableReferenceInf3.published) === null || _loadableReferenceInf4 === void 0 ? void 0 : _loadableReferenceInf4._id);
  const handleRemoveStrengthenOnPublish = (0, _react.useCallback)(() => {
    onChange([type.weak === true ? (0, _patch.set)(true, ['_weak']) : (0, _patch.unset)(['_weak']), (0, _patch.unset)(['_strengthenOnPublish'])]);
  }, [onChange, type]);
  const {
    push
  } = (0, _ui.useToast)();
  const errors = (0, _react.useMemo)(() => validation.filter(item => item.level === 'error'), [validation]);
  const pressed = selectedState === 'pressed';
  const selected = selectedState === 'selected';
  const handleFocus = (0, _react.useCallback)(event => {
    if (event.currentTarget === forwardedRef.current) {
      onFocus(event);
    }
  }, [onFocus, forwardedRef]);
  const handleAutocompleteFocus = (0, _react.useCallback)(event => {
    if (onFocusPath && event.currentTarget === forwardedRef.current) {
      onFocusPath(['_ref']);
    }
  }, [onFocusPath, forwardedRef]);
  const handleQueryChange = (0, _reactRx.useObservableCallback)(inputValue$ => {
    return inputValue$.pipe((0, _operators.filter)(_isNonNullable.isNonNullable), (0, _operators.distinctUntilChanged)(), (0, _operators.switchMap)(searchString => (0, _rxjs.concat)((0, _rxjs.of)({
      isLoading: true
    }), onSearch(searchString).pipe((0, _operators.map)(hits => ({
      hits,
      isLoading: false,
      searchString
    })), (0, _operators.catchError)(error => {
      push({
        title: 'Reference search failed',
        description: error.message,
        status: 'error',
        id: "reference-search-fail-".concat(inputId)
      });
      console.error(error);
      return (0, _rxjs.of)({
        hits: []
      });
    })))), (0, _operators.scan)((prevState, nextState) => ({ ...prevState,
      ...nextState
    }), INITIAL_SEARCH_STATE), (0, _operators.tap)(setSearchState));
  }, []);
  const handleAutocompleteOpenButtonClick = (0, _react.useCallback)(() => {
    handleQueryChange('');
  }, [handleQueryChange]);
  const showWeakRefMismatch = !loadableReferenceInfo.isLoading && hasRef && weakIs !== weakShouldBe && !weakWarningOverride;
  const inputId = (0, _autoId.useId)();
  const handleCreateButtonKeyDown = (0, _react.useCallback)(e => {
    if (e.key === 'Escape') {
      var _forwardedRef$current;

      (_forwardedRef$current = forwardedRef.current) === null || _forwardedRef$current === void 0 ? void 0 : _forwardedRef$current.focus();
    }
  }, [forwardedRef]);
  const renderOption = (0, _react.useCallback)(option => {
    var _option$hit$draft, _option$hit$published;

    const id = ((_option$hit$draft = option.hit.draft) === null || _option$hit$draft === void 0 ? void 0 : _option$hit$draft._id) || ((_option$hit$published = option.hit.published) === null || _option$hit$published === void 0 ? void 0 : _option$hit$published._id);
    return /*#__PURE__*/_react.default.createElement(_PreviewCard.PreviewCard, {
      as: "button",
      type: "button",
      radius: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 3,
      paddingY: 1
    }, /*#__PURE__*/_react.default.createElement(_OptionPreview.OptionPreview, {
      getReferenceInfo: getReferenceInfoMemo,
      id: id,
      renderPreview: renderPreview,
      type: type
    })));
  }, [type, getReferenceInfoMemo, renderPreview]);
  const OpenLink = (0, _react.useMemo)(() =>
  /*#__PURE__*/
  // eslint-disable-next-line @typescript-eslint/no-shadow
  (0, _react.forwardRef)(function OpenLink(restProps, _ref) {
    const template = ((value === null || value === void 0 ? void 0 : value._strengthenOnPublish) || {}).template;
    return /*#__PURE__*/_react.default.createElement(_router.IntentLink, _extends({}, restProps, {
      intent: "edit",
      params: [{
        id: value === null || value === void 0 ? void 0 : value._ref,
        type: refType === null || refType === void 0 ? void 0 : refType.name,
        template: template === null || template === void 0 ? void 0 : template.id
      }, {
        params: template === null || template === void 0 ? void 0 : template.params
      }],
      target: "_blank",
      rel: "noopener noreferrer",
      ref: _ref
    }));
  }), [refType === null || refType === void 0 ? void 0 : refType.name, value === null || value === void 0 ? void 0 : value._ref, value === null || value === void 0 ? void 0 : value._strengthenOnPublish]);
  const preview = ((_loadableReferenceInf5 = loadableReferenceInfo.result) === null || _loadableReferenceInf5 === void 0 ? void 0 : _loadableReferenceInf5.preview.draft) || ((_loadableReferenceInf6 = loadableReferenceInfo.result) === null || _loadableReferenceInf6 === void 0 ? void 0 : _loadableReferenceInf6.preview.published);
  const isEditing = hasFocusAtRef; // --- click outside handling

  const clickOutsideBoundaryRef = (0, _react.useRef)(null);
  const autocompletePortalRef = (0, _react.useRef)(null);
  const createButtonMenuPortalRef = (0, _react.useRef)(null);
  (0, _useOnClickOutside.useOnClickOutside)([clickOutsideBoundaryRef, autocompletePortalRef, createButtonMenuPortalRef], () => {
    if (isEditing) {
      handleCancelEdit();
    }
  });
  const autocompletePopoverReferenceElementRef = (0, _react.useRef)(null);
  return /*#__PURE__*/_react.default.createElement(_RowWrapper.RowWrapper, {
    radius: 2,
    padding: 1,
    tone: isEditing ? ' default' : readOnly ? 'transparent' : loadableReferenceInfo.error || errors.length > 0 ? 'critical' : showWeakRefMismatch ? 'caution' : 'default'
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, !isEditing && isSortable && !readOnly && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    className: "dragHandle",
    tone: "inherit",
    marginRight: 1
  }, dragHandle), isEditing ? /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 1,
    ref: clickOutsideBoundaryRef
  }, /*#__PURE__*/_react.default.createElement(_formField.FormField, {
    validation: validation,
    __unstable_presence: presence,
    inputId: inputId,
    title: type.title,
    level: level,
    description: type.description
  }, /*#__PURE__*/_react.default.createElement(_AutocompleteContainer.AutocompleteContainer, {
    ref: autocompletePopoverReferenceElementRef
  }, /*#__PURE__*/_react.default.createElement(_ReferenceAutocomplete.ReferenceAutocomplete, {
    "data-testid": "autocomplete",
    loading: searchState.isLoading,
    portalRef: autocompletePortalRef,
    ref: forwardedRef,
    id: inputId || '',
    options: searchState.hits.map(hit => ({
      value: hit.id,
      hit: hit
    })),
    onFocus: handleAutocompleteFocus,
    onBlur: onBlur,
    radius: 1,
    placeholder: "Type to search",
    onKeyDown: handleAutocompleteKeyDown,
    readOnly: readOnly,
    disabled: loadableReferenceInfo.isLoading,
    onQueryChange: handleQueryChange,
    searchString: searchState.searchString,
    onChange: handleChange,
    referenceElement: autocompletePopoverReferenceElementRef.current,
    filterOption: NO_FILTER,
    renderOption: renderOption,
    openButton: {
      onClick: handleAutocompleteOpenButtonClick
    }
  }), !readOnly && createOptions.length > 0 && /*#__PURE__*/_react.default.createElement(_CreateButton.CreateButton, {
    menuRef: createButtonMenuPortalRef,
    id: "".concat(inputId, "-selectTypeMenuButton"),
    createOptions: createOptions,
    onCreate: handleCreateNew,
    onKeyDown: handleCreateButtonKeyDown
  })))) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, hasRef ? /*#__PURE__*/_react.default.createElement(StyledPreviewCard, {
    flex: 1,
    padding: 1,
    paddingRight: 3,
    radius: 2,
    forwardedAs: EditReferenceLink,
    documentId: value === null || value === void 0 ? void 0 : value._ref,
    documentType: refType === null || refType === void 0 ? void 0 : refType.name,
    "data-as": "a",
    tone: selected ? 'default' : 'inherit',
    __unstable_focusRing: true,
    tabIndex: 0,
    selected: selected,
    pressed: pressed,
    onKeyPress: handlePreviewKeyPress,
    onFocus: handleFocus,
    "data-selected": selected ? true : undefined,
    "data-pressed": pressed ? true : undefined,
    ref: forwardedRef
  }, /*#__PURE__*/_react.default.createElement(_PreviewReferenceValue.PreviewReferenceValue, {
    value: value,
    referenceInfo: loadableReferenceInfo,
    renderPreview: renderPreview,
    type: type
  })) : /*#__PURE__*/_react.default.createElement(_ui.Card, {
    flex: 1,
    padding: 3,
    marginRight: 1,
    radius: 2,
    tone: "inherit",
    as: "button",
    __unstable_focusRing: true,
    tabIndex: 0,
    onClick: () => onFocusPath(['_ref']),
    ref: forwardedRef
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true
  }, "Empty reference"))), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    marginLeft: !readOnly && presence.length > 0 ? 2 : undefined
  }, !readOnly && presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_formField.FormFieldValidationStatus, {
    validation: validation
  })), !(value !== null && value !== void 0 && value._key) && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "This item is missing the required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property.")),
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution"
  }, "Missing key")))))), !isEditing && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      paddingY: 3,
      paddingX: 2,
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "".concat(inputId, "-menuButton"),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, !readOnly && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Remove",
      tone: "critical",
      icon: _icons.TrashIcon,
      onClick: handleClear
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Replace",
      icon: _icons.SyncIcon,
      onClick: () => {
        onFocusPath(['_ref']);
      }
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Duplicate",
      icon: _icons.CopyIcon,
      onClick: handleDuplicate
    }), /*#__PURE__*/_react.default.createElement(_InsertMenu.InsertMenu, {
      onInsert: handleInsert,
      types: insertableTypes
    })), !readOnly && hasRef && /*#__PURE__*/_react.default.createElement(_ui.MenuDivider, null), hasRef && /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      as: OpenLink,
      "data-as": "a",
      text: "Open in new tab",
      icon: _icons.LaunchIcon
    })),
    placement: "right",
    popover: {
      portal: true,
      tone: 'default'
    }
  }))), liveEdit && referenceExists && value._strengthenOnPublish && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: type.weak ? 'Finalize reference' : 'Convert to strong reference',
    status: "info",
    "data-testid": "alert-reference-published"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("strong", null, (_loadableReferenceInf7 = loadableReferenceInfo.result) === null || _loadableReferenceInf7 === void 0 ? void 0 : (_loadableReferenceInf8 = _loadableReferenceInf7.preview.published) === null || _loadableReferenceInf8 === void 0 ? void 0 : _loadableReferenceInf8.title), " is published and this reference should now be", ' ', type.weak ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "finalized") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "converted to a strong reference"), "."), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleRemoveStrengthenOnPublish,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Convert to strong reference"),
    tone: "positive"
  }))), showWeakRefMismatch && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: "Reference strength mismatch",
    status: "warning",
    "data-testid": "alert-reference-strength-mismatch"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This reference is ", /*#__PURE__*/_react.default.createElement("em", null, weakIs), ", but according to the current schema it should be", ' ', /*#__PURE__*/_react.default.createElement("em", null, weakShouldBe, ".")), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, type.weak ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "It will not be possible to delete the \"", preview === null || preview === void 0 ? void 0 : preview.title, "\"-document without first removing this reference.") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This makes it possible to delete the \"", preview === null || preview === void 0 ? void 0 : preview.title, "\"-document without first deleting this reference, leaving this field referencing a nonexisting document.")), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleFixStrengthMismatch,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Convert to ", weakShouldBe, " reference"),
    tone: "caution"
  }))), loadableReferenceInfo.error && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: "Unable to load reference metadata",
    status: "warning",
    "data-testid": "alert-reference-info-failed"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Error: ", loadableReferenceInfo.error.message), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: loadableReferenceInfo.retry,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Retry"),
    tone: "primary"
  }))));
}
