// sanity/lib/cjs/form/inputs/ReferenceInput/CreateButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateButton = CreateButton;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _InsufficientPermissionsMessage = require("../../../components/InsufficientPermissionsMessage");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ConditionalTooltip(props) {
  const {
    enabled,
    ...rest
  } = props;
  return enabled ? /*#__PURE__*/_react.default.createElement(_ui.Tooltip, rest) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, props.children);
}

const INLINE_BLOCK_STYLE = {
  display: 'inline-flex'
};
const FULL_WIDTH = {
  width: '100%'
};

function CreateButton(props) {
  const {
    createOptions,
    onCreate,
    id,
    ...rest
  } = props;
  const canCreateAny = createOptions.some(option => option.permission.granted);

  if (!canCreateAny) {
    return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
      content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
        padding: 2
      }, /*#__PURE__*/_react.default.createElement(_InsufficientPermissionsMessage.InsufficientPermissionsMessage, {
        operationLabel: "create a new reference"
      }))
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: INLINE_BLOCK_STYLE
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      text: "Create new",
      mode: "ghost",
      disabled: true,
      icon: _icons.AddIcon,
      style: FULL_WIDTH
    })));
  }

  return createOptions.length > 1 ? /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, _extends({}, rest, {
      text: "Create new\u2026",
      mode: "ghost",
      icon: _icons.AddIcon
    })),
    id: id,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, {
      ref: props.menuRef
    }, createOptions.map(createOption => /*#__PURE__*/_react.default.createElement(ConditionalTooltip, {
      enabled: !createOption.permission.granted,
      key: createOption.id,
      content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
        padding: 2
      }, /*#__PURE__*/_react.default.createElement(_InsufficientPermissionsMessage.InsufficientPermissionsMessage, {
        operationLabel: "create this type of document"
      })),
      portal: true
    }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      disabled: !createOption.permission.granted,
      icon: createOption.icon,
      text: createOption.title,
      onClick: () => onCreate(createOption)
    }))))),
    placement: "right",
    popover: {
      portal: true,
      tone: 'default',
      constrainSize: true
    }
  }) : /*#__PURE__*/_react.default.createElement(_ui.Button, _extends({}, rest, {
    text: "Create new",
    mode: "ghost",
    disabled: !createOptions[0].permission.granted,
    onClick: () => onCreate(createOptions[0]),
    icon: _icons.AddIcon
  }));
}
