// sanity/lib/cjs/form/components/Alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Alert = Alert;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const STATUS_TONES = {
  warning: 'caution',
  error: 'critical'
};
const SuffixBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  border-top: 1px solid var(--card-border-color);\n"])));

function Alert(props) {
  const {
    children,
    status = 'warning',
    suffix,
    title,
    ...rest
  } = props;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, _extends({
    radius: 2,
    tone: STATUS_TONES[status]
  }, rest, {
    "data-ui": "Alert"
  }), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, status === 'warning' && /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null), status === 'error' && /*#__PURE__*/_react.default.createElement(_icons.ErrorOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    marginLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold"
  }, title), children && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, children))), suffix && /*#__PURE__*/_react.default.createElement(SuffixBox, null, suffix));
}
