// sanity/lib/cjs/form/components/AlertStrip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertStrip = AlertStrip;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _Details = require("./Details");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const STATUS_TONES = {
  warning: 'caution',
  error: 'critical',
  info: 'positive'
};
const STATUS_ICONS = {
  warning: /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null),
  error: /*#__PURE__*/_react.default.createElement(_icons.ErrorOutlineIcon, null),
  info: /*#__PURE__*/_react.default.createElement(_icons.InfoOutlineIcon, null)
};

function AlertStrip(props) {
  const {
    children,
    status = 'warning',
    title,
    ...rest
  } = props;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, _extends({
    radius: 2,
    tone: STATUS_TONES[status]
  }, rest, {
    "data-ui": "Alert"
  }), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 1
  }, children && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_Details.Details, {
    icon: STATUS_ICONS[status],
    title: title
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 3,
    marginTop: 3
  }, children)))));
}
