// sanity/lib/cjs/form/components/Details.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Details = Details;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const HeaderButton = _styledComponents.default.button(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: block;\n  -webkit-font-smoothing: inherit;\n  appearance: none;\n  font: inherit;\n  background: none;\n  width: 100%;\n  text-align: left;\n  border: 0;\n  margin: 0;\n  padding: 0;\n  outline: none;\n"])));

const ToggleArrow = (0, _styledComponents.default)(_icons.ToggleArrowRightIcon)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  transform: ", ";\n"])), props => props.open ? 'rotate(90deg)' : '');
const Header = (0, _styledComponents.default)(_ui.Flex)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  cursor: default;\n"])));
const IconBox = (0, _styledComponents.default)(_ui.Box)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  & > div > svg {\n    transform: rotate(0);\n    transition: transform 100ms;\n  }\n\n  &[data-open] > div > svg {\n    transform: rotate(90deg);\n  }\n"])));

function Details(props) {
  const {
    children,
    open: openProp,
    icon,
    title = 'Details',
    ...restProps
  } = props;
  const [open, setOpen] = (0, _react.useState)(openProp || false);
  const handleToggle = (0, _react.useCallback)(() => setOpen(v => !v), []);
  (0, _react.useEffect)(() => setOpen(openProp || false), [openProp]);
  return /*#__PURE__*/_react.default.createElement(_ui.Box, restProps, /*#__PURE__*/_react.default.createElement(HeaderButton, {
    type: "button",
    onClick: handleToggle
  }, /*#__PURE__*/_react.default.createElement(Header, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, null, /*#__PURE__*/_react.default.createElement(IconBox, {
    "data-open": open ? '' : undefined
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(ToggleArrow, {
    open: open
  }))), icon && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, icon), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "medium"
  }, title))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    hidden: !open,
    marginTop: 3
  }, children));
}
