// sanity/lib/cjs/form/components/EditPortal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditPortal = EditPortal;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _popoverDialog = require("../../components/popoverDialog");

var _presence = require("../../presence");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const PRESENCE_MARGINS = [0, 0, 1, 0];

function EditPortal(props) {
  const {
    type,
    id,
    onClose,
    children,
    legacy_referenceElement: referenceElement,
    header
  } = props;

  if (type === 'inline') {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
  }

  if (type === 'fullscreen') {
    return /*#__PURE__*/_react.default.createElement(_ui.Layer, null, /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      width: "auto",
      id: id || '',
      onClose: onClose,
      header: header,
      __unstable_autoFocus: false
    }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
      margins: PRESENCE_MARGINS
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, children))));
  } // @ts-expect-error fold is deprecated and should not be visible though types


  if (type === 'popover' || type === 'fold') {
    // @ts-expect-error fold is deprecated and should not be visible though types
    if (type === 'fold') {
      console.warn("The option named `editItem: \"fold\"` is no longer supported");
    }

    return /*#__PURE__*/_react.default.createElement(_popoverDialog.PopoverDialog, {
      onClose: onClose,
      referenceElement: referenceElement,
      placement: "auto",
      title: header
    }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
      margins: PRESENCE_MARGINS
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, children)));
  }

  return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    width: 1,
    id: id || '',
    onClose: onClose,
    header: header,
    __unstable_autoFocus: false
  }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
    margins: PRESENCE_MARGINS
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, children)));
}
