// sanity/lib/cjs/field/diff/components/ChangeBreadcrumb.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChangeBreadcrumb = ChangeBreadcrumb;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _ChangeTitleSegment = require("./ChangeTitleSegment");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ChangeBreadcrumb(props) {
  const {
    change,
    titlePath
  } = props;
  return /*#__PURE__*/_react.default.createElement(_ui.Breadcrumbs, {
    separator: /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "/")
  }, titlePath.map((titleSegment, idx) => {
    const showSegment = typeof titleSegment === 'string' || !change || change.showIndex;

    if (!showSegment) {
      return null;
    }

    return /*#__PURE__*/_react.default.createElement(_ChangeTitleSegment.ChangeTitleSegment, {
      change: change,
      key: idx,
      segment: titleSegment
    });
  }));
}
