// sanity/lib/cjs/field/diff/components/DiffTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffTooltip = DiffTooltip;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _transitional = require("../../../components/transitional");

var _UserAvatar = require("../../../components/UserAvatar");

var _hooks = require("../../../hooks");

var _annotations = require("../annotations");

var _datastores = require("../../../datastores");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function DiffTooltip(props) {
  if ('diff' in props) {
    const {
      diff,
      path = [],
      ...restProps
    } = props;
    const annotation = (0, _annotations.getAnnotationAtPath)(diff, path);
    return /*#__PURE__*/_react.default.createElement(DiffTooltipWithAnnotation, _extends({}, restProps, {
      annotations: annotation ? [annotation] : []
    }));
  }

  return /*#__PURE__*/_react.default.createElement(DiffTooltipWithAnnotation, props);
}

function DiffTooltipWithAnnotation(props) {
  const {
    annotations,
    children,
    description = 'Changed',
    ...restProps
  } = props;

  if (!annotations) {
    return children;
  }

  const content = /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    padding: 3,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 1,
    style: {
      textTransform: 'uppercase'
    },
    muted: true
  }, description), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, annotations.map((annotation, idx) => /*#__PURE__*/_react.default.createElement(AnnotationItem, {
    annotation: annotation,
    key: idx
  }))));

  return /*#__PURE__*/_react.default.createElement(_transitional.LegacyLayerProvider, {
    zOffset: "paneFooter"
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, _extends({
    content: content // data-placement={restProps.placement}
    ,
    portal: true,
    allowedAutoPlacements: ['top', 'bottom']
  }, restProps), children));
}

function AnnotationItem(_ref) {
  let {
    annotation
  } = _ref;
  const {
    author,
    timestamp
  } = annotation;
  const [user] = (0, _datastores.useUser)(author);
  const color = (0, _annotations.useAnnotationColor)(annotation);
  const timeAgo = (0, _hooks.useTimeAgo)(timestamp, {
    minimal: true
  });
  return /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    paddingRight: 3,
    style: {
      backgroundColor: color.background,
      color: color.text,
      borderRadius: 'calc(23px / 2)'
    }
  }, /*#__PURE__*/_react.default.createElement(_UserAvatar.UserAvatar, {
    user: author
  }), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    paddingLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    style: {
      color: color.text
    }
  }, user ? user.displayName : 'Loading…'))), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "time",
    muted: true,
    size: 1,
    dateTime: timestamp
  }, timeAgo));
}
