// sanity/lib/cjs/field/diff/components/DiffErrorBoundary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffErrorBoundary = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _environment = require("../../../environment");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DiffErrorBoundary extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {});
  }

  static getDerivedStateFromError(error) {
    return {
      error
    };
  }

  // eslint-disable-next-line class-methods-use-this
  componentDidCatch(error) {
    console.error('Error rendering diff component: ');
    console.error(error);
  }

  render() {
    const {
      error
    } = this.state;

    if (!error) {
      return this.props.children;
    }

    return /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 3,
      radius: 2,
      tone: "critical"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.ErrorOutlineIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "h3",
      size: 1,
      weight: "medium"
    }, "Rendering the changes to this field caused an error"), _environment.isDev && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p",
      size: 1
    }, "Check the developer console for more information")))));
  }

}

exports.DiffErrorBoundary = DiffErrorBoundary;
