// sanity/lib/cjs/field/diff/components/FallbackDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FallbackDiff = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _preview = require("../../../preview");

var _DiffFromTo = require("./DiffFromTo");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const FallbackPreview = _ref => {
  let {
    value,
    schemaType
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_preview.SanityPreview, {
    schemaType: schemaType,
    value: value,
    layout: "default"
  }));
};

const FallbackDiff = props => {
  const {
    diff,
    schemaType
  } = props;
  return /*#__PURE__*/_react.default.createElement(_DiffFromTo.DiffFromTo, {
    diff: diff,
    schemaType: schemaType,
    previewComponent: FallbackPreview,
    layout: "grid"
  });
};

exports.FallbackDiff = FallbackDiff;
