// sanity/lib/cjs/form/inputs/PortableText/__workshop__/editObjects/schemaType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = void 0;

var _types = require("@sanity/types");

var _schema = require("../../../../../schema");

const linkType = {
  type: 'object',
  name: 'link',
  title: 'Link',
  fields: [{
    type: 'string',
    name: 'url',
    title: 'URL'
  }],
  options: {
    modal: {
      type: 'popover'
    }
  }
};
const myObjectBlockType = (0, _types.defineType)({
  type: 'object',
  name: 'myObjectBlock',
  title: 'My object block',
  fields: [{
    type: 'string',
    name: 'title',
    title: 'Title'
  }],
  options: {
    modal: {
      type: 'dialog',
      width: 'auto'
    }
  }
});
const myInlineObjectType = {
  type: 'object',
  name: 'myInlineObject',
  title: 'My inline object',
  fields: [{
    type: 'string',
    name: 'title',
    title: 'Title'
  }],
  options: {
    modal: {
      type: 'popover'
    }
  }
};
const blockType = {
  type: 'block',
  marks: {
    annotations: [linkType]
  },
  of: [myInlineObjectType, myObjectBlockType]
};
const bodyType = {
  type: 'array',
  name: 'body',
  title: 'Body',
  of: [blockType, myObjectBlockType]
};
const schema = (0, _schema.createSchema)({
  name: 'default',
  types: [bodyType]
});
exports.schema = schema;
