// sanity/lib/esm/form/inputs/ObjectInput/UnknownFields.js
import React, { useCallback } from 'react';
import { Box, Button, Card, Code, Stack, Text } from '@sanity/ui';
import { TrashIcon } from '@sanity/icons';
import { isDev } from '../../../environment';
import { unset } from '../../patch';
import { Details } from '../../components/Details';
import { Alert } from '../../components/Alert';
export function UnknownFields(props) {
  const {
    fieldNames,
    onChange,
    readOnly,
    value
  } = props;
  const fieldsLen = fieldNames.length;
  const handleUnsetClick = useCallback(fieldName => {
    onChange(unset([fieldName]));
  }, [onChange]);
  return /*#__PURE__*/React.createElement(Alert, {
    status: "warning",
    title: /*#__PURE__*/React.createElement(React.Fragment, null, fieldsLen === 1 && /*#__PURE__*/React.createElement(React.Fragment, null, "Unknown field found"), fieldsLen > 1 && /*#__PURE__*/React.createElement(React.Fragment, null, "Unknown fields found"))
  }, /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, fieldsLen === 1 ? /*#__PURE__*/React.createElement(React.Fragment, null, "Encountered a field that is not defined in the schema.") : /*#__PURE__*/React.createElement(React.Fragment, null, "Encountered ", fieldsLen, " fields that are not defined in the schema.")), /*#__PURE__*/React.createElement(Details, {
    marginTop: 4,
    open: isDev,
    title: /*#__PURE__*/React.createElement(React.Fragment, null, "Developer info")
  }, /*#__PURE__*/React.createElement(Box, {
    marginBottom: 3
  }, /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, fieldsLen === 1 ? /*#__PURE__*/React.createElement(React.Fragment, null, "This field is not defined in the schema, which could mean that the field definition has been removed or that someone else has added it to their own local project and have not deployed their changes yet.") : /*#__PURE__*/React.createElement(React.Fragment, null, "These fields are not defined in the document\u2019s schema, which could mean that the field definitions have been removed or that someone else has added them to their own local project and have not deployed their changes yet."))), /*#__PURE__*/React.createElement(Stack, {
    as: "ul",
    space: 3
  }, fieldNames.map(fieldName => {
    return /*#__PURE__*/React.createElement(UnknownField, {
      key: fieldName,
      fieldName: fieldName,
      onUnsetClick: handleUnsetClick,
      readOnly: readOnly,
      value: value === null || value === void 0 ? void 0 : value[fieldName]
    });
  }))));
}

function UnknownField(_ref) {
  let {
    fieldName,
    onUnsetClick,
    readOnly,
    value
  } = _ref;
  const handleUnsetClick = useCallback(() => {
    onUnsetClick(fieldName);
  }, [fieldName, onUnsetClick]);
  return /*#__PURE__*/React.createElement(Card, {
    as: "li",
    padding: 3,
    radius: 2,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 2
  }, /*#__PURE__*/React.createElement(Card, {
    border: true,
    radius: 1
  }, /*#__PURE__*/React.createElement(Card, {
    borderBottom: true,
    padding: 3
  }, /*#__PURE__*/React.createElement(Code, {
    weight: "medium"
  }, fieldName)), /*#__PURE__*/React.createElement(Box, {
    overflow: "auto",
    padding: 3
  }, /*#__PURE__*/React.createElement(Code, {
    language: "json"
  }, JSON.stringify(value, null, 2)))), readOnly && /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This field is ", /*#__PURE__*/React.createElement("strong", null, "read only"), " according to the document\u2019s schema and cannot be unset. If you want to be able to unset this in Studio, make sure you remove the", ' ', /*#__PURE__*/React.createElement("code", null, "readOnly"), " field from the enclosing type in the schema."), !readOnly && /*#__PURE__*/React.createElement(Button, {
    icon: TrashIcon,
    onClick: handleUnsetClick,
    tone: "critical",
    text: /*#__PURE__*/React.createElement(React.Fragment, null, "Remove field")
  })));
}
