// sanity/lib/dts/src/components/NoJavascript.js
import React from 'react';
const NoJsStyles = `
.sanity-app-no-js__root {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background: #fff;
}

.sanity-app-no-js__content {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  font-family: helvetica, arial, sans-serif;
}
`;
export function NoJavascript() {
    return (React.createElement("noscript", null,
        React.createElement("div", { className: "sanity-app-no-js__root" },
            React.createElement("div", { className: "sanity-app-no-js__content" },
                React.createElement("style", { type: "text/css" }, NoJsStyles),
                React.createElement("h1", null, "JavaScript disabled"),
                React.createElement("p", null,
                    "Please ",
                    React.createElement("a", { href: "https://www.enable-javascript.com/" }, "enable JavaScript"),
                    " in your browser and reload the page to proceed.")))));
}
