// sanity/lib/cjs/form/inputs/EmailInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmailInput = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _patch = require("../patch");

var _changeIndicators = require("../../components/changeIndicators");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const EmailInput = /*#__PURE__*/_react.default.forwardRef(function EmailInput(props, forwardedRef) {
  const {
    changed,
    focused,
    id,
    onBlur,
    onChange,
    onFocus,
    path,
    readOnly,
    schemaType,
    validationError,
    value
  } = props;

  const handleChange = _react.default.useCallback(event => {
    const inputValue = event.currentTarget.value;
    onChange(inputValue ? (0, _patch.set)(inputValue) : (0, _patch.unset)());
  }, [onChange]);

  return /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/_react.default.createElement(_ui.TextInput, {
    type: "email",
    inputMode: "email",
    id: id,
    customValidity: validationError,
    value: value || '',
    readOnly: Boolean(readOnly),
    placeholder: schemaType.placeholder,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    ref: forwardedRef
  }));
});

exports.EmailInput = EmailInput;
