// sanity/lib/cjs/form/inputs/StringInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StringInput = StringInput;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _patch = require("../patch");

var _changeIndicators = require("../../components/changeIndicators");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function StringInput(props) {
  const {
    value,
    readOnly,
    id,
    path,
    focused,
    focusRef,
    schemaType,
    validationError,
    onFocus,
    onBlur,
    onChange,
    changed
  } = props;
  const placeholder = schemaType.placeholder;

  const handleChange = _react.default.useCallback(event => {
    const inputValue = event.currentTarget.value;
    onChange(inputValue ? (0, _patch.set)(inputValue) : (0, _patch.unset)());
  }, [onChange]);

  return /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/_react.default.createElement(_ui.TextInput, {
    id: id,
    customValidity: validationError,
    value: value || '',
    readOnly: Boolean(readOnly),
    placeholder: placeholder,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    ref: focusRef
  }));
}
