// @sanity/dashboard/lib/widgets/sanityTutorials/Tutorial.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

// eslint-disable-next-line no-useless-escape
var youtubeRegex = /youtu(?:.*\/v\/|.*v\=|\.be\/)([A-Za-z0-9_-]{11})/;
var PlayIconBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  transform: translate(-50%, -50%);\n\n  &:before {\n    content: '';\n    position: absolute;\n    top: 50%;\n    left: 50%;\n    transform: translate(-50%, -50%);\n    width: 2.75em;\n    height: 2.75em;\n    border-radius: 50%;\n    background: ", ";\n    opacity: 0.75;\n  }\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.color.card.enabled.bg;
});
var Root = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  &:hover {\n    ", " {\n      &:before {\n        opacity: 1;\n      }\n    }\n  }\n"])), PlayIconBox);
var PosterCard = (0, _styledComponents.default)(_ui.Card)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  width: 100%;\n  padding-bottom: calc(9 / 16 * 100%);\n  position: relative;\n  border-bottom-left-radius: 0;\n  border-bottom-right-radius: 0;\n"])));

var Poster = _styledComponents.default.img(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  height: 100%;\n  width: 100%;\n  object-fit: cover;\n  display: block;\n  border-radius: inherit;\n\n  &:not([src]) {\n    display: none;\n  }\n"])));

var YoutubeContainer = (0, _styledComponents.default)(_ui.Card)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  position: relative;\n  padding-bottom: 56.25%;\n  overflow: hidden;\n  border-bottom-left-radius: 0;\n  border-bottom-right-radius: 0;\n\n  iframe {\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n  }\n"])));

var YoutubeEmbed = _ref2 => {
  var embedId = _ref2.embedId;
  return /*#__PURE__*/_react.default.createElement(YoutubeContainer, {
    radius: 3
  }, /*#__PURE__*/_react.default.createElement("iframe", {
    width: "853",
    height: "480",
    src: "https://www.youtube.com/embed/".concat(embedId),
    frameBorder: "0",
    allow: "accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture",
    allowFullScreen: true
  }));
};

class Tutorial extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        posterURL = _this$props.posterURL,
        showPlayIcon = _this$props.showPlayIcon,
        href = _this$props.href,
        presenterSubtitle = _this$props.presenterSubtitle;
    var isYoutube = showPlayIcon && href && href.match(youtubeRegex);
    return isYoutube ? /*#__PURE__*/_react.default.createElement(_ui.Card, {
      space: 2,
      sizing: "border",
      flex: 1,
      radius: 3,
      style: {
        position: 'relative'
      },
      border: true,
      paddingBottom: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2,
      height: "fill"
    }, /*#__PURE__*/_react.default.createElement(YoutubeEmbed, {
      embedId: href.match(youtubeRegex)[1]
    }), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 3,
      flex: 1,
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "h3",
      size: 1,
      weight: "bold"
    }, title), presenterSubtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      muted: true
    }, presenterSubtitle)))) : /*#__PURE__*/_react.default.createElement(Root, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      sizing: "border",
      flex: 1,
      radius: 3,
      as: "a",
      href: href,
      target: "_blank",
      style: {
        position: 'relative'
      },
      border: true,
      paddingBottom: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2,
      height: "fill"
    }, posterURL && /*#__PURE__*/_react.default.createElement(PosterCard, {
      radius: 3
    }, /*#__PURE__*/_react.default.createElement(Poster, {
      src: posterURL
    })), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 3,
      flex: 1,
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "h3",
      size: 1,
      weight: "bold"
    }, title), presenterSubtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      muted: true
    }, presenterSubtitle)))));
  }

}

_defineProperty(Tutorial, "defaultProps", {
  posterURL: null,
  showPlayIcon: false
});

var _default = Tutorial;
exports.default = _default;
