// sanity/lib/esm/studio/components/navbar/changelog/ChangelogList.js
import { Stack, Flex, Heading, Badge, Text } from '@sanity/ui';
import React from 'react';
import { ChangelogItem } from './ChangelogItem';
export function ChangelogList(props) {
  const {
    changeItems,
    isLatest,
    version
  } = props;
  const features = changeItems.filter(c => c.changeType === 'feature');
  const bugfixes = changeItems.filter(c => c.changeType === 'bugfix');
  const bugFixesDescription = bugfixes.map(_ref => {
    let {
      description
    } = _ref;
    return description;
  }).flat();
  return /*#__PURE__*/React.createElement(Stack, {
    space: 4
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    gap: 2
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 1
  }, "v", version), isLatest && /*#__PURE__*/React.createElement(Badge, {
    tone: "positive"
  }, "Latest")), /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, features === null || features === void 0 ? void 0 : features.map(change => /*#__PURE__*/React.createElement(ChangelogItem, {
    key: change.title,
    changeType: change.changeType,
    description: change === null || change === void 0 ? void 0 : change.description,
    title: change.title
  })), bugFixesDescription.length > 0 && /*#__PURE__*/React.createElement(ChangelogItem, {
    changeType: "bugfix",
    description: bugFixesDescription
  })), /*#__PURE__*/React.createElement(Flex, {
    justify: "flex-end"
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, /*#__PURE__*/React.createElement("a", {
    href: "https://github.com/sanity-io/sanity/releases",
    rel: "noreferrer noopener",
    target: "_blank"
  }, "See full changelog on GitHub"))));
}
