// sanity/lib/esm/studio/components/navbar/changelog/ChangelogAccordion.js
var _templateObject;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { ToggleArrowRightIcon } from '@sanity/icons';
import { Card, Stack, Button, Text, Code, Box } from '@sanity/ui';
import React, { useCallback, useState } from 'react';
import styled from 'styled-components';
const COMMANDS = ['npm i @sanity/cli -g', 'sanity upgrade'];
const Icon = styled(ToggleArrowRightIcon)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-rotate='true'] {\n    transform: rotate(90deg);\n  }\n"])));
export function ChangelogAccordion(props) {
  const {
    defaultOpen
  } = props;
  const [open, setOpen] = useState(defaultOpen);
  const handleClick = useCallback(() => setOpen(v => !v), []);
  return /*#__PURE__*/React.createElement(Card, {
    overflow: "hidden"
  }, /*#__PURE__*/React.createElement(Stack, null, /*#__PURE__*/React.createElement(Button, {
    "aria-controls": "upgrade-commands",
    icon: /*#__PURE__*/React.createElement(Icon, {
      "data-rotate": open
    }),
    justify: "flex-start",
    mode: "bleed",
    onClick: handleClick,
    padding: 4,
    radius: 0,
    space: 2,
    text: "How to upgrade"
  })), /*#__PURE__*/React.createElement(Card, {
    "aria-hidden": !open,
    hidden: !open,
    id: "upgrade-commands",
    overflow: "auto",
    paddingX: 4,
    paddingY: 2,
    role: "region",
    tone: "inherit"
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 2
  }, /*#__PURE__*/React.createElement(Card, {
    tone: "transparent",
    padding: 3,
    radius: 2
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, /*#__PURE__*/React.createElement(Code, {
    language: "bash",
    size: 1
  }, ["# Run these commands to upgrade", ...COMMANDS].join('\n')))), /*#__PURE__*/React.createElement(Box, {
    paddingX: 2,
    paddingY: 3
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1,
    muted: true
  }, "If you have problems upgrading, please get in touch with us in the", ' ', /*#__PURE__*/React.createElement("a", {
    href: "https://slack.sanity.io/",
    rel: "noopener noreferrer",
    target: "_blank"
  }, "Sanity Community"), ".")))));
}
