// sanity/lib/cjs/field/types/portableText/diff/components/Annotation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Annotation = Annotation;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _paths = require("@sanity/util/paths");

var _types = require("@sanity/types");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _changeIndicators = require("../../../../../components/changeIndicators");

var _diff = require("../../../../diff");

var _helpers = require("../helpers");

var _styledComponents2 = require("./styledComponents");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const AnnotationWrapper = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  text-decoration: none;\n  display: inline;\n  position: relative;\n  border: 0;\n  padding: 0;\n  border-bottom: 2px dotted currentColor;\n  box-shadow: inset 0 0 0 1px var(--card-border-color);\n  white-space: nowrap;\n  align-items: center;\n  background-color: color(var(--card-fg-color) a(10%));\n\n  &[data-changed] {\n    cursor: pointer;\n  }\n\n  &[data-removed] {\n    text-decoration: line-through;\n  }\n\n  &:hover ", " {\n    opacity: 1;\n  }\n"])), _styledComponents2.PreviewContainer);

function Annotation(_ref) {
  let {
    children,
    diff,
    object,
    schemaType,
    path,
    ...restProps
  } = _ref;

  if (!schemaType) {
    return /*#__PURE__*/_react.default.createElement(AnnotationWrapper, restProps, "Unknown schema type");
  }

  if (diff && diff.action !== 'unchanged') {
    return /*#__PURE__*/_react.default.createElement(AnnnotationWithDiff, _extends({}, restProps, {
      diff: diff,
      object: object,
      schemaType: schemaType,
      path: path
    }), children);
  }

  return /*#__PURE__*/_react.default.createElement(AnnotationWrapper, null, children);
}

function AnnnotationWithDiff(_ref2) {
  let {
    diff,
    children,
    object,
    schemaType,
    path,
    ...restProps
  } = _ref2;
  const {
    onSetFocus
  } = (0, _react.useContext)(_changeIndicators.ConnectorContext);
  const {
    path: fullPath
  } = (0, _react.useContext)(_diff.DiffContext);
  const [popoverElement, setPopoverElement] = (0, _react.useState)(null);
  const color = (0, _diff.useDiffAnnotationColor)(diff, []);
  const style = (0, _react.useMemo)(() => color ? {
    background: color.background,
    color: color.text
  } : {}, [color]);
  const isRemoved = diff.action === 'removed';
  const [open, setOpen] = (0, _react.useState)(false);
  const emptyObject = object && (0, _helpers.isEmptyObject)(object);
  const markDefPath = (0, _react.useMemo)(() => [path[0]].concat(['markDefs', {
    _key: object._key
  }]), [object._key, path]);
  const prefix = (0, _react.useMemo)(() => fullPath.slice(0, fullPath.findIndex(seg => (0, _types.isKeySegment)(seg) && seg._key === object._key)), [fullPath, object._key]);
  const annotationPath = (0, _react.useMemo)(() => prefix.concat(path), [path, prefix]);
  const myPath = (0, _react.useMemo)(() => prefix.concat(markDefPath), [markDefPath, prefix]);
  const myValue = "field-".concat((0, _paths.toString)(myPath));
  const values = (0, _changeIndicators.useReportedValues)();
  const isEditing = (0, _react.useMemo)(() => values.filter(_ref3 => {
    let [p] = _ref3;
    return p.startsWith(myValue);
  }).length > 0, [myValue, values]);
  (0, _react.useEffect)(() => {
    if (!open && isEditing) {
      setOpen(true);
      onSetFocus(myPath);
    }
  }, [isEditing, myPath, onSetFocus, open]);
  const handleOpenPopup = (0, _react.useCallback)(event => {
    event.stopPropagation();
    setOpen(true);

    if (!isRemoved) {
      event.preventDefault();
      onSetFocus(annotationPath); // Go to span first

      setTimeout(() => onSetFocus(myPath), 10); // Open edit object interface
    }
  }, [annotationPath, isRemoved, myPath, onSetFocus]);
  const handleClickOutside = (0, _react.useCallback)(() => {
    if (!isEditing) {
      setOpen(false);
    }
  }, [isEditing]);
  (0, _ui.useClickOutside)(handleClickOutside, [popoverElement]);
  const annotation = diff.action !== 'unchanged' && diff.annotation || null;
  const annotations = (0, _react.useMemo)(() => annotation ? [annotation] : [], [annotation]);

  const popoverContent = /*#__PURE__*/_react.default.createElement(_diff.DiffContext.Provider, {
    value: {
      path: myPath
    }
  }, /*#__PURE__*/_react.default.createElement(_styledComponents2.PopoverContainer, {
    padding: 3
  }, /*#__PURE__*/_react.default.createElement("div", null, emptyObject && /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 1,
    muted: true
  }, "Empty ", schemaType.title), !emptyObject && /*#__PURE__*/_react.default.createElement(_diff.ChangeList, {
    diff: diff,
    schemaType: schemaType
  }))));

  return /*#__PURE__*/_react.default.createElement(AnnotationWrapper, _extends({}, restProps, {
    onClick: handleOpenPopup,
    style: style,
    "data-changed": "",
    "data-removed": diff.action === 'removed' ? '' : undefined
  }), /*#__PURE__*/_react.default.createElement(_ui.Popover, {
    content: popoverContent,
    open: open,
    ref: setPopoverElement,
    portal: true
  }, /*#__PURE__*/_react.default.createElement(_styledComponents2.PreviewContainer, {
    paddingLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    annotations: annotations,
    description: "".concat(diff.action, " annotation")
  }, /*#__PURE__*/_react.default.createElement(_styledComponents2.InlineBox, {
    style: {
      display: 'inline-flex'
    }
  }, /*#__PURE__*/_react.default.createElement("span", null, children), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    paddingX: 1
  }, /*#__PURE__*/_react.default.createElement(_styledComponents2.InlineText, {
    size: 0
  }, /*#__PURE__*/_react.default.createElement(_icons.ChevronDownIcon, null))))))));
}
