// sanity/lib/cjs/field/types/portableText/diff/components/Block.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Block;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _changeIndicators = require("../../../../../components/changeIndicators");

var _diff = require("../../../../diff");

var _helpers = require("../helpers");

var _Blockquote = _interopRequireDefault(require("./Blockquote"));

var _Header = _interopRequireDefault(require("./Header"));

var _Paragraph = _interopRequireDefault(require("./Paragraph"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const EMPTY_PATH = [];

function Block(props) {
  const {
    diff,
    block,
    children
  } = props;
  const color = (0, _diff.useDiffAnnotationColor)(diff, EMPTY_PATH);
  const {
    path: fullPath
  } = (0, _react.useContext)(_diff.DiffContext);
  const {
    onSetFocus
  } = (0, _react.useContext)(_changeIndicators.ConnectorContext);
  const isRemoved = diff.action === 'removed';
  let returned = children;
  const handleClick = (0, _react.useCallback)(event => {
    event.stopPropagation();

    if (!isRemoved) {
      onSetFocus(fullPath);
    }
  }, [fullPath, isRemoved, onSetFocus]);

  if (block.style === 'blockquote') {
    returned = /*#__PURE__*/_react.default.createElement(_Blockquote.default, null, returned);
  } else if (block.style && (0, _helpers.isHeader)(block)) {
    returned = /*#__PURE__*/_react.default.createElement(_Header.default, {
      style: block.style
    }, returned);
  } else {
    returned = /*#__PURE__*/_react.default.createElement(_Paragraph.default, null, returned);
  }

  let fromStyle; // If style was changed, indicate that

  if (diff.origin.action === 'changed' && diff.origin.fields.style && diff.origin.fields.style.action === 'changed' && diff.origin.fields.style.annotation) {
    var _diff$origin, _diff$origin$fromValu, _diff$origin$fields$s;

    fromStyle = diff === null || diff === void 0 ? void 0 : (_diff$origin = diff.origin) === null || _diff$origin === void 0 ? void 0 : (_diff$origin$fromValu = _diff$origin.fromValue) === null || _diff$origin$fromValu === void 0 ? void 0 : _diff$origin$fromValu.style;
    const style = color ? {
      background: color.background,
      color: color.text
    } : {};
    returned = /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 3,
      border: true,
      radius: 2,
      style: {
        borderStyle: 'dotted'
      },
      "diff-block-action": diff.action,
      "data-block-note": "changed_from_style_".concat(fromStyle || 'undefined')
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2
    }, /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
      annotations: [(_diff$origin$fields$s = diff.origin.fields.style) === null || _diff$origin$fields$s === void 0 ? void 0 : _diff$origin$fields$s.annotation],
      diff: diff.origin.fields.style
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 0
    }, "Changed block style from '", fromStyle, "'")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      style: style
    }, returned)));
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    onClick: handleClick,
    "diff-block-action": diff.action,
    "data-block-note": "changed_from_style_".concat(fromStyle || 'undefined')
  }, returned);
}
