// sanity/lib/esm/cli/actions/graphql/gen1/generateTypeQueries.js
import _upperFirst from "lodash/upperFirst";
import _startCase from "lodash/startCase";
import pluralize from 'pluralize-esm';
import { isNonUnion } from '../helpers';

function pluralizeTypeName(name) {
  const words = _startCase(name).split(' ');

  const last = words[words.length - 1];
  const plural = pluralize(last.toLowerCase());
  words[words.length - 1] = _upperFirst(plural);
  return words.join('');
}

export function generateTypeQueries(types, filters) {
  const queries = [];
  const queryable = types.filter(isNonUnion).filter(type => type.type === 'Object' && type.interfaces && type.interfaces.includes('Document')); // Single ID-based result lookup queries

  queryable.forEach(type => {
    queries.push({
      fieldName: type.name,
      type: type.name,
      constraints: [{
        field: '_id',
        comparator: 'EQUALS',
        value: {
          kind: 'argumentValue',
          argName: 'id'
        }
      }],
      args: [{
        name: 'id',
        description: "".concat(type.name, " document ID"),
        type: 'ID',
        isNullable: false
      }]
    });
  }); // Fetch all of type

  queryable.forEach(type => {
    const filterName = "".concat(type.name, "Filter");
    const hasFilter = filters.find(filter => filter.name === filterName);
    queries.push({
      fieldName: "all".concat(pluralizeTypeName(type.name)),
      filter: "_type == \"".concat(type.originalName || type.name, "\""),
      type: {
        kind: 'List',
        isNullable: false,
        children: {
          type: type.name,
          isNullable: false
        }
      },
      args: hasFilter ? [{
        name: 'where',
        type: filterName,
        isFieldFilter: true
      }, ...getLimitOffsetArgs()] : getLimitOffsetArgs()
    });
  });
  return queries;
}

function getLimitOffsetArgs() {
  return [{
    name: 'limit',
    type: 'Int',
    description: 'Max documents to return',
    isFieldFilter: false
  }, {
    name: 'offset',
    type: 'Int',
    description: 'Offset at which to start returning documents from',
    isFieldFilter: false
  }];
}
