// sanity/lib/esm/components/previews/__workshop__/InlinePreviewStory.js
var _templateObject;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { DocumentIcon } from '@sanity/icons';
import { Card, Container, Flex, Heading, Stack, Text } from '@sanity/ui';
import { useBoolean, useString } from '@sanity/ui-workshop';
import React from 'react';
import styled from 'styled-components';
import { InlinePreview } from '../portableText/InlinePreview';
const PreviewWrapper = styled.span(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: inline-block;\n  vertical-align: top;\n  background-color: var(--card-bg-color);\n  box-shadow: inset 0 0 0 1px var(--card-border-color);\n  border-radius: 2.66px;\n  padding: 2px;\n  margin-top: 0.0625em;\n  height: calc(1em - 1px);\n\n  & > span {\n    vertical-align: top;\n    margin: 0;\n  }\n"])));
export default function InlinePreviewStory() {
  const title = useString('Title', 'Inline object', 'Props');
  const withImage = useBoolean('With image', true, 'Props');
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 1
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 4
  }, /*#__PURE__*/React.createElement(Card, {
    tone: "primary"
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 5
  }, "Hello", ' ', /*#__PURE__*/React.createElement(PreviewWrapper, null, /*#__PURE__*/React.createElement(InlinePreview // eslint-disable-next-line react/jsx-no-bind
  , {
    media: () => withImage ? /*#__PURE__*/React.createElement("img", {
      src: "https://source.unsplash.com/60x60/?face"
    }) : /*#__PURE__*/React.createElement(DocumentIcon, null),
    title: title
  })), ' ', "world")), /*#__PURE__*/React.createElement(Card, {
    tone: "primary"
  }, /*#__PURE__*/React.createElement(Heading, null, "Hello", ' ', /*#__PURE__*/React.createElement(PreviewWrapper, null, /*#__PURE__*/React.createElement(InlinePreview // eslint-disable-next-line react/jsx-no-bind
  , {
    media: () => withImage ? /*#__PURE__*/React.createElement("img", {
      src: "https://source.unsplash.com/60x60/?face"
    }) : /*#__PURE__*/React.createElement(DocumentIcon, null),
    title: title
  })), ' ', "world")), /*#__PURE__*/React.createElement(Card, {
    tone: "primary"
  }, /*#__PURE__*/React.createElement(Text, null, "Hello", ' ', /*#__PURE__*/React.createElement(PreviewWrapper, null, /*#__PURE__*/React.createElement(InlinePreview // eslint-disable-next-line react/jsx-no-bind
  , {
    media: () => withImage ? /*#__PURE__*/React.createElement("img", {
      src: "https://source.unsplash.com/60x60/?face"
    }) : /*#__PURE__*/React.createElement(DocumentIcon, null),
    title: title
  })), ' ', "world"))))));
}
