// sanity/lib/esm/components/previews/__workshop__/TemplatePreviewStory.js
import { DocumentIcon } from '@sanity/icons';
import { Card, Container, Flex, Text } from '@sanity/ui';
import { useBoolean, useSelect, useString, useText } from '@sanity/ui-workshop';
import React, { createElement } from 'react';
import { TemplatePreview } from '../template/TemplatePreview';
const MEDIA_OPTIONS = {
  None: 'none',
  Image: 'image',
  Icon: 'icon'
};
const LAYOUT_OPTIONS = {
  'Create document': 'createDocument'
};
const mediaValues = {
  image: /*#__PURE__*/React.createElement("img", {
    src: "https://source.unsplash.com/70x70/?abstract"
  }),
  icon: /*#__PURE__*/React.createElement(DocumentIcon, null)
};
const previewComponents = {
  createDocument: TemplatePreview
};
export default function TemplatePreviewStory() {
  const layout = useSelect('Layout', LAYOUT_OPTIONS, 'createDocument');
  const isPlaceholder = useBoolean('Is placeholder', false);
  const mediaKey = useSelect('Media', MEDIA_OPTIONS) || 'none';
  const title = useString('Title', 'Title', 'Props');
  const subtitle = useString('Subtitle', 'Subtitle', 'Props');
  const description = useText('Description', undefined, 'Props');
  const media = mediaValues[mediaKey] || false;
  const component = layout && previewComponents[layout];

  if (!component) {
    return /*#__PURE__*/React.createElement(Flex, {
      align: "center",
      height: "fill",
      justify: "center",
      padding: 4,
      sizing: "border"
    }, /*#__PURE__*/React.createElement(Text, null, "Unknown layout: ", layout));
  }

  return /*#__PURE__*/React.createElement(Card, {
    height: "fill",
    tone: "transparent"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 0
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 3,
    radius: 2
  }, /*#__PURE__*/createElement(component, {
    description,
    isPlaceholder,
    media,
    title,
    subtitle
  })))));
}
