// sanity/lib/cjs/components/popoverDialog/PopoverDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverDialog = PopoverDialog;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function PopoverDialog(props) {
  const {
    children,
    header,
    onClose,
    width,
    ...restProps
  } = props;
  return /*#__PURE__*/_react.default.createElement(_ui.Popover, _extends({}, restProps, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Container, {
      width: width
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      direction: "column"
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: "none"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold"
    }, header)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.CloseIcon,
      onClick: onClose
    })))), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: 1,
      overflow: "auto"
    }, children)))
  }));
}
