// sanity/lib/cjs/studio/screens/CorsOriginErrorScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CorsOriginErrorScreen = CorsOriginErrorScreen;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function CorsOriginErrorScreen(props) {
  const {
    projectId
  } = props;
  const corsUrl = "https://sanity.io/manage/project/".concat(projectId, "/settings/api");
  const origin = window.location.origin;
  const handleRetry = (0, _react.useCallback)(() => {
    window.location.reload();
  }, []);
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    id: "cors-error-dialog",
    header: "Error",
    width: 1,
    footer: /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      paddingX: 3,
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      text: "Retry",
      onClick: handleRetry
    }))
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    paddingX: 4,
    paddingY: 5,
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "It looks like the error is being caused by the current origin (", /*#__PURE__*/_react.default.createElement("code", null, origin), ") not being allowed for this project."), /*#__PURE__*/_react.default.createElement(_ui.Text, null, "If you are a project administrator or developer, you can head to", ' ', /*#__PURE__*/_react.default.createElement("a", {
    rel: "noopener noreferrer",
    target: "_blank",
    href: corsUrl
  }, "the project management"), ' ', "interface to configure CORS origins for this project."), /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement("a", {
    href: "https://www.sanity.io/docs/front-ends/cors",
    target: "_blank",
    rel: "noopener noreferrer"
  }, "Read more about CORS Origins"), "."))));
}
