// sanity/lib/cjs/studio/screens/NotFoundScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotFoundScreen = NotFoundScreen;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function NotFoundScreen(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    sizing: "border",
    tone: "caution",
    display: "flex"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "row",
    justify: "center",
    flex: 1,
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    as: "h1"
  }, "Workspace not found"), /*#__PURE__*/_react.default.createElement(_ui.Inline, null, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Go to default workspace",
    onClick: props.onNavigateToDefaultWorkspace,
    mode: "ghost"
  })))));
}
