// sanity/lib/esm/desk/panes/unknown/UnknownPaneType.js
import { Box, Text } from '@sanity/ui';
import React from 'react';
import { isRecord } from '../../../util';
import { Pane, PaneContent, PaneHeader } from '../../components/pane';

/**
 * @internal
 */
export function UnknownPane(props) {
  const {
    isSelected,
    pane,
    paneKey
  } = props;
  const type = isRecord(pane) && pane.type || null;
  return /*#__PURE__*/React.createElement(Pane, {
    id: paneKey,
    selected: isSelected
  }, /*#__PURE__*/React.createElement(PaneHeader, {
    title: "Unknown pane type"
  }), /*#__PURE__*/React.createElement(PaneContent, null, /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, typeof type === 'string' ? /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true
  }, "Structure item of type ", /*#__PURE__*/React.createElement("code", null, type), " is not a known entity.") : /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true
  }, "Structure item is missing required ", /*#__PURE__*/React.createElement("code", null, "type"), " property."))));
}
