// sanity/lib/esm/util/draftUtils.js
import { isNonNullable } from './isNonNullable'; // nominal/opaque type hack

export const DRAFTS_FOLDER = 'drafts';
const DRAFTS_PREFIX = "".concat(DRAFTS_FOLDER, ".");
export function isDraft(document) {
  return isDraftId(document._id);
}
export function isDraftId(id) {
  return id.startsWith(DRAFTS_PREFIX);
}
export function getIdPair(id) {
  return {
    draftId: getDraftId(id),
    publishedId: getPublishedId(id)
  };
}
export function isPublishedId(id) {
  return !isDraftId(id);
}
export function getDraftId(id) {
  return isDraftId(id) ? id : DRAFTS_PREFIX + id;
}
export function getPublishedId(id) {
  return isDraftId(id) ? id.slice(DRAFTS_PREFIX.length) : id;
}
export function createDraftFrom(document) {
  return { ...document,
    _id: getDraftId(document._id)
  };
}
export function newDraftFrom(document) {
  return { ...document,
    _id: DRAFTS_PREFIX
  };
}
export function createPublishedFrom(document) {
  return { ...document,
    _id: getPublishedId(document._id)
  };
}
/**
 * Takes a list of documents and collates draft/published pairs into single entries
 * `{id: <published id>, draft?: <draft document>, published?: <published document>}`
 *
 * Note: because Map is ordered by insertion key the resulting array will be ordered by whichever
 * version appeared first
 */

export function collate(documents) {
  const byId = documents.reduce((res, doc) => {
    const publishedId = getPublishedId(doc._id);
    let entry = res.get(publishedId);

    if (!entry) {
      entry = {
        id: publishedId,
        type: doc._type,
        published: undefined,
        draft: undefined
      };
      res.set(publishedId, entry);
    }

    entry[publishedId === doc._id ? 'published' : 'draft'] = doc;
    return res;
  }, new Map());
  return Array.from(byId.values());
} // Removes published documents that also has a draft

export function removeDupes(documents) {
  return collate(documents).map(entry => entry.draft || entry.published).filter(isNonNullable);
}
