// sanity/lib/esm/form/components/formField/FormFieldHeaderText.js
/* eslint-disable camelcase */
import { Box, Flex, Stack, Text } from '@sanity/ui';
import React, { memo } from 'react';
import { FormFieldValidationStatus } from './FormFieldValidationStatus';
const EMPTY_ARRAY = [];
export const FormFieldHeaderText = /*#__PURE__*/memo(function FormFieldHeaderText(props) {
  const {
    description,
    inputId,
    title,
    validation = EMPTY_ARRAY
  } = props;
  const hasValidations = validation.length > 0;
  return /*#__PURE__*/React.createElement(Stack, {
    space: 2
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Text, {
    as: "label",
    htmlFor: inputId,
    weight: "semibold",
    size: 1
  }, title || /*#__PURE__*/React.createElement("span", {
    style: {
      color: 'var(--card-muted-fg-color)'
    }
  }, "Untitled")), hasValidations && /*#__PURE__*/React.createElement(Box, {
    marginLeft: 2
  }, /*#__PURE__*/React.createElement(FormFieldValidationStatus, {
    fontSize: 1,
    validation: validation
  }))), description && /*#__PURE__*/React.createElement(Text, {
    muted: true,
    size: 1
  }, description));
});
