// sanity/lib/cjs/form/__workshop__/_common/FilterFieldInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilterFieldInput = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const ExampleCode = (0, _styledComponents.default)(_ui.Code)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  cursor: pointer;\n"])));
const EXAMPLE_FILTER_FUNCTION = "(type, field) => field.name === 'title'";

const FilterFieldInput = /*#__PURE__*/_react.default.forwardRef(function FilterFieldInput(props, ref) {
  const {
    value,
    onChange,
    onFilter
  } = props;
  const handleChange = (0, _react.useCallback)(event => {
    onChange(event.target.value);
  }, [onChange]);
  const handleFilterExample = (0, _react.useCallback)(() => {
    onChange(EXAMPLE_FILTER_FUNCTION);
  }, [onChange]);
  const handleFilter = (0, _react.useCallback)(() => {
    onFilter(value);
  }, [onFilter, value]);
  const handleReset = (0, _react.useCallback)(() => {
    onFilter("");
  }, [onFilter]);
  const isDisabled = (0, _react.useMemo)(() => {
    return (value === null || value === void 0 ? void 0 : value.length) === 0;
  }, [value]);
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    tone: "default",
    border: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 0
  }, "Function value"), /*#__PURE__*/_react.default.createElement(_ui.TextArea, {
    rows: 4,
    ref: ref,
    onChange: handleChange,
    value: value || ''
  }), /*#__PURE__*/_react.default.createElement(ExampleCode, {
    title: "Use example code",
    size: 1,
    onClick: handleFilterExample
  }, "Example: ".concat(EXAMPLE_FILTER_FUNCTION)), /*#__PURE__*/_react.default.createElement(_ui.Grid, {
    columns: 2,
    gap: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Filter",
    onClick: handleFilter,
    disabled: isDisabled
  }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Reset",
    mode: "ghost",
    onClick: handleReset,
    disabled: isDisabled
  }))));
});

exports.FilterFieldInput = FilterFieldInput;
