// sanity/lib/cjs/form/__workshop__/_common/TypeTester.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeTester = TypeTester;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _typer = require("./typer");

var _templateObject, _templateObject2;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const HIGHLIGHT_CLASSNAME = 'js-highlight-selected-input';
const GlobalStyle = (0, _styledComponents.createGlobalStyle)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  .", " {\n    outline: 2px solid red;\n  }\n"])), HIGHLIGHT_CLASSNAME);

function getNameFromInput(input) {
  const classNames = input.classList.toString().split(' ').filter(name => name !== HIGHLIGHT_CLASSNAME).join('.');
  return "".concat(input.tagName.toLowerCase(), ".").concat(classNames);
}

const Sticky = _styledComponents.default.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: sticky;\n  top: 1rem;\n  z-index: 2;\n"])));

function TypeTester(_ref) {
  let {
    readOnly
  } = _ref;
  const textAreaRef = (0, _react.useRef)(null);
  const [focusedInput, setFocusedInput] = (0, _react.useState)();
  const [testOutput, setTestOutput] = (0, _react.useState)('');
  const [isRunning, setIsRunning] = (0, _react.useState)(false);
  const cancelTypeTester = (0, _react.useRef)(() => true);
  const handleFocus = (0, _react.useCallback)(e => {
    if (textAreaRef.current && e.target === textAreaRef.current) {
      return;
    }

    if (e.type === 'focusin' && e.target instanceof HTMLInputElement || e.target instanceof HTMLTextAreaElement) {
      setFocusedInput(e.target);
    }
  }, [textAreaRef]);
  (0, _react.useEffect)(() => {
    document.addEventListener('focusin', handleFocus);
    return () => {
      document.removeEventListener('focusin', handleFocus);

      if (cancelTypeTester.current) {
        cancelTypeTester.current();
      }
    };
  }, [handleFocus]);
  const handleRunTest = (0, _react.useCallback)(() => {
    setIsRunning(true);
    setTestOutput('Running test...');
    cancelTypeTester.current = (0, _typer.runTest)({
      inputElement: focusedInput,
      times: 4,
      gracePeriod: 4000,
      onRun: output => {
        setTestOutput(existing => "".concat(existing, "\n").concat(output));
      },
      onFinished: () => {
        setIsRunning(false);
      }
    });
  }, [focusedInput]);
  const handleHighlightMouseOver = (0, _react.useCallback)(event => {
    if (event.type === 'mouseover') {
      focusedInput.classList.add(HIGHLIGHT_CLASSNAME);
    }

    if (event.type === 'mouseout') {
      focusedInput.classList.remove(HIGHLIGHT_CLASSNAME);
    }
  }, [focusedInput]);
  const handleHighlightClick = (0, _react.useCallback)(() => {
    if (focusedInput) {
      focusedInput.focus();
    }
  }, [focusedInput]);
  const isDisabled = (0, _react.useMemo)(() => {
    return isRunning || !focusedInput || !['INPUT', 'TEXTAREA'].includes(focusedInput.tagName) || readOnly;
  }, [focusedInput, isRunning, readOnly]);
  return /*#__PURE__*/_react.default.createElement(Sticky, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    tone: "default",
    border: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 0
  }, "Type Performance Tester"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    title: isDisabled ? undefined : "Click to focus input",
    size: 1,
    onClick: handleHighlightClick,
    onMouseOver: handleHighlightMouseOver,
    onMouseOut: handleHighlightMouseOver
  }, isDisabled && !isRunning ? 'No input selected' : "Selected input: ".concat(focusedInput ? getNameFromInput(focusedInput) : '<none>')), /*#__PURE__*/_react.default.createElement(_ui.TextArea, {
    ref: textAreaRef,
    readOnly: true,
    rows: 6,
    value: testOutput
  }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Run test",
    onClick: handleRunTest,
    disabled: isDisabled,
    loading: isRunning
  }))), /*#__PURE__*/_react.default.createElement(GlobalStyle, null));
}
