// sanity/lib/cjs/form/studio/inputs/reference/StudioReferenceInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StudioReferenceInput = StudioReferenceInput;

var PathUtils = _interopRequireWildcard(require("@sanity/util/paths"));

var _react = _interopRequireWildcard(require("react"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _util = require("../../../../util");

var adapter = _interopRequireWildcard(require("../client-adapters/reference"));

var _ReferenceInput = require("../../../inputs/ReferenceInput/ReferenceInput");

var _datastores = require("../../../../datastores");

var _hooks = require("../../../../hooks");

var _contexts = require("../../contexts");

var _useFormValue = require("../../../useFormValue");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

// eslint-disable-next-line require-await
async function resolveUserDefinedFilter(options, document, valuePath) {
  if (!options) {
    return {};
  }

  if (typeof options.filter === 'function') {
    const parentPath = valuePath.slice(0, -1);
    const parent = (0, PathUtils.get)(document, parentPath);
    return options.filter({
      document,
      parentPath,
      parent
    });
  }

  return {
    filter: options.filter,
    params: 'filterParams' in options ? options.filterParams : undefined
  };
}

function useValueRef(value) {
  const ref = (0, _react.useRef)(value);
  ref.current = value;
  return ref;
}

function StudioReferenceInput(props) {
  var _documentRef$current, _schemaType$options, _props$value, _props$value$_strengt;

  const client = (0, _hooks.useClient)();
  const schema = (0, _hooks.useSchema)();
  const documentPreviewStore = (0, _datastores.useDocumentPreviewStore)();
  const searchClient = (0, _react.useMemo)(() => client.withConfig({
    apiVersion: '2021-03-25'
  }), [client]);
  const {
    path,
    schemaType
  } = props;
  const {
    EditReferenceLinkComponent,
    onEditReference,
    activePath,
    initialValueTemplateItems
  } = (0, _contexts.useReferenceInputOptions)();
  const documentValue = (0, _useFormValue.useFormValue)([]);
  const documentRef = useValueRef(documentValue);
  const documentTypeName = (_documentRef$current = documentRef.current) === null || _documentRef$current === void 0 ? void 0 : _documentRef$current._type;
  const refType = schema.get(documentTypeName);
  const isDocumentLiveEdit = (0, _react.useMemo)(() => refType === null || refType === void 0 ? void 0 : refType.liveEdit, [refType]);
  const disableNew = ((_schemaType$options = schemaType.options) === null || _schemaType$options === void 0 ? void 0 : _schemaType$options.disableNew) === true;
  const handleSearch = (0, _react.useCallback)(searchString => (0, _rxjs.from)(resolveUserDefinedFilter(schemaType.options, documentRef.current, path)).pipe((0, _operators.mergeMap)(_ref => {
    let {
      filter,
      params
    } = _ref;
    return adapter.referenceSearch(searchClient, searchString, schemaType, { ...schemaType.options,
      filter,
      params,
      tag: 'search.reference'
    });
  }), (0, _operators.catchError)(err => {
    var _schemaType$options2;

    const isQueryError = err.details && err.details.type === 'queryParseError';

    if ((_schemaType$options2 = schemaType.options) !== null && _schemaType$options2 !== void 0 && _schemaType$options2.filter && isQueryError) {
      err.message = "Invalid reference filter, please check the custom \"filter\" option";
    }

    return (0, _rxjs.throwError)(err);
  })), [documentRef, path, searchClient, schemaType]);
  const template = (_props$value = props.value) === null || _props$value === void 0 ? void 0 : (_props$value$_strengt = _props$value._strengthenOnPublish) === null || _props$value$_strengt === void 0 ? void 0 : _props$value$_strengt.template;
  const EditReferenceLink = (0, _react.useMemo)(() => /*#__PURE__*/(0, _react.forwardRef)(function EditReferenceLink_(_props, forwardedRef) {
    return EditReferenceLinkComponent ? /*#__PURE__*/_react.default.createElement(EditReferenceLinkComponent, _extends({}, _props, {
      ref: forwardedRef,
      parentRefPath: path,
      template: template
    })) : null;
  }), [EditReferenceLinkComponent, path, template]);
  const handleEditReference = (0, _react.useCallback)(event => {
    onEditReference === null || onEditReference === void 0 ? void 0 : onEditReference({
      parentRefPath: path,
      id: event.id,
      type: event.type,
      template: event.template
    });
  }, [onEditReference, path]);
  const selectedState = PathUtils.startsWith(path, (activePath === null || activePath === void 0 ? void 0 : activePath.path) || []) ? activePath === null || activePath === void 0 ? void 0 : activePath.state : 'none';
  const createOptions = (0, _react.useMemo)(() => {
    if (disableNew) {
      return [];
    }

    return (initialValueTemplateItems || [] // eslint-disable-next-line max-nested-callbacks
    ).filter(i => {
      return schemaType.to.some(_refType => {
        var _i$template;

        return _refType.name === ((_i$template = i.template) === null || _i$template === void 0 ? void 0 : _i$template.schemaType);
      });
    }).map(item => {
      var _item$template, _item$template2, _item$template3;

      return (_item$template = item.template) !== null && _item$template !== void 0 && _item$template.schemaType ? {
        id: item.id,
        title: item.title || "".concat(item.template.schemaType, " from template ").concat((_item$template2 = item.template) === null || _item$template2 === void 0 ? void 0 : _item$template2.id),
        type: item.template.schemaType,
        icon: item.icon,
        template: {
          id: (_item$template3 = item.template) === null || _item$template3 === void 0 ? void 0 : _item$template3.id,
          params: item.parameters
        },
        permission: {
          granted: item.granted,
          reason: item.reason
        }
      } : undefined;
    }).filter(_util.isNonNullable);
  }, [disableNew, initialValueTemplateItems, schemaType.to]);
  const getReferenceInfo = (0, _react.useCallback)((id, _type) => adapter.getReferenceInfo(documentPreviewStore, id, _type), [documentPreviewStore]);
  return /*#__PURE__*/_react.default.createElement(_ReferenceInput.ReferenceInput, _extends({}, props, {
    onSearch: handleSearch,
    liveEdit: isDocumentLiveEdit,
    getReferenceInfo: getReferenceInfo,
    selectedState: selectedState,
    editReferenceLinkComponent: EditReferenceLink,
    createOptions: createOptions,
    onEditReference: handleEditReference
  }));
}
