// sanity/lib/cjs/desk/structureBuilder/views/ComponentView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentViewBuilder = void 0;

var _SerializeError = require("../SerializeError");

var _util = require("../../../util");

var _View = require("./View");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const isComponentSpec = spec => (0, _util.isRecord)(spec) && spec.type === 'component';

class ComponentViewBuilder extends _View.GenericViewBuilder {
  constructor(componentOrSpec) {
    const spec = isComponentSpec(componentOrSpec) ? { ...componentOrSpec
    } : {
      options: {}
    };
    super();

    _defineProperty(this, "spec", void 0);

    this.spec = spec;
    const userComponent = typeof componentOrSpec === 'function' ? componentOrSpec : this.spec.component;

    if (userComponent) {
      // Because we're cloning, this'll return a new instance, so grab the spec from it
      this.spec = this.component(userComponent).spec;
    }
  }

  component(component) {
    return this.clone({
      component
    });
  }

  getComponent() {
    return this.spec.component;
  }

  options(options) {
    return this.clone({
      options
    });
  }

  getOptions() {
    return this.spec.options || {};
  }

  serialize() {
    let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    const base = super.serialize(options);
    const component = this.spec.component;

    if (typeof component !== 'function') {
      throw new _SerializeError.SerializeError('`component` is required and must be a function for `component()` view item', options.path, options.index).withHelpUrl(_SerializeError.HELP_URL.COMPONENT_REQUIRED);
    }

    return { ...base,
      component,
      options: this.spec.options || {},
      type: 'component'
    };
  }

  clone(withSpec) {
    const builder = new ComponentViewBuilder();
    builder.spec = { ...this.spec,
      ...(withSpec || {})
    };
    return builder;
  }

}

exports.ComponentViewBuilder = ComponentViewBuilder;
