// sanity/lib/cjs/form/studio/StudioFormBuilderProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StudioFormBuilderProvider = StudioFormBuilderProvider;

var _react = _interopRequireDefault(require("react"));

var _studio = require("../../studio");

var _FormBuilderProvider = require("../FormBuilderProvider");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable camelcase */

/**
 * Default wiring for `FormBuilderProvider` when used with Sanity
 *
 * @alpha This API might change.
 */
function StudioFormBuilderProvider(props) {
  const {
    __internal_patchChannel: patchChannel,
    autoFocus,
    changesOpen,
    children,
    collapsedFieldSets,
    collapsedPaths,
    focusPath,
    focused,
    groups,
    id,
    members,
    onChange,
    onPathBlur,
    onPathFocus,
    onPathOpen,
    onFieldGroupSelect,
    onSetPathCollapsed,
    onSetFieldSetCollapsed,
    presence,
    readOnly,
    schemaType,
    validation,
    value
  } = props;
  const {
    file,
    image,
    renderField,
    renderInput,
    renderItem,
    renderPreview
  } = (0, _studio.useSource)().form;
  return /*#__PURE__*/_react.default.createElement(_FormBuilderProvider.FormBuilderProvider, {
    __internal_patchChannel: patchChannel,
    autoFocus: autoFocus,
    changesOpen: changesOpen,
    collapsedFieldSets: collapsedFieldSets,
    collapsedPaths: collapsedPaths,
    file: file,
    focusPath: focusPath,
    focused: focused,
    groups: groups,
    id: id,
    image: image,
    members: members,
    onChange: onChange,
    onPathBlur: onPathBlur,
    onPathFocus: onPathFocus,
    onPathOpen: onPathOpen,
    onFieldGroupSelect: onFieldGroupSelect,
    onSetPathCollapsed: onSetPathCollapsed,
    onSetFieldSetCollapsed: onSetFieldSetCollapsed,
    presence: presence,
    readOnly: readOnly,
    renderField: renderField,
    renderInput: renderInput,
    renderItem: renderItem,
    renderPreview: renderPreview,
    schemaType: schemaType,
    validation: validation,
    value: value
  }, children);
}
