// sanity/lib/esm/components/__workshop__/PreviewCardStory.js
import { EditIcon, PublishIcon } from '@sanity/icons';
import { Container, Flex } from '@sanity/ui';
import { useBoolean } from '@sanity/ui-workshop';
import React from 'react';
import { DocumentPreviewPresence } from '../../presence';
import { PreviewCard } from '../PreviewCard';
import { TextWithTone } from '../TextWithTone';
export default function PreviewCardStory() {
  const selected = useBoolean('Selected', true);
  return /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 0
  }, /*#__PURE__*/React.createElement(PreviewCard, {
    as: "button",
    paddingY: 2,
    paddingX: 3,
    radius: 2,
    selected: selected
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    justify: "center",
    gap: 4
  }, /*#__PURE__*/React.createElement(DocumentPreviewPresence, {
    presence: ['id-1', 'id-2', 'id-3', 'id-4'].map(num => ({
      status: 'online',
      lastActiveAt: '',
      user: {
        imageUrl: 'https://source.unsplash.com/96x96/?face',
        id: num
      }
    }))
  }), /*#__PURE__*/React.createElement(TextWithTone, {
    tone: "positive"
  }, /*#__PURE__*/React.createElement(PublishIcon, null)), /*#__PURE__*/React.createElement(TextWithTone, {
    tone: "caution"
  }, /*#__PURE__*/React.createElement(EditIcon, null))))));
}
