// sanity/lib/esm/components/__workshop__/TextWithToneStory.js
import { InfoOutlineIcon } from '@sanity/icons';
import { Box, Card, Flex, Grid, Stack } from '@sanity/ui';
import { useBoolean, useSelect } from '@sanity/ui-workshop';
import React from 'react';
import { TextWithTone } from '../TextWithTone';
const TEXT_SIZE_OPTIONS = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4
};
const CARD_TONE_OPTIONS = {
  Default: 'default',
  Primary: 'primary',
  Positive: 'positive',
  Caution: 'caution',
  Critical: 'critical'
};
const tones = ['default', 'primary', 'positive', 'caution', 'critical'];
export default function TextWithToneStory() {
  const size = useSelect('Size', TEXT_SIZE_OPTIONS, 2);
  const dimmed = useBoolean('Dimmed', false);
  const muted = useBoolean('Muted', false);
  const cardTone = useSelect('Card tone', CARD_TONE_OPTIONS);
  return /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 4,
    border: true,
    radius: 2,
    tone: cardTone
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, tones.map(tone => /*#__PURE__*/React.createElement(TextWithTone, {
    key: tone,
    size: size,
    dimmed: dimmed,
    muted: muted,
    tone: tone
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    gap: 2
  }, /*#__PURE__*/React.createElement(InfoOutlineIcon, null), tone))))), /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, tones.map(tone => /*#__PURE__*/React.createElement(TextWithTone, {
    key: tone,
    size: size,
    dimmed: dimmed,
    muted: muted,
    tone: tone
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    gap: 2
  }, /*#__PURE__*/React.createElement(InfoOutlineIcon, null), tone)))))));
}
