// sanity/lib/cjs/form/inputs/PortableText/text/AnnotationToolbarPopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnnotationToolbarPopover = AnnotationToolbarPopover;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _icons = require("@sanity/icons");

var _portableTextEditor = require("@sanity/portable-text-editor");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const ToolbarPopover = (0, _styledComponents.default)(_ui.Popover)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-popper-reference-hidden='true'] {\n    display: none !important;\n  }\n"])));
const POPOVER_FALLBACK_PLACEMENTS = ['top', 'bottom'];

function AnnotationToolbarPopover(props) {
  const {
    scrollElement,
    annotationElement,
    focused,
    textElement,
    title,
    onEdit,
    onDelete
  } = props;
  const [open, setOpen] = (0, _react.useState)(false);
  const [cursorRect, setCursorRect] = (0, _react.useState)(null);
  const [selection, setSelection] = (0, _react.useState)(null);
  const isClosingRef = (0, _react.useRef)(false);
  const rangeRef = (0, _react.useRef)(null);
  const editButtonRef = (0, _react.useRef)(null);
  const isTabbing = (0, _react.useRef)(false);
  const {
    sanity
  } = (0, _ui.useTheme)();
  const editor = (0, _portableTextEditor.usePortableTextEditor)();
  const popoverScheme = sanity.color.dark ? 'light' : 'dark'; // This is a "virtual element" (supported by Popper.js)

  const cursorElement = (0, _react.useMemo)(() => {
    if (!cursorRect) {
      return null;
    }

    return {
      getBoundingClientRect: () => {
        return cursorRect;
      }
    };
  }, [cursorRect]);
  (0, _react.useEffect)(() => {
    if (!open) {
      return undefined;
    }

    const handleScroll = () => {
      if (rangeRef.current) {
        setCursorRect(rangeRef.current.getBoundingClientRect());
      }
    };

    scrollElement === null || scrollElement === void 0 ? void 0 : scrollElement.addEventListener('scroll', handleScroll, {
      passive: true
    });
    return () => {
      scrollElement === null || scrollElement === void 0 ? void 0 : scrollElement.removeEventListener('scroll', handleScroll);
    };
  }, [open, scrollElement]); // Close floating toolbar on Escape

  (0, _ui.useGlobalKeyDown)((0, _react.useCallback)(event => {
    if (!open) {
      return;
    }

    if (event.key === 'Escape') {
      event.preventDefault();
      event.stopPropagation();
      setOpen(false);
      isTabbing.current = false;

      _portableTextEditor.PortableTextEditor.focus(editor);
    }

    if (event.key === 'Tab') {
      if (!isTabbing.current) {
        var _editButtonRef$curren;

        event.preventDefault();
        event.stopPropagation();
        (_editButtonRef$curren = editButtonRef.current) === null || _editButtonRef$curren === void 0 ? void 0 : _editButtonRef$curren.focus();
        isTabbing.current = true;
      }
    }
  }, [editor, open])); // Detect selection changes

  (0, _react.useEffect)(() => {
    function handleSelectionChange() {
      if (!textElement) return;
      const winSelection = window.getSelection();

      if (!winSelection) {
        return;
      }

      const {
        anchorNode,
        anchorOffset,
        focusNode,
        focusOffset
      } = winSelection;
      setSelection({
        anchorNode,
        anchorOffset,
        focusNode,
        focusOffset
      });
    }

    document.addEventListener('selectionchange', handleSelectionChange, {
      passive: true
    });
    return () => {
      document.removeEventListener('selectionchange', handleSelectionChange);
    };
  }, [textElement]); // Open popover when selection is within annotations

  (0, _react.useEffect)(() => {
    if (!selection) return;
    if (isClosingRef.current) return;
    const {
      anchorNode,
      focusNode
    } = selection; // Safari would close the popover by loosing range when button is focused.
    // If we are focused and currently tabbing to the action buttons, just return here.

    if (focused && isTabbing.current) {
      return;
    }

    if (annotationElement && annotationElement.contains(anchorNode) && anchorNode === focusNode) {
      var _window$getSelection;

      const range = (_window$getSelection = window.getSelection()) === null || _window$getSelection === void 0 ? void 0 : _window$getSelection.getRangeAt(0);
      const rect = range === null || range === void 0 ? void 0 : range.getBoundingClientRect();
      rangeRef.current = range || null;

      if (rect) {
        setCursorRect(rect);
      }

      setOpen(true);
    } else {
      setOpen(false);
      setCursorRect(null);
      rangeRef.current = null;
    }
  }, [focused, selection, annotationElement]);
  return /*#__PURE__*/_react.default.createElement(ToolbarPopover, {
    boundaryElement: scrollElement,
    constrainSize: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
      space: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold",
      size: 1
    }, title)), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      ref: editButtonRef,
      icon: _icons.EditIcon,
      mode: "bleed",
      onClick: onEdit,
      padding: 2,
      alt: "Edit annotation"
    }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.TrashIcon,
      mode: "bleed",
      padding: 2,
      onClick: onDelete,
      tone: "critical",
      alt: "Remove annotation"
    }))),
    fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS,
    open: open,
    placement: "top",
    portal: "editor",
    referenceElement: cursorElement,
    scheme: popoverScheme
  });
}
