// sanity/lib/cjs/form/inputs/PortableText/text/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TEXT_STYLE_PADDING = exports.TEXT_NUMBER_FORMATS = exports.TEXT_LEVELS = exports.TEXT_DECORATOR_TAGS = exports.TEXT_BULLET_MARKERS = void 0;
const TEXT_LEVELS = [1, 2, 3, 4, 5, 6, 7, 8, 9];
exports.TEXT_LEVELS = TEXT_LEVELS;
const TEXT_BULLET_MARKERS = ['●', '○', '■'];
exports.TEXT_BULLET_MARKERS = TEXT_BULLET_MARKERS;
const TEXT_NUMBER_FORMATS = ['number', 'lower-alpha', 'lower-roman'];
exports.TEXT_NUMBER_FORMATS = TEXT_NUMBER_FORMATS;
const TEXT_DECORATOR_TAGS = {
  em: 'em',
  'strike-through': 's',
  underline: 'u',
  strong: 'strong',
  code: 'code'
};
exports.TEXT_DECORATOR_TAGS = TEXT_DECORATOR_TAGS;
const TEXT_STYLE_PADDING = {
  h1: {
    paddingTop: 5,
    paddingBottom: 4
  },
  h2: {
    paddingTop: 4,
    paddingBottom: 4
  },
  h3: {
    paddingTop: 4,
    paddingBottom: 3
  },
  h4: {
    paddingTop: 4,
    paddingBottom: 3
  },
  h5: {
    paddingTop: 4,
    paddingBottom: 3
  },
  h6: {
    paddingTop: 4,
    paddingBottom: 2
  },
  normal: {
    paddingTop: 2,
    paddingBottom: 3
  },
  blockquote: {
    paddingTop: 2,
    paddingBottom: 3
  }
};
exports.TEXT_STYLE_PADDING = TEXT_STYLE_PADDING;
