// sanity/lib/cjs/form/inputs/PortableText/__workshop__/customSchema/blockActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderBlockActions = void 0;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _icons = require("@sanity/icons");

var _portableTextEditor = require("@sanity/portable-text-editor");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const BlockActions = /*#__PURE__*/(0, _react.memo)(function BlockActions(props) {
  const {
    block,
    insert
  } = props;
  const handleDuplicate = (0, _react.useCallback)(() => {
    const dupBlock = { ...block,
      _type: 'block',
      _key: (0, _portableTextEditor.keyGenerator)(),
      children: block.children || [],
      markDefs: block.markDefs || []
    };

    if (dupBlock.children) {
      dupBlock.children = dupBlock.children.map(c => ({ ...c,
        _key: (0, _portableTextEditor.keyGenerator)()
      }));
    }

    insert(dupBlock);
  }, [block, insert]);
  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "Duplicate")),
    placement: "right",
    portal: "default"
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    "aria-label": "Duplicate",
    fontSize: 1,
    icon: _icons.CopyIcon,
    onClick: handleDuplicate,
    padding: 2,
    mode: "bleed"
  }));
});

const renderBlockActions = _ref => {
  let {
    block,
    insert
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(BlockActions, {
    block: block,
    insert: insert
  });
};

exports.renderBlockActions = renderBlockActions;
