// sanity/lib/cjs/form/inputs/PortableText/__workshop__/customSchema/Story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Story;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _uiWorkshop = require("@sanity/ui-workshop");

var _values = require("./values");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Story() {
  const readOnly = (0, _uiWorkshop.useBoolean)('Read only', false);
  const withError = (0, _uiWorkshop.useBoolean)('With error', false);
  const withWarning = (0, _uiWorkshop.useBoolean)('With warning', false);
  const withCustomMarkers = (0, _uiWorkshop.useBoolean)('With custom markers', false);
  const selectedValue = (0, _uiWorkshop.useSelect)('Values', _values.valueOptions) || 'empty';
  const value = _values.values[selectedValue];
  return /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, {
    zOffset: 100
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "TODO")))));
}
