// @sanity/color-input/lib/cjs/index.js
var $k7rGe$sanity = require("sanity");
var $k7rGe$reactjsxruntime = require("react/jsx-runtime");
var $k7rGe$react = require("react");
var $k7rGe$lodash = require("lodash");
var $k7rGe$sanityui = require("@sanity/ui");
var $k7rGe$sanityicons = require("@sanity/icons");
var $k7rGe$reactcolorlibcomponentscommon = require("react-color/lib/components/common");
var $k7rGe$reactcolor = require("react-color");
var $k7rGe$styledcomponents = require("styled-components");
var $k7rGe$reactcolorlibhelperscolor = require("react-color/lib/helpers/color");

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

$parcel$export(module.exports, "colorInput", () => $244e63ca53592e4d$export$1d1c20ca26e20eb8);
$parcel$export(module.exports, "hslaColor", () => $f424886802e2a804$export$4e4b8c12b77623c0);
$parcel$export(module.exports, "rgbaColor", () => $6707e7c5509edbba$export$cc215b4b2c268dfd);
$parcel$export(module.exports, "color", () => $1ba04456571fe4ef$export$35e9368ef982300f);
$parcel$export(module.exports, "hsvaColor", () => $ebd6b32193e371d0$export$633851ea6f85c81b);
$parcel$export(module.exports, "ColorInput", () => $50859b401fad6cb5$export$5a1d7ca0a925d9c2);


const $f424886802e2a804$export$4e4b8c12b77623c0 = (0, $k7rGe$sanity.defineType)({
    title: "Hue Saturation Lightness",
    name: "hslaColor",
    type: "object",
    fields: [
        {
            name: "h",
            type: "number",
            title: "Hue"
        },
        {
            name: "s",
            type: "number",
            title: "Saturation"
        },
        {
            name: "l",
            type: "number",
            title: "Lightness"
        },
        {
            name: "a",
            type: "number",
            title: "Alpha"
        }, 
    ]
});



const $6707e7c5509edbba$export$cc215b4b2c268dfd = (0, $k7rGe$sanity.defineType)({
    title: "Red Green Blue (rgb)",
    name: "rgbaColor",
    type: "object",
    fields: [
        {
            name: "r",
            type: "number",
            title: "Red"
        },
        {
            name: "g",
            type: "number",
            title: "Green"
        },
        {
            name: "b",
            type: "number",
            title: "Blue"
        },
        {
            name: "a",
            type: "number",
            title: "Alpha"
        }, 
    ]
});






















const $bb8e678a4bd9bec1$export$5367735792921e7d = ({ onChange: onChange , rgb: rgb , hsl: hsl , hex: hex , disableAlpha: disableAlpha  })=>{
    const { sanity: sanity  } = (0, $k7rGe$sanityui.useTheme)();
    const inputStyles = (0, $k7rGe$react.useMemo)(()=>({
            input: {
                width: "80%",
                padding: "4px 10% 3px",
                border: "none",
                boxShadow: `inset 0 0 0 1px ${sanity.color.input.default.enabled.border}`,
                color: sanity.color.input.default.enabled.fg,
                backgroundColor: sanity.color.input.default.enabled.bg,
                fontSize: sanity.fonts.text.sizes[0].fontSize,
                textAlign: "center"
            },
            label: {
                display: "block",
                textAlign: "center",
                fontSize: sanity.fonts.label.sizes[0].fontSize,
                color: sanity.color.base.fg,
                paddingTop: "3px",
                paddingBottom: "4px",
                textTransform: "capitalize"
            }
        }), [
        sanity
    ]);
    const handleChange = (0, $k7rGe$react.useCallback)((data)=>{
        if ("hex" in data && data.hex && (0, $k7rGe$reactcolorlibhelperscolor.isValidHex)(data.hex)) onChange({
            hex: data.hex,
            source: "hex"
        });
        else if (rgb && ("r" in data && data.r || "g" in data && data.g || "b" in data && data.b)) onChange({
            r: Number(data.r) || rgb.r,
            g: Number(data.g) || rgb.g,
            b: Number(data.b) || rgb.b,
            a: rgb.a,
            source: "rgb"
        });
        else if (hsl && "a" in data && data.a) {
            let alpha = Number(data.a);
            if (alpha < 0) alpha = 0;
            else if (alpha > 100) alpha = 100;
            alpha /= 100;
            onChange({
                h: hsl.h,
                s: hsl.s,
                l: hsl.l,
                a: alpha,
                source: "hsl"
            });
        }
    }, [
        onChange,
        hsl,
        rgb
    ]);
    return /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Flex), {
        children: [
            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                flex: 2,
                marginRight: 1,
                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.EditableInput), {
                    style: inputStyles,
                    label: "hex",
                    value: hex?.replace("#", ""),
                    onChange: handleChange
                })
            }),
            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                flex: 1,
                marginRight: 1,
                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.EditableInput), {
                    style: inputStyles,
                    label: "r",
                    value: rgb?.r,
                    onChange: handleChange,
                    dragLabel: true,
                    dragMax: 255
                })
            }),
            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                flex: 1,
                marginRight: 1,
                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.EditableInput), {
                    style: inputStyles,
                    label: "g",
                    value: rgb?.g,
                    onChange: handleChange,
                    dragLabel: true,
                    dragMax: 255
                })
            }),
            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                flex: 1,
                marginRight: 1,
                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.EditableInput), {
                    style: inputStyles,
                    label: "b",
                    value: rgb?.b,
                    onChange: handleChange,
                    dragLabel: true,
                    dragMax: 255
                })
            }),
            !disableAlpha && /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                flex: 1,
                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.EditableInput), {
                    style: inputStyles,
                    label: "a",
                    value: Math.round((rgb?.a ?? 1) * 100),
                    onChange: handleChange,
                    dragLabel: true,
                    dragMax: 100
                })
            })
        ]
    });
};


const $9cf89bc516c02185$var$ColorBox = (0, ($parcel$interopDefault($k7rGe$styledcomponents)))((0, $k7rGe$sanityui.Box))`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;
const $9cf89bc516c02185$var$ReadOnlyContainer = (0, ($parcel$interopDefault($k7rGe$styledcomponents)))((0, $k7rGe$sanityui.Flex))`
  margin-top: 6rem;
  background-color: var(--card-bg-color);
  position: relative;
  width: 100%;
`;
const $9cf89bc516c02185$var$ColorPickerInner = (props)=>{
    const { width: width , color: { rgb: rgb , hex: hex , hsv: hsv , hsl: hsl  } , onChange: onChange , onUnset: onUnset , disableAlpha: disableAlpha , readOnly: readOnly ,  } = props;
    return /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)("div", {
        style: {
            width: width
        },
        children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Card), {
            padding: 1,
            border: true,
            radius: 1,
            children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Stack), {
                space: 2,
                children: [
                    !readOnly && /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$reactjsxruntime.Fragment), {
                        children: [
                            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Card), {
                                overflow: "hidden",
                                style: {
                                    position: "relative",
                                    height: "5em"
                                },
                                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.Saturation), {
                                    onChange: onChange,
                                    hsl: hsl,
                                    hsv: hsv
                                })
                            }),
                            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Card), {
                                shadow: 1,
                                radius: 3,
                                overflow: "hidden",
                                style: {
                                    position: "relative",
                                    height: "10px"
                                },
                                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.Hue), {
                                    hsl: hsl,
                                    onChange: !readOnly && onChange
                                })
                            }),
                            !disableAlpha && /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Card), {
                                shadow: 1,
                                radius: 3,
                                overflow: "hidden",
                                style: {
                                    position: "relative",
                                    height: "10px"
                                },
                                children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.Alpha), {
                                    rgb: rgb,
                                    hsl: hsl,
                                    onChange: onChange
                                })
                            })
                        ]
                    }),
                    /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Flex), {
                        children: [
                            /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Card), {
                                flex: 1,
                                radius: 2,
                                overflow: "hidden",
                                style: {
                                    position: "relative",
                                    minWidth: "4em"
                                },
                                children: [
                                    /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactcolorlibcomponentscommon.Checkboard), {}),
                                    /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)($9cf89bc516c02185$var$ColorBox, {
                                        style: {
                                            backgroundColor: `rgba(${rgb?.r},${rgb?.g},${rgb?.b},${rgb?.a})`
                                        }
                                    }),
                                    readOnly && /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)($9cf89bc516c02185$var$ReadOnlyContainer, {
                                        padding: 2,
                                        paddingBottom: 1,
                                        sizing: "border",
                                        justify: "space-between",
                                        children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Stack), {
                                            space: 3,
                                            marginTop: 1,
                                            children: [
                                                /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Text), {
                                                    size: 3,
                                                    weight: "bold",
                                                    children: hex
                                                }),
                                                /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Inline), {
                                                    space: 3,
                                                    children: [
                                                        /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Text), {
                                                            size: 1,
                                                            children: [
                                                                /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)("strong", {
                                                                    children: "RGB: "
                                                                }),
                                                                rgb?.r,
                                                                " ",
                                                                rgb?.g,
                                                                " ",
                                                                rgb?.b
                                                            ]
                                                        }),
                                                        /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Text), {
                                                            size: 1,
                                                            children: [
                                                                /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)("strong", {
                                                                    children: "HSL: "
                                                                }),
                                                                " ",
                                                                Math.round(hsl?.h ?? 0),
                                                                " ",
                                                                Math.round(hsl?.s ?? 0),
                                                                "%",
                                                                " ",
                                                                Math.round(hsl?.l ?? 0)
                                                            ]
                                                        })
                                                    ]
                                                })
                                            ]
                                        })
                                    })
                                ]
                            }),
                            !readOnly && /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsxs)((0, $k7rGe$sanityui.Flex), {
                                align: "flex-start",
                                marginLeft: 2,
                                children: [
                                    /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                                        style: {
                                            width: 200
                                        },
                                        children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $bb8e678a4bd9bec1$export$5367735792921e7d), {
                                            rgb: rgb,
                                            hsl: hsl,
                                            hex: hex,
                                            onChange: onChange,
                                            disableAlpha: disableAlpha
                                        })
                                    }),
                                    /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Box), {
                                        marginLeft: 2,
                                        children: /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Button), {
                                            onClick: onUnset,
                                            title: "Delete color",
                                            icon: (0, $k7rGe$sanityicons.TrashIcon),
                                            tone: "critical"
                                        })
                                    })
                                ]
                            })
                        ]
                    })
                ]
            })
        })
    });
};
const $9cf89bc516c02185$export$9feb1bc2e5f1ccb3 = (0, $k7rGe$reactcolor.CustomPicker)($9cf89bc516c02185$var$ColorPickerInner);


const $50859b401fad6cb5$var$DEFAULT_COLOR = {
    hex: "#24a3e3",
    hsl: {
        h: 200,
        s: 0.7732,
        l: 0.5156,
        a: 1
    },
    hsv: {
        h: 200,
        s: 0.8414,
        v: 0.8901,
        a: 1
    },
    rgb: {
        r: 46,
        g: 163,
        b: 227,
        a: 1
    },
    source: "hex"
};
function $50859b401fad6cb5$export$5a1d7ca0a925d9c2(props) {
    const { onChange: onChange , schemaType: type , readOnly: readOnly , value: value  } = props;
    const focusRef = (0, $k7rGe$react.useRef)(null);
    // use local state so we can have instant ui updates while debouncing patch emits
    const [color, setColor] = (0, $k7rGe$react.useState)(value);
    (0, $k7rGe$react.useEffect)(()=>setColor(value), [
        value
    ]);
    const emitSetColor = (0, $k7rGe$react.useCallback)((nextColor)=>{
        const fieldPatches = type.fields.filter((field)=>field.name in nextColor).map((field)=>{
            const nextFieldValue = nextColor[field.name];
            const isObject = field.type.jsonType === "object";
            return (0, $k7rGe$sanity.set)(isObject ? Object.assign({
                _type: field.type.name
            }, nextFieldValue) : nextFieldValue, [
                field.name
            ]);
        });
        onChange([
            (0, $k7rGe$sanity.setIfMissing)({
                _type: type.name
            }),
            (0, $k7rGe$sanity.set)(type.name, [
                "_type"
            ]),
            (0, $k7rGe$sanity.set)(nextColor.rgb?.a, [
                "alpha"
            ]),
            ...fieldPatches, 
        ]);
    }, [
        onChange,
        type
    ]);
    // The color picker emits onChange events continuously while the user is sliding the
    // hue/saturation/alpha selectors. This debounces the event to avoid excessive patches
    const debouncedColorChange = (0, $k7rGe$react.useMemo)(()=>(0, $k7rGe$lodash.debounce)(emitSetColor, 100), [
        emitSetColor
    ]);
    const handleColorChange = (0, $k7rGe$react.useCallback)((nextColor)=>{
        setColor(nextColor);
        debouncedColorChange(nextColor);
    }, [
        debouncedColorChange,
        setColor
    ]);
    const handleCreateColor = (0, $k7rGe$react.useCallback)(()=>{
        setColor($50859b401fad6cb5$var$DEFAULT_COLOR);
        emitSetColor($50859b401fad6cb5$var$DEFAULT_COLOR);
    }, [
        emitSetColor
    ]);
    const handleUnset = (0, $k7rGe$react.useCallback)(()=>{
        setColor(undefined);
        onChange((0, $k7rGe$sanity.unset)());
    }, [
        onChange
    ]);
    return /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$reactjsxruntime.Fragment), {
        children: value ? /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $9cf89bc516c02185$export$9feb1bc2e5f1ccb3), {
            /*            ref={this.focusRef}*/ color: color,
            onChange: handleColorChange,
            readOnly: readOnly || typeof type.readOnly === "boolean" && type.readOnly,
            disableAlpha: !!type.options?.disableAlpha,
            onUnset: handleUnset
        }) : /*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)((0, $k7rGe$sanityui.Button), {
            icon: (0, $k7rGe$sanityicons.AddIcon),
            mode: "ghost",
            text: "Create color",
            ref: focusRef,
            disabled: Boolean(readOnly),
            onClick: handleCreateColor
        })
    });
}



const $1ba04456571fe4ef$var$round = (val = 1)=>Math.round(val * 100);
const $1ba04456571fe4ef$var$colorTypeName = "color";
const $1ba04456571fe4ef$export$35e9368ef982300f = (0, $k7rGe$sanity.defineType)({
    name: $1ba04456571fe4ef$var$colorTypeName,
    type: "object",
    title: "Color",
    components: {
        input: (0, $50859b401fad6cb5$export$5a1d7ca0a925d9c2)
    },
    fields: [
        {
            title: "Hex",
            name: "hex",
            type: "string"
        },
        {
            title: "Alpha",
            name: "alpha",
            type: "number"
        },
        {
            title: "Hue Saturation Lightness",
            name: "hsl",
            type: "hslaColor"
        },
        {
            title: "Hue Saturation Value",
            name: "hsv",
            type: "hsvaColor"
        },
        {
            title: "Red Green Blue (rgb)",
            name: "rgb",
            type: "rgbaColor"
        }, 
    ],
    preview: {
        select: {
            title: "hex",
            alpha: "alpha",
            hex: "hex",
            hsl: "hsl"
        },
        prepare ({ title: title , hex: hex , hsl: hsl , alpha: alpha  }) {
            let subtitle = hex || "No color set";
            if (hsl) subtitle = `H:${$1ba04456571fe4ef$var$round(hsl.h)} S:${$1ba04456571fe4ef$var$round(hsl.s)} L:${$1ba04456571fe4ef$var$round(hsl.l)} A:${$1ba04456571fe4ef$var$round(alpha)}`;
            return {
                title: title,
                subtitle: subtitle,
                media: ()=>/*#__PURE__*/ (0, $k7rGe$reactjsxruntime.jsx)("div", {
                        style: {
                            backgroundColor: hex ?? "#000",
                            opacity: alpha ?? 1,
                            position: "absolute",
                            height: "100%",
                            width: "100%",
                            top: "0",
                            left: "0"
                        }
                    })
            };
        }
    }
});



const $ebd6b32193e371d0$export$633851ea6f85c81b = (0, $k7rGe$sanity.defineType)({
    title: "Hue Saturation Value",
    name: "hsvaColor",
    type: "object",
    fields: [
        {
            name: "h",
            type: "number",
            title: "Hue"
        },
        {
            name: "s",
            type: "number",
            title: "Saturation"
        },
        {
            name: "v",
            type: "number",
            title: "Value"
        },
        {
            name: "a",
            type: "number",
            title: "Alpha"
        }, 
    ]
});



const $244e63ca53592e4d$export$1d1c20ca26e20eb8 = (0, $k7rGe$sanity.createPlugin)({
    name: "@sanity/color-input",
    schema: {
        types: [
            (0, $f424886802e2a804$export$4e4b8c12b77623c0),
            (0, $ebd6b32193e371d0$export$633851ea6f85c81b),
            (0, $6707e7c5509edbba$export$cc215b4b2c268dfd),
            (0, $1ba04456571fe4ef$export$35e9368ef982300f)
        ]
    }
});
