// sanity/lib/dts/test/testUtils/getMockWorkspaceFromConfig.js
import { createWorkspaceFromConfig, } from '../../src/config';
import { createMockSanityClient } from '../mocks/mockSanityClient';
const defaultMockUser = {
    id: 'doug',
    name: 'Doug',
    email: 'doug@sanity.io',
    role: 'admin',
    roles: [{ name: 'administrator', title: 'Administrator' }],
};
const defaultMockSchema = {
    name: 'mock',
    types: [
        {
            name: 'author',
            title: 'Author',
            type: 'document',
            fields: [
                {
                    name: 'name',
                    title: 'Name',
                    type: 'string',
                },
            ],
        },
    ],
};
const defaultMockConfig = {
    projectId: 'mock-project-id',
    dataset: 'mock-data-set',
    schema: defaultMockSchema,
};
export function getMockWorkspace({ config: userConfig, currentUser = defaultMockUser, client = createMockSanityClient(), } = {}) {
    return createWorkspaceFromConfig({ ...defaultMockConfig, ...userConfig, currentUser, client });
}
export async function getMockSource(options = {}) {
    const workspace = await getMockWorkspace(options);
    return workspace.unstable_sources[0];
}
