// sanity/lib/esm/form/inputs/PortableText/__workshop__/customSchema/Story.js
import { Card, Container, Flex, LayerProvider } from '@sanity/ui';
import React from 'react';
import { useBoolean, useSelect } from '@sanity/ui-workshop';
import { values, valueOptions } from './values';
export default function Story() {
  const readOnly = useBoolean('Read only', false);
  const withError = useBoolean('With error', false);
  const withWarning = useBoolean('With warning', false);
  const withCustomMarkers = useBoolean('With custom markers', false);
  const selectedValue = useSelect('Values', valueOptions) || 'empty';
  const value = values[selectedValue];
  return /*#__PURE__*/React.createElement(LayerProvider, {
    zOffset: 100
  }, /*#__PURE__*/React.createElement(Card, {
    height: "fill",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 1
  }, /*#__PURE__*/React.createElement(React.Fragment, null, "TODO")))));
}
