// sanity/lib/esm/desk/components/pane/__workshop__/SplitPanesStory/DocumentPane/DocumentActions.js
import { ChevronDownIcon, PublishIcon } from '@sanity/icons';
import { Button, Flex, Menu, MenuButton, MenuItem, Stack, useToast } from '@sanity/ui';
import React, { useCallback } from 'react';
const MENU_BUTTON_POPOVER_PROPS = {
  constrainSize: true,
  portal: true
};
export function DocumentActions() {
  const {
    push: pushToast
  } = useToast();
  const publish = useCallback(() => pushToast({
    status: 'success',
    title: 'Successfully published'
  }), [pushToast]);
  return /*#__PURE__*/React.createElement(Flex, {
    gap: 1
  }, /*#__PURE__*/React.createElement(Stack, {
    flex: 1
  }, /*#__PURE__*/React.createElement(Button, {
    icon: PublishIcon,
    onClick: publish,
    text: "Publish",
    tone: "positive"
  })), /*#__PURE__*/React.createElement(MenuButton, {
    button: /*#__PURE__*/React.createElement(Button, {
      icon: ChevronDownIcon,
      mode: "ghost"
    }),
    id: "actions",
    menu: /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(MenuItem, {
      text: "Discard changes"
    })),
    popover: MENU_BUTTON_POPOVER_PROPS
  }));
}
