// sanity/lib/esm/form/inputs/TagsArrayInput.js
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef } from 'react';
import { set, unset } from '../patch';
import { TagInput } from '../components/tagInput';
import { ChangeIndicator } from '../../components/changeIndicators';
export const TagsArrayInput = /*#__PURE__*/forwardRef(function TagsArrayInput(props, ref) {
  const {
    id,
    onChange,
    onFocus,
    readOnly,
    value = [],
    path,
    focused,
    changed
  } = props;
  const tagInputValue = useMemo(() => value === null || value === void 0 ? void 0 : value.map(v => ({
    value: v
  })), [value]);
  const inputRef = useRef(null);
  const handleChange = useCallback(nextValue => {
    onChange(nextValue.length === 0 ? unset() : set(nextValue.map(v => v.value)));
  }, [onChange]);
  useImperativeHandle(ref, () => ({
    focus: () => {
      var _inputRef$current;

      return (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  }));
  return /*#__PURE__*/React.createElement(ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/React.createElement(TagInput, {
    id: id,
    onChange: handleChange,
    onFocus: onFocus,
    readOnly: readOnly,
    ref: inputRef,
    value: tagInputValue
  }));
});
