// sanity/lib/esm/form/inputs/TelephoneInput.js
import React from 'react';
import { TextInput } from '@sanity/ui';
import { set, unset } from '../patch';
import { ChangeIndicator } from '../../components/changeIndicators';
export const TelephoneInput = /*#__PURE__*/React.forwardRef(function TelephoneInput(props, forwardedRef) {
  const {
    changed,
    focused,
    id,
    onBlur,
    onChange,
    onFocus,
    path,
    readOnly,
    schemaType,
    validationError,
    value
  } = props;
  const handleChange = React.useCallback(event => {
    const inputValue = event.currentTarget.value;
    onChange(inputValue ? set(inputValue) : unset());
  }, [onChange]);
  return /*#__PURE__*/React.createElement(ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/React.createElement(TextInput, {
    type: "tel",
    inputMode: "tel",
    id: id,
    customValidity: validationError,
    value: value || '',
    readOnly: Boolean(readOnly),
    placeholder: schemaType.placeholder,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    ref: forwardedRef
  }));
});
