// sanity/lib/esm/form/inputs/UrlInput.js
import { TextInput } from '@sanity/ui';
import React from 'react';
import { set, unset } from '../patch';
import { getValidationRule } from '../utils/getValidationRule';
import { ChangeIndicator } from '../../components/changeIndicators';
// @todo Rename to `URLInput`?
export const UrlInput = /*#__PURE__*/React.forwardRef(function UrlInput(props, forwardedRef) {
  var _uriRule$constraint, _uriRule$constraint$o;

  const {
    changed,
    focused,
    id,
    onBlur,
    onChange,
    onFocus,
    path,
    readOnly,
    schemaType,
    validationError,
    value
  } = props;
  const handleChange = React.useCallback(event => {
    const inputValue = event.currentTarget.value;
    onChange(inputValue ? set(inputValue) : unset());
  }, [onChange]);
  const uriRule = getValidationRule(schemaType, 'uri');
  const inputType = uriRule !== null && uriRule !== void 0 && (_uriRule$constraint = uriRule.constraint) !== null && _uriRule$constraint !== void 0 && (_uriRule$constraint$o = _uriRule$constraint.options) !== null && _uriRule$constraint$o !== void 0 && _uriRule$constraint$o.allowRelative ? 'text' : 'url';
  return /*#__PURE__*/React.createElement(ChangeIndicator, {
    path: path,
    isChanged: changed,
    hasFocus: !!focused
  }, /*#__PURE__*/React.createElement(TextInput, {
    type: inputType,
    inputMode: "url",
    id: id,
    customValidity: validationError,
    value: value || '',
    readOnly: Boolean(readOnly),
    placeholder: schemaType.placeholder,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    ref: forwardedRef
  }));
});
