// sanity/lib/esm/desk/components/confirmDeleteDialog/index.js
import React, { useCallback, useState } from 'react';
import { Box, Flex, Button, Dialog, Text, ErrorBoundary } from '@sanity/ui';
import { useId } from '@reach/auto-id';
import { ConfirmDeleteDialog } from './ConfirmDeleteDialog';

function ConfirmDeleteDialogContainer(props) {
  const id = useId();
  const [error, setError] = useState(null);
  const handleRetry = useCallback(() => setError(null), []);
  return error ? /*#__PURE__*/React.createElement(Dialog, {
    id: "dialog-error-".concat(id),
    "data-testid": "confirm-delete-error-dialog",
    header: "Error",
    footer: /*#__PURE__*/React.createElement(Flex, {
      paddingX: 4,
      paddingY: 3,
      direction: "column"
    }, /*#__PURE__*/React.createElement(Button, {
      mode: "ghost",
      text: "Retry",
      onClick: handleRetry
    })),
    onClose: props.onCancel
  }, /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Text, null, "An error occurred while loading referencing documents."))) : /*#__PURE__*/React.createElement(ErrorBoundary, {
    onCatch: setError
  }, /*#__PURE__*/React.createElement(ConfirmDeleteDialog, props));
}

export { ConfirmDeleteDialogContainer as ConfirmDeleteDialog };
