// sanity/lib/cjs/desk/components/confirmDeleteDialog/ReferencePreviewLink.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReferencePreviewLink = ReferencePreviewLink;

var _react = _interopRequireWildcard(require("react"));

var _util = require("../../../util");

var _PreviewCard = require("../../../components/PreviewCard");

var _datastores = require("../../../datastores");

var _PaneItemPreview = require("../paneItem/PaneItemPreview");

var _paneRouter = require("../paneRouter");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const EMPTY_ARRAY = [];

function ReferencePreviewLink(props) {
  const {
    onClick,
    type,
    value
  } = props;
  const publishedId = (0, _util.getPublishedId)(value === null || value === void 0 ? void 0 : value._id);
  const documentPresence = (0, _datastores.useDocumentPresence)(publishedId);
  const documentPreviewStore = (0, _datastores.useDocumentPreviewStore)();
  const {
    ReferenceChildLink
  } = (0, _paneRouter.usePaneRouter)();
  const Link = (0, _react.useCallback)(linkProps => /*#__PURE__*/_react.default.createElement(ReferenceChildLink, _extends({
    documentId: value === null || value === void 0 ? void 0 : value._id,
    documentType: type === null || type === void 0 ? void 0 : type.name,
    parentRefPath: EMPTY_ARRAY
  }, linkProps)), [ReferenceChildLink, type === null || type === void 0 ? void 0 : type.name, value === null || value === void 0 ? void 0 : value._id]);
  return /*#__PURE__*/_react.default.createElement(_PreviewCard.PreviewCard, {
    __unstable_focusRing: true,
    as: Link,
    "data-as": "a",
    onClick: onClick,
    padding: 2,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_PaneItemPreview.PaneItemPreview, {
    documentPreviewStore: documentPreviewStore,
    icon: type === null || type === void 0 ? void 0 : type.icon,
    layout: "default",
    presence: (documentPresence === null || documentPresence === void 0 ? void 0 : documentPresence.length) > 0 ? documentPresence : EMPTY_ARRAY,
    schemaType: type,
    value: value
  }));
}
