// sanity/lib/esm/desk/panes/document/statusBar/sparkline/DocumentBadges.js
import { Badge, Box, Inline, Text, Tooltip } from '@sanity/ui';
import React from 'react';
import { RenderBadgeCollectionState } from '../../../../badges';
import { useDocumentPane } from '../../useDocumentPane';
const BADGE_TONES = {
  primary: 'primary',
  success: 'positive',
  warning: 'caution',
  danger: 'critical'
};

function DocumentBadgesInner(_ref) {
  let {
    states
  } = _ref;
  return /*#__PURE__*/React.createElement(Inline, {
    space: 1
  }, states.map((badge, index) => /*#__PURE__*/React.createElement(Tooltip, {
    content: badge.title && /*#__PURE__*/React.createElement(Box, {
      padding: 2
    }, /*#__PURE__*/React.createElement(Text, {
      size: 1
    }, badge.title)),
    disabled: !badge.title,
    key: String(index),
    placement: "top",
    portal: true
  }, /*#__PURE__*/React.createElement(Badge, {
    fontSize: 1,
    mode: "outline",
    paddingX: 2,
    paddingY: 1,
    radius: 4,
    tone: badge.color ? BADGE_TONES[badge.color] : undefined,
    style: {
      whiteSpace: 'nowrap'
    }
  }, badge.label))));
}

export function DocumentBadges() {
  const {
    badges,
    editState
  } = useDocumentPane();
  if (!editState || !badges) return null;
  return /*#__PURE__*/React.createElement(RenderBadgeCollectionState, {
    badges: badges,
    badgeProps: editState
  }, _ref2 => {
    let {
      states
    } = _ref2;
    return /*#__PURE__*/React.createElement(DocumentBadgesInner, {
      states: states
    });
  });
}
