// sanity/lib/cjs/form/members/MemberField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MemberField = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ArrayOfObjectsField = require("./fields/ArrayOfObjectsField");

var _PrimitiveField = require("./fields/PrimitiveField");

var _ObjectField = require("./fields/ObjectField");

var _ArrayOfPrimitivesField = require("./fields/ArrayOfPrimitivesField");

var _asserters = require("./fields/asserters");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const MemberField = /*#__PURE__*/(0, _react.memo)(function MemberField(props) {
  const {
    member,
    renderField,
    renderInput,
    renderItem,
    renderPreview
  } = props;

  if ((0, _asserters.isMemberObject)(member)) {
    // this field is of an object type
    return /*#__PURE__*/_react.default.createElement(_ObjectField.ObjectField, {
      member: member,
      renderField: renderField,
      renderInput: renderInput,
      renderItem: renderItem,
      renderPreview: renderPreview
    });
  }

  if ((0, _asserters.isMemberArrayOfPrimitives)(member)) {
    return /*#__PURE__*/_react.default.createElement(_ArrayOfPrimitivesField.ArrayOfPrimitivesField, {
      member: member,
      renderField: renderField,
      renderInput: renderInput // todo: these have different signatures, so renderItem for a primitive input should not be the same as renderItem in array of object inputs
      ,
      renderItem: renderItem,
      renderPreview: renderPreview
    });
  }

  if ((0, _asserters.isMemberArrayOfObjects)(member)) {
    return /*#__PURE__*/_react.default.createElement(_ArrayOfObjectsField.ArrayOfObjectsField, {
      member: member,
      renderField: renderField,
      renderInput: renderInput,
      renderItem: renderItem,
      renderPreview: renderPreview
    });
  }

  return /*#__PURE__*/_react.default.createElement(_PrimitiveField.PrimitiveField, {
    member: member,
    renderField: renderField,
    renderInput: renderInput
  });
});
exports.MemberField = MemberField;
