// sanity/lib/cjs/form/members/MemberFieldset.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MemberFieldSet = void 0;

var _react = _interopRequireWildcard(require("react"));

var _formField = require("../components/formField");

var _FormCallbacks = require("../studio/contexts/FormCallbacks");

var _MemberField = require("./MemberField");

var _MemberFieldError = require("./MemberFieldError");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const MemberFieldSet = /*#__PURE__*/(0, _react.memo)(function MemberFieldSet(props) {
  var _member$fieldSet;

  const {
    member,
    renderField,
    renderInput,
    renderItem,
    renderPreview
  } = props;
  const {
    onSetFieldSetCollapsed
  } = (0, _FormCallbacks.useFormCallbacks)();
  const handleCollapse = (0, _react.useCallback)(() => {
    onSetFieldSetCollapsed(member.fieldSet.path, true);
  }, [member.fieldSet.path, onSetFieldSetCollapsed]);
  const handleExpand = (0, _react.useCallback)(() => {
    onSetFieldSetCollapsed(member.fieldSet.path, false);
  }, [member.fieldSet.path, onSetFieldSetCollapsed]);
  return /*#__PURE__*/_react.default.createElement(_formField.FormFieldSet, {
    title: member.fieldSet.title,
    description: member.fieldSet.description,
    level: member.fieldSet.level,
    collapsible: member.fieldSet.collapsible,
    collapsed: member.fieldSet.collapsed,
    onCollapse: handleCollapse,
    onExpand: handleExpand,
    columns: member === null || member === void 0 ? void 0 : (_member$fieldSet = member.fieldSet) === null || _member$fieldSet === void 0 ? void 0 : _member$fieldSet.columns,
    "data-testid": "fieldset-".concat(member.fieldSet.name)
  }, member.fieldSet.members.map(fieldsetMember => {
    if (fieldsetMember.kind === 'error') {
      return /*#__PURE__*/_react.default.createElement(_MemberFieldError.MemberFieldError, {
        key: member.key,
        member: fieldsetMember
      });
    }

    return /*#__PURE__*/_react.default.createElement(_MemberField.MemberField, {
      member: fieldsetMember,
      renderField: renderField,
      renderInput: renderInput,
      renderItem: renderItem,
      renderPreview: renderPreview,
      key: fieldsetMember.key
    });
  }));
});
exports.MemberFieldSet = MemberFieldSet;
