// sanity/lib/cjs/form/members/ObjectMembers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ObjectMembers = ObjectMembers;

var _react = _interopRequireDefault(require("react"));

var _MemberField = require("./MemberField");

var _MemberFieldError = require("./MemberFieldError");

var _MemberFieldset = require("./MemberFieldset");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ObjectMembers(props) {
  const {
    members,
    renderInput,
    renderField,
    renderItem,
    renderPreview
  } = props;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, members.map(member => {
    if (member.kind === 'field') {
      return /*#__PURE__*/_react.default.createElement(_MemberField.MemberField, {
        key: member.key,
        member: member,
        renderInput: renderInput,
        renderField: renderField,
        renderItem: renderItem,
        renderPreview: renderPreview
      });
    }

    if (member.kind === 'error') {
      return /*#__PURE__*/_react.default.createElement(_MemberFieldError.MemberFieldError, {
        key: member.key,
        member: member
      });
    }

    if (member.kind === 'fieldSet') {
      return /*#__PURE__*/_react.default.createElement(_MemberFieldset.MemberFieldSet, {
        key: member.key,
        member: member,
        renderInput: renderInput,
        renderField: renderField,
        renderItem: renderItem,
        renderPreview: renderPreview
      });
    } //@ts-expect-error The branching above should cover all possible cases


    console.warn(new Error("Unhandled member kind ".concat(member.kind)));
    return null;
  }));
}
