// sanity/lib/esm/form/components/tagInput/tagInput.js
var _templateObject;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { CloseIcon } from '@sanity/icons';
import { Box, Button, Card, Flex, isHTMLElement, rem, Text, useForwardedRef } from '@sanity/ui';
import React, { forwardRef, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import styled, { css } from 'styled-components';
import { focusRingBorderStyle, focusRingStyle } from './styles';
const Root = styled(Box)(props => {
  const {
    theme
  } = props;
  const {
    focusRing,
    input,
    radius
  } = theme.sanity;
  const color = theme.sanity.color.input;
  const space = rem(theme.sanity.space[1]);
  return {
    position: 'relative',
    borderRadius: "".concat(radius[1], "px"),
    color: color.default.enabled.fg,
    backgroundColor: color.default.enabled.bg,
    boxShadow: focusRingBorderStyle({
      color: color.default.enabled.border,
      width: input.border.width
    }),
    '& > .content': {
      position: 'relative',
      lineHeight: 0,
      margin: "-".concat(space, " 0 0 -").concat(space)
    },
    '& > .content > div': {
      display: 'inline-block',
      verticalAlign: 'top',
      padding: "".concat(space, " 0 0 ").concat(space)
    },
    // enabled
    '&:not([data-read-only])': {
      cursor: 'text'
    },
    // hovered
    '@media(hover:hover)': {
      '&:not([data-disabled]):not([data-read-only]):hover': {
        borderColor: color.default.hovered.border
      }
    },
    // focused
    '&:not([data-disabled]):not([data-read-only])[data-focused]': {
      boxShadow: focusRingStyle({
        border: {
          color: color.default.enabled.border,
          width: input.border.width
        },
        focusRing
      })
    },
    // disabled
    '*:disabled + &': {
      color: color.default.disabled.fg,
      backgroundColor: color.default.disabled.bg,
      boxShadow: focusRingBorderStyle({
        color: color.default.disabled.border,
        width: input.border.width
      })
    }
  };
});
const Input = styled.input(props => {
  const {
    theme
  } = props;
  const font = theme.sanity.fonts.text;
  const color = theme.sanity.color.input;
  const p = theme.sanity.space[2];
  const size = theme.sanity.fonts.text.sizes[2];
  return {
    appearance: 'none',
    background: 'none',
    border: 0,
    borderRadius: 0,
    outline: 'none',
    fontSize: rem(size.fontSize),
    lineHeight: size.lineHeight / size.fontSize,
    fontFamily: font.family,
    fontWeight: font.weights.regular,
    margin: 0,
    display: 'block',
    minWidth: '1px',
    maxWidth: '100%',
    boxSizing: 'border-box',
    paddingTop: rem(p - size.ascenderHeight),
    paddingRight: rem(p),
    paddingBottom: rem(p - size.descenderHeight),
    paddingLeft: rem(p),
    // enabled
    '&:not(:invalid):not(:disabled)': {
      color: color.default.enabled.fg
    },
    // disabled
    '&:not(:invalid):disabled': {
      color: color.default.disabled.fg
    }
  };
});
const Placeholder = styled(Box)(props => {
  const {
    theme
  } = props;
  const color = theme.sanity.color.input;
  return css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    position: absolute;\n    top: 0;\n    left: 0;\n    right: 0;\n    pointer-events: none;\n    --card-fg-color: ", ";\n  "])), color.default.enabled.placeholder);
});
export const TagInput = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    disabled,
    onChange,
    onFocus,
    placeholder: placeholderProp,
    readOnly,
    value = [],
    ...restProps
  } = props;
  const placeholder = useMemo(() => {
    if (placeholderProp) return placeholderProp;

    if (typeof window !== 'undefined' && 'ontouchstart' in window) {
      return 'Enter tag…';
    }

    return 'Enter tag and press ENTER…';
  }, [placeholderProp]);
  const [inputValue, setInputValue] = useState('');
  const enabled = !disabled && !readOnly;
  const [focused, setFocused] = useState(false);
  const forwardedRef = useForwardedRef(ref);
  const rootRef = useRef(null);
  const handleRootPointerDown = useCallback(event => {
    const isTagElement = isHTMLElement(event.target) && event.target.closest('[data-ui="Tag"]');
    if (isTagElement) return;
    const inputElement = forwardedRef.current;

    if (inputElement) {
      setTimeout(() => inputElement.focus(), 0);
    }
  }, [forwardedRef]);
  const handleInputBlur = useCallback(() => {
    setFocused(false);
  }, []);
  const handleInputChange = useCallback(event => {
    setInputValue(event.currentTarget.value);
  }, []);
  const handleInputFocus = useCallback(event => {
    setFocused(true);
    if (onFocus) onFocus(event);
  }, [onFocus]);
  const handleInputKeyDown = useCallback(event => {
    if (event.key === 'Enter') {
      event.preventDefault();
      event.stopPropagation();

      if (onChange && inputValue) {
        const newValue = value.concat([{
          value: inputValue
        }]);
        setInputValue('');
        if (onChange) onChange(newValue);
      }
    }
  }, [inputValue, onChange, value]);
  const handleTagRemove = useCallback(index => {
    if (!onChange) return;
    const newValue = value.slice(0);
    newValue.splice(index, 1);
    onChange(newValue);
  }, [onChange, value]);
  useEffect(() => {
    const inputElement = forwardedRef.current;

    if (inputElement) {
      inputElement.style.width = '0';
      inputElement.style.width = "".concat(inputElement.scrollWidth, "px");
    }
  }, [forwardedRef, inputValue]);
  return /*#__PURE__*/React.createElement(Root, {
    "data-disabled": disabled ? '' : undefined,
    "data-focused": focused ? '' : undefined,
    "data-read-only": readOnly ? '' : undefined,
    "data-ui": "TagInput",
    onPointerDown: handleRootPointerDown,
    padding: 1,
    ref: rootRef
  }, enabled && /*#__PURE__*/React.createElement(Placeholder, {
    hidden: Boolean(inputValue || value.length),
    padding: 3
  }, /*#__PURE__*/React.createElement(Text, {
    textOverflow: "ellipsis"
  }, placeholder)), /*#__PURE__*/React.createElement("div", {
    className: "content"
  }, value.map((tag, tagIndex) =>
  /*#__PURE__*/
  // eslint-disable-next-line react/no-array-index-key
  React.createElement("div", {
    key: "tag-".concat(tagIndex)
  }, /*#__PURE__*/React.createElement(Tag, {
    enabled: enabled,
    index: tagIndex,
    muted: disabled,
    onRemove: handleTagRemove,
    tag: tag
  }))), /*#__PURE__*/React.createElement("div", {
    key: "tag-input"
  }, /*#__PURE__*/React.createElement(Input, _extends({}, restProps, {
    disabled: !enabled,
    onBlur: handleInputBlur,
    onChange: handleInputChange,
    onFocus: handleInputFocus,
    onKeyDown: handleInputKeyDown,
    ref: forwardedRef,
    type: "text",
    value: inputValue
  })))));
});
TagInput.displayName = 'TagInput';

function Tag(props) {
  const {
    enabled,
    index,
    muted,
    onRemove,
    tag
  } = props;
  const handleRemoveClick = useCallback(() => {
    onRemove(index);
  }, [index, onRemove]);
  return /*#__PURE__*/React.createElement(Card, {
    "data-ui": "Tag",
    padding: 1,
    radius: 2,
    tone: "transparent"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center"
  }, /*#__PURE__*/React.createElement(Box, {
    flex: 1,
    padding: 1
  }, /*#__PURE__*/React.createElement(Text, {
    muted: muted
  }, tag.value)), enabled && /*#__PURE__*/React.createElement(Box, {
    marginLeft: 1
  }, /*#__PURE__*/React.createElement(Button, {
    icon: CloseIcon,
    mode: "bleed",
    onClick: handleRemoveClick,
    padding: 1
  }))));
}
