// sanity/lib/cjs/desk/panes/userComponent/UserComponentPaneHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserComponentPaneHeader = UserComponentPaneHeader;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _components = require("../../components");

var _useDeskTool = require("../../useDeskTool");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function UserComponentPaneHeader(props) {
  const {
    actionHandlers,
    index,
    menuItems,
    menuItemGroups,
    title
  } = props;
  const {
    features
  } = (0, _useDeskTool.useDeskTool)();

  if (!(menuItems !== null && menuItems !== void 0 && menuItems.length) && !title) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_components.PaneHeader, {
    actions: /*#__PURE__*/_react.default.createElement(_components.PaneHeaderActions, {
      menuItems: menuItems,
      menuItemGroups: menuItemGroups,
      actionHandlers: actionHandlers
    }),
    backButton: features.backButton && index > 0 && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      as: _components.BackLink,
      "data-as": "a",
      icon: _icons.ArrowLeftIcon,
      mode: "bleed"
    }),
    title: title
  });
}
