// sanity/lib/cjs/form/members/fields/PrimitiveField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PrimitiveField = PrimitiveField;

var _react = _interopRequireWildcard(require("react"));

var _patch = require("../../patch");

var _useDidUpdate = require("../../hooks/useDidUpdate");

var _FormCallbacks = require("../../studio/contexts/FormCallbacks");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Responsible for creating inputProps and fieldProps to pass to ´renderInput´ and ´renderField´ for a primitive field/input
 * @param props - Component props
 */
function PrimitiveField(props) {
  const {
    member,
    renderInput,
    renderField
  } = props;
  const focusRef = (0, _react.useRef)();
  const {
    onPathBlur,
    onPathFocus,
    onChange
  } = (0, _FormCallbacks.useFormCallbacks)();
  (0, _useDidUpdate.useDidUpdate)(member.field.focused, (hadFocus, hasFocus) => {
    if (!hadFocus && hasFocus) {
      var _focusRef$current;

      (_focusRef$current = focusRef.current) === null || _focusRef$current === void 0 ? void 0 : _focusRef$current.focus();
    }
  });
  const handleBlur = (0, _react.useCallback)(event => {
    onPathBlur(member.field.path);
  }, [member.field.path, onPathBlur]);
  const handleFocus = (0, _react.useCallback)(event => {
    onPathFocus(member.field.path);
  }, [member.field.path, onPathFocus]);
  const handleChange = (0, _react.useCallback)(event => {
    onChange(_patch.PatchEvent.from(event).prefixAll(member.name));
  }, [onChange, member.name]);
  const validationError = (0, _react.useMemo)(() => member.field.validation.filter(item => item.level === 'error').map(item => item.message).join('\n'), [member.field.validation]) || undefined;
  const inputProps = (0, _react.useMemo)(() => {
    return {
      onBlur: handleBlur,
      value: member.field.value,
      readOnly: member.field.readOnly,
      schemaType: member.field.schemaType,
      changed: member.field.changed,
      focusRef: focusRef,
      id: member.field.id,
      onFocus: handleFocus,
      path: member.field.path,
      focused: member.field.focused,
      level: member.field.level,
      onChange: handleChange,
      validation: member.field.validation,
      presence: member.field.presence,
      validationError
    };
  }, [handleBlur, member.field.value, member.field.readOnly, member.field.schemaType, member.field.id, member.field.path, member.field.focused, member.field.level, member.field.validation, member.field.presence, member.field.changed, handleFocus, handleChange, validationError]);
  const renderedInput = (0, _react.useMemo)(() => renderInput(inputProps), [inputProps, renderInput]);
  const fieldProps = (0, _react.useMemo)(() => {
    return {
      name: member.name,
      index: member.index,
      level: member.field.level,
      value: member.field.value,
      schemaType: member.field.schemaType,
      title: member.field.schemaType.title,
      description: member.field.schemaType.description,
      inputId: member.field.id,
      path: member.field.path,
      validation: member.field.validation,
      presence: member.field.presence,
      children: renderedInput,
      changed: member.field.changed
    };
  }, [member.name, member.index, member.field.level, member.field.value, member.field.schemaType, member.field.id, member.field.path, member.field.validation, member.field.presence, member.field.changed, renderedInput]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, renderField(fieldProps));
}
