// next-sanity/lib/cjs/studio.js
var $cerUD$reactjsxruntime = require("react/jsx-runtime");
var $cerUD$react = require("react");
var $cerUD$sanity = require("sanity");
var $cerUD$styledcomponents = require("styled-components");
var $cerUD$nexthead = require("next/head");
var $cerUD$nextdocument = require("next/document");
var $cerUD$nextrouter = require("next/router");

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
var $5aa79282f535e8b1$exports = {};
var $c8bf9e0a6313dcb9$exports = {};

$parcel$export($c8bf9e0a6313dcb9$exports, "NextStudio", () => $c8bf9e0a6313dcb9$export$eefb95cccd1bb6c5);




/**
 * Intended to render at the root of a page, letting the Studio own that page and render much like it would if you used `npx sanity start` to render
 * It's a drop-in replacement for `import {Studio} from 'sanity'`
 */ const $c8bf9e0a6313dcb9$var$NextStudioComponent = ({ children: children , config: config , unstable__noGlobalStyle: unstable__noGlobalStyle , unstable__noTailwindSvgFix: unstable__noTailwindSvgFix , unstable__head: unstable__head , unstable__document_title: unstable__document_title , unstable__bg: unstable__bg , unstable__noFavicons: unstable__noFavicons , unstable__noNoScript: unstable__noNoScript , ...props })=>{
    const theme = (0, $b07d91633f3d677b$export$93d4e7f90805808f)(config);
    const { themeColorLight: themeColorLight , themeColorDark: themeColorDark  } = (0, $b07d91633f3d677b$export$d3a2146a441ec034)(theme);
    return /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsxs)((0, $cerUD$reactjsxruntime.Fragment), {
        children: [
            children || /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)((0, $cerUD$sanity.Studio), {
                config: config,
                ...props
            }),
            /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)((0, $9a339c4a487c3478$export$79723eee2c1bf75e), {
                themeColorLight: themeColorLight,
                themeColorDark: themeColorDark,
                title: unstable__document_title,
                favicons: !unstable__noFavicons,
                children: unstable__head
            }),
            !unstable__noGlobalStyle && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)((0, $1ac20522861d7f1e$export$c8fda7ee7699d07e), {
                bg: unstable__bg !== null && unstable__bg !== void 0 ? unstable__bg : themeColorLight,
                unstable__tailwindSvgFix: !unstable__noTailwindSvgFix
            }),
            !unstable__noNoScript && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)((0, $817646c15207d6a3$export$39af4e536de5c6ed), {})
        ]
    });
};
const $c8bf9e0a6313dcb9$export$eefb95cccd1bb6c5 = /*#__PURE__*/ (0, $cerUD$react.memo)($c8bf9e0a6313dcb9$var$NextStudioComponent);


var $1ac20522861d7f1e$exports = {};

$parcel$export($1ac20522861d7f1e$exports, "NextStudioGlobalStyle", () => $1ac20522861d7f1e$export$c8fda7ee7699d07e);

let $1ac20522861d7f1e$var$_ = (t)=>t, $1ac20522861d7f1e$var$t, $1ac20522861d7f1e$var$t1, $1ac20522861d7f1e$var$t2;
const $1ac20522861d7f1e$export$c8fda7ee7699d07e = (0, $cerUD$styledcomponents.createGlobalStyle)($1ac20522861d7f1e$var$t2 || ($1ac20522861d7f1e$var$t2 = $1ac20522861d7f1e$var$_`
${0}
html,
body,
#__next {
  height: 100%;
}
body {
  margin: 0;
  overscroll-behavior: none;
  -webkit-font-smoothing: antialiased;
}
${0}`), ({ bg: bg  })=>bg ? (0, $cerUD$styledcomponents.css)($1ac20522861d7f1e$var$t || ($1ac20522861d7f1e$var$t = $1ac20522861d7f1e$var$_`
        html {
          background-color: ${0};
        }
      `), bg) : "", ({ unstable__tailwindSvgFix: unstable__tailwindSvgFix  })=>unstable__tailwindSvgFix ? (0, $cerUD$styledcomponents.css)($1ac20522861d7f1e$var$t1 || ($1ac20522861d7f1e$var$t1 = $1ac20522861d7f1e$var$_`
        /* override tailwind reset */
        :root svg {
          display: inline;
        }
      `)) : "");


var $9a339c4a487c3478$exports = {};

$parcel$export($9a339c4a487c3478$exports, "NextStudioHead", () => $9a339c4a487c3478$export$79723eee2c1bf75e);



var $f8b95ae7ca517cc1$exports = {};
$f8b95ae7ca517cc1$exports = new URL("apple-touch-icon.abc4aca2.png", "file:" + __filename).toString();


var $f55f7bae1fab74eb$exports = {};
$f55f7bae1fab74eb$exports = new URL("favicon.648fc2f2.ico", "file:" + __filename).toString();


var $6951ae923d5c6ed1$exports = {};
$6951ae923d5c6ed1$exports = new URL("favicon.8f76876f.svg", "file:" + __filename).toString();


var $20032f35397ac709$exports = {};
$20032f35397ac709$exports = new URL("favicon-192.40c8bab4.png", "file:" + __filename).toString();


var $0f7b7d4813cad0d8$exports = {};
$0f7b7d4813cad0d8$exports = new URL("favicon-512.5e08f951.png", "file:" + __filename).toString();


var $c480472b608d459a$exports = {};
$c480472b608d459a$exports = JSON.parse('{"icons":[{"src":"./favicon-192.png","type":"image/png","sizes":"192x192"},{"src":"./favicon-512.png","type":"image/png","sizes":"512x512"}]}');


// Interop between how Parcel and Next deals with asset imports
const $9a339c4a487c3478$var$interop = (href)=>typeof href === "string" ? href : href.src;
const $9a339c4a487c3478$var$NextStudioHeadComponent = ({ children: children , themeColorDark: themeColorDark , themeColorLight: themeColorLight , title: title = "Sanity Studio" , favicons: favicons  })=>{
    const inlineWebmanifest = (0, $cerUD$react.useCallback)(()=>{
        const manifest = JSON.parse(JSON.stringify((0, (/*@__PURE__*/$parcel$interopDefault($c480472b608d459a$exports)))));
        const icons = manifest.icons.map((icon)=>{
            // Inline manifests works best when URLs are absolute
            const src = // eslint-disable-next-line no-nested-ternary
            icon.src === "./favicon-192.png" ? $9a339c4a487c3478$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($20032f35397ac709$exports)))) : icon.src === "./favicon-512.png" ? $9a339c4a487c3478$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($0f7b7d4813cad0d8$exports)))) : icon.src;
            return {
                ...icon,
                src: src
            };
        });
        return `data:application/manifest+json,${encodeURIComponent(JSON.stringify({
            ...manifest,
            icons: icons
        }))}`;
    }, []);
    return /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsxs)((0, ($parcel$interopDefault($cerUD$nexthead))), {
        children: [
            /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("meta", {
                name: "viewport",
                // Studio implements display cutouts CSS (The iPhone Notch ™ ) and needs `viewport-fit=covered` for it to work correctly
                content: "width=device-width, initial-scale=1, viewport-fit=cover"
            }),
            /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("meta", {
                name: "robots",
                content: "noindex"
            }),
            /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("meta", {
                name: "referrer",
                content: "same-origin"
            }),
            title && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("title", {
                children: title
            }),
            favicons && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("link", {
                rel: "icon",
                href: $9a339c4a487c3478$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($f55f7bae1fab74eb$exports)))),
                sizes: "any"
            }),
            favicons && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("link", {
                rel: "icon",
                href: $9a339c4a487c3478$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($6951ae923d5c6ed1$exports)))),
                type: "image/svg+xml"
            }),
            favicons && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("link", {
                rel: "apple-touch-icon",
                href: $9a339c4a487c3478$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($f8b95ae7ca517cc1$exports))))
            }),
            favicons && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("link", {
                rel: "manifest",
                // eslint-disable-next-line no-warning-comments
                // @TODO until parcel fixes https://github.com/parcel-bundler/parcel/issues/8025 and stops stripping process.env.NEXT_PUBLIC_VERCEL_URL from the compiled code, use the remove webmanifest
                href: "https://next.sanity.build/manifest.webmanifest"
            }),
            themeColorLight && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("meta", {
                name: "theme-color",
                content: themeColorLight,
                media: "(prefers-color-scheme: light)"
            }, "theme-color-light"),
            themeColorDark && /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("meta", {
                name: "theme-color",
                content: themeColorDark,
                media: "(prefers-color-scheme: dark)"
            }, "theme-color-dark"),
            children
        ]
    });
};
const $9a339c4a487c3478$export$79723eee2c1bf75e = /*#__PURE__*/ (0, $cerUD$react.memo)($9a339c4a487c3478$var$NextStudioHeadComponent);


var $817646c15207d6a3$exports = {};

$parcel$export($817646c15207d6a3$exports, "NextStudioNoScript", () => $817646c15207d6a3$export$39af4e536de5c6ed);

/* eslint-disable react/no-danger */ const $817646c15207d6a3$var$style = {
    __html: `
.sanity-app-no-js__root {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background: #fff;
}

.sanity-app-no-js__content {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  font-family: helvetica, arial, sans-serif;
}
`
};
const $817646c15207d6a3$export$39af4e536de5c6ed = ()=>/*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("noscript", {
        children: /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("div", {
            className: "sanity-app-no-js__root",
            children: /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsxs)("div", {
                className: "sanity-app-no-js__content",
                children: [
                    /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("style", {
                        type: "text/css",
                        dangerouslySetInnerHTML: $817646c15207d6a3$var$style
                    }),
                    /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("h1", {
                        children: "JavaScript disabled"
                    }),
                    /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsxs)("p", {
                        children: [
                            "Please ",
                            /*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)("a", {
                                href: "https://www.enable-javascript.com/",
                                children: "enable JavaScript"
                            }),
                            " in your browser and reload the page to proceed."
                        ]
                    })
                ]
            })
        })
    });


var $4dc132240adf8d31$exports = {};

$parcel$export($4dc132240adf8d31$exports, "ServerStyleSheetDocument", () => $4dc132240adf8d31$export$f64e2f55a1dca172);



class $4dc132240adf8d31$export$f64e2f55a1dca172 extends (0, ($parcel$interopDefault($cerUD$nextdocument))) {
    static async getInitialProps(ctx) {
        const sheet = new (0, $cerUD$styledcomponents.ServerStyleSheet)();
        const originalRenderPage = ctx.renderPage;
        try {
            ctx.renderPage = ()=>originalRenderPage({
                    enhanceApp: (App)=>(props)=>sheet.collectStyles(/*#__PURE__*/ (0, $cerUD$reactjsxruntime.jsx)(App, {
                                ...props
                            }))
                });
            const initialProps = await (0, ($parcel$interopDefault($cerUD$nextdocument))).getInitialProps(ctx);
            return {
                ...initialProps,
                styles: [
                    initialProps.styles,
                    sheet.getStyleElement()
                ]
            };
        } finally{
            sheet.seal();
        }
    }
}


var $b07d91633f3d677b$exports = {};

$parcel$export($b07d91633f3d677b$exports, "isWorkspaces", () => $b07d91633f3d677b$export$b7c6924067135ce);
$parcel$export($b07d91633f3d677b$exports, "isWorkspaceWithTheme", () => $b07d91633f3d677b$export$31236b9ee07dd150);
$parcel$export($b07d91633f3d677b$exports, "useTheme", () => $b07d91633f3d677b$export$93d4e7f90805808f);
$parcel$export($b07d91633f3d677b$exports, "useBackgroundColorsFromTheme", () => $b07d91633f3d677b$export$d3a2146a441ec034);
$parcel$export($b07d91633f3d677b$exports, "useBasePath", () => $b07d91633f3d677b$export$53bdd69a9f9bd75);
$parcel$export($b07d91633f3d677b$exports, "useConfigWithBasePath", () => $b07d91633f3d677b$export$297ba38aa0562c6e);



function $b07d91633f3d677b$export$b7c6924067135ce(config) {
    return Array.isArray(config);
}
function $b07d91633f3d677b$export$31236b9ee07dd150(workspace) {
    return Boolean(workspace.theme);
}
function $b07d91633f3d677b$export$93d4e7f90805808f(config) {
    const workspace = (0, $cerUD$react.useMemo)(()=>$b07d91633f3d677b$export$b7c6924067135ce(config) ? config[0] : config, [
        config
    ]);
    return (0, $cerUD$react.useMemo)(()=>$b07d91633f3d677b$export$31236b9ee07dd150(workspace) ? workspace.theme : (0, $cerUD$sanity.defaultTheme), [
        workspace
    ]);
}
const $b07d91633f3d677b$export$d3a2146a441ec034 = (theme)=>{
    return (0, $cerUD$react.useMemo)(()=>({
            themeColorLight: theme.color.light.default.base.bg,
            themeColorDark: theme.color.dark.default.base.bg
        }), [
        theme
    ]);
};
function $b07d91633f3d677b$export$53bdd69a9f9bd75() {
    const router = (0, $cerUD$nextrouter.useRouter)();
    return (0, $cerUD$react.useMemo)(()=>{
        const [basePath = "/"] = router.route.split("/[");
        return basePath;
    }, [
        router.route
    ]);
}
function $b07d91633f3d677b$export$297ba38aa0562c6e(config) {
    const basePath = $b07d91633f3d677b$export$53bdd69a9f9bd75();
    return (0, $cerUD$react.useMemo)(()=>{
        if ($b07d91633f3d677b$export$b7c6924067135ce(config)) return config.map((workspace)=>({
                ...workspace,
                basePath: workspace.basePath === "/" ? basePath : `${basePath}${workspace.basePath || ""}`
            }));
        return {
            ...config,
            basePath: config.basePath === "/" ? basePath : `${basePath}${config.basePath || ""}`
        };
    }, [
        config,
        basePath
    ]);
}


$parcel$exportWildcard($5aa79282f535e8b1$exports, $c8bf9e0a6313dcb9$exports);
$parcel$exportWildcard($5aa79282f535e8b1$exports, $1ac20522861d7f1e$exports);
$parcel$exportWildcard($5aa79282f535e8b1$exports, $9a339c4a487c3478$exports);
$parcel$exportWildcard($5aa79282f535e8b1$exports, $817646c15207d6a3$exports);
$parcel$exportWildcard($5aa79282f535e8b1$exports, $4dc132240adf8d31$exports);
$parcel$exportWildcard($5aa79282f535e8b1$exports, $b07d91633f3d677b$exports);


$parcel$exportWildcard(module.exports, $5aa79282f535e8b1$exports);
