// sanity/lib/cjs/desk/panes/document/statusBar/sparkline/DocumentBadges.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentBadges = DocumentBadges;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _badges = require("../../../../badges");

var _useDocumentPane = require("../../useDocumentPane");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BADGE_TONES = {
  primary: 'primary',
  success: 'positive',
  warning: 'caution',
  danger: 'critical'
};

function DocumentBadgesInner(_ref) {
  let {
    states
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 1
  }, states.map((badge, index) => /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: badge.title && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, badge.title)),
    disabled: !badge.title,
    key: String(index),
    placement: "top",
    portal: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    fontSize: 1,
    mode: "outline",
    paddingX: 2,
    paddingY: 1,
    radius: 4,
    tone: badge.color ? BADGE_TONES[badge.color] : undefined,
    style: {
      whiteSpace: 'nowrap'
    }
  }, badge.label))));
}

function DocumentBadges() {
  const {
    badges,
    editState
  } = (0, _useDocumentPane.useDocumentPane)();
  if (!editState || !badges) return null;
  return /*#__PURE__*/_react.default.createElement(_badges.RenderBadgeCollectionState, {
    badges: badges,
    badgeProps: editState
  }, _ref2 => {
    let {
      states
    } = _ref2;
    return /*#__PURE__*/_react.default.createElement(DocumentBadgesInner, {
      states: states
    });
  });
}
