// sanity/lib/esm/form/inputs/arrays/ArrayOfObjectsInput/InsertMenu.js
/* eslint-disable react/jsx-no-bind */
import React, { memo } from 'react';
import { MenuGroup, MenuItem } from '@sanity/ui';
import { InsertAboveIcon, InsertBelowIcon } from '@sanity/icons';
const MENU_POPOVER_PROPS = {
  portal: true,
  tone: 'default',
  placement: 'left',
  constrainSize: true
};
export const InsertMenu = /*#__PURE__*/memo(function InsertMenu(props) {
  const {
    types,
    onInsert
  } = props;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(InsertMenuGroup, {
    pos: "before",
    types: types,
    onInsert: onInsert,
    text: "Add item before",
    icon: InsertAboveIcon
  }), /*#__PURE__*/React.createElement(InsertMenuGroup, {
    pos: "after",
    types: types,
    onInsert: onInsert,
    text: "Add item after",
    icon: InsertBelowIcon
  }));
});

function InsertMenuGroup(props) {
  const {
    types,
    onInsert,
    pos,
    text,
    icon
  } = props;

  if ((types === null || types === void 0 ? void 0 : types.length) === 1) {
    return /*#__PURE__*/React.createElement(MenuItem, {
      key: pos,
      text: text,
      icon: icon,
      onClick: () => onInsert(pos, types[0])
    });
  }

  return /*#__PURE__*/React.createElement(MenuGroup, {
    text: text,
    key: pos,
    popover: MENU_POPOVER_PROPS
  }, types === null || types === void 0 ? void 0 : types.map(insertableType => /*#__PURE__*/React.createElement(MenuItem, {
    key: insertableType.name,
    icon: insertableType.icon,
    text: insertableType.title,
    onClick: () => onInsert(pos, insertableType)
  })));
}
