// sanity/lib/esm/datastores/__workshop__/DocumentPairStory.js
import { Box, Button, Code, Text } from '@sanity/ui';
import React, { useEffect, useState } from 'react';
import { useDocumentStore } from '../datastores';
export default function DocumentPairStory() {
  const documentStore = useDocumentStore();
  const [draftSnapshot, setDraftSnapshot] = useState(null);
  const [publishedSnapshot, setPublishedSnapshot] = useState(null);
  useEffect(() => {
    const {
      draft,
      published
    } = documentStore.checkoutPair({
      draftId: 'drafts.test',
      publishedId: 'test'
    });
    const draftSub = draft.remoteSnapshot$.subscribe(setDraftSnapshot);
    const publishedSub = published.remoteSnapshot$.subscribe(setPublishedSnapshot);
    return () => {
      draftSub.unsubscribe();
      publishedSub.unsubscribe();
    };
  }, [documentStore]);
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1,
    weight: "semibold"
  }, /*#__PURE__*/React.createElement("code", null, "documentStore.checkoutPair({draftId: string, publishedId: string})")), /*#__PURE__*/React.createElement(Box, {
    hidden: true
  }, /*#__PURE__*/React.createElement(Button, {
    text: "Subscribe"
  })), /*#__PURE__*/React.createElement(Box, {
    marginTop: 3
  }, /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, JSON.stringify({
    draft: {
      snapshot: draftSnapshot
    },
    published: {
      snapshot: publishedSnapshot
    }
  }, null, 2))));
}
