// sanity/lib/esm/datastores/__workshop__/DocumentPermissionStory.js
import { Box, Code, Text } from '@sanity/ui';
import React, { useEffect, useState } from 'react';
import { useGrantsStore } from '../datastores';
export default function DocumentPermissionStory() {
  const grantsStore = useGrantsStore();
  const [permission, setPermission] = useState(null);
  useEffect(() => {
    const permission$ = grantsStore.checkDocumentPermission('update', {
      _id: 'test'
    });
    const sub = permission$.subscribe(setPermission);
    return () => {
      sub.unsubscribe();
    };
  }, [grantsStore]);
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1,
    weight: "semibold"
  }, /*#__PURE__*/React.createElement("code", null, "grantsStore.checkDocumentPermission(permissionName: DocumentValuePermission, document: SanityDocument)")), /*#__PURE__*/React.createElement(Box, {
    marginTop: 3
  }, /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, JSON.stringify(permission, null, 2))));
}
