// sanity/lib/esm/datastores/__workshop__/HistoryStory.js
import { Box, Code, Text } from '@sanity/ui';
import React, { useEffect, useState } from 'react';
import { useHistoryStore } from '../datastores';
const SECOND = 1000;
const MINUTE = SECOND * 60;
const HOUR = MINUTE * 60;
const DAY = HOUR * 24;
export default function HistoryStory() {
  const historyStore = useHistoryStore();
  const [history, setHistory] = useState(null);
  useEffect(() => {
    // The time at which to view the document’s value
    const time = new Date(Date.now() - DAY).toISOString();
    historyStore.getHistory(['drafts.test', 'test'], {
      time
    }).then(setHistory);
  }, [historyStore]);
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1,
    weight: "semibold"
  }, /*#__PURE__*/React.createElement("code", null, "historyStore.getHistory")), /*#__PURE__*/React.createElement(Box, {
    marginTop: 3
  }, /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, JSON.stringify(history, null, 2))));
}
