// sanity/lib/cjs/form/components/formField/FormFieldSet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormFieldSet = void 0;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _presence = require("../../../presence");

var _FormFieldValidationStatus = require("./FormFieldValidationStatus");

var _FormFieldSetLegend = require("./FormFieldSetLegend");

var _styles = require("./styles");

var _templateObject, _templateObject2;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function getChildren(children) {
  return typeof children === 'function' ? children() : children;
}

const Root = (0, _styledComponents.default)(_ui.Box).attrs({
  forwardedAs: 'fieldset'
})(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  border: none;\n\n  /* See: https://thatemil.com/blog/2015/01/03/reset-your-fieldset/ */\n  body:not(:-moz-handler-blocked) & {\n    display: table-cell;\n  }\n"])));
const Content = (0, _styledComponents.default)(_ui.Box)(props => {
  const {
    $borderLeft,
    theme
  } = props;
  const {
    focusRing,
    radius
  } = theme.sanity;
  const {
    base
  } = theme.sanity.color;
  return (0, _styledComponents.css)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    outline: none;\n    border-left: ", ";\n    border-radius: ", ";\n\n    &:focus {\n      box-shadow: ", ";\n    }\n\n    &:focus:not(:focus-visible) {\n      box-shadow: none;\n    }\n  "])), $borderLeft ? '1px solid var(--card-border-color)' : undefined, (0, _ui.rem)(radius[1]), (0, _styles.focusRingStyle)({
    base,
    focusRing: { ...focusRing,
      offset: 2
    }
  }));
});
const EMPTY_ARRAY = [];
const FormFieldSet = /*#__PURE__*/(0, _react.forwardRef)(function FormFieldSet(props, ref) {
  const {
    validation = EMPTY_ARRAY,
    __unstable_presence: presence = EMPTY_ARRAY,
    children,
    collapsible,
    columns,
    description,
    level = 0,
    onFocus,
    onCollapse,
    onExpand,
    collapsed,
    tabIndex,
    title,
    ...restProps
  } = props;
  const hasValidationMarkers = validation.length > 0;
  const forwardedRef = (0, _ui.useForwardedRef)(ref);
  const handleFocus = (0, _react.useCallback)(event => {
    const element = forwardedRef.current;

    if (element === event.target) {
      if (onFocus) onFocus(event);
    }
  }, [forwardedRef, onFocus]);
  const handleToggle = (0, _react.useCallback)(() => collapsed ? onExpand === null || onExpand === void 0 ? void 0 : onExpand() : onCollapse === null || onCollapse === void 0 ? void 0 : onCollapse(), [collapsed, onCollapse, onExpand]);
  const content = (0, _react.useMemo)(() => {
    if (collapsed) {
      return null;
    }

    return /*#__PURE__*/_react.default.createElement(_ui.Grid, {
      columns: columns,
      gapX: 4,
      gapY: 5
    }, getChildren(children));
  }, [children, collapsed, columns]);
  return /*#__PURE__*/_react.default.createElement(Root, _extends({
    "data-level": level
  }, restProps), title && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "flex-end"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    paddingY: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_FormFieldSetLegend.FormFieldSetLegend, {
    collapsed: Boolean(collapsed),
    collapsible: collapsible,
    onClick: collapsible ? handleToggle : undefined,
    title: title
  }), hasValidationMarkers && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_FormFieldValidationStatus.FormFieldValidationStatus, {
    fontSize: 1,
    validation: validation
  }))), description && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1
  }, description))), presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    maxAvatars: 4,
    presence: presence
  }))), /*#__PURE__*/_react.default.createElement(Content, {
    $borderLeft: level > 0,
    hidden: collapsed,
    marginTop: 1,
    paddingLeft: level === 0 ? 0 : 3,
    onFocus: typeof tabIndex === 'number' && tabIndex > -1 ? handleFocus : undefined,
    ref: forwardedRef,
    tabIndex: tabIndex
  }, !collapsed && content));
});
exports.FormFieldSet = FormFieldSet;
