// sanity/lib/cjs/components/react-track-elements/createScope.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createScope = createScope;

var _react = _interopRequireDefault(require("react"));

var _createUseReporter = require("./createUseReporter");

var _createStore = require("./createStore");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Todo: consider memozing individual functions or move the context assertion/guard to a separate step.
let didWarn = false;

const useReporterGuard = id => {
  if (!didWarn) {
    // eslint-disable-next-line no-console
    console.warn(new Error("No context provided for reporter. Make sure that the component calling \"useReporter(".concat(id, ", ...)\", is wrapped in a <Tracker> element")));
  }

  didWarn = true;
};

function useReportedValueGuard() {
  if (!didWarn) {
    // eslint-disable-next-line no-console
    console.warn(new Error('No context provided for reporter. Make sure that the component calling "useReportedValues()", is wrapped inside a <Tracker> element'));
  }

  didWarn = true;
  return [];
}

const useSubscribeGuard = () => {
  if (!didWarn) {
    // eslint-disable-next-line no-console
    console.warn(new Error('No context provided for reporter. Make sure that the component calling "useReportedValues()", is wrapped inside a <Tracker> element'));
  }

  didWarn = true; // eslint-disable-next-line @typescript-eslint/no-empty-function

  return () => {};
};

const DEFAULT_CONTEXT = {
  add: useReporterGuard,
  update: useReporterGuard,
  remove: useReporterGuard,
  subscribe: useSubscribeGuard,
  read: useReportedValueGuard
};
let id = 0;

const getNextId = () => ++id;

function createScope() {
  const Context = /*#__PURE__*/_react.default.createContext(DEFAULT_CONTEXT);

  function useReportedValues() {
    const context = _react.default.useContext(Context);

    const [values, setValues] = _react.default.useState(context.read());

    _react.default.useLayoutEffect(() => {
      setValues(context.read());
      return context.subscribe(setValues);
    }, [context]);

    return values;
  }

  function Tracker(props) {
    const store = _react.default.useMemo(() => (0, _createStore.createStore)(), []);

    return /*#__PURE__*/_react.default.createElement(Context.Provider, {
      value: store
    }, props.children);
  }

  const useReporter = (0, _createUseReporter.createUseReporter)(Context);
  return {
    Tracker,
    useReportedValues,
    useReporter,
    useAutoIdReporter: function (value) {
      let isEqual = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Object.is;
      return useReporter("element-".concat(_react.default.useRef(getNextId()).current), value, isEqual);
    }
  };
}
