// sanity/lib/cjs/components/__workshop__/TextWithToneStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TextWithToneStory;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _uiWorkshop = require("@sanity/ui-workshop");

var _react = _interopRequireDefault(require("react"));

var _TextWithTone = require("../TextWithTone");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const TEXT_SIZE_OPTIONS = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4
};
const CARD_TONE_OPTIONS = {
  Default: 'default',
  Primary: 'primary',
  Positive: 'positive',
  Caution: 'caution',
  Critical: 'critical'
};
const tones = ['default', 'primary', 'positive', 'caution', 'critical'];

function TextWithToneStory() {
  const size = (0, _uiWorkshop.useSelect)('Size', TEXT_SIZE_OPTIONS, 2);
  const dimmed = (0, _uiWorkshop.useBoolean)('Dimmed', false);
  const muted = (0, _uiWorkshop.useBoolean)('Muted', false);
  const cardTone = (0, _uiWorkshop.useSelect)('Card tone', CARD_TONE_OPTIONS);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    border: true,
    radius: 2,
    tone: cardTone
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, tones.map(tone => /*#__PURE__*/_react.default.createElement(_TextWithTone.TextWithTone, {
    key: tone,
    size: size,
    dimmed: dimmed,
    muted: muted,
    tone: tone
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    gap: 2
  }, /*#__PURE__*/_react.default.createElement(_icons.InfoOutlineIcon, null), tone))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, tones.map(tone => /*#__PURE__*/_react.default.createElement(_TextWithTone.TextWithTone, {
    key: tone,
    size: size,
    dimmed: dimmed,
    muted: muted,
    tone: tone
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    gap: 2
  }, /*#__PURE__*/_react.default.createElement(_icons.InfoOutlineIcon, null), tone)))))));
}
