// sanity/lib/esm/field/types/image/diff/HotspotCropSVG.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { DiffTooltip, useDiffAnnotationColor } from '../../../diff';
import { hexToRgba } from './helpers';
export function HotspotCropSVG(props) {
  const {
    crop,
    diff,
    hash,
    hotspot,
    width = 100,
    height = 100,
    ...restProps
  } = props;
  const cropColor = useDiffAnnotationColor(diff, 'crop');
  const hotspotColor = useDiffAnnotationColor(diff, 'hotspot');
  return /*#__PURE__*/React.createElement("svg", _extends({}, restProps, {
    fill: "none",
    width: width,
    height: height,
    viewBox: "0 0 ".concat(width, " ").concat(height)
  }), /*#__PURE__*/React.createElement("defs", null, crop && hotspot && /*#__PURE__*/React.createElement("mask", {
    id: "mask-hotspot-".concat(hash)
  }, /*#__PURE__*/React.createElement("rect", {
    x: 0,
    y: 0,
    width: width,
    height: height,
    fill: "#fff"
  }), /*#__PURE__*/React.createElement(HotspotSVG, {
    hotspot: hotspot,
    fill: "#000",
    offset: 1,
    width: width,
    height: height,
    stroke: "#000",
    strokeWidth: 3
  }))), crop && /*#__PURE__*/React.createElement(DiffTooltip, {
    diff: diff,
    path: "crop",
    description: "Crop changed"
  }, /*#__PURE__*/React.createElement("g", null, /*#__PURE__*/React.createElement(CropSVG, {
    crop: crop,
    fill: hexToRgba(cropColor.border, 0.25),
    mask: hotspot ? "url(#mask-hotspot-".concat(hash, ")") : undefined,
    stroke: cropColor.border,
    strokeWidth: 1,
    width: width,
    height: height
  }))), hotspot && /*#__PURE__*/React.createElement(DiffTooltip, {
    diff: diff,
    path: "hotspot",
    description: "Hotspot changed"
  }, /*#__PURE__*/React.createElement("g", null, /*#__PURE__*/React.createElement(HotspotSVG, {
    hotspot: hotspot,
    fill: hexToRgba(hotspotColor.border, 0.25),
    stroke: hotspotColor.border,
    strokeWidth: 1,
    width: width,
    height: height
  }))));
}

function CropSVG(_ref) {
  let {
    crop,
    width,
    height,
    ...restProps
  } = _ref;
  const rectProps = {
    x: crop.left * width,
    y: crop.top * height,
    width: (1 - crop.right - crop.left) * width,
    height: (1 - crop.bottom - crop.top) * height
  };
  return /*#__PURE__*/React.createElement("rect", _extends({}, restProps, rectProps, {
    style: {
      vectorEffect: 'non-scaling-stroke'
    }
  }));
}

function HotspotSVG(_ref2) {
  let {
    hotspot,
    offset = 0,
    width,
    height,
    ...restProps
  } = _ref2;
  const ellipseProps = {
    cx: hotspot.x * width,
    cy: hotspot.y * height,
    rx: hotspot.width / 2 * width + offset,
    ry: hotspot.height / 2 * height + offset
  };
  return /*#__PURE__*/React.createElement("ellipse", _extends({}, restProps, ellipseProps, {
    style: {
      vectorEffect: 'non-scaling-stroke'
    }
  }));
}
