// sanity/lib/esm/desk/panes/document/statusBar/dialogs/PopoverDialog.js
import { Popover, useClickOutside, useGlobalKeyDown, useLayer } from '@sanity/ui';
import React, { useCallback, useState } from 'react';
import { POPOVER_FALLBACK_PLACEMENTS } from './constants';
export function PopoverDialog(props) {
  const {
    modal,
    referenceElement
  } = props;
  return /*#__PURE__*/React.createElement(Popover, {
    content: /*#__PURE__*/React.createElement(PopoverDialogContent, {
      modal: modal
    }),
    fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS,
    open: true,
    placement: "top",
    portal: true,
    preventOverflow: true,
    referenceElement: referenceElement
  });
}

function PopoverDialogContent(props) {
  const {
    modal
  } = props;
  const {
    content,
    onClose
  } = modal;
  const {
    isTopLayer
  } = useLayer();
  const [element, setElement] = useState(null);
  const handleClickOutside = useCallback(() => {
    if (isTopLayer) onClose();
  }, [isTopLayer, onClose]);
  const handleGlobalKeyDown = useCallback(event => {
    if (event.key === 'Escape' && isTopLayer) onClose();
  }, [isTopLayer, onClose]);
  useClickOutside(handleClickOutside, [element]);
  useGlobalKeyDown(handleGlobalKeyDown);
  return /*#__PURE__*/React.createElement("div", {
    ref: setElement
  }, content);
}
