// sanity/lib/esm/presence/PresenceTooltip.js
import React, { useMemo } from 'react';
import { Flex, Stack, Text, Tooltip } from '@sanity/ui';
import { UserAvatar } from '../components/UserAvatar';

/**
 * The "documentScrollElement" is being passed to the PortalProvider in DocumentPanel.
 * The default portal element provided by the PortalProvider causes some layout issues with the Tooltip.
 * Therefore, another portal element (i.e. the "documentScrollElement") is being used to solve this.
 */
export function PresenceTooltip(props) {
  const {
    children,
    items
  } = props;
  const content = useMemo(() => /*#__PURE__*/React.createElement(Stack, {
    padding: 1
  }, items.map(item => /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    gap: 2,
    padding: 1,
    key: item.user.id
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(UserAvatar, {
    user: item.user,
    status: "online"
  })), /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, item.user.displayName)))), [items]);
  return /*#__PURE__*/React.createElement(Tooltip, {
    content: content,
    placement: "top",
    portal: "documentScrollElement"
  }, children);
}
