// sanity/lib/dts/test/form/renderFileInput.js
import { EMPTY } from 'rxjs';
import { renderObjectInput } from './renderObjectInput';
const STUB_ASSET_SOURCES = [{}]; // @todo
const STUB_OBSERVE_ASSET = () => EMPTY;
const STUB_RESOLVE_UPLOADER = () => ({
    priority: 1,
    type: 'file',
    accepts: 'file/*',
    upload: () => EMPTY,
});
export async function renderFileInput(options) {
    const { assetSources = STUB_ASSET_SOURCES, fieldDefinition, observeAsset = STUB_OBSERVE_ASSET, props, render: initialRender, resolveUploader = STUB_RESOLVE_UPLOADER, } = options;
    function transformProps(inputProps, context) {
        const { schemaType, value, ...restProps } = inputProps;
        const { client } = context;
        return {
            ...restProps,
            assetSources,
            client,
            directUploads: true,
            observeAsset,
            resolveUploader,
            schemaType: schemaType,
            value: value,
        };
    }
    const result = await renderObjectInput({
        fieldDefinition: fieldDefinition,
        props,
        render: (inputProps, context) => initialRender(transformProps(inputProps, context), context),
    });
    function rerender(subsequentRender) {
        result.rerender((inputProps, context) => subsequentRender(transformProps(inputProps, context), context));
    }
    return { ...result, rerender };
}
