// sanity/lib/dts/test/form/TestProvider.js
import React from 'react';
import { LayerProvider, studioTheme, ThemeProvider, ToastProvider } from '@sanity/ui';
import { createWorkspaceFromConfig } from '../../src/config';
import { SourceProvider, WorkspaceProvider } from '../../src/studio';
export async function createTestProvider({ client, config }) {
    const currentUser = {
        id: 'doug',
        name: 'Doug',
        email: 'doug@sanity.io',
        role: 'admin',
        roles: [{ name: 'admin', title: 'Admin' }],
    };
    const workspace = await createWorkspaceFromConfig({ ...config, currentUser, client });
    function TestProvider({ children }) {
        return (React.createElement(ThemeProvider, { theme: studioTheme },
            React.createElement(ToastProvider, null,
                React.createElement(LayerProvider, null,
                    React.createElement(WorkspaceProvider, { workspace: workspace },
                        React.createElement(SourceProvider, { source: workspace.unstable_sources[0] }, children))))));
    }
    return TestProvider;
}
