// sanity/lib/esm/studio/screens/schemaErrors/SchemaErrorsScreen.js
import { Card, Container, Heading, Stack } from '@sanity/ui';
import React, { useEffect } from 'react';
import { SchemaProblemGroups } from './SchemaProblemGroups';
import { reportWarnings } from './reportWarnings';
export function SchemaErrorsScreen(_ref) {
  var _schema$_validation;

  let {
    schema
  } = _ref;
  const groupsWithErrors = ((_schema$_validation = schema._validation) === null || _schema$_validation === void 0 ? void 0 : _schema$_validation.filter(group => group.problems.some(problem => problem.severity === 'error'))) || [];
  useEffect(() => reportWarnings(schema), [schema]);
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill",
    overflow: "auto",
    paddingY: [4, 5, 6, 7],
    paddingX: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 1
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 5
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h1"
  }, "Schema errors"), /*#__PURE__*/React.createElement(SchemaProblemGroups, {
    problemGroups: groupsWithErrors
  }))));
}
