// @sanity/dashboard/lib/components/NotFoundWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Root = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  flex-direction: column;\n  justify-content: stretch;\n  height: 100%;\n"])));

function NotFoundWidget(props) {
  var title = props.title,
      children = props.children;
  return /*#__PURE__*/_react.default.createElement(Root, {
    radius: 3,
    paddingX: 3,
    paddingY: 4,
    tone: "critical"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, title && /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    size: 1,
    as: "h2"
  }, title), children && /*#__PURE__*/_react.default.createElement(_ui.Box, null, children)));
}

NotFoundWidget.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  children: _propTypes.default.any,
  // eslint-disable-next-line react/forbid-prop-types
  title: _propTypes.default.any
};
NotFoundWidget.defaultProps = {
  children: null,
  title: null
};
var _default = NotFoundWidget;
exports.default = _default;
