// @sanity/dashboard/lib/components/WidgetGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _ui = require("@sanity/ui");

var _legacyParts = require("../legacyParts");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var media = {
  small: function small() {
    return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n      @media (min-width: ", "px) {\n        ", "\n      }\n    "])), _ref => {
      var theme = _ref.theme;
      return theme.sanity.media[0];
    }, (0, _styledComponents.css)(...arguments));
  },
  medium: function medium() {
    return (0, _styledComponents.css)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n      @media (min-width: ", "px) {\n        ", "\n      }\n    "])), _ref2 => {
      var theme = _ref2.theme;
      return theme.sanity.media[2];
    }, (0, _styledComponents.css)(...arguments));
  }
};
var Root = (0, _styledComponents.default)(_ui.Grid)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));\n\n  & > div {\n    overflow: hidden;\n  }\n\n  & > div[data-width='medium'] {\n    ", "\n  }\n\n  & > div[data-width='large'] {\n    ", "\n\n    ", "\n  }\n\n  & > div[data-width='full'] {\n    ", "\n  }\n\n  & > div[data-height='medium'] {\n    ", "\n  }\n\n  & > div[data-height='large'] {\n    ", "\n\n    ", "\n  }\n\n  & > div[data-height='full'] {\n    ", "\n  }\n"])), media.small(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n      grid-column: span 2;\n    "]))), media.small(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n      grid-column: span 2;\n    "]))), media.medium(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n      grid-column: span 3;\n    "]))), media.small(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n      grid-column: 1 / -1;\n    "]))), media.small(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["\n      grid-row: span 2;\n    "]))), media.small(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["\n      grid-row: span 2;\n    "]))), media.medium(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["\n      grid-row: span 3;\n    "]))), media.medium(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["\n      grid-row: 1 / -1;\n    "]))));

function WidgetGroup(props) {
  var config = props.config || {};
  var widgets = config.widgets || [];
  var layout = config.layout || {};
  return /*#__PURE__*/_react.default.createElement(Root, {
    autoFlow: "dense",
    "data-width": layout.width || 'auto',
    "data-height": layout.height || 'auto',
    "data-name": "sanity-dashboard-widget-group",
    gap: 4
  }, widgets.map((widgetConfig, index) => {
    if (widgetConfig.type === '__experimental_group') {
      return /*#__PURE__*/_react.default.createElement(WidgetGroup, {
        key: String(index),
        config: widgetConfig
      });
    }

    return /*#__PURE__*/_react.default.createElement(_legacyParts.WidgetContainer, {
      key: String(index),
      config: widgetConfig
    });
  }));
}

var _default = WidgetGroup;
exports.default = _default;
