// sanity/lib/cjs/studio/workspaceLoader/ErrorMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorMessage = ErrorMessage;

var _startCase2 = _interopRequireDefault(require("lodash/startCase"));

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject, _templateObject2;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const ListItem = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral([""])));
const ErrorMessageRoot = (0, _styledComponents.default)(_ui.Box).attrs({
  padding: 4
})(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral([""])));

function ErrorMessage(_ref) {
  let {
    error,
    message,
    path,
    stack
  } = _ref;
  (0, _react.useEffect)(() => {
    console.error(error);
  }, [error]);
  const last = path[path.length - 1];
  return /*#__PURE__*/_react.default.createElement(ErrorMessageRoot, {
    forwardedAs: _ui.Flex,
    direction: "column",
    gap: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    gap: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "semibold",
    size: 3
  }, (0, _startCase2.default)(last.type), " Error")), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "critical",
    overflow: "auto",
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, null, message)), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    as: "ul",
    direction: "column",
    gap: 2
  }, path.map((_ref2, index) => {
    let {
      name,
      type
    } = _ref2;
    return /*#__PURE__*/_react.default.createElement(ListItem, {
      forwardedAs: "li",
      gap: 2,
      align: "center",
      key: index
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Code, null, name)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, type)));
  })), stack && /*#__PURE__*/_react.default.createElement("details", null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "summary"
  }, "Stack Trace"), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    overflow: "auto",
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, null, stack))));
}
