// sanity/lib/cjs/form/inputs/PortableText/object/renderers/DefaultEditDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultEditDialog = DefaultEditDialog;

var _react = _interopRequireDefault(require("react"));

var _autoId = require("@reach/auto-id");

var _ui = require("@sanity/ui");

var _presence = require("../../../../../presence");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultEditDialog(props) {
  var _portal$elements;

  const {
    onClose,
    children,
    title,
    width = 1
  } = props;
  const dialogId = (0, _autoId.useId)();
  const portal = (0, _ui.usePortal)();
  return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    __unstable_autoFocus: true,
    id: dialogId || '',
    onClose: onClose,
    onClickOutside: onClose,
    header: title,
    portal: "default",
    width: width
  }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
    margins: [0, 0, 1, 0]
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.PortalProvider, {
    element: (_portal$elements = portal.elements) === null || _portal$elements === void 0 ? void 0 : _portal$elements.default
  }, children))));
}
