// sanity/lib/cjs/form/inputs/PortableText/object/renderers/PopoverEditDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverEditDialog = PopoverEditDialog;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _presence = require("../../../../../presence");

var _debug = require("./debug");

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const RootPopover = (0, _styledComponents.default)(_ui.Popover)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-popper-reference-hidden='true'] {\n    visibility: hidden;\n    pointer-events: none;\n  }\n\n  & > div {\n    overflow: hidden;\n  }\n"])));
const ContentContainer = (0, _styledComponents.default)(_ui.Container)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  &:not([hidden]) {\n    display: flex;\n  }\n  direction: column;\n"])));
const ContentScrollerBox = (0, _styledComponents.default)(_ui.Box)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  /* Prevent overflow caused by change indicator */\n  overflow-x: hidden;\n  overflow-y: auto;\n"])));
const ContentHeaderBox = (0, _styledComponents.default)(_ui.Box)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  background-color: var(--card-bg-color);\n  box-shadow: 0 1px 0 var(--card-shadow-outline-color);\n  position: relative;\n  z-index: 10;\n  min-height: auto;\n"])));
const POPOVER_FALLBACK_PLACEMENTS = ['top', 'bottom'];

function PopoverEditDialog(props) {
  const {
    width,
    elementRef,
    onClose,
    scrollElement
  } = props;
  (0, _ui.useGlobalKeyDown)((0, _react.useCallback)(event => {
    if (event.key === 'Escape') {
      onClose();
    }
  }, [onClose]));
  const [forceUpdate, setForceUpdate] = (0, _react.useState)(0);
  const virtualElement = (0, _react.useMemo)(() => {
    var _elementRef$current;

    if (!(elementRef !== null && elementRef !== void 0 && (_elementRef$current = elementRef.current) !== null && _elementRef$current !== void 0 && _elementRef$current.getBoundingClientRect())) {
      return null;
    }

    return {
      contextElement: elementRef.current || undefined,
      getBoundingClientRect: () => {
        var _elementRef$current2;

        return ((_elementRef$current2 = elementRef.current) === null || _elementRef$current2 === void 0 ? void 0 : _elementRef$current2.getBoundingClientRect()) || null;
      }
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [elementRef === null || elementRef === void 0 ? void 0 : elementRef.current, forceUpdate]);
  const [rootElement, setRootElement] = (0, _react.useState)(null);
  const boundaryElement = (0, _ui.useBoundaryElement)();
  const boundaryElementRect = (0, _ui.useElementRect)(boundaryElement.element);
  const contentStyle = (0, _react.useMemo)(() => ({
    opacity: boundaryElementRect ? undefined : 0,
    width: boundaryElementRect ? "".concat(boundaryElementRect.width - 16, "px") : undefined
  }), [boundaryElementRect]);
  const handleScrollOrResize = (0, _react.useCallback)(() => {
    setForceUpdate(forceUpdate + 1);
  }, [forceUpdate]);
  (0, _react.useEffect)(() => {
    if (scrollElement) {
      scrollElement.addEventListener('scroll', handleScrollOrResize, true);
    }

    return () => {
      if (scrollElement) {
        scrollElement.removeEventListener('scroll', handleScrollOrResize, true);
      }
    };
  }, [handleScrollOrResize, scrollElement]);
  return /*#__PURE__*/_react.default.createElement(RootPopover, {
    constrainSize: true,
    content: /*#__PURE__*/_react.default.createElement(Content, _extends({}, props, {
      rootElement: rootElement,
      style: contentStyle,
      width: width
    })),
    fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS,
    placement: "bottom",
    open: true,
    portal: "default",
    ref: setRootElement,
    referenceElement: virtualElement || _debug.debugElement
  });
}

function Content(props) {
  var _portal$elements;

  const {
    onClose,
    rootElement,
    style,
    width = 0,
    title
  } = props;
  const {
    element: boundaryElement
  } = (0, _ui.useBoundaryElement)();
  const portal = (0, _ui.usePortal)();
  (0, _ui.useClickOutside)(onClose, [rootElement], boundaryElement);
  return /*#__PURE__*/_react.default.createElement(ContentContainer, {
    style: style,
    width: width
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(ContentHeaderBox, {
    padding: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "semibold"
  }, title)), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    icon: _icons.CloseIcon,
    mode: "bleed",
    onClick: onClose,
    padding: 2
  }))), /*#__PURE__*/_react.default.createElement(ContentScrollerBox, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
    margins: [0, 0, 1, 0]
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.PortalProvider, {
    element: (_portal$elements = portal.elements) === null || _portal$elements === void 0 ? void 0 : _portal$elements.default
  }, props.children))))));
}
