// sanity/lib/esm/desk/panes/document/menuItems.js
import { EarthAmericasIcon, BinaryDocumentIcon, RestoreIcon } from '@sanity/icons';

const getHistoryMenuItem = params => {
  const {
    features,
    hasValue,
    changesOpen
  } = params;
  if (!features.reviewChanges) return null;
  return {
    action: 'reviewChanges',
    title: 'Review changes',
    icon: RestoreIcon,
    isDisabled: changesOpen || !hasValue
  };
};

const getInspectItem = _ref => {
  let {
    hasValue
  } = _ref;
  return {
    action: 'inspect',
    title: 'Inspect',
    icon: BinaryDocumentIcon,
    isDisabled: !hasValue,
    shortcut: 'Ctrl+Alt+I'
  };
};

export const getProductionPreviewItem = _ref2 => {
  let {
    previewUrl
  } = _ref2;
  if (!previewUrl) return null;
  return {
    action: 'production-preview',
    title: 'Open preview',
    icon: EarthAmericasIcon,
    shortcut: 'Ctrl+Alt+O'
  };
};
export const getMenuItems = params => {
  const items = [getProductionPreviewItem, getHistoryMenuItem, getInspectItem].filter(Boolean).map(fn => fn(params));
  return items.filter(i => i !== null);
};
