// sanity/lib/cjs/field/types/image/diff/ImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImagePreview = ImagePreview;
exports.NoImagePreview = void 0;

var _react = _interopRequireWildcard(require("react"));

var _assetUtils = require("@sanity/asset-utils");

var _imageUrl = _interopRequireDefault(require("@sanity/image-url"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _color = require("@sanity/color");

var _hooks = require("../../../../hooks");

var _diff = require("../../../diff");

var _datastores = require("../../../../datastores");

var _helpers = require("./helpers");

var _HotspotCropSVG = require("./HotspotCropSVG");

var _templateObject, _templateObject2, _templateObject3;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const ASSET_FIELDS = ['originalFilename']; // To trigger error state, change `src` attribute to random string ("foo")
// To trigger slow loading, use a throttling proxy (charles) or browser devtools
// To trigger deleted state, set `id` to valid, non-existant image asset ID,
// eg: 'image-1217bc35db5030739b7be571c79d3c401551911d-300x200-png'

const NoImagePreview = () => /*#__PURE__*/_react.default.createElement(_ui.Card, {
  flex: 1,
  tone: "transparent",
  padding: 4,
  radius: 2,
  height: "stretch"
}, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
  align: "center",
  justify: "center",
  height: "fill"
}, /*#__PURE__*/_react.default.createElement(_ui.Text, {
  size: 1,
  muted: true
}, "(no image)")));

exports.NoImagePreview = NoImagePreview;

const ImageWrapper = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 100%;\n  max-height: 190px;\n  position: relative;\n\n  /* Ideally the checkerboard component currently in the form builder should be made available and used here */\n  background-color: ", ";\n  background-image: linear-gradient(45deg, ", " 25%, transparent 25%),\n    linear-gradient(-45deg, ", " 25%, transparent 25%),\n    linear-gradient(45deg, transparent 75%, ", " 75%),\n    linear-gradient(-45deg, transparent 75%, ", " 75%);\n  background-size: 16px 16px;\n  background-position: 0 0, 0 8px, 8px -8px, -8px 0;\n\n  &::after {\n    content: '';\n    display: block;\n    position: absolute;\n    top: 0;\n    left: 0;\n    right: 0;\n    bottom: 0;\n    box-shadow: inset 0 0 0 1px var(--card-border-color);\n    pointer-events: none;\n  }\n\n  &[data-changed] {\n    opacity: 0.45;\n  }\n"])), _color.hues.gray[100].hex, _color.hues.gray[50].hex, _color.hues.gray[50].hex, _color.hues.gray[50].hex, _color.hues.gray[50].hex);

const Image = _styledComponents.default.img(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  display: block;\n  flex: 1;\n  min-height: 0;\n  object-fit: contain;\n  width: 100%;\n  height: 100%;\n\n  &[data-action='removed'] {\n    opacity: 0.45;\n  }\n"])));

const HotspotDiff = _styledComponents.default.div(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  svg {\n    display: block;\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n  }\n"])));

function ImagePreview(props) {
  var _diff$fromValue, _diff$fromValue$asset, _diff$toValue, _diff$toValue$asset;

  const {
    id,
    action,
    diff,
    hotspot,
    crop,
    is
  } = props;
  const client = (0, _hooks.useClient)();

  const [imageError, setImageError] = _react.default.useState();

  const {
    value: asset
  } = (0, _datastores.useDocumentValues)(id, ASSET_FIELDS);
  const dimensions = (0, _assetUtils.getImageDimensions)(id);
  const imageBuilder = (0, _react.useMemo)(() => (0, _imageUrl.default)(client), [client]); // undefined = still loading, null = its gone

  const assetIsDeleted = asset === null;
  const title = asset && asset.originalFilename || 'Untitled';
  const imageSource = imageBuilder.image(id).height(190) // Should match container max-height
  .dpr((0, _helpers.getDeviceDpr)()).fit('max');
  const assetChanged = ((_diff$fromValue = diff.fromValue) === null || _diff$fromValue === void 0 ? void 0 : (_diff$fromValue$asset = _diff$fromValue.asset) === null || _diff$fromValue$asset === void 0 ? void 0 : _diff$fromValue$asset._ref) !== ((_diff$toValue = diff.toValue) === null || _diff$toValue === void 0 ? void 0 : (_diff$toValue$asset = _diff$toValue.asset) === null || _diff$toValue$asset === void 0 ? void 0 : _diff$toValue$asset._ref);
  const metaAction = action === 'changed' ? undefined : action;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    height: "fill",
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 2,
    paddingBottom: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    as: ImageWrapper,
    direction: "column",
    "data-changed": is === 'from' && assetChanged ? '' : undefined,
    "data-error": imageError ? '' : undefined
  }, !assetIsDeleted && !imageError && /*#__PURE__*/_react.default.createElement(Image, {
    src: imageSource.toString() || '',
    alt: title,
    "data-action": metaAction,
    onError: setImageError,
    width: dimensions.width,
    height: dimensions.height
  }), (assetIsDeleted || imageError) && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingY: 5
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true,
    align: "center"
  }, assetIsDeleted ? 'Image is deleted' : 'Error loading image')), /*#__PURE__*/_react.default.createElement(HotspotDiff, null, /*#__PURE__*/_react.default.createElement(_HotspotCropSVG.HotspotCropSVG, {
    crop: crop && !(0, _assetUtils.isDefaultCrop)(crop) ? crop : undefined,
    diff: diff,
    hash: (0, _helpers.simpleHash)("".concat(imageSource.toString() || '', "-").concat(is)),
    hotspot: hotspot && !(0, _assetUtils.isDefaultHotspot)(hotspot) ? hotspot : undefined,
    width: dimensions.width,
    height: dimensions.height
  })))), /*#__PURE__*/_react.default.createElement(_diff.MetaInfo, {
    title: title,
    icon: _icons.ImageIcon,
    markRemoved: assetChanged && is === 'from'
  }, metaAction ? /*#__PURE__*/_react.default.createElement("div", null, metaAction) : /*#__PURE__*/_react.default.createElement("div", null, dimensions.width, " \xD7 ", dimensions.height)));
}
