// sanity/lib/cjs/components/DefaultDocument.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultDocument = DefaultDocument;

var _react = _interopRequireDefault(require("react"));

var _Favicons = require("./Favicons");

var _globalErrorHandler = require("./globalErrorHandler");

var _NoJavascript = require("./NoJavascript");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const globalStyles = "\n  html {\n    background-color: #f1f3f6;\n  }\n  html,\n  body,\n  #sanity {\n    height: 100%;\n  }\n  body {\n    margin: 0;\n    -webkit-font-smoothing: antialiased;\n  }\n";
const EMPTY_ARRAY = [];

function DefaultDocument(props) {
  const {
    entryPath,
    css = EMPTY_ARRAY,
    basePath = '/'
  } = props;
  return /*#__PURE__*/_react.default.createElement("html", {
    lang: "en"
  }, /*#__PURE__*/_react.default.createElement("head", null, /*#__PURE__*/_react.default.createElement("meta", {
    charSet: "utf-8"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1, viewport-fit=cover"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "robots",
    content: "noindex"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "referrer",
    content: "same-origin"
  }), /*#__PURE__*/_react.default.createElement(_Favicons.Favicons, {
    basePath: basePath
  }), /*#__PURE__*/_react.default.createElement("title", null, "Sanity Studio"), /*#__PURE__*/_react.default.createElement(_globalErrorHandler.GlobalErrorHandler, null), css.map(href => /*#__PURE__*/_react.default.createElement("link", {
    key: href,
    rel: "stylesheet",
    href: href
  })), /*#__PURE__*/_react.default.createElement("style", null, globalStyles)), /*#__PURE__*/_react.default.createElement("body", null, /*#__PURE__*/_react.default.createElement("div", {
    id: "sanity"
  }), /*#__PURE__*/_react.default.createElement("script", {
    type: "module",
    src: entryPath
  }), /*#__PURE__*/_react.default.createElement(_NoJavascript.NoJavascript, null)));
}
