// sanity/lib/cjs/components/InsufficientPermissionsMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InsufficientPermissionsMessage = InsufficientPermissionsMessage;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function InsufficientPermissionsMessage(props) {
  const {
    currentUser,
    title = 'Insufficient permissions',
    operationLabel = 'access this feature'
  } = props;
  const roles = (currentUser === null || currentUser === void 0 ? void 0 : currentUser.roles) || [];
  const plural = roles.length !== 1;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.AccessDeniedIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "semibold"
  }, title)), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, roles.length === 0 ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "You have no role that grants you permission to ", operationLabel) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Your role", plural && 's', ' ', join(roles.map(r => /*#__PURE__*/_react.default.createElement("code", {
    key: r.name
  }, r.title)), /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ", ")), ' ', "do", plural || 'es', " not have permissions to ", operationLabel))));
}

function join(array, sep) {
  return array.reduce((result, item) => result === null ? [item] : [...result, sep, item], null);
}
