// sanity/lib/cjs/components/WithReferringDocuments.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WithReferringDocuments = void 0;

var _reactPropsStream = require("react-props-stream");

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

const WithReferringDocuments = (0, _reactPropsStream.withPropsStream)(connect, function _WithReferringDocuments(_ref) {
  let {
    children,
    ...props
  } = _ref;
  return children(props);
});
exports.WithReferringDocuments = WithReferringDocuments;

function connect(receivedProps$) {
  return receivedProps$.pipe((0, _operators.distinctUntilChanged)((prev, next) => prev.id === next.id), (0, _operators.switchMap)(receivedProps => (0, _rxjs.concat)((0, _rxjs.of)({ ...receivedProps,
    referringDocuments: [],
    isLoading: true
  }), receivedProps.documentStore.listenQuery('*[references($docId)] [0...101]', {
    docId: receivedProps.id
  }, {
    tag: 'with-referring-documents'
  }).pipe((0, _operators.map)(docs => ({ ...receivedProps,
    referringDocuments: docs,
    isLoading: false
  }))))));
}
