// sanity/lib/esm/form/inputs/files/ImageInput/ImagePreview.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useCallback, useEffect, useState } from 'react';
import { AccessDeniedIcon, ImageIcon, ReadOnlyIcon } from '@sanity/icons';
import { Box, Card, Heading, Text, useElementRect } from '@sanity/ui';
import { MAX_DEFAULT_HEIGHT, RatioBox, Overlay, FlexOverlay, SpinnerWrapper } from './ImagePreview.styled';

/*
  Used for setting the initial image height - specifically for images
  that are small and so can take less space in the document
*/
const getImageSize = src => {
  const imageUrlParams = new URLSearchParams(src.split('?')[1]);
  const rect = imageUrlParams.get('rect');

  if (rect) {
    return [rect.split(',')[2], rect.split(',')[3]].map(Number);
  }

  return src.split('-')[1].split('.')[0].split('x').map(Number);
};

export function ImagePreview(props) {
  const {
    drag,
    readOnly,
    isRejected,
    src,
    ...rest
  } = props;
  const [isLoaded, setLoaded] = useState(false);
  const [rootElement, setRootElement] = useState(null);
  const rootRect = useElementRect(rootElement);
  const rootWidth = (rootRect === null || rootRect === void 0 ? void 0 : rootRect.width) || 0;
  const acceptTone = isRejected || readOnly ? 'critical' : 'primary';
  const tone = drag ? acceptTone : 'default';
  const maxHeightToPx = MAX_DEFAULT_HEIGHT * document.documentElement.clientHeight / 100; // convert from vh to px, max height of the input

  const [imageWidth, imageHeight] = getImageSize(src);
  const imageRatio = imageWidth / imageHeight; // is the image wider than root? if so calculate the resized height

  const renderedImageHeight = imageWidth > rootWidth ? rootWidth / imageRatio : imageHeight;
  /*
    if the rendered image is smaller than the max height then it doesn't require a height set
    otherwise, set the max height (to prevent a large image in the document)
  */

  const rootHeight = renderedImageHeight < maxHeightToPx ? null : "".concat(MAX_DEFAULT_HEIGHT, "vh");
  useEffect(() => {
    /* set for when the src is being switched when the image input already had a image src
    - meaning it already had an asset */
    setLoaded(false);
  }, [src]);
  const onLoadChange = useCallback(() => {
    setLoaded(true);
  }, []);
  return /*#__PURE__*/React.createElement(RatioBox, _extends({}, rest, {
    ref: setRootElement,
    style: {
      height: rootHeight
    },
    tone: "transparent"
  }), /*#__PURE__*/React.createElement(Card, {
    "data-container": true,
    tone: "inherit"
  }, !isLoaded && /*#__PURE__*/React.createElement(OverlayComponent, {
    cardTone: "transparent",
    drag: true,
    content: /*#__PURE__*/React.createElement(SpinnerWrapper, null)
  }), /*#__PURE__*/React.createElement("img", {
    src: src,
    "data-testid": "hotspot-image-input",
    alt: props.alt,
    onLoad: onLoadChange,
    referrerPolicy: "strict-origin-when-cross-origin"
  })), drag && /*#__PURE__*/React.createElement(OverlayComponent, {
    cardTone: tone,
    drag: drag,
    content: /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Box, {
      marginBottom: 3
    }, /*#__PURE__*/React.createElement(Heading, null, /*#__PURE__*/React.createElement(HoverIcon, {
      isRejected: isRejected,
      readOnly: readOnly
    }))), /*#__PURE__*/React.createElement(HoverText, {
      isRejected: isRejected,
      readOnly: readOnly
    }))
  }));
}

function HoverIcon(_ref) {
  let {
    isRejected,
    readOnly
  } = _ref;

  if (isRejected) {
    return /*#__PURE__*/React.createElement(AccessDeniedIcon, null);
  }

  if (readOnly) {
    return /*#__PURE__*/React.createElement(ReadOnlyIcon, null);
  }

  return /*#__PURE__*/React.createElement(ImageIcon, null);
}

function HoverText(_ref2) {
  let {
    isRejected,
    readOnly
  } = _ref2;
  let message = 'Drop image to upload';

  if (isRejected) {
    message = 'Cannot upload this file here';
  }

  if (readOnly) {
    message = 'This field is read only';
  }

  return /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, message);
}

function OverlayComponent(_ref3) {
  let {
    cardTone,
    drag,
    content
  } = _ref3;
  return /*#__PURE__*/React.createElement(Overlay, {
    justify: "flex-end",
    padding: 3,
    tone: cardTone,
    drag: drag
  }, /*#__PURE__*/React.createElement(FlexOverlay, {
    direction: "column",
    align: "center",
    justify: "center"
  }, content));
}
