// sanity/lib/cjs/datastores/__workshop__/CurrentUserStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CurrentUserStory;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _studio = require("../../studio");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CurrentUserStory() {
  const {
    currentUser
  } = (0, _studio.useSource)();
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold"
  }, /*#__PURE__*/_react.default.createElement("code", null, "userStore.me")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    hidden: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Subscribe"
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json",
    size: 1
  }, JSON.stringify(currentUser, null, 2))));
}
