// sanity/lib/cjs/datastores/__workshop__/DocumentPairStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DocumentPairStory;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _datastores = require("../datastores");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function DocumentPairStory() {
  const documentStore = (0, _datastores.useDocumentStore)();
  const [draftSnapshot, setDraftSnapshot] = (0, _react.useState)(null);
  const [publishedSnapshot, setPublishedSnapshot] = (0, _react.useState)(null);
  (0, _react.useEffect)(() => {
    const {
      draft,
      published
    } = documentStore.checkoutPair({
      draftId: 'drafts.test',
      publishedId: 'test'
    });
    const draftSub = draft.remoteSnapshot$.subscribe(setDraftSnapshot);
    const publishedSub = published.remoteSnapshot$.subscribe(setPublishedSnapshot);
    return () => {
      draftSub.unsubscribe();
      publishedSub.unsubscribe();
    };
  }, [documentStore]);
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold"
  }, /*#__PURE__*/_react.default.createElement("code", null, "documentStore.checkoutPair({draftId: string, publishedId: string})")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    hidden: true
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    text: "Subscribe"
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
    language: "json",
    size: 1
  }, JSON.stringify({
    draft: {
      snapshot: draftSnapshot
    },
    published: {
      snapshot: publishedSnapshot
    }
  }, null, 2))));
}
