// sanity/lib/cjs/cli/commands/hook/createHookCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _open = _interopRequireDefault(require("open"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const createHookCommand = {
  name: 'create',
  group: 'hook',
  signature: '',
  helpText: '',
  description: 'Create a new hook for the given dataset',
  action: async (args, context) => {
    const {
      apiClient,
      output
    } = context;
    const client = apiClient();
    const {
      projectId
    } = client.config();

    if (!projectId) {
      throw new Error('No project ID found');
    }

    const projectInfo = (await client.projects.getById(projectId)) || {};
    const organizationId = projectInfo.organizationId || 'personal';
    const manageUrl = "https://www.sanity.io/organizations/".concat(organizationId, "/project/").concat(projectId, "/api/webhooks/new");
    output.print("Opening ".concat(manageUrl));
    (0, _open.default)(manageUrl);
  }
};
var _default = createHookCommand;
exports.default = _default;
