// sanity/lib/cjs/cli/commands/hook/listHooksCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const listHooksCommand = {
  name: 'list',
  group: 'hook',
  signature: '',
  helpText: '',
  description: 'List hooks for a given project',
  action: async (args, context) => {
    const {
      apiClient,
      output
    } = context;
    const client = apiClient();
    let hooks;

    try {
      hooks = await client.clone().config({
        apiVersion: '2021-10-04'
      }).request({
        uri: '/hooks'
      });
    } catch (err) {
      throw new Error("Hook list retrieval failed:\n".concat(err.message));
    }

    hooks.forEach(hook => {
      output.print("Name: ".concat(hook.name));
      output.print("Dataset: ".concat(hook.dataset));
      output.print("URL: ".concat(hook.url));

      if (hook.type === 'document') {
        output.print("HTTP method: ".concat(hook.httpMethod));

        if (hook.description) {
          output.print("Description: ".concat(hook.description));
        }
      }

      output.print('');
    });
  }
};
var _default = listHooksCommand;
exports.default = _default;
