// sanity/lib/esm/form/patch/PatchEvent.js
import _flatten from "lodash/flatten";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { prefixPath } from './patch';

/**
 * @internal
 */
export class PatchEvent {
  static from(input) {
    if (input instanceof PatchEvent) {
      return input;
    }

    return new PatchEvent(Array.isArray(input) ? _flatten(input) : [input]);
  }

  constructor(patches) {
    _defineProperty(this, "patches", void 0);

    this.patches = patches;
  }

  prepend() {
    for (var _len = arguments.length, patches = new Array(_len), _key = 0; _key < _len; _key++) {
      patches[_key] = arguments[_key];
    }

    return PatchEvent.from([..._flatten(patches), ...this.patches]);
  }

  append() {
    for (var _len2 = arguments.length, patches = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      patches[_key2] = arguments[_key2];
    }

    return PatchEvent.from([...this.patches, ..._flatten(patches)]);
  }

  prefixAll(segment) {
    return PatchEvent.from(this.patches.map(patch => prefixPath(patch, segment)));
  }

}
