// sanity/lib/esm/router/findMatchingRoutes.js
import _pick from "lodash/pick";
import _isPlainObject from "lodash/isPlainObject";
import _intersection from "lodash/intersection";
import _difference from "lodash/difference";
import { arrayify } from './utils/arrayify';

function isRecord(value) {
  return _isPlainObject(value);
}

function createMatchResult(nodes, missing, remaining) {
  return {
    nodes,
    missing,
    remaining
  };
}

export function findMatchingRoutes(node, _state) {
  if (!_state) {
    return createMatchResult([], [], []);
  }

  const state = node.scope ? _state[node.scope] : _state;
  const requiredParams = node.route.segments.filter(seg => seg.type === 'param').map(seg => seg.name);
  const stateKeys = isRecord(state) ? Object.keys(state) : [];

  const consumedParams = _intersection(stateKeys, requiredParams);

  const missingParams = _difference(requiredParams, consumedParams);

  const remainingParams = _difference(stateKeys, consumedParams);

  if (missingParams.length > 0) {
    return createMatchResult([], missingParams, []);
  }

  if (remainingParams.length === 0) {
    return createMatchResult([node], [], []);
  }

  const children = arrayify((typeof node.children === 'function' ? node.children(isRecord(state) ? state : {}) : node.children) || []);

  if (remainingParams.length > 0 && children.length === 0) {
    return createMatchResult([], remainingParams, []);
  }

  const remainingState = _pick(state, remainingParams);

  let matchingChild = {
    nodes: [],
    remaining: [],
    missing: []
  };
  arrayify(children).some(childNode => {
    matchingChild = findMatchingRoutes(childNode, remainingState);
    return matchingChild.nodes.length > 0;
  });

  if (matchingChild.nodes.length === 0) {
    return createMatchResult([], missingParams, remainingParams);
  }

  return createMatchResult([node, ...matchingChild.nodes], matchingChild.missing, matchingChild.remaining);
}
