// sanity/lib/esm/datastores/authStore/createAuthStore.js
import _memoize from "lodash/memoize";
import createClient from '@sanity/client';
import { defer } from 'rxjs';
import { map, shareReplay, startWith, switchMap } from 'rxjs/operators';
import { checkCors, CorsOriginError } from '../cors';
import { createBroadcastChannel } from './createBroadcastChannel';
import { sessionId } from './sessionId';
import * as storage from './storage';
import { createLoginComponent } from './createLoginComponent';

const getStorageKey = projectId => {
  // Project ID is part of the localStorage key so that different projects can
  // store their separate tokens, and it's easier to do book keeping.
  if (!projectId) throw new Error('Invalid project id');
  return "__studio_auth_token_".concat(projectId);
};

const getToken = projectId => {
  try {
    const item = storage.getItem(getStorageKey(projectId));

    if (item) {
      const {
        token
      } = JSON.parse(item);

      if (token && typeof token === 'string') {
        return token;
      }
    }
  } catch (err) {
    console.error(err);
  }

  return null;
};

const clearToken = projectId => {
  try {
    storage.removeItem(getStorageKey(projectId));
  } catch (err) {
    console.error(err);
  }
};

const saveToken = _ref => {
  let {
    token,
    projectId
  } = _ref;

  try {
    storage.setItem(getStorageKey(projectId), JSON.stringify({
      token,
      time: new Date().toISOString()
    }));
  } catch (err) {
    console.error(err);
  }
};

const getCurrentUser = async client => {
  const result = await checkCors(client);

  if (result !== null && result !== void 0 && result.isCorsError) {
    var _client$config;

    throw new CorsOriginError({ ...result,
      projectId: (_client$config = client.config()) === null || _client$config === void 0 ? void 0 : _client$config.projectId
    });
  }

  try {
    const user = await client.request({
      uri: '/users/me',
      withCredentials: true,
      tag: 'users.get-current'
    }); // if the user came back with an id, assume it's a full CurrentUser

    return typeof (user === null || user === void 0 ? void 0 : user.id) === 'string' ? user : null;
  } catch (err) {
    if (err.statusCode === 401) return null;
    throw err;
  }
};
/**
 * @internal
 */


export function _createAuthStore(_ref2) {
  let {
    projectId,
    dataset,
    loginMethod = 'dual',
    ...providerOptions
  } = _ref2;
  // this broadcast channel receives either a token as a `string` or `null`.
  // a new client will be created from it, otherwise, it'll only trigger a retry
  // for cookie-based auth
  const {
    broadcast,
    messages
  } = createBroadcastChannel("dual_mode_auth_".concat(projectId)); // // TODO: there is currently a bug where the AuthBoundary flashes the
  // // `NotAuthenticatedComponent` on the first load after a login with
  // // cookieless mode. A potential solution to fix this bug is to delay
  // // emitting `state$` until the session ID has been converted to a token
  // const firstMessage = messages.pipe(first())

  const token$ = messages.pipe(startWith(loginMethod === 'dual' ? getToken(projectId) : null));
  const state$ = token$.pipe( // // see above
  // debounce(() => firstMessage),
  map(token => createClient({
    projectId,
    dataset,
    apiVersion: '2021-06-07',
    useCdn: false,
    ...(token && {
      token
    }),
    withCredentials: true,
    requestTagPrefix: 'sanity.studio',
    ignoreBrowserTokenWarning: true
  })), switchMap(client => defer(async () => {
    const currentUser = await getCurrentUser(client);
    return {
      currentUser,
      client,
      authenticated: !!currentUser
    };
  })), shareReplay(1));

  async function handleCallbackUrl() {
    if (sessionId && loginMethod === 'dual') {
      const requestClient = createClient({
        projectId,
        dataset,
        useCdn: true,
        withCredentials: true,
        apiVersion: '2021-06-07',
        requestTagPrefix: 'sanity.studio'
      }); // try to get the current user by using the cookie credentials

      const currentUser = await getCurrentUser(requestClient);

      if (currentUser) {
        // if that worked, then we don't need to fetch a token
        broadcast(null);
      } else {
        // if that didn't work, then we need to trade the session ID for a token
        const {
          token
        } = await requestClient.request({
          method: 'GET',
          uri: "/auth/fetch",
          query: {
            sid: sessionId
          },
          tag: 'auth.fetch-token'
        });
        saveToken({
          token,
          projectId
        });
        broadcast(token);
      }
    } else {
      broadcast(loginMethod === 'dual' ? getToken(projectId) : null);
    }
  }

  async function logout() {
    const requestClient = createClient({
      projectId,
      dataset,
      useCdn: true,
      withCredentials: true,
      apiVersion: '2021-06-07',
      requestTagPrefix: 'sanity.studio'
    });
    clearToken(projectId);
    await requestClient.auth.logout();
    broadcast(null);
  }

  const LoginComponent = createLoginComponent({ ...providerOptions,
    getClient: () => state$.pipe(map(state => state.client)),
    loginMethod
  });
  return {
    handleCallbackUrl,
    token: token$,
    state: state$,
    LoginComponent,
    logout
  };
}

function hash(value) {
  if (typeof value !== 'object' || value === null) return "".concat(value); // note: this code path works for arrays as well as objects

  return JSON.stringify(Object.fromEntries(Object.entries(value).sort((_ref3, _ref4) => {
    let [a] = _ref3;
    let [b] = _ref4;
    return a.localeCompare(b, 'en');
  }).map(_ref5 => {
    let [k, v] = _ref5;
    return [k, hash(v)];
  })));
}
/**
 * @public
 */


const createAuthStore = _memoize(_createAuthStore, hash);

export { createAuthStore };
