// sanity/lib/esm/field/__workshop__/ChangeResolverStory.js
import { Box, Card } from '@sanity/ui';
import React, { useCallback, useMemo } from 'react';
import { useSchema } from '../../hooks';
import { ChangeResolver, DocumentChangeContext } from '../diff';
export default function ChangeResolverStory() {
  const documentId = 'test';
  const documentType = 'author';
  const schema = useSchema();
  const schemaType = schema.get(documentType);
  const nameDiff = useMemo(() => ({
    type: 'string',
    action: 'changed',
    isChanged: true,
    fromValue: '',
    toValue: 'Test',
    annotation: {
      chunk: {
        index: 0,
        id: 'foo',
        type: 'editDraft',
        start: 0,
        end: 0,
        startTimestamp: new Date('2021-01-01').toJSON(),
        endTimestamp: new Date('2021-01-02').toJSON(),
        authors: new Set(['p27ewL8aM']),
        draftState: 'present',
        publishedState: 'missing'
      },
      timestamp: new Date('2021-01-02').toJSON(),
      author: 'p27ewL8aM'
    },
    segments: []
  }), []); // export interface FieldChangeNode {
  //   type: 'field'
  //   diff: Diff
  //   itemDiff?: ItemDiff
  //   parentDiff?: ObjectDiff | ArrayDiff
  //   key: string
  //   path: Path
  //   error?: FieldValueError
  //   titlePath: ChangeTitlePath
  //   schemaType: ObjectFieldType
  //   showHeader: boolean
  //   showIndex: boolean
  //   diffComponent?: DiffComponent
  //   parentSchema?: ArraySchemaType | ObjectSchemaType
  //   readOnly?: ConditionalProperty
  //   hidden?: ConditionalProperty
  // }

  const change = useMemo(() => ({
    type: 'field',
    diff: nameDiff,
    key: 'name',
    path: ['name'],
    parentSchema: schemaType,
    schemaType: schemaType.fields.find(f => f.name === 'name'),
    titlePath: ['Name'],
    showIndex: true,
    showHeader: true
  }), [nameDiff, schemaType]);
  const FieldWrapper = useCallback(_props => {
    // console.log('props', _props)
    return /*#__PURE__*/React.createElement(Card, null, _props.children);
  }, []);
  const diff = useMemo(() => ({
    type: 'object',
    action: 'changed',
    isChanged: true,
    fields: {
      name: nameDiff
    },
    fromValue: {
      name: ''
    },
    toValue: {
      name: 'Test'
    },
    annotation: {
      chunk: {
        index: 0,
        id: 'foo',
        type: 'editDraft',
        start: 0,
        end: 0,
        startTimestamp: new Date('2021-01-01').toJSON(),
        endTimestamp: new Date('2021-01-02').toJSON(),
        authors: new Set(['p27ewL8aM']),
        draftState: 'present',
        publishedState: 'missing'
      },
      timestamp: new Date('2021-01-02').toJSON(),
      author: 'p27ewL8aM'
    }
  }), [nameDiff]);
  const documentContext = useMemo(() => ({
    documentId,
    FieldWrapper,
    isComparingCurrent: true,
    rootDiff: diff,
    schemaType,
    value: {
      name: 'Test'
    }
  }), [diff, documentId, FieldWrapper, schemaType]);
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(DocumentChangeContext.Provider, {
    value: documentContext
  }, /*#__PURE__*/React.createElement(ChangeResolver, {
    change: change
  })));
}
