// sanity/lib/esm/field/diff/components/DiffInspectWrapper.js
var _templateObject, _templateObject2;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { Box, Card, Code, Label, Stack } from '@sanity/ui';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import { pathToString } from '../../paths';
import { FromToArrow } from './FromToArrow';
const CodeWrapper = styled.pre(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  overflow-x: auto;\n  position: relative;\n"])));
const Meta = styled.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  right: 0;\n"])));
export function DiffInspectWrapper(props) {
  const {
    children,
    as,
    change,
    ...restProps
  } = props;
  const isHovering = useRef(false);
  const [isInspecting, setIsInspecting] = useState(false);
  const toggleInspect = useCallback(() => setIsInspecting(state => !state), [setIsInspecting]);
  const handleMouseEnter = useCallback(() => isHovering.current = true, []);
  const handleMouseLeave = useCallback(() => isHovering.current = false, [isHovering]);
  useEffect(() => {
    function onKeyDown(evt) {
      const {
        metaKey,
        key
      } = evt;

      if (metaKey && key === 'i' && isHovering.current) {
        toggleInspect();
      }
    }

    window.addEventListener('keydown', onKeyDown, false);
    return () => window.removeEventListener('keydown', onKeyDown, false);
  }, [toggleInspect]);
  return /*#__PURE__*/React.createElement(Box, _extends({
    as: as,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave
  }, restProps), isInspecting ? /*#__PURE__*/React.createElement(DiffInspector, {
    change: change
  }) : children);
}

const MetaLabel = _ref => {
  let {
    title
  } = _ref;
  return /*#__PURE__*/React.createElement(Box, {
    padding: 3,
    display: "inline-block",
    as: Meta
  }, /*#__PURE__*/React.createElement(Label, {
    size: 1,
    muted: true
  }, title));
};

function DiffInspector(_ref2) {
  var _change$itemDiff, _change$itemDiff2, _change$itemDiff3;

  let {
    change
  } = _ref2;
  return /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 3,
    tone: "transparent",
    as: CodeWrapper,
    radius: 1
  }, /*#__PURE__*/React.createElement(MetaLabel, {
    title: "meta"
  }), /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, printMeta({
    path: pathToString(change.path),
    fromIndex: (_change$itemDiff = change.itemDiff) === null || _change$itemDiff === void 0 ? void 0 : _change$itemDiff.fromIndex,
    toIndex: (_change$itemDiff2 = change.itemDiff) === null || _change$itemDiff2 === void 0 ? void 0 : _change$itemDiff2.toIndex,
    hasMoved: (_change$itemDiff3 = change.itemDiff) === null || _change$itemDiff3 === void 0 ? void 0 : _change$itemDiff3.hasMoved,
    action: change.diff.action,
    isChanged: change.diff.isChanged
  }))), /*#__PURE__*/React.createElement(Card, {
    as: CodeWrapper,
    tone: "critical",
    padding: 3,
    radius: 1
  }, /*#__PURE__*/React.createElement(MetaLabel, {
    title: "from"
  }), /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, jsonify(change.diff.fromValue))), /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(FromToArrow, {
    direction: "down",
    align: "center"
  })), /*#__PURE__*/React.createElement(Card, {
    as: CodeWrapper,
    tone: "positive",
    padding: 3,
    radius: 1
  }, /*#__PURE__*/React.createElement(MetaLabel, {
    title: "to"
  }), /*#__PURE__*/React.createElement(Code, {
    language: "json",
    size: 1
  }, jsonify(change.diff.toValue))));
}

function jsonify(value) {
  if (typeof value === 'undefined') {
    return 'undefined';
  }

  return JSON.stringify(value, null, 2);
}

function printMeta(keys) {
  const lines = [];
  Object.entries(keys).forEach(_ref3 => {
    let [key, value] = _ref3;

    if (typeof value !== 'undefined' && value !== null) {
      lines.push("".concat(key, ": ").concat(value));
    }
  });
  return lines.join('\n');
}
