// sanity/lib/cjs/components/changeIndicators/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STROKE_WIDTH = exports.INTERACTIVE_STROKE_WIDTH = exports.ENABLED = exports.DEBUG_LAYER_BOUNDS = exports.DEBUG = exports.CORNER_RADIUS = exports.CONNECTOR_MARGIN = exports.ARROW_THRESHOLD = exports.ARROW_SIZE = exports.ARROW_MARGIN_Y = exports.ARROW_MARGIN_X = void 0;
// Toggle tracking of change indicator DOM positions and connectors overlay rendering
const ENABLED = true;
exports.ENABLED = ENABLED;
const CORNER_RADIUS = 4;
exports.CORNER_RADIUS = CORNER_RADIUS;
const INTERACTIVE_STROKE_WIDTH = 16;
exports.INTERACTIVE_STROKE_WIDTH = INTERACTIVE_STROKE_WIDTH;
const CONNECTOR_MARGIN = 8;
exports.CONNECTOR_MARGIN = CONNECTOR_MARGIN;
const ARROW_MARGIN_X = 8;
exports.ARROW_MARGIN_X = ARROW_MARGIN_X;
const ARROW_MARGIN_Y = 2;
exports.ARROW_MARGIN_Y = ARROW_MARGIN_Y;
const ARROW_SIZE = 4;
exports.ARROW_SIZE = ARROW_SIZE;
const ARROW_THRESHOLD = 12;
exports.ARROW_THRESHOLD = ARROW_THRESHOLD;
const STROKE_WIDTH = 2;
exports.STROKE_WIDTH = STROKE_WIDTH;
const DEBUG = false;
exports.DEBUG = DEBUG;
const DEBUG_LAYER_BOUNDS = DEBUG || false;
exports.DEBUG_LAYER_BOUNDS = DEBUG_LAYER_BOUNDS;
