// sanity/lib/esm/preview/__workshop__/SanityPreviewStory.js
import { Box, Card, Container } from '@sanity/ui';
import { useSelect } from '@sanity/ui-workshop';
import React, { useMemo } from 'react';
import { useSchema } from '../../hooks';
import { SanityPreview } from '../components/SanityPreview';
const LAYOUT_OPTIONS = {
  Default: 'default',
  Detail: 'detail',
  Media: 'media',
  Inline: 'inline',
  Block: 'block',
  BlockImage: 'blockImage'
};
const VALUE_TYPES = {
  Document: 'document',
  Image: 'image'
};
export default function SanityPreviewStory() {
  const layout = useSelect('Layout', LAYOUT_OPTIONS);
  const type = useSelect('Value type', VALUE_TYPES);
  const schema = useSchema();
  const schemaType = useMemo(() => {
    if (type === 'image') {
      return schema.get('imagesTest').fields.find(f => f.name === 'mainImage').type;
    }

    return schema.get('author');
  }, [schema, type]);
  const value = useMemo(() => {
    if (type === 'image') {
      return {
        _type: 'image',
        asset: {
          _ref: 'image-4af4353791af3fd4594c59f5bdc9f5f4a4aba3db-6240x4160-jpg',
          _type: 'reference'
        }
      };
    }

    return {
      _id: 'grrm',
      _type: 'author'
    };
  }, [type]);
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Container, {
    width: 1
  }, /*#__PURE__*/React.createElement(Card, {
    border: true,
    padding: 2,
    radius: 2,
    style: {
      lineHeight: 0
    }
  }, /*#__PURE__*/React.createElement(SanityPreview, {
    layout: layout,
    schemaType: schemaType,
    value: value
  }))));
}
