// sanity/lib/esm/form/inputs/arrays/errors/DuplicateKeysAlert.js
import { Button, Stack, Text } from '@sanity/ui';
import React, { useCallback } from 'react';
import { FormField } from '../../../components/formField';
import { Alert } from '../../../components/Alert';
import { PatchEvent, set } from '../../../patch';
import { Details } from '../../../components/Details';
import { isDev } from '../../../../environment';
export function DuplicateKeysAlert(props) {
  const {
    error,
    onChange
  } = props;
  const handleFixDuplicateKeys = useCallback(() => {
    onChange(PatchEvent.from((error.duplicates || []).map(_ref => {
      let [index, key] = _ref;
      return set("".concat(key, "_deduped_").concat(index), [index, '_key']);
    })));
  }, [error.duplicates, onChange]);
  return /*#__PURE__*/React.createElement(FormField, {
    title: error.schemaType.title,
    description: error.schemaType.description
  }, /*#__PURE__*/React.createElement(Alert, {
    status: "warning",
    suffix: /*#__PURE__*/React.createElement(Stack, {
      padding: 2
    }, /*#__PURE__*/React.createElement(Button, {
      onClick: handleFixDuplicateKeys,
      text: "Generate unique keys",
      tone: "caution"
    })),
    title: /*#__PURE__*/React.createElement(React.Fragment, null, "Non-unique keys")
  }, /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Several items in this list share the same identifier (key). Every item must have an unique identifier."), /*#__PURE__*/React.createElement(Details, {
    marginTop: 4,
    open: isDev,
    title: /*#__PURE__*/React.createElement(React.Fragment, null, "Developer info")
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 3
  }, /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This usually happens when items are created using an API client, and the", ' ', /*#__PURE__*/React.createElement("code", null, "_key"), " property of each elements has been generated non-uniquely."), /*#__PURE__*/React.createElement(Text, {
    as: "p",
    muted: true,
    size: 1
  }, "The value of the ", /*#__PURE__*/React.createElement("code", null, "_key"), " property must be a unique string.")))));
}
