// sanity/lib/cjs/components/previews/portableText/BlockImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockImagePreview = BlockImagePreview;

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _Media = require("../_common/Media");

var _constants = require("../constants");

var _helpers = require("../helpers");

var _BlockImagePreview = require("./BlockImagePreview.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_MEDIA_DIMENSIONS = { ..._constants.PREVIEW_MEDIA_SIZE.blockImage,
  fit: 'fillmax',
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
};

const getRatio = dimensions => {
  const {
    height,
    width
  } = dimensions;

  if (!height || !width) {
    return 1;
  }

  return height / width * 100;
};

function BlockImagePreview(props) {
  const {
    actions,
    title,
    subtitle,
    description,
    fallbackTitle = 'Untitled',
    mediaDimensions = DEFAULT_MEDIA_DIMENSIONS,
    media,
    children,
    status
  } = props;
  return /*#__PURE__*/_react.default.createElement(_BlockImagePreview.RootBox, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_BlockImagePreview.HeaderFlex, {
    paddingLeft: 2,
    paddingRight: 1,
    paddingY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    textOverflow: "ellipsis",
    weight: "semibold"
  }, title || fallbackTitle), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, (0, _helpers.renderPreviewNode)(subtitle, 'block'))), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: 1,
    paddingLeft: 1
  }, status && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 2,
    paddingY: 3
  }, (0, _helpers.renderPreviewNode)(status, 'block')), actions)), /*#__PURE__*/_react.default.createElement(_BlockImagePreview.MediaCard, {
    $ratio: getRatio(mediaDimensions),
    __unstable_checkered: true,
    display: "flex",
    sizing: "border",
    tone: "inherit"
  }, /*#__PURE__*/_react.default.createElement(_Media.Media, {
    border: false,
    dimensions: mediaDimensions,
    layout: "blockImage",
    media: media,
    radius: 0,
    responsive: true
  }))), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 2,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1
  }, (0, _helpers.renderPreviewNode)(description, 'block'))), children && /*#__PURE__*/_react.default.createElement("div", null, children));
}
