// sanity/lib/esm/field/types/boolean/diff/BooleanFieldDiff.js
import { Box, Flex, Text } from '@sanity/ui';
import React from 'react';
import { DiffTooltip, FromToArrow, useDiffAnnotationColor } from '../../../diff';
import { Checkbox, Switch } from '../preview';
export const BooleanFieldDiff = _ref => {
  let {
    diff,
    schemaType
  } = _ref;
  const {
    fromValue,
    toValue
  } = diff;
  const {
    title,
    options
  } = schemaType;
  const Preview = (options === null || options === void 0 ? void 0 : options.layout) === 'checkbox' ? Checkbox : Switch;
  const userColor = useDiffAnnotationColor(diff, []);
  const showToValue = toValue !== undefined && toValue !== null;
  return /*#__PURE__*/React.createElement(Flex, {
    align: "center"
  }, /*#__PURE__*/React.createElement(DiffTooltip, {
    diff: diff
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center"
  }, /*#__PURE__*/React.createElement(Preview, {
    checked: fromValue,
    color: userColor
  }), showToValue && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Box, {
    marginX: 2
  }, /*#__PURE__*/React.createElement(FromToArrow, null)), /*#__PURE__*/React.createElement(Preview, {
    checked: toValue,
    color: userColor
  })))), showToValue && title && /*#__PURE__*/React.createElement(Box, {
    marginLeft: 2
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1,
    weight: "semibold"
  }, title)));
};
