// sanity/lib/cjs/desk/components/deskTool/NoDocumentTypesScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NoDocumentTypesScreen = NoDocumentTypesScreen;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function NoDocumentTypesScreen() {
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    radius: 2,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    marginLeft: 3,
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "h1",
    size: 1,
    weight: "bold"
  }, "No document types"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Please define at least one document type in your schema."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("a", {
    href: "https://beta.sanity.io/docs/platform/studio/config",
    target: "_blank",
    rel: "noreferrer"
  }, "Learn how to add a document type \u2192"))))))));
}
