// sanity/lib/cjs/form/types/asserters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isArrayOfObjectsInputProps = isArrayOfObjectsInputProps;
exports.isArrayOfPrimitivesInputProps = isArrayOfPrimitivesInputProps;
exports.isBooleanInputProps = isBooleanInputProps;
exports.isNumberInputProps = isNumberInputProps;
exports.isObjectInputProps = isObjectInputProps;
exports.isObjectItemProps = isObjectItemProps;
exports.isStringInputProps = isStringInputProps;

var _types = require("@sanity/types");

function isObjectItemProps(item) {
  return (0, _types.isObjectSchemaType)(item.schemaType);
}

function isObjectInputProps(inputProps) {
  return (0, _types.isObjectSchemaType)(inputProps.schemaType);
}

function isStringInputProps(inputProps) {
  return (0, _types.isStringSchemaType)(inputProps.schemaType);
}

function isNumberInputProps(inputProps) {
  return (0, _types.isNumberSchemaType)(inputProps.schemaType);
}

function isBooleanInputProps(inputProps) {
  return (0, _types.isBooleanSchemaType)(inputProps.schemaType);
}

function isArrayOfObjectsInputProps(inputProps) {
  return (0, _types.isArrayOfObjectsSchemaType)(inputProps.schemaType);
}

function isArrayOfPrimitivesInputProps(inputProps) {
  return (0, _types.isArrayOfPrimitivesSchemaType)(inputProps.schemaType);
}
