// sanity/lib/esm/studio/components/navbar/search/SearchPopoverContent.js
var _templateObject;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { Popover } from '@sanity/ui';
import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { useColorScheme } from '../../../colorScheme';
const ResultsPopover = styled(Popover)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  & > div {\n    min-height: 43px;\n    overflow: auto;\n    -webkit-overflow-scrolling: touch;\n  }\n\n  &[data-popper-reference-hidden='true'] {\n    display: none;\n  }\n"])));
export const SearchPopoverContent = /*#__PURE__*/forwardRef(function SearchPopoverContent(props, ref) {
  const {
    scheme
  } = useColorScheme();
  return /*#__PURE__*/React.createElement(ResultsPopover, _extends({
    arrow: false,
    constrainSize: true,
    matchReferenceWidth: true,
    placement: "bottom",
    portal: true,
    radius: 2,
    ref: ref,
    scheme: scheme
  }, props));
});
