// sanity/lib/esm/components/changeIndicators/overlay/DebugLayers.js
import React from 'react';
export function DebugLayers(_ref) {
  let {
    field,
    change
  } = _ref;
  return /*#__PURE__*/React.createElement("g", {
    style: {
      pointerEvents: 'none'
    }
  }, /*#__PURE__*/React.createElement("rect", {
    x: field.bounds.left,
    y: field.bounds.top,
    height: field.bounds.height,
    width: field.bounds.width,
    stroke: "green",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/React.createElement("rect", {
    x: field.rect.left,
    y: field.rect.top,
    height: field.rect.height,
    width: field.rect.width,
    stroke: "black",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/React.createElement("rect", {
    x: change.bounds.left,
    y: change.bounds.top,
    height: change.bounds.height,
    width: change.bounds.width,
    stroke: "crimson",
    fill: "none",
    strokeWidth: 1
  }), /*#__PURE__*/React.createElement("rect", {
    x: change.rect.left,
    y: change.rect.top,
    height: change.rect.height,
    width: change.rect.width,
    stroke: "black",
    fill: "none",
    strokeWidth: 1
  }));
}
