// sanity/lib/esm/form/__workshop__/_common/resolvePreviewComponent.js
import { Button, Card, Code, ErrorBoundary, Stack } from '@sanity/ui';
import React, { useCallback, useState } from 'react';
import { SanityPreview } from '../../../preview';
export const resolvePreviewComponent = () => TestPreview;

function TestPreview(props) {
  const [errorParams, setErrorParams] = useState(null);
  const handleCatch = useCallback(params => {
    setErrorParams(params);
  }, []);
  const handleRetry = useCallback(() => setErrorParams(null), []);

  if (errorParams) {
    return /*#__PURE__*/React.createElement(Card, {
      padding: 3,
      tone: "critical"
    }, /*#__PURE__*/React.createElement(Code, {
      language: "json"
    }, JSON.stringify(errorParams)), /*#__PURE__*/React.createElement(Stack, {
      marginTop: 3
    }, /*#__PURE__*/React.createElement(Button, {
      onClick: handleRetry,
      text: "Retry"
    })));
  }

  return /*#__PURE__*/React.createElement(ErrorBoundary, {
    onCatch: handleCatch
  }, /*#__PURE__*/React.createElement(SanityPreview, props));
}
