// sanity/lib/esm/form/__workshop__/_common/TypeTester.js
var _templateObject, _templateObject2;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React, { useCallback, useEffect, useRef, useState, useMemo } from 'react';
import { Button, Card, Label, Stack, Text, TextArea } from '@sanity/ui';
import styled, { createGlobalStyle } from 'styled-components';
import { runTest } from './typer';
const HIGHLIGHT_CLASSNAME = 'js-highlight-selected-input';
const GlobalStyle = createGlobalStyle(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  .", " {\n    outline: 2px solid red;\n  }\n"])), HIGHLIGHT_CLASSNAME);

function getNameFromInput(input) {
  const classNames = input.classList.toString().split(' ').filter(name => name !== HIGHLIGHT_CLASSNAME).join('.');
  return "".concat(input.tagName.toLowerCase(), ".").concat(classNames);
}

const Sticky = styled.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: sticky;\n  top: 1rem;\n  z-index: 2;\n"])));
export function TypeTester(_ref) {
  let {
    readOnly
  } = _ref;
  const textAreaRef = useRef(null);
  const [focusedInput, setFocusedInput] = useState();
  const [testOutput, setTestOutput] = useState('');
  const [isRunning, setIsRunning] = useState(false);
  const cancelTypeTester = useRef(() => true);
  const handleFocus = useCallback(e => {
    if (textAreaRef.current && e.target === textAreaRef.current) {
      return;
    }

    if (e.type === 'focusin' && e.target instanceof HTMLInputElement || e.target instanceof HTMLTextAreaElement) {
      setFocusedInput(e.target);
    }
  }, [textAreaRef]);
  useEffect(() => {
    document.addEventListener('focusin', handleFocus);
    return () => {
      document.removeEventListener('focusin', handleFocus);

      if (cancelTypeTester.current) {
        cancelTypeTester.current();
      }
    };
  }, [handleFocus]);
  const handleRunTest = useCallback(() => {
    setIsRunning(true);
    setTestOutput('Running test...');
    cancelTypeTester.current = runTest({
      inputElement: focusedInput,
      times: 4,
      gracePeriod: 4000,
      onRun: output => {
        setTestOutput(existing => "".concat(existing, "\n").concat(output));
      },
      onFinished: () => {
        setIsRunning(false);
      }
    });
  }, [focusedInput]);
  const handleHighlightMouseOver = useCallback(event => {
    if (event.type === 'mouseover') {
      focusedInput.classList.add(HIGHLIGHT_CLASSNAME);
    }

    if (event.type === 'mouseout') {
      focusedInput.classList.remove(HIGHLIGHT_CLASSNAME);
    }
  }, [focusedInput]);
  const handleHighlightClick = useCallback(() => {
    if (focusedInput) {
      focusedInput.focus();
    }
  }, [focusedInput]);
  const isDisabled = useMemo(() => {
    return isRunning || !focusedInput || !['INPUT', 'TEXTAREA'].includes(focusedInput.tagName) || readOnly;
  }, [focusedInput, isRunning, readOnly]);
  return /*#__PURE__*/React.createElement(Sticky, null, /*#__PURE__*/React.createElement(Card, {
    padding: 4,
    tone: "default",
    border: true
  }, /*#__PURE__*/React.createElement(Stack, {
    space: 4
  }, /*#__PURE__*/React.createElement(Label, {
    size: 0
  }, "Type Performance Tester"), /*#__PURE__*/React.createElement(Text, {
    title: isDisabled ? undefined : "Click to focus input",
    size: 1,
    onClick: handleHighlightClick,
    onMouseOver: handleHighlightMouseOver,
    onMouseOut: handleHighlightMouseOver
  }, isDisabled && !isRunning ? 'No input selected' : "Selected input: ".concat(focusedInput ? getNameFromInput(focusedInput) : '<none>')), /*#__PURE__*/React.createElement(TextArea, {
    ref: textAreaRef,
    readOnly: true,
    rows: 6,
    value: testOutput
  }), /*#__PURE__*/React.createElement(Button, {
    text: "Run test",
    onClick: handleRunTest,
    disabled: isDisabled,
    loading: isRunning
  }))), /*#__PURE__*/React.createElement(GlobalStyle, null));
}
