// sanity/lib/dts/src/desk/panes/document/statusBar/sparkline/DocumentBadges.js
import { Badge, Box, Inline, Text, Tooltip } from '@sanity/ui';
import React from 'react';
import { RenderBadgeCollectionState } from '../../../../badges';
import { useDocumentPane } from '../../useDocumentPane';
const BADGE_TONES = {
    primary: 'primary',
    success: 'positive',
    warning: 'caution',
    danger: 'critical',
};
function DocumentBadgesInner({ states }) {
    return (React.createElement(Inline, { space: 1 }, states.map((badge, index) => (React.createElement(Tooltip, { content: badge.title && (React.createElement(Box, { padding: 2 },
            React.createElement(Text, { size: 1 }, badge.title))), disabled: !badge.title, key: String(index), placement: "top", portal: true },
        React.createElement(Badge, { fontSize: 1, mode: "outline", paddingX: 2, paddingY: 1, radius: 4, tone: badge.color ? BADGE_TONES[badge.color] : undefined, style: { whiteSpace: 'nowrap' } }, badge.label))))));
}
export function DocumentBadges() {
    const { badges, editState } = useDocumentPane();
    if (!editState || !badges)
        return null;
    return (React.createElement(RenderBadgeCollectionState, { badges: badges, badgeProps: editState }, ({ states }) => React.createElement(DocumentBadgesInner, { states: states })));
}
