// sanity/lib/esm/desk/panes/document/documentPanel/header/DocumentHeaderTabs.js
import React, { useCallback } from 'react';
import { Tab, TabList } from '@sanity/ui';
import { useDocumentPane } from '../../useDocumentPane';
import { usePaneRouter } from '../../../../components';
export function DocumentHeaderTabs() {
  const {
    activeViewId,
    paneKey,
    views
  } = useDocumentPane();
  const tabPanelId = "".concat(paneKey, "tabpanel");
  return /*#__PURE__*/React.createElement(TabList, {
    space: 1
  }, views.map((view, index) => {
    var _view$id;

    return /*#__PURE__*/React.createElement(DocumentHeaderTab, {
      icon: view.icon,
      id: "".concat(paneKey, "tab-").concat(view.id),
      isActive: activeViewId === view.id,
      key: view.id,
      label: /*#__PURE__*/React.createElement(React.Fragment, null, view.title),
      tabPanelId: tabPanelId,
      viewId: index === 0 ? null : (_view$id = view.id) !== null && _view$id !== void 0 ? _view$id : null
    });
  }));
}

function DocumentHeaderTab(props) {
  const {
    icon,
    id,
    isActive,
    label,
    tabPanelId,
    viewId
  } = props;
  const {
    ready
  } = useDocumentPane();
  const {
    setView
  } = usePaneRouter();
  const handleClick = useCallback(() => setView(viewId), [setView, viewId]);
  return /*#__PURE__*/React.createElement(Tab, {
    "aria-controls": tabPanelId,
    disabled: !ready,
    fontSize: 1,
    icon: icon,
    id: id,
    selected: isActive,
    label: label,
    onClick: handleClick
  });
}
