// next-sanity/lib/esm/studio.js
import {jsxs as $4289s$jsxs, Fragment as $4289s$Fragment, jsx as $4289s$jsx} from "react/jsx-runtime";
import {memo as $4289s$memo, useCallback as $4289s$useCallback, useMemo as $4289s$useMemo} from "react";
import {Studio as $4289s$Studio, defaultTheme as $4289s$defaultTheme} from "sanity";
import {createGlobalStyle as $4289s$createGlobalStyle, css as $4289s$css, ServerStyleSheet as $4289s$ServerStyleSheet} from "styled-components";
import $4289s$nexthead from "next/head";
import $4289s$nextdocument from "next/document";
import {useRouter as $4289s$useRouter} from "next/router";

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
var $2e0e6a6d8bb1bf3b$exports = {};
var $1512e3a8033a35cc$exports = {};

$parcel$export($1512e3a8033a35cc$exports, "NextStudio", () => $1512e3a8033a35cc$export$eefb95cccd1bb6c5);




/**
 * Intended to render at the root of a page, letting the Studio own that page and render much like it would if you used `npx sanity start` to render
 * It's a drop-in replacement for `import {Studio} from 'sanity'`
 */ const $1512e3a8033a35cc$var$NextStudioComponent = ({ children: children , config: config , unstable__noGlobalStyle: unstable__noGlobalStyle , unstable__noTailwindSvgFix: unstable__noTailwindSvgFix , unstable__head: unstable__head , unstable__document_title: unstable__document_title , unstable__bg: unstable__bg , unstable__noFavicons: unstable__noFavicons , unstable__noNoScript: unstable__noNoScript , ...props })=>{
    const theme = (0, $47ea254c66a1f1fc$export$93d4e7f90805808f)(config);
    const { themeColorLight: themeColorLight , themeColorDark: themeColorDark  } = (0, $47ea254c66a1f1fc$export$d3a2146a441ec034)(theme);
    return /*#__PURE__*/ (0, $4289s$jsxs)((0, $4289s$Fragment), {
        children: [
            children || /*#__PURE__*/ (0, $4289s$jsx)((0, $4289s$Studio), {
                config: config,
                ...props
            }),
            /*#__PURE__*/ (0, $4289s$jsx)((0, $42e0ba93e8ca4228$export$79723eee2c1bf75e), {
                themeColorLight: themeColorLight,
                themeColorDark: themeColorDark,
                title: unstable__document_title,
                favicons: !unstable__noFavicons,
                children: unstable__head
            }),
            !unstable__noGlobalStyle && /*#__PURE__*/ (0, $4289s$jsx)((0, $96bfa0bd9b2d4d7b$export$c8fda7ee7699d07e), {
                bg: unstable__bg !== null && unstable__bg !== void 0 ? unstable__bg : themeColorLight,
                unstable__tailwindSvgFix: !unstable__noTailwindSvgFix
            }),
            !unstable__noNoScript && /*#__PURE__*/ (0, $4289s$jsx)((0, $30581564ea022575$export$39af4e536de5c6ed), {})
        ]
    });
};
const $1512e3a8033a35cc$export$eefb95cccd1bb6c5 = /*#__PURE__*/ (0, $4289s$memo)($1512e3a8033a35cc$var$NextStudioComponent);


var $96bfa0bd9b2d4d7b$exports = {};

$parcel$export($96bfa0bd9b2d4d7b$exports, "NextStudioGlobalStyle", () => $96bfa0bd9b2d4d7b$export$c8fda7ee7699d07e);

let $96bfa0bd9b2d4d7b$var$_ = (t)=>t, $96bfa0bd9b2d4d7b$var$t, $96bfa0bd9b2d4d7b$var$t1, $96bfa0bd9b2d4d7b$var$t2;
const $96bfa0bd9b2d4d7b$export$c8fda7ee7699d07e = (0, $4289s$createGlobalStyle)($96bfa0bd9b2d4d7b$var$t2 || ($96bfa0bd9b2d4d7b$var$t2 = $96bfa0bd9b2d4d7b$var$_`
${0}
html,
body,
#__next {
  height: 100%;
}
body {
  margin: 0;
  overscroll-behavior: none;
  -webkit-font-smoothing: antialiased;
}
${0}`), ({ bg: bg  })=>bg ? (0, $4289s$css)($96bfa0bd9b2d4d7b$var$t || ($96bfa0bd9b2d4d7b$var$t = $96bfa0bd9b2d4d7b$var$_`
        html {
          background-color: ${0};
        }
      `), bg) : "", ({ unstable__tailwindSvgFix: unstable__tailwindSvgFix  })=>unstable__tailwindSvgFix ? (0, $4289s$css)($96bfa0bd9b2d4d7b$var$t1 || ($96bfa0bd9b2d4d7b$var$t1 = $96bfa0bd9b2d4d7b$var$_`
        /* override tailwind reset */
        :root svg {
          display: inline;
        }
      `)) : "");


var $42e0ba93e8ca4228$exports = {};

$parcel$export($42e0ba93e8ca4228$exports, "NextStudioHead", () => $42e0ba93e8ca4228$export$79723eee2c1bf75e);



var $42f2d74a3bdc48ba$exports = {};
$42f2d74a3bdc48ba$exports = new URL("apple-touch-icon.abc4aca2.png", import.meta.url).toString();


var $8bb87a59001ac99d$exports = {};
$8bb87a59001ac99d$exports = new URL("favicon.648fc2f2.ico", import.meta.url).toString();


var $c9a50285122e97f4$exports = {};
$c9a50285122e97f4$exports = new URL("favicon.8f76876f.svg", import.meta.url).toString();


var $f259818e57447008$exports = {};
$f259818e57447008$exports = new URL("favicon-192.40c8bab4.png", import.meta.url).toString();


var $aa2d7d13e3785544$exports = {};
$aa2d7d13e3785544$exports = new URL("favicon-512.5e08f951.png", import.meta.url).toString();


var $c083397bd1950d1f$exports = {};
$c083397bd1950d1f$exports = JSON.parse('{"icons":[{"src":"./favicon-192.png","type":"image/png","sizes":"192x192"},{"src":"./favicon-512.png","type":"image/png","sizes":"512x512"}]}');


// Interop between how Parcel and Next deals with asset imports
const $42e0ba93e8ca4228$var$interop = (href)=>typeof href === "string" ? href : href.src;
const $42e0ba93e8ca4228$var$NextStudioHeadComponent = ({ children: children , themeColorDark: themeColorDark , themeColorLight: themeColorLight , title: title = "Sanity Studio" , favicons: favicons  })=>{
    const inlineWebmanifest = (0, $4289s$useCallback)(()=>{
        const manifest = JSON.parse(JSON.stringify((0, (/*@__PURE__*/$parcel$interopDefault($c083397bd1950d1f$exports)))));
        const icons = manifest.icons.map((icon)=>{
            // Inline manifests works best when URLs are absolute
            const src = // eslint-disable-next-line no-nested-ternary
            icon.src === "./favicon-192.png" ? $42e0ba93e8ca4228$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($f259818e57447008$exports)))) : icon.src === "./favicon-512.png" ? $42e0ba93e8ca4228$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($aa2d7d13e3785544$exports)))) : icon.src;
            return {
                ...icon,
                src: src
            };
        });
        return `data:application/manifest+json,${encodeURIComponent(JSON.stringify({
            ...manifest,
            icons: icons
        }))}`;
    }, []);
    return /*#__PURE__*/ (0, $4289s$jsxs)((0, $4289s$nexthead), {
        children: [
            /*#__PURE__*/ (0, $4289s$jsx)("meta", {
                name: "viewport",
                // Studio implements display cutouts CSS (The iPhone Notch ™ ) and needs `viewport-fit=covered` for it to work correctly
                content: "width=device-width, initial-scale=1, viewport-fit=cover"
            }),
            /*#__PURE__*/ (0, $4289s$jsx)("meta", {
                name: "robots",
                content: "noindex"
            }),
            /*#__PURE__*/ (0, $4289s$jsx)("meta", {
                name: "referrer",
                content: "same-origin"
            }),
            title && /*#__PURE__*/ (0, $4289s$jsx)("title", {
                children: title
            }),
            favicons && /*#__PURE__*/ (0, $4289s$jsx)("link", {
                rel: "icon",
                href: $42e0ba93e8ca4228$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($8bb87a59001ac99d$exports)))),
                sizes: "any"
            }),
            favicons && /*#__PURE__*/ (0, $4289s$jsx)("link", {
                rel: "icon",
                href: $42e0ba93e8ca4228$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($c9a50285122e97f4$exports)))),
                type: "image/svg+xml"
            }),
            favicons && /*#__PURE__*/ (0, $4289s$jsx)("link", {
                rel: "apple-touch-icon",
                href: $42e0ba93e8ca4228$var$interop((0, (/*@__PURE__*/$parcel$interopDefault($42f2d74a3bdc48ba$exports))))
            }),
            favicons && /*#__PURE__*/ (0, $4289s$jsx)("link", {
                rel: "manifest",
                // eslint-disable-next-line no-warning-comments
                // @TODO until parcel fixes https://github.com/parcel-bundler/parcel/issues/8025 and stops stripping process.env.NEXT_PUBLIC_VERCEL_URL from the compiled code, use the remove webmanifest
                href: "https://next.sanity.build/manifest.webmanifest"
            }),
            themeColorLight && /*#__PURE__*/ (0, $4289s$jsx)("meta", {
                name: "theme-color",
                content: themeColorLight,
                media: "(prefers-color-scheme: light)"
            }, "theme-color-light"),
            themeColorDark && /*#__PURE__*/ (0, $4289s$jsx)("meta", {
                name: "theme-color",
                content: themeColorDark,
                media: "(prefers-color-scheme: dark)"
            }, "theme-color-dark"),
            children
        ]
    });
};
const $42e0ba93e8ca4228$export$79723eee2c1bf75e = /*#__PURE__*/ (0, $4289s$memo)($42e0ba93e8ca4228$var$NextStudioHeadComponent);


var $30581564ea022575$exports = {};

$parcel$export($30581564ea022575$exports, "NextStudioNoScript", () => $30581564ea022575$export$39af4e536de5c6ed);

/* eslint-disable react/no-danger */ const $30581564ea022575$var$style = {
    __html: `
.sanity-app-no-js__root {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background: #fff;
}

.sanity-app-no-js__content {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  font-family: helvetica, arial, sans-serif;
}
`
};
const $30581564ea022575$export$39af4e536de5c6ed = ()=>/*#__PURE__*/ (0, $4289s$jsx)("noscript", {
        children: /*#__PURE__*/ (0, $4289s$jsx)("div", {
            className: "sanity-app-no-js__root",
            children: /*#__PURE__*/ (0, $4289s$jsxs)("div", {
                className: "sanity-app-no-js__content",
                children: [
                    /*#__PURE__*/ (0, $4289s$jsx)("style", {
                        type: "text/css",
                        dangerouslySetInnerHTML: $30581564ea022575$var$style
                    }),
                    /*#__PURE__*/ (0, $4289s$jsx)("h1", {
                        children: "JavaScript disabled"
                    }),
                    /*#__PURE__*/ (0, $4289s$jsxs)("p", {
                        children: [
                            "Please ",
                            /*#__PURE__*/ (0, $4289s$jsx)("a", {
                                href: "https://www.enable-javascript.com/",
                                children: "enable JavaScript"
                            }),
                            " in your browser and reload the page to proceed."
                        ]
                    })
                ]
            })
        })
    });


var $056a6fa2cc4f79d2$exports = {};

$parcel$export($056a6fa2cc4f79d2$exports, "ServerStyleSheetDocument", () => $056a6fa2cc4f79d2$export$f64e2f55a1dca172);



class $056a6fa2cc4f79d2$export$f64e2f55a1dca172 extends (0, $4289s$nextdocument) {
    static async getInitialProps(ctx) {
        const sheet = new (0, $4289s$ServerStyleSheet)();
        const originalRenderPage = ctx.renderPage;
        try {
            ctx.renderPage = ()=>originalRenderPage({
                    enhanceApp: (App)=>(props)=>sheet.collectStyles(/*#__PURE__*/ (0, $4289s$jsx)(App, {
                                ...props
                            }))
                });
            const initialProps = await (0, $4289s$nextdocument).getInitialProps(ctx);
            return {
                ...initialProps,
                styles: [
                    initialProps.styles,
                    sheet.getStyleElement()
                ]
            };
        } finally{
            sheet.seal();
        }
    }
}


var $47ea254c66a1f1fc$exports = {};

$parcel$export($47ea254c66a1f1fc$exports, "isWorkspaces", () => $47ea254c66a1f1fc$export$b7c6924067135ce);
$parcel$export($47ea254c66a1f1fc$exports, "isWorkspaceWithTheme", () => $47ea254c66a1f1fc$export$31236b9ee07dd150);
$parcel$export($47ea254c66a1f1fc$exports, "useTheme", () => $47ea254c66a1f1fc$export$93d4e7f90805808f);
$parcel$export($47ea254c66a1f1fc$exports, "useBackgroundColorsFromTheme", () => $47ea254c66a1f1fc$export$d3a2146a441ec034);
$parcel$export($47ea254c66a1f1fc$exports, "useBasePath", () => $47ea254c66a1f1fc$export$53bdd69a9f9bd75);
$parcel$export($47ea254c66a1f1fc$exports, "useConfigWithBasePath", () => $47ea254c66a1f1fc$export$297ba38aa0562c6e);



function $47ea254c66a1f1fc$export$b7c6924067135ce(config) {
    return Array.isArray(config);
}
function $47ea254c66a1f1fc$export$31236b9ee07dd150(workspace) {
    return Boolean(workspace.theme);
}
function $47ea254c66a1f1fc$export$93d4e7f90805808f(config) {
    const workspace = (0, $4289s$useMemo)(()=>$47ea254c66a1f1fc$export$b7c6924067135ce(config) ? config[0] : config, [
        config
    ]);
    return (0, $4289s$useMemo)(()=>$47ea254c66a1f1fc$export$31236b9ee07dd150(workspace) ? workspace.theme : (0, $4289s$defaultTheme), [
        workspace
    ]);
}
const $47ea254c66a1f1fc$export$d3a2146a441ec034 = (theme)=>{
    return (0, $4289s$useMemo)(()=>({
            themeColorLight: theme.color.light.default.base.bg,
            themeColorDark: theme.color.dark.default.base.bg
        }), [
        theme
    ]);
};
function $47ea254c66a1f1fc$export$53bdd69a9f9bd75() {
    const router = (0, $4289s$useRouter)();
    return (0, $4289s$useMemo)(()=>{
        const [basePath = "/"] = router.route.split("/[");
        return basePath;
    }, [
        router.route
    ]);
}
function $47ea254c66a1f1fc$export$297ba38aa0562c6e(config) {
    const basePath = $47ea254c66a1f1fc$export$53bdd69a9f9bd75();
    return (0, $4289s$useMemo)(()=>{
        if ($47ea254c66a1f1fc$export$b7c6924067135ce(config)) return config.map((workspace)=>({
                ...workspace,
                basePath: workspace.basePath === "/" ? basePath : `${basePath}${workspace.basePath || ""}`
            }));
        return {
            ...config,
            basePath: config.basePath === "/" ? basePath : `${basePath}${config.basePath || ""}`
        };
    }, [
        config,
        basePath
    ]);
}


$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $1512e3a8033a35cc$exports);
$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $96bfa0bd9b2d4d7b$exports);
$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $42e0ba93e8ca4228$exports);
$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $30581564ea022575$exports);
$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $056a6fa2cc4f79d2$exports);
$parcel$exportWildcard($2e0e6a6d8bb1bf3b$exports, $47ea254c66a1f1fc$exports);




export {$1512e3a8033a35cc$export$eefb95cccd1bb6c5 as NextStudio, $96bfa0bd9b2d4d7b$export$c8fda7ee7699d07e as NextStudioGlobalStyle, $42e0ba93e8ca4228$export$79723eee2c1bf75e as NextStudioHead, $30581564ea022575$export$39af4e536de5c6ed as NextStudioNoScript, $056a6fa2cc4f79d2$export$f64e2f55a1dca172 as ServerStyleSheetDocument, $47ea254c66a1f1fc$export$b7c6924067135ce as isWorkspaces, $47ea254c66a1f1fc$export$31236b9ee07dd150 as isWorkspaceWithTheme, $47ea254c66a1f1fc$export$93d4e7f90805808f as useTheme, $47ea254c66a1f1fc$export$d3a2146a441ec034 as useBackgroundColorsFromTheme, $47ea254c66a1f1fc$export$53bdd69a9f9bd75 as useBasePath, $47ea254c66a1f1fc$export$297ba38aa0562c6e as useConfigWithBasePath};
