// sanity/lib/esm/components/previews/general/MediaPreview.styled.js
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { Box, Flex, rem, Skeleton, Stack } from '@sanity/ui';
import styled from 'styled-components';
export const RootBox = styled(Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: relative;\n"])));
export const MediaFlex = styled(Flex).attrs({
  align: 'center',
  justify: 'center'
})(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  left: 0;\n  top: 0;\n  right: 0;\n  bottom: 0;\n"])));
export const MediaSkeleton = styled(Skeleton).attrs({
  animated: true,
  radius: 2
})(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  width: 100%;\n  height: 100%;\n"])));
export const ProgressFlex = styled(Flex).attrs({
  align: 'center',
  justify: 'center'
})(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  position: absolute;\n  left: 0;\n  top: 0;\n  right: 0;\n  bottom: 0;\n\n  &:before {\n    background-color: var(--card-bg-color);\n    opacity: 0.75;\n    content: '';\n    display: block;\n    position: absolute;\n    left: 0;\n    top: 0;\n    right: 0;\n    bottom: 0;\n  }\n\n  > svg {\n    position: relative;\n    z-index: 2;\n  }\n"])));
export const TooltipContentStack = styled(Stack).attrs({
  padding: 2,
  space: 2
})(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  max-width: ", ";\n"])), rem(200));
