// sanity/lib/cjs/preview/documentPreviewStore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDocumentPreviewStore = createDocumentPreviewStore;

var _operators = require("rxjs/operators");

var _util = require("../util");

var _createPathObserver = require("./createPathObserver");

var _createPreviewObserver = require("./createPreviewObserver");

var _observeFields = require("./observeFields");

var _availability = require("./availability");

var _documentPair = require("./documentPair");

function createDocumentPreviewStore(_ref) {
  let {
    crossProjectTokenStore,
    client
  } = _ref;
  const versionedClient = client.withConfig({
    apiVersion: '1'
  }); // NOTE: this is workaroudn for circumventing a circular dependency between `observePaths` and
  // `observeFields`.
  // eslint-disable-next-line camelcase

  const __proxy_observePaths = (value, paths, apiConfig) => {
    return observePaths(value, paths, apiConfig);
  };

  const {
    observeFields
  } = (0, _observeFields.create_preview_observeFields)({
    crossProjectTokenStore,
    observePaths: __proxy_observePaths,
    versionedClient
  });
  const {
    observePaths
  } = (0, _createPathObserver.createPathObserver)({
    observeFields
  });

  function observeDocumentTypeFromId(id, apiConfig) {
    return observePaths({
      _type: 'reference',
      _ref: id
    }, ['_type']).pipe((0, _operators.map)(res => (0, _util.isRecord)(res) && typeof res._type === 'string' ? res._type : undefined), (0, _operators.distinctUntilChanged)());
  } // const {createPreviewObserver} = create_preview_createPreviewObserver(observeDocumentTypeFromId)


  const observeForPreview = (0, _createPreviewObserver.createPreviewObserver)({
    observeDocumentTypeFromId,
    observePaths
  });
  const {
    observeDocumentPairAvailability
  } = (0, _availability.create_preview_availability)(versionedClient, observePaths);
  const {
    observePathsDocumentPair
  } = (0, _documentPair.create_preview_documentPair)(versionedClient, observePaths); // @todo: explain why the API is like this now, and that it should not be like this in the future!

  return {
    observePaths,
    observeForPreview,
    observeDocumentTypeFromId,
    // eslint-disable-next-line camelcase
    unstable_observeDocumentPairAvailability: observeDocumentPairAvailability,
    unstable_observePathsDocumentPair: observePathsDocumentPair
  };
}
