// sanity/lib/dts/src/form/__workshop__/_common/data.js
import Schema from '@sanity/schema';
import { keyBy, mapValues } from 'lodash';
import getSimpleDummySchema from './schema/simpleDummySchema';
import getSimpleFieldGroupSchema from './schema/simpleFieldGroupSchema';
export const DUMMY_DOCUMENT_ID = '10053a07-8647-4ebd-9d1d-33a512d30d3a';
export function wrapSchema(schema) {
    return {
        name: 'test',
        types: [schema],
    };
}
const schemaMap = [
    {
        name: 'simple',
        title: 'Simple',
        schema: getSimpleDummySchema,
    },
    {
        name: 'simpleFieldGroup',
        title: 'Simple w/ Field Group',
        schema: getSimpleFieldGroupSchema,
    },
];
export const schemaListOptions = mapValues(keyBy(schemaMap, 'title'), 'name');
export function getDummyDocument() {
    return {
        _createdAt: '2021-11-04T15:41:48Z',
        _id: DUMMY_DOCUMENT_ID,
        _rev: '5hb8s6-k75-ip4-4bq-5ztbf3fbx',
        _type: 'book',
        _updatedAt: '2021-11-05T12:34:29Z',
        title: 'Hello world',
        person: {
            name: 'Fred',
        },
    };
}
export function getDummySchema(props) {
    const { schemaKey = 'simple' } = props;
    const schemaType = schemaMap.find((s) => s.name === schemaKey);
    const schema = schemaType?.schema(props);
    return Schema.compile(schema);
}
