// sanity/lib/cjs/cli/commands/dataset/deleteDatasetCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _validateDatasetName = require("../../actions/dataset/validateDatasetName");

const deleteDatasetCommand = {
  name: 'delete',
  group: 'dataset',
  helpText: '',
  signature: '[datasetName]',
  description: 'Delete a dataset within your project',
  action: async (args, context) => {
    const {
      apiClient,
      prompt,
      output
    } = context;
    const [ds] = args.argsWithoutOptions;

    if (!ds) {
      throw new Error('Dataset name must be provided');
    }

    const dataset = "".concat(ds);
    const dsError = (0, _validateDatasetName.validateDatasetName)(dataset);

    if (dsError) {
      throw dsError;
    }

    await prompt.single({
      type: 'input',
      message: 'Are you ABSOLUTELY sure you want to delete this dataset?\n  Type the name of the dataset to confirm delete:',
      filter: input => "".concat(input).trim(),
      validate: input => {
        return input === dataset || 'Incorrect dataset name. Ctrl + C to cancel delete.';
      }
    });
    await apiClient().datasets.delete(dataset);
    output.print('Dataset deleted successfully');
  }
};
var _default = deleteDatasetCommand;
exports.default = _default;
