// sanity/lib/cjs/form/inputs/files/common/UploadPlaceholder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadPlaceholder = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _FileInputButton = require("./FileInputButton/FileInputButton");

var _PlaceholderText = require("./PlaceholderText");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const UploadPlaceholder = /*#__PURE__*/_react.default.memo(function UploadPlaceholder(_ref) {
  let {
    browse,
    onUpload,
    readOnly,
    type,
    hoveringFiles,
    acceptedFiles,
    rejectedFilesCount,
    accept,
    directUploads
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    justify: "space-between",
    gap: 4,
    direction: ['column', 'column', 'row'],
    paddingY: [2, 2, 0]
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    justify: "center",
    gap: 2,
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_PlaceholderText.PlaceholderText, {
    readOnly: readOnly,
    hoveringFiles: hoveringFiles,
    acceptedFiles: acceptedFiles,
    rejectedFilesCount: rejectedFilesCount,
    type: type,
    directUploads: directUploads
  })), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_FileInputButton.FileInputButton, {
    icon: _icons.UploadIcon,
    mode: "ghost",
    onSelect: onUpload,
    accept: accept,
    text: "Upload",
    "data-testid": "file-input-upload-button",
    disabled: readOnly || !directUploads
  }), browse));
});

exports.UploadPlaceholder = UploadPlaceholder;
