// sanity/lib/cjs/desk/panes/document/timeline/timelineItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimelineItem = TimelineItem;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _hooks = require("../../../../hooks");

var _helpers = require("./helpers");

var _userAvatarStack = require("./userAvatarStack");

var _timelineItem = require("./timelineItem.styled");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const TIMELINE_ITEM_EVENT_TONE = {
  initial: 'primary',
  create: 'primary',
  publish: 'positive',
  editLive: 'caution',
  editDraft: 'caution',
  unpublish: 'critical',
  discardDraft: 'critical',
  delete: 'critical',
  withinSelection: 'primary'
};

function TimelineItem(props) {
  const {
    isSelectionBottom,
    isSelectionTop,
    state,
    onSelect,
    timestamp,
    chunk,
    type
  } = props;
  const iconComponent = (0, _helpers.getTimelineEventIconComponent)(type);
  const authorUserIds = Array.from(chunk.authors);
  const timeAgo = (0, _hooks.useTimeAgo)(timestamp, {
    minimal: true
  });
  const isSelected = state === 'selected';
  const isWithinSelection = state === 'withinSelection';
  const [isHovered, setHovered] = (0, _react.useState)(false);
  const handleClick = (0, _react.useCallback)(evt => {
    evt.preventDefault();
    evt.stopPropagation();
    onSelect(chunk);
  }, [onSelect, chunk]);
  return /*#__PURE__*/_react.default.createElement(_timelineItem.Root, {
    "data-ui": "timelineItem",
    radius: 2,
    "data-chunk-id": chunk.id,
    paddingY: 0,
    paddingX: 2,
    tone: isHovered || isSelected || isWithinSelection ? 'default' : TIMELINE_ITEM_EVENT_TONE[type],
    pressed: isWithinSelection,
    state: state,
    selected: isSelected,
    isHovered: isHovered,
    disabled: state === 'disabled',
    "data-selection-bottom": isSelectionBottom,
    "data-selection-top": isSelectionTop,
    onClick: handleClick
  }, /*#__PURE__*/_react.default.createElement("div", {
    // eslint-disable-next-line react/jsx-no-bind
    onMouseEnter: () => setHovered(true) // eslint-disable-next-line react/jsx-no-bind
    ,
    onMouseLeave: () => setHovered(false)
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "stretch"
  }, /*#__PURE__*/_react.default.createElement(_timelineItem.IconWrapper, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_timelineItem.IconBox, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 2
  }, iconComponent && /*#__PURE__*/(0, _react.createElement)(iconComponent)))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2,
    margin: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_timelineItem.EventLabel, {
    size: 1,
    weight: "medium"
  }, (0, _helpers.formatTimelineEventLabel)(type) || /*#__PURE__*/_react.default.createElement("code", null, type))), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 0,
    muted: true
  }, timeAgo)), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    flex: 1,
    justify: "flex-end",
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_userAvatarStack.UserAvatarStack, {
    maxLength: 3,
    userIds: authorUserIds
  })))));
}
