// sanity/lib/dts/src/field/__workshop__/DiffCardStory.js
import { Box, Text } from '@sanity/ui';
import React, { useMemo } from 'react';
import { DiffCard } from '../diff';
export default function DiffCardStory() {
    const diff = useMemo(() => ({
        type: 'boolean',
        action: 'added',
        isChanged: true,
        fromValue: null,
        toValue: true,
        annotation: {
            chunk: {
                index: 0,
                id: 'foo',
                type: 'editDraft',
                start: 0,
                end: 0,
                startTimestamp: new Date('2021-01-01').toJSON(),
                endTimestamp: new Date('2021-01-02').toJSON(),
                authors: new Set(['p27ewL8aM']),
                draftState: 'present',
                publishedState: 'missing',
            },
            timestamp: new Date('2021-01-02').toJSON(),
            author: 'p27ewL8aM',
        },
    }), []);
    return (React.createElement(Box, { padding: 4 },
        React.createElement(DiffCard, { diff: diff, 
            // disableHoverEffect
            path: [], tooltip: { description: 'Added' } },
            React.createElement(Box, { padding: 3 },
                React.createElement(Text, null, "DiffCard")))));
}
