// sanity/lib/esm/studio/screens/CorsOriginErrorScreen.js
import { Card, Dialog, Stack, Button, Text } from '@sanity/ui';
import React, { useCallback } from 'react';
export function CorsOriginErrorScreen(props) {
  const {
    projectId
  } = props;
  const corsUrl = "https://sanity.io/manage/project/".concat(projectId, "/settings/api");
  const origin = window.location.origin;
  const handleRetry = useCallback(() => {
    window.location.reload();
  }, []);
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill"
  }, /*#__PURE__*/React.createElement(Dialog, {
    id: "cors-error-dialog",
    header: "Error",
    width: 1,
    footer: /*#__PURE__*/React.createElement(Stack, {
      paddingX: 3,
      paddingY: 2
    }, /*#__PURE__*/React.createElement(Button, {
      text: "Retry",
      onClick: handleRetry
    }))
  }, /*#__PURE__*/React.createElement(Stack, {
    paddingX: 4,
    paddingY: 5,
    space: 4
  }, /*#__PURE__*/React.createElement(Text, null, "It looks like the error is being caused by the current origin (", /*#__PURE__*/React.createElement("code", null, origin), ") not being allowed for this project."), /*#__PURE__*/React.createElement(Text, null, "If you are a project administrator or developer, you can head to", ' ', /*#__PURE__*/React.createElement("a", {
    rel: "noopener noreferrer",
    target: "_blank",
    href: corsUrl
  }, "the project management"), ' ', "interface to configure CORS origins for this project."), /*#__PURE__*/React.createElement(Text, null, /*#__PURE__*/React.createElement("a", {
    href: "https://www.sanity.io/docs/front-ends/cors",
    target: "_blank",
    rel: "noopener noreferrer"
  }, "Read more about CORS Origins"), "."))));
}
