// sanity/lib/esm/studio/screens/NotAuthenticatedScreen.js
import { Button, Card, Dialog, Stack, Text } from '@sanity/ui';
import React, { useCallback, useEffect, useState } from 'react';
import { getProviderTitle } from '../../datastores/authStore/providerTitle';
import { useActiveWorkspace } from '../activeWorkspaceMatcher';
export function NotAuthenticatedScreen() {
  const [currentUser, setCurrentUser] = useState(null);
  const [error, handleError] = useState(null);
  if (error) throw error;
  const {
    activeWorkspace
  } = useActiveWorkspace();
  const handleLogout = useCallback(() => {
    var _activeWorkspace$auth, _activeWorkspace$auth2;

    (_activeWorkspace$auth = (_activeWorkspace$auth2 = activeWorkspace.auth).logout) === null || _activeWorkspace$auth === void 0 ? void 0 : _activeWorkspace$auth.call(_activeWorkspace$auth2);
  }, [activeWorkspace]);
  useEffect(() => {
    const subscription = activeWorkspace.auth.state.subscribe({
      next: _ref => {
        let {
          currentUser: user
        } = _ref;
        setCurrentUser(user);
      },
      error: handleError
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [activeWorkspace]);
  const providerTitle = getProviderTitle(currentUser === null || currentUser === void 0 ? void 0 : currentUser.provider);
  const providerHelp = providerTitle ? " through ".concat(providerTitle) : '';
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill"
  }, /*#__PURE__*/React.createElement(Dialog, {
    id: "not-authorized-dialog",
    header: "Not authorized",
    width: 1,
    footer: /*#__PURE__*/React.createElement(Stack, {
      paddingX: 3,
      paddingY: 2
    }, /*#__PURE__*/React.createElement(Button, {
      text: "Sign out",
      onClick: handleLogout
    }))
  }, /*#__PURE__*/React.createElement(Stack, {
    paddingX: 4,
    paddingY: 5,
    space: 4
  }, /*#__PURE__*/React.createElement(Text, null, "You are not authorized to access this studio. Maybe you could ask someone to invite you to collaborate on this project?"), /*#__PURE__*/React.createElement(Text, null, "If you think this is an error, verify that you are signed in with the correct account. You are currently signed in as", ' ', /*#__PURE__*/React.createElement("strong", null, currentUser === null || currentUser === void 0 ? void 0 : currentUser.name, " (", currentUser === null || currentUser === void 0 ? void 0 : currentUser.email, ")"), providerHelp, "."))));
}
