// sanity/lib/esm/studio/screens/ToolNotFoundScreen.js
import { WarningOutlineIcon } from '@sanity/icons';
import { Box, Card, Container, Flex, Stack, Text } from '@sanity/ui';
import React from 'react';
export function ToolNotFoundScreen(props) {
  const {
    toolName
  } = props;
  return /*#__PURE__*/React.createElement(Card, {
    height: "fill"
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/React.createElement(Container, {
    width: 0
  }, /*#__PURE__*/React.createElement(Card, {
    padding: 4,
    radius: 2,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Box, null, /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, /*#__PURE__*/React.createElement(WarningOutlineIcon, null))), /*#__PURE__*/React.createElement(Stack, {
    flex: 1,
    marginLeft: 3,
    space: 3
  }, /*#__PURE__*/React.createElement(Text, {
    as: "h1",
    size: 1,
    weight: "bold"
  }, "Tool not found: ", /*#__PURE__*/React.createElement("code", null, toolName))))))));
}
