// sanity/lib/cjs/form/studio/uploads/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CLEANUP_EVENT = void 0;
exports.createInitialUploadEvent = createInitialUploadEvent;
exports.createUploadEvent = createUploadEvent;

var _patch = require("../../patch");

var _constants = require("./constants");

const UNSET_UPLOAD_PATCH = (0, _patch.unset)([_constants.UPLOAD_STATUS_KEY]);

function createUploadEvent() {
  let patches = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  return {
    type: 'uploadEvent',
    patches
  };
}

const CLEANUP_EVENT = createUploadEvent([UNSET_UPLOAD_PATCH]);
exports.CLEANUP_EVENT = CLEANUP_EVENT;

function createInitialUploadEvent(file) {
  const now = new Date().toISOString();
  const value = {
    progress: 2,
    initiated: now,
    updated: now,
    file: {
      name: file.name,
      type: file.type
    }
  };
  return createUploadEvent([(0, _patch.setIfMissing)({
    [_constants.UPLOAD_STATUS_KEY]: value
  }, []), (0, _patch.set)(value, [_constants.UPLOAD_STATUS_KEY])]);
}
