// sanity/lib/cjs/desk/structureBuilder/GenericList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shallowIntentChecker = exports.GenericListBuilder = void 0;

var _camelCase2 = _interopRequireDefault(require("lodash/camelCase"));

var _Layout = require("./Layout");

var _MenuItem = require("./MenuItem");

var _MenuItemGroup = require("./MenuItemGroup");

var _Intent = require("./Intent");

var _SerializeError = require("./SerializeError");

var _InitialValueTemplateItem = require("./InitialValueTemplateItem");

var _validateId = require("./util/validateId");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function noChildResolver() {
  return undefined;
}

const shallowIntentChecker = (intentName, params, _ref) => {
  let {
    pane,
    index
  } = _ref;
  return index <= 1 && (0, _Intent.defaultIntentChecker)(intentName, params, {
    pane,
    index
  });
};

exports.shallowIntentChecker = shallowIntentChecker;

class GenericListBuilder {
  constructor() {
    _defineProperty(this, "initialValueTemplatesSpecified", false);

    _defineProperty(this, "spec", {});
  }

  id(id) {
    return this.clone({
      id
    });
  }

  getId() {
    return this.spec.id;
  }

  title(title) {
    return this.clone({
      title,
      id: this.spec.id || (0, _camelCase2.default)(title)
    });
  }

  getTitle() {
    return this.spec.title;
  }

  defaultLayout(defaultLayout) {
    return this.clone({
      defaultLayout
    });
  }

  getDefaultLayout() {
    return this.spec.defaultLayout;
  }

  menuItems(menuItems) {
    return this.clone({
      menuItems
    });
  }

  getMenuItems() {
    return this.spec.menuItems;
  }

  menuItemGroups(menuItemGroups) {
    return this.clone({
      menuItemGroups
    });
  }

  getMenuItemGroups() {
    return this.spec.menuItemGroups;
  }

  child(child) {
    return this.clone({
      child
    });
  }

  getChild() {
    return this.spec.child;
  }

  canHandleIntent(canHandleIntent) {
    return this.clone({
      canHandleIntent
    });
  }

  getCanHandleIntent() {
    return this.spec.canHandleIntent;
  }

  showIcons() {
    let enabled = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    return this.clone({
      displayOptions: { ...(this.spec.displayOptions || {}),
        showIcons: enabled
      }
    });
  }

  getShowIcons() {
    return this.spec.displayOptions ? this.spec.displayOptions.showIcons : undefined;
  }

  initialValueTemplates(templates) {
    this.initialValueTemplatesSpecified = true;
    return this.clone({
      initialValueTemplates: Array.isArray(templates) ? templates : [templates]
    });
  }

  getInitialValueTemplates() {
    return this.spec.initialValueTemplates;
  }

  serialize() {
    let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    const id = this.spec.id || '';
    const path = options.path;
    const defaultLayout = this.spec.defaultLayout;

    if (defaultLayout && !_Layout.layoutOptions.includes(defaultLayout)) {
      throw new _SerializeError.SerializeError("`layout` must be one of ".concat(_Layout.layoutOptions.map(item => "\"".concat(item, "\"")).join(', ')), path, id || options.index, this.spec.title);
    }

    const initialValueTemplates = (this.spec.initialValueTemplates || []).map((item, i) => (0, _InitialValueTemplateItem.maybeSerializeInitialValueTemplateItem)(item, i, path));
    return {
      id: (0, _validateId.validateId)(id, options.path, id || options.index),
      title: this.spec.title,
      type: 'genericList',
      defaultLayout,
      child: this.spec.child || noChildResolver,
      canHandleIntent: this.spec.canHandleIntent || shallowIntentChecker,
      displayOptions: this.spec.displayOptions,
      initialValueTemplates,
      menuItems: (this.spec.menuItems || []).map((item, i) => (0, _MenuItem.maybeSerializeMenuItem)(item, i, path)),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) => (0, _MenuItemGroup.maybeSerializeMenuItemGroup)(item, i, path))
    };
  }

}

exports.GenericListBuilder = GenericListBuilder;
