// sanity/lib/esm/field/types/boolean/preview/BooleanPreview.js
import React from 'react';
export const BooleanPreview = function BooleanPreview(_ref) {
  var _schemaType$options;

  let {
    value,
    schemaType,
    color
  } = _ref;
  const Preview = ((_schemaType$options = schemaType.options) === null || _schemaType$options === void 0 ? void 0 : _schemaType$options.layout) === 'checkbox' ? Checkbox : Switch;
  return /*#__PURE__*/React.createElement(Preview, {
    checked: value,
    color: color
  });
};
export function Checkbox(_ref2) {
  let {
    checked,
    color
  } = _ref2;
  return /*#__PURE__*/React.createElement("svg", {
    width: "17",
    height: "17",
    viewBox: "0 0 17 17",
    xmlns: "http://www.w3.org/2000/svg",
    fill: color === null || color === void 0 ? void 0 : color.background
  }, /*#__PURE__*/React.createElement("rect", {
    x: "0",
    y: "0",
    width: "17",
    height: "17",
    rx: "2.5"
  }), typeof checked === 'undefined' && /*#__PURE__*/React.createElement("path", {
    d: "M4.07996 8.5H12.92",
    stroke: color === null || color === void 0 ? void 0 : color.text,
    strokeWidth: "2"
  }), checked && /*#__PURE__*/React.createElement("path", {
    d: "M3.5 8L7 11.5L13.5 5",
    stroke: color === null || color === void 0 ? void 0 : color.text,
    strokeWidth: "2"
  }));
}
export function Switch(_ref3) {
  let {
    checked,
    color
  } = _ref3;
  return /*#__PURE__*/React.createElement("svg", {
    width: "38",
    height: "22",
    viewBox: "0 0 38 22",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "38",
    height: "22",
    rx: "11",
    fill: checked ? color === null || color === void 0 ? void 0 : color.border : color === null || color === void 0 ? void 0 : color.background
  }), typeof checked === 'undefined' && /*#__PURE__*/React.createElement("rect", {
    x: "11",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }), checked && /*#__PURE__*/React.createElement("rect", {
    x: "18",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }), typeof checked === 'boolean' && !checked && /*#__PURE__*/React.createElement("rect", {
    x: "4",
    y: "3",
    width: "16",
    height: "16",
    rx: "8",
    fill: "white"
  }));
}
