// sanity/lib/esm/components/DefaultDocument.js
import React from 'react';
import { Favicons } from './Favicons';
import { GlobalErrorHandler } from './globalErrorHandler';
import { NoJavascript } from './NoJavascript';
const globalStyles = "\n  html {\n    background-color: #f1f3f6;\n  }\n  html,\n  body,\n  #sanity {\n    height: 100%;\n  }\n  body {\n    margin: 0;\n    -webkit-font-smoothing: antialiased;\n  }\n";
const EMPTY_ARRAY = [];
export function DefaultDocument(props) {
  const {
    entryPath,
    css = EMPTY_ARRAY,
    basePath = '/'
  } = props;
  return /*#__PURE__*/React.createElement("html", {
    lang: "en"
  }, /*#__PURE__*/React.createElement("head", null, /*#__PURE__*/React.createElement("meta", {
    charSet: "utf-8"
  }), /*#__PURE__*/React.createElement("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1, viewport-fit=cover"
  }), /*#__PURE__*/React.createElement("meta", {
    name: "robots",
    content: "noindex"
  }), /*#__PURE__*/React.createElement("meta", {
    name: "referrer",
    content: "same-origin"
  }), /*#__PURE__*/React.createElement(Favicons, {
    basePath: basePath
  }), /*#__PURE__*/React.createElement("title", null, "Sanity Studio"), /*#__PURE__*/React.createElement(GlobalErrorHandler, null), css.map(href => /*#__PURE__*/React.createElement("link", {
    key: href,
    rel: "stylesheet",
    href: href
  })), /*#__PURE__*/React.createElement("style", null, globalStyles)), /*#__PURE__*/React.createElement("body", null, /*#__PURE__*/React.createElement("div", {
    id: "sanity"
  }), /*#__PURE__*/React.createElement("script", {
    type: "module",
    src: entryPath
  }), /*#__PURE__*/React.createElement(NoJavascript, null)));
}
