// sanity/lib/esm/components/InsufficientPermissionsMessage.js
import { AccessDeniedIcon } from '@sanity/icons';
import { Text, Inline, Box } from '@sanity/ui';
import React from 'react';
export function InsufficientPermissionsMessage(props) {
  const {
    currentUser,
    title = 'Insufficient permissions',
    operationLabel = 'access this feature'
  } = props;
  const roles = (currentUser === null || currentUser === void 0 ? void 0 : currentUser.roles) || [];
  const plural = roles.length !== 1;
  return /*#__PURE__*/React.createElement(Box, {
    padding: 2
  }, /*#__PURE__*/React.createElement(Inline, {
    space: 2
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, /*#__PURE__*/React.createElement(AccessDeniedIcon, null)), /*#__PURE__*/React.createElement(Text, {
    weight: "semibold"
  }, title)), /*#__PURE__*/React.createElement(Inline, {
    marginTop: 4
  }, /*#__PURE__*/React.createElement(Text, {
    size: 1
  }, roles.length === 0 ? /*#__PURE__*/React.createElement(React.Fragment, null, "You have no role that grants you permission to ", operationLabel) : /*#__PURE__*/React.createElement(React.Fragment, null, "Your role", plural && 's', ' ', join(roles.map(r => /*#__PURE__*/React.createElement("code", {
    key: r.name
  }, r.title)), /*#__PURE__*/React.createElement(React.Fragment, null, ", ")), ' ', "do", plural || 'es', " not have permissions to ", operationLabel))));
}

function join(array, sep) {
  return array.reduce((result, item) => result === null ? [item] : [...result, sep, item], null);
}
