// sanity/lib/cjs/form/inputs/arrays/errors/DuplicateKeysAlert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DuplicateKeysAlert = DuplicateKeysAlert;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _formField = require("../../../components/formField");

var _Alert = require("../../../components/Alert");

var _patch = require("../../../patch");

var _Details = require("../../../components/Details");

var _environment = require("../../../../environment");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function DuplicateKeysAlert(props) {
  const {
    error,
    onChange
  } = props;
  const handleFixDuplicateKeys = (0, _react.useCallback)(() => {
    onChange(_patch.PatchEvent.from((error.duplicates || []).map(_ref => {
      let [index, key] = _ref;
      return (0, _patch.set)("".concat(key, "_deduped_").concat(index), [index, '_key']);
    })));
  }, [error.duplicates, onChange]);
  return /*#__PURE__*/_react.default.createElement(_formField.FormField, {
    title: error.schemaType.title,
    description: error.schemaType.description
  }, /*#__PURE__*/_react.default.createElement(_Alert.Alert, {
    status: "warning",
    suffix: /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      onClick: handleFixDuplicateKeys,
      text: "Generate unique keys",
      tone: "caution"
    })),
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Non-unique keys")
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Several items in this list share the same identifier (key). Every item must have an unique identifier."), /*#__PURE__*/_react.default.createElement(_Details.Details, {
    marginTop: 4,
    open: _environment.isDev,
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Developer info")
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This usually happens when items are created using an API client, and the", ' ', /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property of each elements has been generated non-uniquely."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "The value of the ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property must be a unique string.")))));
}
