// sanity/lib/cjs/form/inputs/arrays/errors/MissingKeysAlert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MissingKeysAlert = MissingKeysAlert;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _content = require("@sanity/util/content");

var _Details = require("../../../components/Details");

var _environment = require("../../../../environment");

var _Alert = require("../../../components/Alert");

var _patch = require("../../../patch");

var _formField = require("../../../components/formField");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function MissingKeysAlert(props) {
  const {
    error,
    onChange
  } = props;
  const handleFixMissingKeys = (0, _react.useCallback)(() => {
    onChange(_patch.PatchEvent.from((error.value || []).map((val, i) => (0, _patch.setIfMissing)((0, _content.randomKey)(), [i, '_key']))));
  }, [error.value, onChange]);
  return /*#__PURE__*/_react.default.createElement(_formField.FormField, {
    title: error.schemaType.title,
    description: error.schemaType.description
  }, /*#__PURE__*/_react.default.createElement(_Alert.Alert, {
    status: "warning",
    suffix: /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      onClick: handleFixMissingKeys,
      text: "Add missing keys",
      tone: "caution"
    })),
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Missing keys")
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Some items in the list are missing their keys. This must be fixed in order to edit the list."), /*#__PURE__*/_react.default.createElement(_Details.Details, {
    marginTop: 4,
    open: _environment.isDev,
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Developer info")
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This usually happens when items are created using an API client, and the", ' ', /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property has not been included."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "The value of the ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property must be a unique string.")))));
}
