// sanity/lib/esm/components/collapseMenu/CollapseMenu.js
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { EllipsisVerticalIcon } from '@sanity/icons';
import { Box, Button, Flex, Text, Tooltip, useElementRect } from '@sanity/ui';
import React, { cloneElement, forwardRef, useCallback, useMemo, useState } from 'react';
import styled, { css } from 'styled-components';
import { CollapseOverflowMenu } from './CollapseOverflowMenu';
import { ObserveElement } from './ObserveElement';
import { CollapseMenuDivider } from './CollapseMenuDivider';
const FOCUS_RING_PADDING = 3;
const OPTION_STYLE = css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  list-style: none;\n  display: flex;\n  white-space: nowrap;\n\n  &[data-hidden='true'] {\n    opacity: 0;\n    visibility: hidden;\n  }\n"])));
const OuterFlex = styled(Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  padding: ", "px;\n  margin: -", "px;\n  box-sizing: border-box;\n"])), FOCUS_RING_PADDING, FOCUS_RING_PADDING);
const RootFlex = styled(Flex)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  border-radius: inherit;\n  position: relative;\n"])));
const RowFlex = styled(Flex)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  width: max-content;\n\n  &[data-hidden='true'] {\n    height: 0px;\n    visibility: hidden;\n  }\n"])));
const OptionObserveElement = styled(ObserveElement)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  ", "\n"])), OPTION_STYLE);
const OptionHiddenFlex = styled(Flex)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n  ", "\n"])), OPTION_STYLE);

function _isReactElement(node) {
  return Boolean(node);
}

export const CollapseMenu = /*#__PURE__*/forwardRef(function CollapseMenu(props, ref) {
  const {
    children: childrenProp,
    collapsed,
    collapseText = true,
    disableRestoreFocusOnClose,
    gap,
    menuButtonProps,
    onMenuClose
  } = props;
  const [rootEl, setRootEl] = useState(null);
  const [hiddenRowEl, setHiddenRowEl] = useState(null);
  const rootRect = useElementRect(rootEl);
  const [menuOptions, setMenuOptions] = useState([]);
  const hasOverflow = useMemo(() => {
    if (rootRect && rootEl && hiddenRowEl) {
      return rootRect.width < hiddenRowEl.scrollWidth;
    }

    return false;
  }, [hiddenRowEl, rootEl, rootRect]);
  const menuButton = useMemo(() => (menuButtonProps === null || menuButtonProps === void 0 ? void 0 : menuButtonProps.button) || /*#__PURE__*/React.createElement(Button, {
    icon: EllipsisVerticalIcon,
    mode: "bleed"
  }), [menuButtonProps]);
  const intersectionOptions = useMemo(() => ({
    root: rootEl,
    threshold: 1,
    rootMargin: '2px'
  }), [rootEl]);
  const children = useMemo(() => React.Children.toArray(childrenProp).filter(_isReactElement), [childrenProp]);
  const menuOptionsArray = useMemo( // eslint-disable-next-line max-nested-callbacks
  () => children.filter(_ref => {
    let {
      key
    } = _ref;
    return menuOptions.find(o => o.key === key);
  }), [children, menuOptions]);
  const menuIsVisible = useMemo(() => collapsed || menuOptionsArray.length > 0, [collapsed, menuOptionsArray.length]);
  const isInMenu = useCallback(childKey => menuOptionsArray.some(o => o.key === childKey), [menuOptionsArray]);
  const handleIntersection = useCallback((e, child) => {
    const exists = isInMenu(child.key);

    if (!e.isIntersecting && !exists) {
      setMenuOptions(prev => [child, ...prev]);
    }

    if (e.isIntersecting && exists) {
      const updatedOptions = menuOptionsArray.filter(_ref2 => {
        let {
          key
        } = _ref2;
        return key !== child.key;
      });
      setMenuOptions(updatedOptions);
    }
  }, [isInMenu, menuOptionsArray]);
  const items = useMemo(() => children.map(child => {
    const {
      collapsedProps,
      expandedProps
    } = child.props;
    const modeProps = hasOverflow ? collapsedProps : expandedProps;
    const text = hasOverflow && collapseText ? undefined : child.props.text;
    return /*#__PURE__*/cloneElement(child, { ...modeProps,
      text: text
    });
  }), [children, collapseText, hasOverflow]);

  if (collapsed) {
    return /*#__PURE__*/React.createElement(CollapseOverflowMenu, {
      disableRestoreFocusOnClose: disableRestoreFocusOnClose,
      menuButton: menuButton,
      menuButtonProps: menuButtonProps,
      menuOptionsArray: children,
      onMenuClose: onMenuClose
    });
  }

  return /*#__PURE__*/React.createElement(OuterFlex, {
    align: "center",
    "data-ui": "CollapseMenu",
    overflow: "hidden",
    sizing: "border",
    ref: ref
  }, /*#__PURE__*/React.createElement(RootFlex, {
    direction: "column",
    flex: 1,
    justify: "center",
    ref: setRootEl
  }, /*#__PURE__*/React.createElement(RowFlex, {
    gap: gap
  }, items.map((child, index) => {
    const {
      dividerBefore,
      tooltipText = '',
      tooltipProps = {}
    } = child.props;
    const hidden = isInMenu(child.key);
    return /*#__PURE__*/React.createElement(React.Fragment, {
      key: child.key
    }, dividerBefore && index !== 0 && /*#__PURE__*/React.createElement(CollapseMenuDivider, {
      hidden: hidden
    }), /*#__PURE__*/React.createElement(OptionObserveElement, {
      options: intersectionOptions // eslint-disable-next-line react/jsx-no-bind
      ,
      callback: e => handleIntersection(e[0], child),
      "aria-hidden": hidden,
      "data-hidden": hidden
    }, /*#__PURE__*/React.createElement(Tooltip, _extends({
      portal: true,
      disabled: !tooltipText,
      content: /*#__PURE__*/React.createElement(Box, {
        padding: 2,
        sizing: "border"
      }, /*#__PURE__*/React.createElement(Text, {
        size: 1
      }, tooltipText))
    }, tooltipProps), /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/cloneElement(child, {
      disabled: child.props.disabled || hidden,
      'aria-hidden': hidden
    })))));
  })), /*#__PURE__*/React.createElement(RowFlex, {
    "data-hidden": true,
    "aria-hidden": "true",
    gap: gap,
    ref: setHiddenRowEl
  }, children.map((child, index) => {
    const {
      dividerBefore
    } = child.props;
    return /*#__PURE__*/React.createElement(React.Fragment, {
      key: child.key
    }, dividerBefore && index !== 0 && /*#__PURE__*/React.createElement(CollapseMenuDivider, null), /*#__PURE__*/React.createElement(OptionHiddenFlex, {
      key: child.key
    }, /*#__PURE__*/cloneElement(child, {
      disabled: true,
      'aria-hidden': true
    })));
  }))), menuIsVisible && /*#__PURE__*/React.createElement(Flex, {
    marginLeft: gap
  }, /*#__PURE__*/React.createElement(CollapseOverflowMenu, {
    disableRestoreFocusOnClose: disableRestoreFocusOnClose,
    menuButton: menuButton,
    menuButtonProps: menuButtonProps,
    menuOptionsArray: menuOptionsArray,
    onMenuClose: onMenuClose
  })));
});
