// sanity/lib/esm/components/previews/_common/Media.js
import { Text } from '@sanity/ui';
import React, { createElement, isValidElement } from 'react';
import { isValidElementType } from 'react-is';
import { MediaWrapper } from './Media.styled';
export function Media(props) {
  const {
    border = true,
    dimensions,
    layout,
    media,
    radius = 1,
    responsive = false,
    styles
  } = props;
  return /*#__PURE__*/React.createElement(MediaWrapper, {
    $dimensions: dimensions,
    $layout: layout,
    $radius: radius,
    $responsive: responsive,
    className: styles === null || styles === void 0 ? void 0 : styles.media,
    "data-testid": "Media"
  }, renderMedia({
    dimensions,
    layout,
    media
  }), border && /*#__PURE__*/React.createElement("span", null));
}

function renderMedia(props) {
  const {
    dimensions,
    layout,
    media,
    styles
  } = props;

  if (isValidElementType(media)) {
    return /*#__PURE__*/createElement(media, {
      dimensions,
      layout
    });
  }

  if (typeof media === 'string') {
    return /*#__PURE__*/React.createElement(Text, {
      as: "span",
      className: styles === null || styles === void 0 ? void 0 : styles.mediaString
    }, media);
  }

  if ( /*#__PURE__*/isValidElement(media)) {
    return media;
  }

  return null;
}
