// sanity/lib/cjs/form/inputs/DateInputs/base/DatePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePicker = void 0;

var _react = _interopRequireDefault(require("react"));

var _Calendar = require("./calendar/Calendar");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const DatePicker = /*#__PURE__*/_react.default.forwardRef(function DatePicker(props, ref) {
  const {
    value = new Date(),
    onChange,
    ...rest
  } = props;

  const [focusedDate, setFocusedDay] = _react.default.useState();

  const handleSelect = _react.default.useCallback(nextDate => {
    onChange(nextDate);
    setFocusedDay(undefined);
  }, [onChange]);

  return /*#__PURE__*/_react.default.createElement(_Calendar.Calendar, _extends({}, rest, {
    ref: ref,
    selectedDate: value,
    onSelect: handleSelect,
    focusedDate: focusedDate || value,
    onFocusedDateChange: setFocusedDay
  }));
});

exports.DatePicker = DatePicker;
