// sanity/lib/esm/studio/workspaceLoader/ErrorMessage.js
import _startCase from "lodash/startCase";

var _templateObject, _templateObject2;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React, { useEffect } from 'react';
import { Flex, Box, Card, Code, Text } from '@sanity/ui';
import styled from 'styled-components';
const ListItem = styled(Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral([""])));
const ErrorMessageRoot = styled(Box).attrs({
  padding: 4
})(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral([""])));
export function ErrorMessage(_ref) {
  let {
    error,
    message,
    path,
    stack
  } = _ref;
  useEffect(() => {
    console.error(error);
  }, [error]);
  const last = path[path.length - 1];
  return /*#__PURE__*/React.createElement(ErrorMessageRoot, {
    forwardedAs: Flex,
    direction: "column",
    gap: 4
  }, /*#__PURE__*/React.createElement(Flex, {
    direction: "column",
    gap: 2
  }, /*#__PURE__*/React.createElement(Text, {
    weight: "semibold",
    size: 3
  }, _startCase(last.type), " Error")), /*#__PURE__*/React.createElement(Card, {
    tone: "critical",
    overflow: "auto",
    padding: 4
  }, /*#__PURE__*/React.createElement(Code, null, message)), /*#__PURE__*/React.createElement(Flex, {
    as: "ul",
    direction: "column",
    gap: 2
  }, path.map((_ref2, index) => {
    let {
      name,
      type
    } = _ref2;
    return /*#__PURE__*/React.createElement(ListItem, {
      forwardedAs: "li",
      gap: 2,
      align: "center",
      key: index
    }, /*#__PURE__*/React.createElement(Box, null, /*#__PURE__*/React.createElement(Code, null, name)), /*#__PURE__*/React.createElement(Box, null, /*#__PURE__*/React.createElement(Text, {
      muted: true,
      size: 1
    }, type)));
  })), stack && /*#__PURE__*/React.createElement("details", null, /*#__PURE__*/React.createElement(Text, {
    as: "summary"
  }, "Stack Trace"), /*#__PURE__*/React.createElement(Box, {
    overflow: "auto",
    marginTop: 4
  }, /*#__PURE__*/React.createElement(Code, null, stack))));
}
