// sanity/lib/dts/src/datastores/grants/debug/exampleGrants.js
export const administrator = [
    {
        filter: '_id in path("**")',
        permissions: ['read', 'create', 'history', 'update'],
    },
];
export const editor = [
    {
        filter: '_id in path("**")',
        permissions: ['read', 'create', 'history', 'update'],
    },
];
export const developer = [
    {
        filter: '_id in path("**")',
        permissions: ['read', 'create', 'history', 'update'],
    },
];
export const contributor = [
    {
        filter: '_id in path("**")',
        permissions: ['read'],
    },
    {
        filter: '_id in path("drafts.**")',
        permissions: ['create', 'history', 'update'],
    },
];
export const viewer = [
    {
        filter: '_id in path("**")',
        permissions: ['read', 'history'],
    },
];
export const requiresApproval = [
    {
        filter: '!locked',
        permissions: ['read', 'create', 'update'],
    },
];
export const restricted = [
    {
        filter: '_id in path("drafts.**") && _type in ["stringsTest", "book", "author", "referenceTest"]',
        permissions: ['read', 'create', 'history', 'update'],
    },
    {
        filter: '_id in path("**")',
        permissions: ['read'],
    },
];
