// sanity/lib/cjs/cli/commands/dataset/alias/datasetAliasesClient.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ALIAS_PREFIX = void 0;
exports.createAlias = createAlias;
exports.listAliases = listAliases;
exports.removeAlias = removeAlias;
exports.unlinkAlias = unlinkAlias;
exports.updateAlias = updateAlias;

var _validateDatasetAliasName = require("../../../actions/dataset/alias/validateDatasetAliasName");

const ALIAS_PREFIX = '~';
exports.ALIAS_PREFIX = ALIAS_PREFIX;

function listAliases(client) {
  return client.request({
    uri: '/aliases'
  });
}

function createAlias(client, aliasName, datasetName) {
  return modify(client, 'PUT', aliasName, datasetName ? {
    datasetName
  } : undefined);
}

function updateAlias(client, aliasName, datasetName) {
  return modify(client, 'PATCH', aliasName, datasetName ? {
    datasetName
  } : undefined);
}

function unlinkAlias(client, aliasName) {
  (0, _validateDatasetAliasName.validateDatasetAliasName)(aliasName);
  return modify(client, 'PATCH', "".concat(aliasName, "/unlink"), {});
}

function removeAlias(client, aliasName) {
  return modify(client, 'DELETE', aliasName);
}

function modify(client, method, aliasName, body) {
  return client.request({
    method,
    uri: "/aliases/".concat(aliasName),
    body
  });
}
