// @sanity/server/lib/vite/plugin-sanity-dot-workaround.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sanityDotWorkaroundPlugin = sanityDotWorkaroundPlugin;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _connectHistoryApiFallback = _interopRequireDefault(require("connect-history-api-fallback"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This is a Vite plugin for supporting locations containing `.` in their pathname.
 *
 * @see https://github.com/vitejs/vite/issues/2245
 */
function sanityDotWorkaroundPlugin() {
  return {
    name: '@sanity/server/dot-workaround',

    configureServer(server) {
      const {
        root
      } = server.config;
      return () => {
        const handler = (0, _connectHistoryApiFallback.default)({
          disableDotRule: true,
          rewrites: [{
            from: /\/index.html$/,
            to: _ref => {
              let {
                parsedUrl
              } = _ref;
              const pathname = parsedUrl.pathname;

              if (pathname && _fs.default.existsSync(_path.default.join(root, pathname))) {
                return pathname;
              }

              return "/index.html";
            }
          }]
        });
        server.middlewares.use((req, res, next) => {
          handler(req, res, next);
        });
      };
    }

  };
}
