// sanity/lib/esm/form/studio/DefaultAssetSource/AssetMenu.js
import React from 'react';
import { LinkIcon, EllipsisVerticalIcon, TrashIcon } from '@sanity/icons';
import { Button, Menu, MenuItem, MenuButton } from '@sanity/ui';
export function AssetMenu(_ref) {
  let {
    isSelected,
    border = true,
    onAction
  } = _ref;
  const triggerButtonMode = isSelected ? 'default' : 'bleed';
  const triggerButtonTone = isSelected ? 'primary' : 'default';
  return /*#__PURE__*/React.createElement(MenuButton, {
    button: /*#__PURE__*/React.createElement(Button, {
      padding: 2,
      mode: border ? 'ghost' : triggerButtonMode,
      icon: EllipsisVerticalIcon,
      tone: border ? 'default' : triggerButtonTone
    }),
    id: "asset-menu",
    portal: true,
    menu: /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(MenuItem, {
      text: "Show uses",
      icon: LinkIcon,
      onClick: () => {
        onAction({
          type: 'showUsage'
        });
      }
    }), /*#__PURE__*/React.createElement(MenuItem, {
      text: "Delete",
      icon: TrashIcon,
      tone: "critical",
      onClick: () => {
        onAction({
          type: 'delete'
        });
      }
    })),
    placement: "right"
  });
}
