// sanity/lib/esm/form/studio/DefaultAssetSource/ConfirmMessage.js
import React from 'react';
import { WarningOutlineIcon } from '@sanity/icons';
import { Card, Text, Grid, Flex } from '@sanity/ui';
const STYLE_ASSET_IMAGE = {
  maxWidth: '100%',
  height: '120px',
  objectFit: 'contain',
  objectPosition: 'center'
};
const STYLE_CONFIRM_CARD = {
  gridColumn: 'span 1',
  overflow: 'hidden',
  display: 'flex',
  alignSelf: 'center',
  justifyContent: 'center'
};
const STYLE_IMAGE_WRAPPER = {
  height: '100%'
};
export const ConfirmMessage = _ref => {
  let {
    asset,
    assetType,
    hasResults = false
  } = _ref;
  const isImage = assetType === 'image';
  const filenamePlaceholder = asset.originalFilename ? /*#__PURE__*/React.createElement(React.Fragment, null, "The ", assetType, " ", /*#__PURE__*/React.createElement("strong", null, asset.originalFilename)) : "this ".concat(assetType);

  if (hasResults) {
    return /*#__PURE__*/React.createElement(Card, {
      tone: "caution",
      padding: [3, 3, 4],
      border: true,
      radius: 2,
      marginBottom: 3
    }, /*#__PURE__*/React.createElement(Grid, {
      columns: 3,
      gap: [2, 3, 4]
    }, /*#__PURE__*/React.createElement(Flex, {
      gap: [3, 4],
      align: "center",
      style: {
        gridColumn: isImage ? 'span 2' : 'span 3'
      }
    }, /*#__PURE__*/React.createElement(Text, null, /*#__PURE__*/React.createElement(WarningOutlineIcon, null)), /*#__PURE__*/React.createElement(Text, {
      size: 1
    }, filenamePlaceholder, " cannot be deleted because it's being used. In order to delete the ", assetType, " you need the remove all uses of it.")), isImage && /*#__PURE__*/React.createElement(Card, {
      __unstable_checkered: true,
      border: true,
      radius: 1,
      style: STYLE_CONFIRM_CARD
    }, /*#__PURE__*/React.createElement(Flex, {
      align: "center",
      justify: "center",
      style: STYLE_IMAGE_WRAPPER
    }, /*#__PURE__*/React.createElement("img", {
      src: "".concat(asset.url, "?w=200"),
      style: STYLE_ASSET_IMAGE,
      alt: "Preview of image",
      referrerPolicy: "strict-origin-when-cross-origin"
    })))));
  }

  return /*#__PURE__*/React.createElement(Card, {
    paddingX: [2, 3, 4],
    paddingY: [3, 3, 3, 4]
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: 3,
    gap: 3
  }, /*#__PURE__*/React.createElement(Flex, {
    style: {
      gridColumn: isImage ? 'span 2' : 'span 3'
    },
    align: "center"
  }, /*#__PURE__*/React.createElement(Text, null, "You are about to delete the ", assetType, asset.originalFilename && /*#__PURE__*/React.createElement(React.Fragment, null, ' ', /*#__PURE__*/React.createElement("strong", null, asset.originalFilename)), ' ', "and its metadata.", /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("br", null), "Are you sure?")), isImage && /*#__PURE__*/React.createElement(Card, {
    __unstable_checkered: true,
    border: true,
    radius: 1,
    style: STYLE_CONFIRM_CARD
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center",
    justify: "center",
    style: STYLE_IMAGE_WRAPPER
  }, /*#__PURE__*/React.createElement("img", {
    src: "".concat(asset.url, "?w=200"),
    style: STYLE_ASSET_IMAGE,
    alt: "Preview of image",
    referrerPolicy: "strict-origin-when-cross-origin"
  })))));
};
