// sanity/lib/esm/form/studio/DefaultAssetSource/DefaultSource.js
import _uniqueId from "lodash/uniqueId";

var _templateObject, _templateObject2;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React, { useState, useRef, useCallback, useMemo, useEffect } from 'react';
import { DownloadIcon } from '@sanity/icons';
import { Box, Button, Card, Dialog, Flex, Grid, Spinner, Text } from '@sanity/ui';
import styled from 'styled-components';
import { useClient } from '../../../hooks';
import { AssetThumb } from './AssetThumb';
import { TableList } from './TableList';
const PER_PAGE = 200;
const ASSET_TYPE_IMAGE = 'sanity.imageAsset';
const ASSET_TYPE_FILE = 'sanity.fileAsset';

const buildQuery = function () {
  let start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  let end = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : PER_PAGE;
  let assetType = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : ASSET_TYPE_IMAGE;
  return "\n  *[_type == \"".concat(assetType, "\"] | order(_updatedAt desc) [").concat(start, "...").concat(end, "] {\n    _id,\n    _updatedAt,\n    _createdAt,\n    url,\n    originalFilename,\n    mimeType,\n    extension,\n    size,\n    metadata {dimensions}\n  }\n");
};

const ThumbGrid = styled(Grid)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  grid-template-columns: repeat(auto-fill, minmax(120px, 1fr));\n"])));
const CardLoadMore = styled(Card)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  border-top: 1px solid var(--card-border-color);\n  position: sticky;\n  bottom: 0;\n  z-index: 200;\n"])));

const DefaultAssetSource = function DefaultAssetSource(props, ref) {
  const client = useClient();
  const versionedClient = useMemo(() => client.withConfig({
    apiVersion: '1'
  }), [client]);

  const _elementId = useRef("default-asset-source-".concat(_uniqueId()));

  const currentPageNumber = useRef(0);
  const fetch$ = useRef();
  const [assets, setAssets] = useState([]);
  const [isLastPage, setIsLastPage] = useState(false);
  const [hasResetAutoFocus, setHasResetFocus] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const {
    selectedAssets,
    assetType = 'image',
    dialogHeaderTitle = 'Select image',
    onClose,
    onSelect
  } = props;
  const fetchPage = useCallback(pageNumber => {
    const start = pageNumber * PER_PAGE;
    const end = start + PER_PAGE;
    const isImageAssetType = assetType === 'image';
    const tag = isImageAssetType ? 'asset.image-list' : 'asset.file-list';
    const assetTypeParam = isImageAssetType ? ASSET_TYPE_IMAGE : ASSET_TYPE_FILE;
    setIsLoading(true);
    fetch$.current = versionedClient.observable.fetch(buildQuery(start, end, assetTypeParam), {}, {
      tag
    }).subscribe(result => {
      setIsLastPage(result.length < PER_PAGE); // eslint-disable-next-line max-nested-callbacks

      setAssets(prevState => prevState.concat(result));
      setIsLoading(false);
    });
  }, [assetType, setIsLoading, setAssets, setIsLastPage, versionedClient]);
  const handleDeleteFinished = useCallback(id => {
    // eslint-disable-next-line max-nested-callbacks
    setAssets(prevState => prevState.filter(asset => asset._id !== id));
  }, [setAssets]);
  const select = useCallback(id => {
    const selected = assets.find(doc => doc._id === id);

    if (selected) {
      const selectedSource = [{
        kind: 'assetDocumentId',
        value: id
      }];
      onSelect(selectedSource);
    }
  }, [assets, onSelect]);
  const handleItemClick = useCallback(event => {
    event.preventDefault();
    select(event.currentTarget.getAttribute('data-id'));
  }, [select]);
  const handleItemKeyPress = useCallback(event => {
    if (event.key === 'Enter') {
      event.preventDefault();
      select(event.currentTarget.getAttribute('data-id'));
    }
  }, [select]);
  const handleClose = useCallback(() => {
    if (onClose) {
      onClose();
    }
  }, [onClose]);
  const handleFetchNextPage = useCallback(event => {
    event.preventDefault();
    fetchPage(++currentPageNumber.current);
  }, [fetchPage]);
  useEffect(() => {
    fetchPage(currentPageNumber.current);
    return () => {
      if (fetch$.current) {
        fetch$.current.unsubscribe();
      }
    };
  }, [fetchPage]);
  useEffect(() => {
    // We focus on the first item after we're doing loading, but only on initial load, as
    // this will reset the scroll position to the top if we do it on the second page
    if (!isLoading && (!currentPageNumber.current || currentPageNumber.current === 0)) {
      setHasResetFocus(true);
    }
  }, [isLoading]);
  const renderedThumbView = useMemo(() => {
    return /*#__PURE__*/React.createElement(Box, {
      padding: 4
    }, /*#__PURE__*/React.createElement(ThumbGrid, {
      gap: 2
    }, assets.map(asset => /*#__PURE__*/React.createElement(AssetThumb, {
      key: asset._id,
      asset: asset,
      isSelected: selectedAssets.some(selected => selected._id === asset._id),
      onClick: handleItemClick,
      onKeyPress: handleItemKeyPress,
      onDeleteFinished: handleDeleteFinished
    }))), isLoading && assets.length === 0 && /*#__PURE__*/React.createElement(Flex, {
      justify: "center"
    }, /*#__PURE__*/React.createElement(Spinner, {
      muted: true
    })), !isLoading && assets.length === 0 && /*#__PURE__*/React.createElement(Text, {
      align: "center",
      muted: true
    }, "No images"));
  }, [assets, handleDeleteFinished, handleItemClick, handleItemKeyPress, isLoading, selectedAssets]);
  const renderedTableView = useMemo(() => {
    return /*#__PURE__*/React.createElement(TableList, {
      isLoading: isLoading,
      assets: assets,
      selectedAssets: selectedAssets,
      onClick: handleItemClick,
      onKeyPress: handleItemKeyPress,
      onDeleteFinished: handleDeleteFinished
    });
  }, [isLoading, assets, selectedAssets, handleItemClick, handleItemKeyPress, handleDeleteFinished]);
  return /*#__PURE__*/React.createElement(Dialog, {
    ref: ref,
    id: _elementId.current,
    header: dialogHeaderTitle,
    width: 2,
    onClose: handleClose,
    __unstable_autoFocus: hasResetAutoFocus
  }, assetType === 'image' && renderedThumbView, assetType === 'file' && renderedTableView, assets.length > 0 && !isLastPage && /*#__PURE__*/React.createElement(CardLoadMore, {
    tone: "default",
    padding: 4
  }, /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, /*#__PURE__*/React.createElement(Button, {
    type: "button",
    icon: DownloadIcon,
    loading: isLoading,
    onClick: handleFetchNextPage,
    text: "Load more",
    tone: "primary"
  }))));
};

export const DefaultSource = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(DefaultAssetSource));
