// sanity/lib/esm/form/studio/DefaultAssetSource/TableList.js
import React from 'react';
import { Box, Card, Flex, Grid, Label, Spinner, Stack, useMediaIndex } from '@sanity/ui';
import { AssetRow } from './AssetRow';
const STYLES_FILENAME = {
  paddingLeft: '2.2rem'
};
const STYLES_GRID = {
  gridTemplateColumns: '3fr 1fr 1fr 2fr 30px'
};
export function TableList(props) {
  const mediaIndex = useMediaIndex();
  const isMobile = mediaIndex < 2;
  const {
    assets,
    onClick,
    onKeyPress,
    onDeleteFinished,
    selectedAssets,
    isLoading
  } = props;
  return /*#__PURE__*/React.createElement(Box, {
    padding: 4
  }, /*#__PURE__*/React.createElement(Card, {
    borderBottom: true,
    paddingBottom: 2,
    marginBottom: 1
  }, isMobile ? /*#__PURE__*/React.createElement(Grid, {
    style: STYLES_GRID
  }, /*#__PURE__*/React.createElement(Box, {
    flex: 2,
    paddingLeft: 5
  }, /*#__PURE__*/React.createElement(Label, {
    muted: true,
    size: 1
  }, "Filename"))) : /*#__PURE__*/React.createElement(Grid, {
    gap: 1,
    style: STYLES_GRID
  }, /*#__PURE__*/React.createElement(Box, {
    flex: 2,
    style: STYLES_FILENAME
  }, /*#__PURE__*/React.createElement(Label, {
    muted: true,
    size: 1
  }, "Filename")), /*#__PURE__*/React.createElement(Box, {
    flex: 1
  }, /*#__PURE__*/React.createElement(Label, {
    muted: true,
    size: 1
  }, "Size")), /*#__PURE__*/React.createElement(Box, {
    flex: 1
  }, /*#__PURE__*/React.createElement(Label, {
    muted: true,
    size: 1
  }, "Type")), /*#__PURE__*/React.createElement(Box, {
    flex: 1
  }, /*#__PURE__*/React.createElement(Label, {
    muted: true,
    size: 1
  }, "Date added")))), /*#__PURE__*/React.createElement(Stack, null, isLoading && assets.length === 0 && /*#__PURE__*/React.createElement(Box, {
    paddingTop: 4,
    paddingBottom: 2
  }, /*#__PURE__*/React.createElement(Flex, {
    justify: "center"
  }, /*#__PURE__*/React.createElement(Spinner, {
    muted: true
  }))), assets.map(asset => /*#__PURE__*/React.createElement(AssetRow, {
    key: asset._id,
    asset: asset,
    isMobile: isMobile,
    isSelected: selectedAssets.some(selected => selected._id === asset._id),
    onClick: onClick,
    onKeyPress: onKeyPress,
    onDeleteFinished: onDeleteFinished
  }))));
}
