// sanity/lib/cjs/presence/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SNAP_TO_DOCK_DISTANCE_TOP = exports.SNAP_TO_DOCK_DISTANCE_BOTTOM = exports.SLIDE_RIGHT_THRESHOLD_TOP = exports.SLIDE_RIGHT_THRESHOLD_BOTTOM = exports.MAX_AVATARS_GLOBAL = exports.MAX_AVATARS_DOCK = exports.INTERSECTION_THRESHOLDS = exports.INTERSECTION_ELEMENT_PADDING = exports.DISABLE_OVERLAY = exports.DEFAULT_MAX_AVATARS_FIELDS = exports.DEBUG = exports.AVATAR_SIZE = exports.AVATAR_DISTANCE = exports.AVATAR_ARROW_HEIGHT = void 0;
// disable overlay rendering entirely and only display presence in fields
const DISABLE_OVERLAY = false; // number of avatars to display on each fields (will be stacked and have a counter if more than 3 users in a field)

exports.DISABLE_OVERLAY = DISABLE_OVERLAY;
const MAX_AVATARS_DOCK = 3;
exports.MAX_AVATARS_DOCK = MAX_AVATARS_DOCK;
const MAX_AVATARS_GLOBAL = 3;
exports.MAX_AVATARS_GLOBAL = MAX_AVATARS_GLOBAL;
const DEFAULT_MAX_AVATARS_FIELDS = 3; // distance between avatars in pixels (negative number means overlap)

exports.DEFAULT_MAX_AVATARS_FIELDS = DEFAULT_MAX_AVATARS_FIELDS;
const AVATAR_DISTANCE = -4; // height/width of the avatars in pixels

exports.AVATAR_DISTANCE = AVATAR_DISTANCE;
const AVATAR_SIZE = 23;
exports.AVATAR_SIZE = AVATAR_SIZE;
const AVATAR_ARROW_HEIGHT = 4; // these are the thresholds in which the intersection observers will notify about regions going in/out of viewport
// these will affect the thresholds below, so make sure to coordinate edits

exports.AVATAR_ARROW_HEIGHT = AVATAR_ARROW_HEIGHT;
const INTERSECTION_THRESHOLDS = [0, 0.25, 0.75, 1]; // The elements that reports intersection changes are padded top+bottom in order to detect when entering leaving

exports.INTERSECTION_THRESHOLDS = INTERSECTION_THRESHOLDS;
const INTERSECTION_ELEMENT_PADDING = 23; // Make the avatars snap to the top dock when they are closer than this from the top

exports.INTERSECTION_ELEMENT_PADDING = INTERSECTION_ELEMENT_PADDING;
const SNAP_TO_DOCK_DISTANCE_TOP = 8; // Make the avatars snap to the bottom dock when they are closer than this from the bottom

exports.SNAP_TO_DOCK_DISTANCE_TOP = SNAP_TO_DOCK_DISTANCE_TOP;
const SNAP_TO_DOCK_DISTANCE_BOTTOM = 8; // The avatar will move to the right when this close (in pixels) to the top

exports.SNAP_TO_DOCK_DISTANCE_BOTTOM = SNAP_TO_DOCK_DISTANCE_BOTTOM;
const SLIDE_RIGHT_THRESHOLD_TOP = 20; // The avatar will move to the right when this close (in pixels) to the bottom

exports.SLIDE_RIGHT_THRESHOLD_TOP = SLIDE_RIGHT_THRESHOLD_TOP;
const SLIDE_RIGHT_THRESHOLD_BOTTOM = 20; // Switch on debug mode (will display regions)

exports.SLIDE_RIGHT_THRESHOLD_BOTTOM = SLIDE_RIGHT_THRESHOLD_BOTTOM;
const DEBUG = false;
exports.DEBUG = DEBUG;
