// sanity/lib/esm/desk/panes/list/ListPane.js
var _templateObject;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React from 'react';
import { Box, Button, Card, Code, Stack } from '@sanity/ui';
import styled from 'styled-components';
import { ArrowLeftIcon } from '@sanity/icons';
import { Pane, PaneContent, PaneHeader, usePaneLayout, PaneHeaderActions, BackLink, PaneItem } from '../../components';
import { _DEBUG } from '../../constants';
import { useDeskTool } from '../../useDeskTool';
const Divider = styled.hr(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  background-color: var(--card-border-color);\n  height: 1px;\n  margin: 0;\n  border: none;\n"])));
/**
 * @internal
 */

export function ListPane(props) {
  const {
    childItemId,
    index,
    isActive,
    isSelected,
    pane,
    paneKey
  } = props;
  const {
    features
  } = useDeskTool();
  const {
    collapsed: layoutCollapsed
  } = usePaneLayout();
  const {
    defaultLayout,
    displayOptions,
    items,
    menuItems,
    menuItemGroups,
    title
  } = pane;
  const paneShowIcons = displayOptions === null || displayOptions === void 0 ? void 0 : displayOptions.showIcons;

  const shouldShowIconForItem = item => {
    var _item$displayOptions;

    const itemShowIcon = (_item$displayOptions = item.displayOptions) === null || _item$displayOptions === void 0 ? void 0 : _item$displayOptions.showIcon; // Specific true/false on item should have precedence over list setting

    if (typeof itemShowIcon !== 'undefined') {
      return itemShowIcon !== false; // Boolean(item.icon)
    } // If no item setting is defined, defer to the pane settings


    return paneShowIcons !== false; // Boolean(item.icon)
  };

  return /*#__PURE__*/React.createElement(Pane, {
    currentMaxWidth: 350,
    "data-testid": "desk-tool-list-pane",
    id: paneKey,
    maxWidth: 640,
    minWidth: 320,
    selected: isSelected
  }, _DEBUG && /*#__PURE__*/React.createElement(Card, {
    padding: 4,
    tone: "transparent"
  }, /*#__PURE__*/React.createElement(Code, null, pane.source || '(none)')), /*#__PURE__*/React.createElement(PaneHeader, {
    actions: /*#__PURE__*/React.createElement(PaneHeaderActions, {
      menuItems: menuItems,
      menuItemGroups: menuItemGroups
    }),
    backButton: features.backButton && index > 0 && /*#__PURE__*/React.createElement(Button, {
      as: BackLink,
      "data-as": "a",
      icon: ArrowLeftIcon,
      mode: "bleed"
    }),
    title: title
  }), /*#__PURE__*/React.createElement(PaneContent, {
    overflow: layoutCollapsed ? undefined : 'auto'
  }, /*#__PURE__*/React.createElement(Stack, {
    padding: 2,
    space: 1
  }, items && items.map((item, itemIndex) => {
    if (item.type === 'divider') {
      return (
        /*#__PURE__*/
        // eslint-disable-next-line react/no-array-index-key
        React.createElement(Box, {
          key: "divider-".concat(itemIndex),
          paddingY: 1
        }, /*#__PURE__*/React.createElement(Divider, null))
      );
    }

    const pressed = !isActive && childItemId === item.id;
    const selected = isActive && childItemId === item.id;
    return /*#__PURE__*/React.createElement(PaneItem, {
      icon: shouldShowIconForItem(item) ? item.icon : false,
      id: item.id,
      key: item.id,
      layout: defaultLayout,
      pressed: pressed,
      schemaType: item.schemaType,
      selected: selected,
      title: item.title,
      value: // If this is a document list item, pass on the ID and type,
      // otherwise leave it undefined to use the passed title and gang
      item._id && item.schemaType ? {
        _id: item._id,
        _type: item.schemaType.name,
        title: item.title
      } : undefined
    });
  }))));
}
