// sanity/lib/cjs/desk/components/pane/__workshop__/ResizeStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ResizeStory;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _usePaneLayout = require("../usePaneLayout");

var _PaneDivider = require("../PaneDivider");

var _PaneLayoutContext = require("../PaneLayoutContext");

var _constants = require("../constants");

var _paneLayoutController = require("../paneLayoutController");

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const PaneLayoutRoot = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-resizing] {\n    cursor: ew-resize;\n  }\n"])));

function ResizeStory() {
  const [rootElement, setRootElement] = (0, _react.useState)(null);
  const controller = (0, _react.useMemo)(() => (0, _paneLayoutController.createPaneLayoutController)(), []);
  const collapsed = false;
  const [state, setState] = (0, _react.useState)({
    expandedElement: null,
    panes: [],
    resizing: false
  });
  const paneLayout = (0, _react.useMemo)(() => ({
    collapse: controller.collapse,
    collapsed,
    expand: controller.expand,
    expandedElement: state.expandedElement,
    mount: controller.mount,
    panes: state.panes,
    resize: controller.resize,
    resizing: state.resizing
  }), [collapsed, controller, state.expandedElement, state.panes, state.resizing]);
  (0, _react.useEffect)(() => controller.subscribe(setState), [controller]);
  (0, _react.useEffect)(() => {
    if (!rootElement) return undefined;
    const ro = new ResizeObserver(entries => {
      controller.setRootWidth(entries[0].contentRect.width);
    });
    ro.observe(rootElement);
    return () => ro.disconnect();
  }, [controller, rootElement]);
  return /*#__PURE__*/_react.default.createElement(_PaneLayoutContext.PaneLayoutContext.Provider, {
    value: paneLayout
  }, /*#__PURE__*/_react.default.createElement(PaneLayoutRoot, {
    "data-resizing": state.resizing ? '' : undefined,
    height: "fill",
    ref: setRootElement
  }, /*#__PURE__*/_react.default.createElement(Pane, {
    currentMinWidth: undefined // currentMaxWidth={320}
    ,
    flex: 1,
    minWidth: 160,
    maxWidth: 320
  }), /*#__PURE__*/_react.default.createElement(Pane, {
    currentMinWidth: undefined // currentMaxWidth={320}
    ,
    flex: 1,
    minWidth: 160,
    maxWidth: 320
  }), /*#__PURE__*/_react.default.createElement(Pane, {
    currentMinWidth: undefined // currentMaxWidth={320}
    ,
    flex: 1,
    minWidth: 160,
    maxWidth: 320
  }), /*#__PURE__*/_react.default.createElement(Pane, {
    currentMinWidth: 320,
    currentMaxWidth: undefined,
    flex: 2,
    minWidth: 160,
    maxWidth: undefined
  })));
}

function Pane(props) {
  var _pane$flex, _pane$currentMinWidth, _pane$currentMaxWidth;

  const {
    currentMinWidth: currentMinWidthProp,
    currentMaxWidth: currentMaxWidthProp,
    flex: flexProp = 1,
    minWidth: minWidthProp,
    maxWidth: maxWidthProp
  } = props;
  const [rootElement, setRootElement] = (0, _react.useState)(null);
  const {
    collapse,
    expand,
    expandedElement,
    mount,
    panes,
    resizing
  } = (0, _usePaneLayout.usePaneLayout)();
  const pane = panes.find(p => p.element === rootElement);
  const paneIndex = pane && panes.indexOf(pane);
  const nextPane = typeof paneIndex === 'number' ? panes[paneIndex + 1] : undefined;
  const isLast = !nextPane;
  const expanded = expandedElement === rootElement;
  const collapsed = (pane === null || pane === void 0 ? void 0 : pane.collapsed) || false;
  const nextCollapsed = (nextPane === null || nextPane === void 0 ? void 0 : nextPane.collapsed) || false;
  const flex = (_pane$flex = pane === null || pane === void 0 ? void 0 : pane.flex) !== null && _pane$flex !== void 0 ? _pane$flex : flexProp;
  const currentMinWidth = (_pane$currentMinWidth = pane === null || pane === void 0 ? void 0 : pane.currentMinWidth) !== null && _pane$currentMinWidth !== void 0 ? _pane$currentMinWidth : currentMinWidthProp;
  const currentMaxWidth = (_pane$currentMaxWidth = pane === null || pane === void 0 ? void 0 : pane.currentMaxWidth) !== null && _pane$currentMaxWidth !== void 0 ? _pane$currentMaxWidth : currentMaxWidthProp;
  (0, _react.useEffect)(() => {
    if (!rootElement) {
      return undefined;
    }

    return mount(rootElement, {
      currentMinWidth: currentMinWidthProp,
      currentMaxWidth: currentMaxWidthProp,
      flex: flexProp,
      id: String(paneIndex),
      minWidth: minWidthProp,
      maxWidth: maxWidthProp
    });
  }, [currentMinWidthProp, currentMaxWidthProp, flexProp, minWidthProp, maxWidthProp, mount, paneIndex, rootElement]);
  const minWidth = (0, _react.useMemo)(() => {
    if (collapsed) return _constants.PANE_COLLAPSED_WIDTH;

    if (currentMinWidth === 0) {
      return minWidthProp || _constants.PANE_DEFAULT_MIN_WIDTH;
    }

    return currentMinWidth || minWidthProp || _constants.PANE_DEFAULT_MIN_WIDTH;
  }, [collapsed, currentMinWidth, minWidthProp]);
  const maxWidth = (0, _react.useMemo)(() => {
    if (collapsed) return _constants.PANE_COLLAPSED_WIDTH;

    if (maxWidthProp !== undefined) {
      return currentMaxWidth !== null && currentMaxWidth !== void 0 ? currentMaxWidth : maxWidthProp;
    }

    return isLast ? undefined : currentMaxWidth;
  }, [collapsed, currentMaxWidth, isLast, maxWidthProp]);
  const handleCollapse = (0, _react.useCallback)(() => {
    if (!rootElement) return;
    collapse(rootElement);
  }, [collapse, rootElement]);
  const handleExpand = (0, _react.useCallback)(() => {
    if (!rootElement) return;
    expand(rootElement);
  }, [expand, rootElement]);
  return /*#__PURE__*/_react.default.createElement(PaneView, {
    collapsed: collapsed,
    currentMinWidth: currentMinWidth,
    currentMaxWidth: currentMaxWidth,
    expanded: expanded,
    flex: flex,
    index: paneIndex,
    isLast: isLast,
    minWidth: minWidth,
    maxWidth: maxWidth,
    nextCollapsed: nextCollapsed,
    onCollapse: handleCollapse,
    onExpand: handleExpand,
    resizing: resizing,
    rootElement: rootElement,
    setRootElement: setRootElement
  });
}

const PaneViewRoot = (0, _styledComponents.default)(_ui.Card)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  box-shadow: 0 0 0 1px var(--card-border-color) !important;\n\n  &[data-resizing] {\n    pointer-events: none;\n  }\n\n  & + & {\n    margin-left: 1px;\n  }\n"])));
const PaneView = /*#__PURE__*/(0, _react.memo)(function PaneView(props) {
  const {
    collapsed,
    currentMinWidth,
    currentMaxWidth,
    expanded,
    flex,
    index,
    isLast,
    minWidth,
    maxWidth,
    nextCollapsed,
    onCollapse,
    onExpand,
    resizing,
    rootElement,
    setRootElement
  } = props;
  const style = (0, _react.useMemo)(() => ({
    flex,
    minWidth,
    maxWidth: maxWidth === Infinity ? undefined : maxWidth
  }), [flex, minWidth, maxWidth]);
  const handleClick = (0, _react.useCallback)(() => {
    if (collapsed) onExpand();else onCollapse();
  }, [collapsed, onCollapse, onExpand]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(PaneViewRoot, {
    "data-as": "button",
    "data-resizing": resizing ? '' : undefined,
    id: "pane-".concat(index),
    onClick: handleClick,
    overflow: "hidden",
    ref: setRootElement,
    style: style,
    tone: expanded ? 'primary' : undefined
  }, !collapsed && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
    size: 1
  }, ["#".concat(index), "collapsed=".concat(collapsed), "currentMinWidth=".concat(currentMinWidth), "currentMaxWidth=".concat(currentMaxWidth), "flex=".concat(flex), "minWidth=".concat(minWidth), "maxWidth=".concat(maxWidth)].join('\n')))), !isLast && /*#__PURE__*/_react.default.createElement(_PaneDivider.PaneDivider, {
    disabled: collapsed || nextCollapsed,
    element: rootElement
  }));
});
