// sanity/lib/esm/form/inputs/ReferenceInput/OptionPreview.js
/* eslint-disable max-nested-callbacks */

/* eslint-disable no-nested-ternary */
import React from 'react';
import { Stack, Text, TextSkeleton } from '@sanity/ui';
import { Alert } from '../../components/Alert';
import { useReferenceInfo } from './useReferenceInfo';
import { ReferencePreview } from './ReferencePreview';
/**
 * Used to preview a referenced type
 * Takes the reference type as props
 */

export function OptionPreview(props) {
  const {
    getReferenceInfo,
    id: documentId,
    renderPreview
  } = props;
  const {
    isLoading,
    result: referenceInfo,
    error
  } = useReferenceInfo(documentId, getReferenceInfo);

  if (isLoading) {
    return /*#__PURE__*/React.createElement(Stack, {
      space: 2,
      padding: 1
    }, /*#__PURE__*/React.createElement(TextSkeleton, {
      style: {
        maxWidth: 320
      },
      radius: 1,
      animated: true
    }), /*#__PURE__*/React.createElement(TextSkeleton, {
      style: {
        maxWidth: 200
      },
      radius: 1,
      size: 1,
      animated: true
    }));
  }

  if (error) {
    return /*#__PURE__*/React.createElement(Stack, {
      space: 2,
      padding: 1
    }, /*#__PURE__*/React.createElement(Alert, {
      title: "Failed to load referenced document"
    }, /*#__PURE__*/React.createElement(Text, {
      muted: true,
      size: 1
    }, "Error: ", error.message)));
  }

  if (!referenceInfo) {
    return null;
  }

  if (referenceInfo.availability.reason === 'PERMISSION_DENIED') {
    return /*#__PURE__*/React.createElement(Stack, {
      space: 2,
      padding: 1
    }, "Insufficient permissions to view this document");
  }

  const refType = props.type.to.find(toType => toType.name === referenceInfo.type);

  if (!refType) {
    return /*#__PURE__*/React.createElement(Stack, {
      space: 2,
      padding: 1
    }, "Search returned a type that's not valid for this reference: \"$", referenceInfo.type, "\"");
  }

  return referenceInfo && refType && /*#__PURE__*/React.createElement(ReferencePreview, {
    availability: referenceInfo.availability,
    id: referenceInfo.id,
    layout: "default",
    preview: referenceInfo.preview,
    refType: refType,
    renderPreview: renderPreview,
    showTypeLabel: props.type.to.length > 1
  });
}
