// sanity/lib/esm/form/components/ActivateOnFocus/ActivateOnFocus.js
// This is transitional in order to track usage of the ActivateOnFocusPart part from within the form-builder package
import React from 'react';
import { OverlayContainer, FlexContainer, CardContainer, ContentContainer } from './ActivateOnFocus.styles';
/**
 * @internal
 */

/**
 * @internal
 */
export function ActivateOnFocus(props) {
  const {
    children,
    message,
    onActivate,
    isOverlayActive
  } = props;

  function handleClick() {
    if (onActivate) {
      onActivate();
    }
  }

  function handleBlur() {
    if (onActivate && isOverlayActive) {
      onActivate();
    }
  }

  return /*#__PURE__*/React.createElement(OverlayContainer, {
    onClick: handleClick,
    onBlur: handleBlur
  }, isOverlayActive && /*#__PURE__*/React.createElement(FlexContainer, {
    tabIndex: 0,
    align: "center",
    justify: "center"
  }, /*#__PURE__*/React.createElement(CardContainer // Almost all input elements have radius=1, and this component is
  // typically used for overlaying input elements.
  // @todo Consider making `radius` a component property of `ActivateOnFocus`.
  , {
    radius: 1
  }), /*#__PURE__*/React.createElement(ContentContainer, null, message)), children);
}
