// sanity/lib/cjs/desk/components/MissingDocumentTypesMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MissingDocumentTypesMessage = MissingDocumentTypesMessage;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _studio = require("../../studio");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function MissingDocumentTypesMessage() {
  const {
    name: sourceName
  } = (0, _studio.useSource)();
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    radius: 2,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    marginLeft: 3,
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "h1",
    size: 1,
    weight: "bold"
  }, "No schema types in the ", /*#__PURE__*/_react.default.createElement("em", null, sourceName), " source!"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Please add schema types in your source configuration."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("a", {
    href: "https://beta.sanity.io/docs/platform/studio/config",
    target: "_blank",
    rel: "noreferrer"
  }, "Learn how to add a schema types \u2192"))))))));
}
