// sanity/lib/cjs/desk/components/PublishedStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PublishedStatus = PublishedStatus;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _TextWithTone = require("../../components/TextWithTone");

var _TimeAgo = require("./TimeAgo");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PublishedStatus(props) {
  const {
    document
  } = props;
  const updatedAt = document && '_updatedAt' in document && document._updatedAt;
  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    portal: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, document ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Published ", updatedAt && /*#__PURE__*/_react.default.createElement(_TimeAgo.TimeAgo, {
      time: updatedAt
    })) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Not published")))
  }, /*#__PURE__*/_react.default.createElement(_TextWithTone.TextWithTone, {
    tone: "positive",
    dimmed: !document,
    muted: !document,
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.PublishIcon, null)));
}
