// sanity/lib/esm/desk/panes/documentList/DocumentListPaneHeader.js
import { ArrowLeftIcon } from '@sanity/icons';
import { Button } from '@sanity/ui';
import React, { memo, useMemo } from 'react';
import { BackLink, PaneHeader, PaneHeaderActions } from '../../components';
import { useDeskTool } from '../../useDeskTool';
export const DocumentListPaneHeader = /*#__PURE__*/memo(_ref => {
  let {
    index,
    initialValueTemplates = [],
    menuItems = [],
    menuItemGroups = [],
    setLayout,
    setSortOrder,
    title
  } = _ref;
  const {
    features
  } = useDeskTool();
  const actionHandlers = useMemo(() => {
    return {
      setLayout: _ref2 => {
        let {
          layout: value
        } = _ref2;
        setLayout(value);
      },
      setSortOrder: sort => {
        setSortOrder(sort);
      }
    };
  }, [setLayout, setSortOrder]);
  return /*#__PURE__*/React.createElement(PaneHeader, {
    backButton: features.backButton && index > 0 && /*#__PURE__*/React.createElement(Button, {
      as: BackLink,
      "data-as": "a",
      icon: ArrowLeftIcon,
      mode: "bleed"
    }),
    title: title,
    actions: /*#__PURE__*/React.createElement(PaneHeaderActions, {
      initialValueTemplateItems: initialValueTemplates,
      actionHandlers: actionHandlers,
      menuItemGroups: menuItemGroups,
      menuItems: menuItems
    })
  });
});
DocumentListPaneHeader.displayName = 'DocumentListPaneHeader';
