// sanity/lib/cjs/studio/components/navbar/workspace/WorkspaceAuth/Layout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Layout = Layout;

var _logos = require("@sanity/logos");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const LINKS = [{
  url: 'https://slack.sanity.io/',
  title: 'Community'
}, {
  url: 'https://www.sanity.io/docs',
  title: 'Docs'
}, {
  url: 'https://www.sanity.io/legal/privacy',
  title: 'Privacy'
}, {
  url: 'https://www.sanity.io',
  title: 'sanity.io'
}];
const StyledText = (0, _styledComponents.default)(_ui.Text)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  a {\n    color: inherit;\n  }\n"])));

function Layout(props) {
  const {
    children,
    footer,
    header
  } = props;
  return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 6
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    border: true,
    radius: 3,
    overflow: "auto"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, typeof header === 'object' && /*#__PURE__*/_react.default.createElement(_ui.Box, null, header), typeof header === 'string' && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingY: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    align: "center",
    size: 1
  }, header)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    borderTop: Boolean(header),
    borderBottom: Boolean(footer)
  }, children)), footer && /*#__PURE__*/_react.default.createElement(_ui.Box, null, footer))), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    gap: 4,
    justify: "center",
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 3
  }, /*#__PURE__*/_react.default.createElement(_logos.SanityLogo, null)), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    gap: 2
  }, LINKS.map((link, index) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: link.title
  }, /*#__PURE__*/_react.default.createElement(StyledText, {
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("a", {
    href: link.url,
    target: "_blank",
    rel: "noopener noreferrer"
  }, link.title)), index < LINKS.length - 1 && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true
  }, "\u2022"))))));
}
