// sanity/lib/dts/src/form/inputs/arrays/ArrayOfObjectsInput/InsertMenu.js
import React, { memo } from 'react';
import { MenuGroup, MenuItem } from '@sanity/ui';
import { InsertAboveIcon, InsertBelowIcon } from '@sanity/icons';
const MENU_POPOVER_PROPS = {
    portal: true,
    tone: 'default',
    placement: 'left',
    constrainSize: true,
};
export const InsertMenu = memo(function InsertMenu(props) {
    const { types, onInsert } = props;
    return (React.createElement(React.Fragment, null,
        React.createElement(InsertMenuGroup, { pos: "before", types: types, onInsert: onInsert, text: "Add item before", icon: InsertAboveIcon }),
        React.createElement(InsertMenuGroup, { pos: "after", types: types, onInsert: onInsert, text: "Add item after", icon: InsertBelowIcon })));
});
function InsertMenuGroup(props) {
    const { types, onInsert, pos, text, icon } = props;
    if (types?.length === 1) {
        return React.createElement(MenuItem, { key: pos, text: text, icon: icon, onClick: () => onInsert(pos, types[0]) });
    }
    return (React.createElement(MenuGroup, { text: text, key: pos, popover: MENU_POPOVER_PROPS }, types?.map((insertableType) => (React.createElement(MenuItem, { key: insertableType.name, icon: insertableType.icon, text: insertableType.title, onClick: () => onInsert(pos, insertableType) })))));
}
