// sanity/lib/cjs/desk/components/pane/PaneContextMenuButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneContextMenuButton = PaneContextMenuButton;

var _autoId = require("@reach/auto-id");

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _router = require("../../../router");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const CONTEXT_MENU_POPOVER_PROPS = {
  constrainSize: true,
  placement: 'bottom',
  portal: true
};
/**
 * @beta This API will change. DO NOT USE IN PRODUCTION.
 */

function PaneContextMenuButton(props) {
  const {
    items,
    itemGroups,
    onAction
  } = props;
  const id = (0, _autoId.useId)() || '';
  const groups = (0, _react.useMemo)(() => {
    if (!itemGroups || itemGroups.length === 0) {
      return [{
        id: '$default',
        items
      }];
    }

    const defaultGroup = {
      id: '$default',
      items: []
    };
    const groupMap = itemGroups.reduce((acc, group) => {
      acc[group.id] = {
        id: group.id,
        title: group.title,
        items: []
      };
      return acc;
    }, {});

    for (const item of items) {
      const group = groupMap[item.group || '$default'] || defaultGroup;
      group.items.push(item);
    }

    return Object.values(groupMap).concat([defaultGroup]).filter(g => g.items.length > 0);
  }, [items, itemGroups]);
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.EllipsisVerticalIcon,
      mode: "bleed",
      padding: 3,
      title: "Show menu"
    }),
    id: id,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, groups.map((group, groupIndex) =>
    /*#__PURE__*/
    // eslint-disable-next-line react/no-array-index-key
    _react.default.createElement(_react.Fragment, {
      key: groupIndex
    }, groupIndex > 0 && /*#__PURE__*/_react.default.createElement(_ui.MenuDivider, null), group.title && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 3,
      paddingTop: 3,
      paddingBottom: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Label, {
      muted: true
    }, group.title)), group.items.map((item, itemIndex) => /*#__PURE__*/_react.default.createElement(PaneContextMenuItem, {
      item: item // eslint-disable-next-line react/no-array-index-key
      ,
      key: "".concat(itemIndex, "-").concat(item.title),
      onAction: onAction
    }))))),
    popover: CONTEXT_MENU_POPOVER_PROPS,
    portal: true
  });
}

function PaneContextMenuItem(props) {
  const {
    item,
    onAction
  } = props;
  const handleClick = (0, _react.useCallback)(() => {
    onAction(item);
  }, [item, onAction]);
  const hotkeys = (0, _react.useMemo)(() => {
    if (!item.shortcut) return undefined;
    return item.shortcut.split('+');
  }, [item]);
  const IntentButtonOrActionButton = /*#__PURE__*/(0, _react.forwardRef)((linkProps, linkRef) => item.intent ? /*#__PURE__*/_react.default.createElement(_router.IntentLink, _extends({}, linkProps, {
    intent: item.intent.type,
    params: item.intent.params,
    ref: linkRef
  })) : /*#__PURE__*/_react.default.createElement("button", _extends({
    type: "button"
  }, linkProps)));
  IntentButtonOrActionButton.displayName = 'Link';
  return /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
    as: IntentButtonOrActionButton,
    "data-as": item.intent ? 'a' : 'button',
    hotkeys: hotkeys,
    icon: item.icon,
    onClick: handleClick,
    text: item.title
  });
}
