// sanity/lib/cjs/desk/components/pane/PaneHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneHeader = void 0;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _transitional = require("../../../components/transitional");

var _usePane = require("./usePane");

var _PaneHeader = require("./PaneHeader.styles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @beta This API will change. DO NOT USE IN PRODUCTION.
 */
const PaneHeader = /*#__PURE__*/(0, _react.forwardRef)(function PaneHeader(props, ref) {
  const {
    actions,
    backButton,
    loading,
    subActions,
    tabs,
    title
  } = props;
  const {
    collapse,
    collapsed,
    expand,
    rootElement: paneElement
  } = (0, _usePane.usePane)();
  const paneRect = (0, _ui.useElementRect)(paneElement || null);
  const layoutStyle = (0, _react.useMemo)(() => ({
    width: collapsed ? (paneRect === null || paneRect === void 0 ? void 0 : paneRect.height) || window.innerHeight : undefined
  }), [collapsed, paneRect]);
  const handleTitleClick = (0, _react.useCallback)(() => {
    if (collapsed) return;
    collapse();
  }, [collapse, collapsed]);
  const handleLayoutClick = (0, _react.useCallback)(() => {
    if (!collapsed) return;
    expand();
  }, [collapsed, expand]);
  return /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, {
    zOffset: 100
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.Root, {
    "data-collapsed": collapsed ? '' : undefined,
    "data-testid": "pane-header",
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_transitional.LegacyLayerProvider, {
    zOffset: "paneHeader"
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-collapsed": collapsed ? '' : undefined,
    tone: "inherit"
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.Layout, {
    onClick: handleLayoutClick,
    padding: 2,
    paddingBottom: tabs || subActions ? 1 : 2,
    sizing: "border",
    style: layoutStyle
  }, backButton, /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleBox, {
    flex: 1,
    onClick: handleTitleClick,
    paddingY: 3,
    paddingLeft: backButton ? 1 : 3
  }, loading && /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleTextSkeleton, {
    animated: true,
    radius: 1
  }), !loading && /*#__PURE__*/_react.default.createElement(_PaneHeader.TitleText, {
    tabIndex: 0,
    textOverflow: "ellipsis",
    weight: "semibold"
  }, title)), actions && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    hidden: collapsed,
    paddingLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_transitional.LegacyLayerProvider, {
    zOffset: "paneHeader"
  }, actions))), (tabs || subActions) && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    hidden: collapsed,
    paddingTop: 0,
    paddingRight: 2,
    paddingBottom: 2,
    paddingLeft: 3,
    overflow: "auto"
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.TabsBox, {
    flex: 1,
    marginRight: subActions ? 3 : 0
  }, /*#__PURE__*/_react.default.createElement("div", null, tabs)), subActions && /*#__PURE__*/_react.default.createElement(_ui.Box, null, subActions))))));
});
exports.PaneHeader = PaneHeader;
