// sanity/lib/esm/form/inputs/PortableText/__workshop__/editObjects/formBuilder.js
var _templateObject;

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import { TextArea } from '@sanity/ui';
import React, { useCallback } from 'react';
import styled, { css } from 'styled-components';
import { set } from '../../../../patch';
const DebugTextArea = styled(TextArea)(_ref => {
  let {
    theme
  } = _ref;
  return css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    font-family: ", ";\n  "])), theme.sanity.fonts.code.family);
});

function DebugInput(props) {
  const {
    onChange,
    onBlur,
    onFocus,
    readOnly,
    focusRef
  } = props;
  const handleChange = useCallback(() => {
    onChange(set({}));
  }, [onChange]);
  return /*#__PURE__*/React.createElement(DebugTextArea, {
    onBlur: onBlur,
    onChange: handleChange,
    onFocus: onFocus,
    padding: 3,
    radius: 1,
    readOnly: readOnly,
    ref: focusRef,
    rows: 100,
    value: JSON.stringify(props.value, null, 2)
  });
} // export const resolveInputComponent: FormInputComponentResolver = () => {
//   return DebugInput
// }
// export const resolvePreviewComponent = () => {
//   return function PreviewAny() {
//     return <div>preview</div>
//   }
// }
