// sanity/lib/cjs/form/inputs/arrays/ArrayOfObjectsInput/item/ItemWithMissingType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ItemWithMissingType = ItemWithMissingType;

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _icons = require("@sanity/icons");

var _content = require("@sanity/util/content");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const PopoverCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  max-width: ", "px;\n"])), _ref => {
  let {
    theme
  } = _ref;
  return theme.sanity.container[1];
});

function ItemWithMissingType(props) {
  const {
    value,
    onFocus,
    vertical,
    ...rest
  } = props;

  const [showDetails, setShowDetails] = _react.default.useState(false);

  const [popoverRef, setPopoverRef] = _react.default.useState(null);

  (0, _ui.useClickOutside)(() => setShowDetails(false), [popoverRef]);

  const handleKeyDown = _react.default.useCallback(e => {
    if (e.key === 'Escape' || e.key === 'Tab') {
      setShowDetails(false);
    }
  }, []);

  const handleShowDetails = (0, _react.useCallback)(() => {
    setShowDetails(v => !v);
  }, []);
  const typeName = (0, _content.resolveTypeName)(value);
  return /*#__PURE__*/_react.default.createElement(_ui.Popover, {
    open: showDetails,
    ref: setPopoverRef,
    onKeyDown: handleKeyDown,
    portal: true,
    constrainSize: true,
    tone: "default",
    content: /*#__PURE__*/_react.default.createElement(PopoverCard, {
      margin: 1,
      padding: 3,
      onKeyDown: handleKeyDown,
      tabIndex: 0,
      overflow: "auto"
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold"
    }, "Why is this happening?")), /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "The current schema does not declare items of type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " as valid for this list. This could mean that the type has been removed as a valid item type, or that someone else has added it to their own local schema that is not yet deployed."), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.BulbOutlineIcon, null), " You can still move or delete this item, but it cannot be edited since the schema definition for its type is nowhere to be found.")), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      weight: "semibold"
    }, "JSON representation of this item:"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 2,
      overflow: "auto",
      border: true
    }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 1,
      as: "pre"
    }, JSON.stringify(value, null, 2))))))
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, _extends({
    as: "button",
    type: "button",
    radius: 2,
    tone: "inherit",
    padding: 2,
    onFocus: onFocus,
    onClick: handleShowDetails,
    onKeyDown: handleKeyDown
  }, rest), vertical && /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    align: "center"
  }, "Item type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " not defined for this list")), !vertical && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginRight: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    textOverflow: "ellipsis"
  }, "Item type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " not defined for this list")))));
}
