// sanity/lib/cjs/form/inputs/arrays/ArrayOfObjectsInput/item/RowItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RowItem = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _autoId = require("@reach/auto-id");

var _formField = require("../../../../components/formField");

var _presence = require("../../../../../presence");

var _DragHandle = require("../../common/DragHandle");

var _randomKey = require("../../common/randomKey");

var _ArrayInput = require("../ArrayInput");

var _InsertMenu = require("../InsertMenu");

var _empty = require("../../../../utils/empty");

var _ItemWithMissingType = require("./ItemWithMissingType");

var _RowWrapper = require("./components/RowWrapper");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  paddingX: 1,
  paddingY: 3
});

const MENU_POPOVER_PROPS = {
  portal: true,
  tone: 'default'
};

const RowItem = /*#__PURE__*/_react.default.forwardRef(function RegularItem(props, ref) {
  var _type$options;

  const focusRef = _react.default.useRef(null);

  const {
    isSortable,
    value,
    onClick,
    onFocus,
    type,
    readOnly,
    presence,
    onInsert,
    insertableTypes,
    onRemove,
    renderPreview,
    validation = _empty.EMPTY_ARRAY,
    ...rest
  } = props;
  const hasErrors = validation.some(v => v.level === 'error');
  const hasWarnings = validation.some(v => v.level === 'warning');
  const handleDuplicate = (0, _react.useCallback)(() => {
    onInsert({
      items: [{ ...value,
        _key: (0, _randomKey.randomKey)()
      }],
      position: 'after'
    });
  }, [onInsert, value]);
  const handleInsert = (0, _react.useCallback)((pos, insertType) => {
    onInsert({
      items: [{ ...(0, _ArrayInput.createProtoValue)(insertType),
        _key: (0, _randomKey.randomKey)()
      }],
      position: pos
    });
  }, [onInsert]);
  const id = (0, _autoId.useId)();
  return /*#__PURE__*/_react.default.createElement(_RowWrapper.RowWrapper, _extends({}, rest, {
    ref: ref,
    radius: 2,
    padding: 1,
    tone: readOnly ? 'transparent' : hasErrors ? 'critical' : hasWarnings ? 'caution' : 'default'
  }), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, isSortable && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    className: "dragHandle",
    tone: "inherit",
    marginRight: 1
  }, dragHandle), type ? /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    type: "button",
    tone: "inherit",
    radius: 2,
    padding: 1,
    flex: 1,
    onClick: onClick,
    ref: focusRef,
    onFocus: onFocus,
    __unstable_focusRing: true
  }, renderPreview({
    layout: type.options && 'layout' in type.options && ((_type$options = type.options) === null || _type$options === void 0 ? void 0 : _type$options.layout) === 'grid' ? 'media' : 'default',
    schemaType: type,
    value
  })) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ItemWithMissingType.ItemWithMissingType, {
    value: value,
    onFocus: onFocus
  })), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, !readOnly && presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_formField.FormFieldValidationStatus, {
    __unstable_showSummary: !(value !== null && value !== void 0 && value._ref)
  })), !readOnly && /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      padding: 2,
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "".concat(id, "-menuButton"),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Remove",
      tone: "critical",
      icon: _icons.TrashIcon,
      onClick: onRemove
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Duplicate",
      icon: _icons.CopyIcon,
      onClick: handleDuplicate
    }), /*#__PURE__*/_react.default.createElement(_InsertMenu.InsertMenu, {
      types: insertableTypes,
      onInsert: handleInsert
    })),
    placement: "right",
    popover: MENU_POPOVER_PROPS
  }), !value._key && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "This item is missing the required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property.")),
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution"
  }, "Missing key"))))));
});

exports.RowItem = RowItem;
