// sanity/lib/esm/form/inputs/DateInputs/base/calendar/CalendarDay.js
import { Card, Text } from '@sanity/ui';
import React, { useCallback } from 'react';
export function CalendarDay(props) {
  const {
    date,
    focused,
    isCurrentMonth,
    isToday,
    onSelect,
    selected
  } = props;
  const handleClick = useCallback(() => {
    onSelect(date);
  }, [date, onSelect]);
  return /*#__PURE__*/React.createElement("div", {
    "aria-selected": selected,
    "data-ui": "CalendarDay"
  }, /*#__PURE__*/React.createElement(Card, {
    "aria-label": date.toDateString(),
    "aria-pressed": selected,
    as: "button",
    __unstable_focusRing: true,
    "data-weekday": true,
    "data-focused": focused ? 'true' : '',
    role: "button",
    tabIndex: -1,
    onClick: handleClick,
    padding: 3,
    radius: 2,
    selected: selected,
    tone: isToday || selected ? 'primary' : 'default'
  }, /*#__PURE__*/React.createElement(Text, {
    muted: !selected && !isCurrentMonth,
    style: {
      textAlign: 'center'
    },
    weight: isCurrentMonth ? 'medium' : 'regular'
  }, date.getDate())));
}
