// sanity/lib/cjs/search/__workshop__/DevStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DevStory;

var _ui = require("@sanity/ui");

var _uiWorkshop = require("@sanity/ui-workshop");

var _react = _interopRequireDefault(require("react"));

var _hooks = require("../../hooks");

var _preview = require("../../preview");

var _useDocumentSearchResults = require("../useDocumentSearchResults");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DevStory() {
  var _useString;

  const schema = (0, _hooks.useSchema)();
  const includeDrafts = (0, _uiWorkshop.useBoolean)('Include drafts', true);
  const query = (_useString = (0, _uiWorkshop.useString)('Query', 'grrm', 'Props')) !== null && _useString !== void 0 ? _useString : '';
  const limit = (0, _uiWorkshop.useNumber)('Limit', 10, 'Props') || 1;
  const results = (0, _useDocumentSearchResults.useDocumentSearchResults)({
    includeDrafts,
    limit,
    query
  });
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, results.value.map(result => {
    const type = schema.get(result.hit._type);

    if (!type) {
      return /*#__PURE__*/_react.default.createElement(_ui.Card, {
        border: true,
        key: result.hit._id,
        padding: 3,
        tone: "critical"
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1
      }, "Unknown type: ", /*#__PURE__*/_react.default.createElement("code", null, result.hit._type)));
    }

    return /*#__PURE__*/_react.default.createElement(_ui.Card, {
      border: true,
      key: result.hit._id
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_preview.SanityPreview, {
      schemaType: type,
      value: {
        _type: result.hit._type,
        _id: result.hit._id
      }
    })), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      tone: "transparent"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      padding: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 1
    }, result.hit._id)), /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 1
    }, result.score.toFixed(2)))));
  }))));
}
