// sanity/lib/cjs/studio/components/navbar/workspace/WorkspacePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaCard = void 0;
exports.WorkspacePreview = WorkspacePreview;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireWildcard(require("react"));

var _reactIs = require("react-is");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const STATE_TITLES = {
  'logged-in': '',
  'logged-out': 'Signed out',
  'no-access': ''
};
const MediaCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  width: 35px;\n  height: 35px;\n\n  svg {\n    width: 100%;\n    height: 100%;\n  }\n"])));
exports.MediaCard = MediaCard;

const createIcon = icon => {
  if ((0, _reactIs.isValidElementType)(icon)) return /*#__PURE__*/(0, _react.createElement)(icon);
  if ( /*#__PURE__*/(0, _react.isValidElement)(icon)) return icon;
  return undefined;
};

function WorkspacePreview(props) {
  const {
    state,
    subtitle,
    selected,
    title,
    icon,
    iconRight
  } = props;
  const iconComponent = (0, _react.useMemo)(() => createIcon(icon), [icon]);
  const iconRightComponent = (0, _react.useMemo)(() => createIcon(iconRight), [iconRight]);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    flex: "none",
    gap: 3
  }, /*#__PURE__*/_react.default.createElement(MediaCard, {
    radius: 2,
    tone: "transparent"
  }, iconComponent), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis",
    weight: "medium"
  }, title), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, subtitle)), state && STATE_TITLES[state] && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true,
    textOverflow: "ellipsis"
  }, STATE_TITLES[state])), (selected || iconRightComponent) && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    gap: 4,
    paddingLeft: 3,
    paddingRight: 2
  }, selected && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true
  }, /*#__PURE__*/_react.default.createElement(_icons.CheckmarkIcon, null)), iconRightComponent && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true
  }, iconRightComponent)));
}
