// sanity/lib/cjs/form/patch/PatchEvent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PatchEvent = void 0;

var _flatten2 = _interopRequireDefault(require("lodash/flatten"));

var _patch = require("./patch");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @internal
 */
class PatchEvent {
  static from(input) {
    if (input instanceof PatchEvent) {
      return input;
    }

    return new PatchEvent(Array.isArray(input) ? (0, _flatten2.default)(input) : [input]);
  }

  constructor(patches) {
    _defineProperty(this, "patches", void 0);

    this.patches = patches;
  }

  prepend() {
    for (var _len = arguments.length, patches = new Array(_len), _key = 0; _key < _len; _key++) {
      patches[_key] = arguments[_key];
    }

    return PatchEvent.from([...(0, _flatten2.default)(patches), ...this.patches]);
  }

  append() {
    for (var _len2 = arguments.length, patches = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      patches[_key2] = arguments[_key2];
    }

    return PatchEvent.from([...this.patches, ...(0, _flatten2.default)(patches)]);
  }

  prefixAll(segment) {
    return PatchEvent.from(this.patches.map(patch => (0, _patch.prefixPath)(patch, segment)));
  }

}

exports.PatchEvent = PatchEvent;
